{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm = pd.read_csv('../EL.data.model.csv').drop(columns = ['Unnamed: 0'], errors = 'ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ValidValue', 'Component', 'AnalysisType', 'MeasurementUnit',\n",
       "       'ManifestColumn', 'Ontology', 'Repository', 'Demographics',\n",
       "       'BaseAnnotation'], dtype=object)"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm['Parent'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": [
    "other_values = [\"Other\",\"Unknown\",\"Not collected\",\"Not applicable\",\"Not specified\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_df = pd.read_excel(\"../_data/RFC Tables/EL RFC_ Assay_proteomics Data Model.xlsx\")\n",
    "\n",
    "p_df = p_df.rename(\n",
    "    columns={\n",
    "        \"key\": \"Attribute\",\n",
    "        \"description\": \"Description\",\n",
    "        \"required\": \"Required\",\n",
    "        \"requires\": \"Module\",\n",
    "        \"concept source ontology\": \"Ontology\",\n",
    "        \"valid values\": \"Valid Values\",\n",
    "        \"type\": \"columnType\",\n",
    "        \"note\": \"Notes\",\n",
    "    }\n",
    ")\n",
    "p_df = p_df.fillna(\"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>Module</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>columnType</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>specimenID</td>\n",
       "      <td>Identifying string linked to a particular samp...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Biospecimen,\\nproteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>values unique to each data contributor)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sampleType</td>\n",
       "      <td>The type of sample collected or the term used ...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,\\n</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>specifySampleType</td>\n",
       "      <td>If \"other\" is selected list the type of sample</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics,\\nsampleType = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>experimentType</td>\n",
       "      <td>The type of experiment used.\\n\\nProvide a valu...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomic Data Commons</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The name of the measurement technique describi...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the me...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics,\\nmeasurementTechnique = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>technologyPlatformVersion</td>\n",
       "      <td>The specific version (application, manufacture...</td>\n",
       "      <td>\\n</td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C45378</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>specifyPlatformVersion</td>\n",
       "      <td>If \"other\" list the name of the platform version</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics,\\ntechnologyPlatformVersion = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>platformLocation</td>\n",
       "      <td>The name of the laboratory, facility, vendor, ...</td>\n",
       "      <td>\\n</td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,\\nDSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>specifyPlatformLocation</td>\n",
       "      <td>If \"other\" list the name of the platform location</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics,\\nplatformLocation = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>instrumentModel</td>\n",
       "      <td>The model of the instrument used.\\n\\nProvide a...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,\\nDSLWG,\\nProposed minimum me...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>samplePrepProtocol</td>\n",
       "      <td>An internet address that may provide more deta...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,\\nDSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>dataFile</td>\n",
       "      <td>Name of additional file(s) accompanying the da...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>databaseSource</td>\n",
       "      <td>The name of the organization, project, or labo...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001012</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>databaseWeblink</td>\n",
       "      <td>An internet address that may provide details o...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,\\nDSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>databaseName</td>\n",
       "      <td>The name of the search database (nr, SwissProt...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001013</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>modificationParameters</td>\n",
       "      <td>Modification parameters for a search engine ru...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/terms?...</td>\n",
       "      <td>Possible values are listed under modification ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>quantationStrategy</td>\n",
       "      <td>The general strategy used for differential ana...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>labelQuantitation</td>\n",
       "      <td>Type of labeling used \\n\\nProvide a value OR p...</td>\n",
       "      <td>AminoxyTMT,\\nDiART,\\nDiLeu,\\nGlyco-TMT,\\nHydra...</td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,\\nDSLWG,\\nhttps://www.ebi.ac....</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>labelFreeQuantitation</td>\n",
       "      <td>Type of label-free data analysis strategy \\n\\n...</td>\n",
       "      <td>Label-free protein level quantitation,\\nLabel-...</td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics\\nlabelQuantiation = Label Free</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Data Commons</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>specifyLabelFreeQuantitation</td>\n",
       "      <td>If \"other\" is selected please specify</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics\\nlabelFreeQuantitation = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>isobaricLabelingReagent</td>\n",
       "      <td>Reagent set used for isobaric labeling.\\n\\nExa...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Data Commons</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>msTarget</td>\n",
       "      <td>Specifies whether or not a specific molecule(s...</td>\n",
       "      <td>Targeted,\\nUntargeted\\nUnknown,\\nNot collected...</td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,\\nDSLWG</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>msAnalyzerTypeMS1</td>\n",
       "      <td>Describes the mass spectrometer analyzer used ...</td>\n",
       "      <td>LTQ Orbitrap XL,\\nOrbitrap Fusion,\\nOrbitrap F...</td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,\\nProposed minimum metadata r...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>specifyMSAnalyzerTypeMS1</td>\n",
       "      <td>If \"other\" is selected list the name of the ma...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics,\\nmsAnalyzerTypeMS1 = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>msAnalyzerTypeMS2</td>\n",
       "      <td>Describes the mass spectrometer analyzer used ...</td>\n",
       "      <td>LTQ Orbitrap XL,\\nOrbitrap Fusion,\\nOrbitrap F...</td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,\\nProposed minimum metadata r...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>specifyMSAnalyzerTypeMS2</td>\n",
       "      <td>If \"other\" is selected list the name of the ma...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics,\\nmsAnalyzerTypeMS2 = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>msAssayTechnique</td>\n",
       "      <td>Specified the mass spectrometer technique used...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,\\nDSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>specifyMSAssayTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the ma...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,\\nmsAssayTechnique = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>msAnalyteType</td>\n",
       "      <td>The type of biospecimen subjected to analysis ...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,\\n</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C156434\\nP...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>specifyMSAnalyteType</td>\n",
       "      <td>If \"other\" list the type of analyte</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,\\nmsAnalyteType = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>specifyProteomicsAssayType</td>\n",
       "      <td>List the name of the type of peptide quantific...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,\\nproteomicsAssayType = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>specifyMSInstrumentModel</td>\n",
       "      <td>List the name of the instrument model</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,\\nmsInstrumentModel = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>digestionMethod</td>\n",
       "      <td>The type of digestion (ex. in-solution, in-gel...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,\\nDSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>specifyDigestionMethod</td>\n",
       "      <td>If \"other\" list the type of digestion</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,\\ndigestionMethod = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>digestionReagent</td>\n",
       "      <td>Enzymes or reagents used for digestion. \\n\\n</td>\n",
       "      <td>2-iodobenzoate,\\nArg-C, Asp-N,\\nAsp-N_ambic,\\n...</td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Data Commons</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>specifyDigestionReagent</td>\n",
       "      <td>If \"other\" is selected specify the name of the...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics digestionReagent=other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>fractionIdentifier</td>\n",
       "      <td>Number of fractions in the experimental run\\n\\...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,\\nThe Proteomics Sample Metad...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>enrichmentStrategy</td>\n",
       "      <td>Type of peptide or protein enrichment\\n\\n</td>\n",
       "      <td>Acetylome\\nProteome,\\nPhosphoproteome,\\nGlycop...</td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomic Data Commons</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>specifyEnrichmentStrategy</td>\n",
       "      <td>If \"other\" is selected specify the name</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomic Data Commons</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>enrichment</td>\n",
       "      <td>Protocol or reagent used for enrichment (ex. F...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomic Data Commons</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>fractionIdentifier</td>\n",
       "      <td>An identifier string that describes the sample...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>number</td>\n",
       "      <td>Sage Bionetworks,\\nThe Proteomics Sample Metad...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>acquisitionMode</td>\n",
       "      <td>The specific aspect of a mass spectrometer met...</td>\n",
       "      <td>Unknown,\\nNot collected,\\nNot applicable,\\nNot...</td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/terms?...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>acquisitionSoftware</td>\n",
       "      <td>The name of the acquisition software used</td>\n",
       "      <td>Unknown,\\nNot collected,\\nNot applicable,\\nNot...</td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>acquistionSoftwareVersion</td>\n",
       "      <td>The version number of software used</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics,\\nacquisitionSoftware</td>\n",
       "      <td>False</td>\n",
       "      <td>number</td>\n",
       "      <td>Sage Bionetworks,\\nDSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>databaseName</td>\n",
       "      <td>The name of the search database (nr, SwissProt...</td>\n",
       "      <td>MassBank of North America (MoNA),\\nNIST17,\\nMe...</td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001013</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>specifyDatabaseName</td>\n",
       "      <td>If \"other\" list the name of the search database</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics,\\ndatabaseName = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>databaseSource</td>\n",
       "      <td>The name of the organization, project, or labo...</td>\n",
       "      <td>Other\\nUnknown,\\nNot collected,\\nNot applicabl...</td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001012</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>specifyDatabaseSource</td>\n",
       "      <td>If \"other\" list the name of the database source</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics,\\ndatabaseSource = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>spectrometerFrequency</td>\n",
       "      <td>The frequency at which a spectrometer causes h...</td>\n",
       "      <td>\\n</td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>number</td>\n",
       "      <td>Sage Bionetworks,\\nDSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>specifyOtherDataFile</td>\n",
       "      <td>If \"other\" is selected, list the name of the d...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics, dataFile = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>specifyOtherModificationParameters</td>\n",
       "      <td>If \"other\" is selected, list the custom modifi...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,\\nmodificationParameters = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>cleavageAgents</td>\n",
       "      <td>Name(s) of the enzyme used (possible values ar...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>specifyOtherCleavageAgent</td>\n",
       "      <td>If \"other\" is selected, list the enzyme name.</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,\\ncleavageAgents = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>fragmentMassTolerance</td>\n",
       "      <td>Units of measure are Da or ppm (ex. \"0.6 Da\" o...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>precursorMassTolerance</td>\n",
       "      <td>Units of measure are Da or ppm (ex. \"0.6 Da\" o...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Attribute  \\\n",
       "0                           specimenID   \n",
       "1                           sampleType   \n",
       "2                    specifySampleType   \n",
       "3                       experimentType   \n",
       "4                 measurementTechnique   \n",
       "5          specifyMeasurementTechnique   \n",
       "6            technologyPlatformVersion   \n",
       "7               specifyPlatformVersion   \n",
       "8                     platformLocation   \n",
       "9              specifyPlatformLocation   \n",
       "10                     instrumentModel   \n",
       "11                  samplePrepProtocol   \n",
       "12                            dataFile   \n",
       "13                      databaseSource   \n",
       "14                     databaseWeblink   \n",
       "15                        databaseName   \n",
       "16              modificationParameters   \n",
       "17                  quantationStrategy   \n",
       "18                   labelQuantitation   \n",
       "19               labelFreeQuantitation   \n",
       "20        specifyLabelFreeQuantitation   \n",
       "21             isobaricLabelingReagent   \n",
       "22                            msTarget   \n",
       "23                   msAnalyzerTypeMS1   \n",
       "24            specifyMSAnalyzerTypeMS1   \n",
       "25                   msAnalyzerTypeMS2   \n",
       "26            specifyMSAnalyzerTypeMS2   \n",
       "27                    msAssayTechnique   \n",
       "28             specifyMSAssayTechnique   \n",
       "29                       msAnalyteType   \n",
       "30                specifyMSAnalyteType   \n",
       "31          specifyProteomicsAssayType   \n",
       "32            specifyMSInstrumentModel   \n",
       "33                     digestionMethod   \n",
       "34              specifyDigestionMethod   \n",
       "35                    digestionReagent   \n",
       "36             specifyDigestionReagent   \n",
       "37                  fractionIdentifier   \n",
       "38                  enrichmentStrategy   \n",
       "39           specifyEnrichmentStrategy   \n",
       "40                          enrichment   \n",
       "41                  fractionIdentifier   \n",
       "42                     acquisitionMode   \n",
       "43                 acquisitionSoftware   \n",
       "44           acquistionSoftwareVersion   \n",
       "45                        databaseName   \n",
       "46                 specifyDatabaseName   \n",
       "47                      databaseSource   \n",
       "48               specifyDatabaseSource   \n",
       "49               spectrometerFrequency   \n",
       "50                specifyOtherDataFile   \n",
       "51  specifyOtherModificationParameters   \n",
       "52                      cleavageAgents   \n",
       "53           specifyOtherCleavageAgent   \n",
       "54               fragmentMassTolerance   \n",
       "55              precursorMassTolerance   \n",
       "\n",
       "                                          Description  \\\n",
       "0   Identifying string linked to a particular samp...   \n",
       "1   The type of sample collected or the term used ...   \n",
       "2      If \"other\" is selected list the type of sample   \n",
       "3   The type of experiment used.\\n\\nProvide a valu...   \n",
       "4   The name of the measurement technique describi...   \n",
       "5   If \"other\" is selected list the name of the me...   \n",
       "6   The specific version (application, manufacture...   \n",
       "7    If \"other\" list the name of the platform version   \n",
       "8   The name of the laboratory, facility, vendor, ...   \n",
       "9   If \"other\" list the name of the platform location   \n",
       "10  The model of the instrument used.\\n\\nProvide a...   \n",
       "11  An internet address that may provide more deta...   \n",
       "12  Name of additional file(s) accompanying the da...   \n",
       "13  The name of the organization, project, or labo...   \n",
       "14  An internet address that may provide details o...   \n",
       "15  The name of the search database (nr, SwissProt...   \n",
       "16  Modification parameters for a search engine ru...   \n",
       "17  The general strategy used for differential ana...   \n",
       "18  Type of labeling used \\n\\nProvide a value OR p...   \n",
       "19  Type of label-free data analysis strategy \\n\\n...   \n",
       "20              If \"other\" is selected please specify   \n",
       "21  Reagent set used for isobaric labeling.\\n\\nExa...   \n",
       "22  Specifies whether or not a specific molecule(s...   \n",
       "23  Describes the mass spectrometer analyzer used ...   \n",
       "24  If \"other\" is selected list the name of the ma...   \n",
       "25  Describes the mass spectrometer analyzer used ...   \n",
       "26  If \"other\" is selected list the name of the ma...   \n",
       "27  Specified the mass spectrometer technique used...   \n",
       "28  If \"other\" is selected list the name of the ma...   \n",
       "29  The type of biospecimen subjected to analysis ...   \n",
       "30               If \"other\" list the type of analyte    \n",
       "31  List the name of the type of peptide quantific...   \n",
       "32              List the name of the instrument model   \n",
       "33  The type of digestion (ex. in-solution, in-gel...   \n",
       "34              If \"other\" list the type of digestion   \n",
       "35       Enzymes or reagents used for digestion. \\n\\n   \n",
       "36  If \"other\" is selected specify the name of the...   \n",
       "37  Number of fractions in the experimental run\\n\\...   \n",
       "38          Type of peptide or protein enrichment\\n\\n   \n",
       "39            If \"other\" is selected specify the name   \n",
       "40  Protocol or reagent used for enrichment (ex. F...   \n",
       "41  An identifier string that describes the sample...   \n",
       "42  The specific aspect of a mass spectrometer met...   \n",
       "43          The name of the acquisition software used   \n",
       "44                The version number of software used   \n",
       "45  The name of the search database (nr, SwissProt...   \n",
       "46    If \"other\" list the name of the search database   \n",
       "47  The name of the organization, project, or labo...   \n",
       "48    If \"other\" list the name of the database source   \n",
       "49  The frequency at which a spectrometer causes h...   \n",
       "50  If \"other\" is selected, list the name of the d...   \n",
       "51  If \"other\" is selected, list the custom modifi...   \n",
       "52  Name(s) of the enzyme used (possible values ar...   \n",
       "53     If \"other\" is selected, list the enzyme name.    \n",
       "54  Units of measure are Da or ppm (ex. \"0.6 Da\" o...   \n",
       "55  Units of measure are Da or ppm (ex. \"0.6 Da\" o...   \n",
       "\n",
       "                                         Valid Values  Required  \\\n",
       "0                                                          True   \n",
       "1                                                          True   \n",
       "2                                                         False   \n",
       "3                                                          True   \n",
       "4                                                          True   \n",
       "5                                                         False   \n",
       "6                                                  \\n      True   \n",
       "7                                                         False   \n",
       "8                                                  \\n      True   \n",
       "9                                                         False   \n",
       "10                                                         True   \n",
       "11                                                         True   \n",
       "12                                                         True   \n",
       "13                                                         True   \n",
       "14                                                         True   \n",
       "15                                                         True   \n",
       "16                                                         True   \n",
       "17                                                         True   \n",
       "18  AminoxyTMT,\\nDiART,\\nDiLeu,\\nGlyco-TMT,\\nHydra...      True   \n",
       "19  Label-free protein level quantitation,\\nLabel-...      True   \n",
       "20                                                        False   \n",
       "21                                                         True   \n",
       "22  Targeted,\\nUntargeted\\nUnknown,\\nNot collected...      True   \n",
       "23  LTQ Orbitrap XL,\\nOrbitrap Fusion,\\nOrbitrap F...      True   \n",
       "24                                                         True   \n",
       "25  LTQ Orbitrap XL,\\nOrbitrap Fusion,\\nOrbitrap F...      True   \n",
       "26                                                         True   \n",
       "27                                                         True   \n",
       "28                                                         True   \n",
       "29                                                         True   \n",
       "30                                                         True   \n",
       "31                                                         True   \n",
       "32                                                         True   \n",
       "33                                                         True   \n",
       "34                                                         True   \n",
       "35  2-iodobenzoate,\\nArg-C, Asp-N,\\nAsp-N_ambic,\\n...      True   \n",
       "36                                                        False   \n",
       "37                                                         True   \n",
       "38  Acetylome\\nProteome,\\nPhosphoproteome,\\nGlycop...      True   \n",
       "39                                                        False   \n",
       "40                                                         True   \n",
       "41                                                         True   \n",
       "42  Unknown,\\nNot collected,\\nNot applicable,\\nNot...      True   \n",
       "43  Unknown,\\nNot collected,\\nNot applicable,\\nNot...      True   \n",
       "44                                                         True   \n",
       "45  MassBank of North America (MoNA),\\nNIST17,\\nMe...      True   \n",
       "46                                                         True   \n",
       "47  Other\\nUnknown,\\nNot collected,\\nNot applicabl...      True   \n",
       "48                                                         True   \n",
       "49                                                 \\n      True   \n",
       "50                                                         True   \n",
       "51                                                         True   \n",
       "52                                                         True   \n",
       "53                                                         True   \n",
       "54                                                         True   \n",
       "55                                                         True   \n",
       "\n",
       "                                              Module  multivalue columnType  \\\n",
       "0                           Biospecimen,\\nproteomics       False     string   \n",
       "1                                         proteomics       False     string   \n",
       "2                    proteomics,\\nsampleType = other       False     string   \n",
       "3                                         proteomics        True     string   \n",
       "4                                         proteomics        True     string   \n",
       "5          proteomics,\\nmeasurementTechnique = other        True     string   \n",
       "6                                         proteomics        True     string   \n",
       "7     proteomics,\\ntechnologyPlatformVersion = other        True     string   \n",
       "8                                         proteomics        True     string   \n",
       "9              proteomics,\\nplatformLocation = other        True     string   \n",
       "10                                        proteomics        True     string   \n",
       "11                                        proteomics        True     string   \n",
       "12                                        proteomics        True     string   \n",
       "13                                        proteomics        True     string   \n",
       "14                                        proteomics        True     string   \n",
       "15                                        proteomics        True     string   \n",
       "16                                        proteomics        True     string   \n",
       "17                                        proteomics        True     string   \n",
       "18                                        proteomics        True     string   \n",
       "19         proteomics\\nlabelQuantiation = Label Free        True     string   \n",
       "20         proteomics\\nlabelFreeQuantitation = other       False     string   \n",
       "21                                        proteomics        True     string   \n",
       "22                                        proteomics        True     string   \n",
       "23                              mass spec proteomics        True     string   \n",
       "24  mass spec proteomics,\\nmsAnalyzerTypeMS1 = other        True     string   \n",
       "25                              mass spec proteomics        True     string   \n",
       "26  mass spec proteomics,\\nmsAnalyzerTypeMS2 = other        True     string   \n",
       "27                                        proteomics        True     string   \n",
       "28             proteomics,\\nmsAssayTechnique = other        True     string   \n",
       "29                                     proteomics,\\n        True     string   \n",
       "30                proteomics,\\nmsAnalyteType = other        True     string   \n",
       "31          proteomics,\\nproteomicsAssayType = other        True     string   \n",
       "32            proteomics,\\nmsInstrumentModel = other        True     string   \n",
       "33                                        proteomics        True     string   \n",
       "34              proteomics,\\ndigestionMethod = other       False     string   \n",
       "35                                        proteomics        True     string   \n",
       "36                 proteomics digestionReagent=other       False     string   \n",
       "37                                        proteomics        True     string   \n",
       "38                                        proteomics        True     string   \n",
       "39                                        proteomics        True     string   \n",
       "40                                        proteomics        True     string   \n",
       "41                                        proteomics       False     number   \n",
       "42                              mass spec proteomics       False     string   \n",
       "43                              mass spec proteomics       False     string   \n",
       "44        mass spec proteomics,\\nacquisitionSoftware       False     number   \n",
       "45                              mass spec proteomics       False     string   \n",
       "46       mass spec proteomics,\\ndatabaseName = other       False     string   \n",
       "47                              mass spec proteomics       False     string   \n",
       "48     mass spec proteomics,\\ndatabaseSource = other       False     string   \n",
       "49                              mass spec proteomics       False     number   \n",
       "50                      proteomics, dataFile = other       False     string   \n",
       "51       proteomics,\\nmodificationParameters = other       False     string   \n",
       "52                                        proteomics        True     string   \n",
       "53               proteomics,\\ncleavageAgents = other        True     string   \n",
       "54                                        proteomics       False     string   \n",
       "55                                        proteomics       False     string   \n",
       "\n",
       "                                             Ontology  \\\n",
       "0                                    Sage Bionetworks   \n",
       "1                                 Sage Bionetworks,\\n   \n",
       "2                                    Sage Bionetworks   \n",
       "3                              Proteomic Data Commons   \n",
       "4                                    Sage Bionetworks   \n",
       "5                                    Sage Bionetworks   \n",
       "6          http://purl.obolibrary.org/obo/NCIT_C45378   \n",
       "7                                    Sage Bionetworks   \n",
       "8                            Sage Bionetworks,\\nDSLWG   \n",
       "9                                    Sage Bionetworks   \n",
       "10  Sage Bionetworks,\\nDSLWG,\\nProposed minimum me...   \n",
       "11                           Sage Bionetworks,\\nDSLWG   \n",
       "12                         Proteomics Sample Metadata   \n",
       "13          http://purl.obolibrary.org/obo/MS_1001012   \n",
       "14                           Sage Bionetworks,\\nDSLWG   \n",
       "15          http://purl.obolibrary.org/obo/MS_1001013   \n",
       "16  https://www.ebi.ac.uk/ols/ontologies/ms/terms?...   \n",
       "17                                   Sage Bionetworks   \n",
       "18  Sage Bionetworks,\\nDSLWG,\\nhttps://www.ebi.ac....   \n",
       "19                            Proteomics Data Commons   \n",
       "20                                   Sage Bionetworks   \n",
       "21                            Proteomics Data Commons   \n",
       "22                           Sage Bionetworks,\\nDSLWG   \n",
       "23  Sage Bionetworks,\\nProposed minimum metadata r...   \n",
       "24                                   Sage Bionetworks   \n",
       "25  Sage Bionetworks,\\nProposed minimum metadata r...   \n",
       "26                                   Sage Bionetworks   \n",
       "27                           Sage Bionetworks,\\nDSLWG   \n",
       "28                                   Sage Bionetworks   \n",
       "29  http://purl.obolibrary.org/obo/NCIT_C156434\\nP...   \n",
       "30                                   Sage Bionetworks   \n",
       "31                                   Sage Bionetworks   \n",
       "32                                   Sage Bionetworks   \n",
       "33                           Sage Bionetworks,\\nDSLWG   \n",
       "34                                   Sage Bionetworks   \n",
       "35                            Proteomics Data Commons   \n",
       "36                                   Sage Bionetworks   \n",
       "37  Sage Bionetworks,\\nThe Proteomics Sample Metad...   \n",
       "38                             Proteomic Data Commons   \n",
       "39                             Proteomic Data Commons   \n",
       "40                             Proteomic Data Commons   \n",
       "41  Sage Bionetworks,\\nThe Proteomics Sample Metad...   \n",
       "42  https://www.ebi.ac.uk/ols/ontologies/ms/terms?...   \n",
       "43          http://purl.obolibrary.org/obo/MS_1001455   \n",
       "44                           Sage Bionetworks,\\nDSLWG   \n",
       "45          http://purl.obolibrary.org/obo/MS_1001013   \n",
       "46                                   Sage Bionetworks   \n",
       "47          http://purl.obolibrary.org/obo/MS_1001012   \n",
       "48                                   Sage Bionetworks   \n",
       "49                           Sage Bionetworks,\\nDSLWG   \n",
       "50                                   Sage Bionetworks   \n",
       "51                                   Sage Bionetworks   \n",
       "52                         Proteomics Sample Metadata   \n",
       "53                                   Sage Bionetworks   \n",
       "54                         Proteomics Sample Metadata   \n",
       "55                         Proteomics Sample Metadata   \n",
       "\n",
       "                                                Notes  \n",
       "0             values unique to each data contributor)  \n",
       "1   A data contributor should be able to write in ...  \n",
       "2                                                      \n",
       "3   A data contributor should be able to write in ...  \n",
       "4   A data contributor should be able to write in ...  \n",
       "5   A data contributor should be able to write in ...  \n",
       "6   A data contributor should be able to write in ...  \n",
       "7   A data contributor should be able to write in ...  \n",
       "8   A data contributor should be able to write in ...  \n",
       "9                                                      \n",
       "10  A data contributor should be able to write in ...  \n",
       "11  A data contributor should be able to write in ...  \n",
       "12  A data contributor should be able to write in ...  \n",
       "13  A data contributor should be able to write in ...  \n",
       "14  A data contributor should be able to write in ...  \n",
       "15  A data contributor should be able to write in ...  \n",
       "16  Possible values are listed under modification ...  \n",
       "17  A data contributor should be able to write in ...  \n",
       "18                                                     \n",
       "19                                                     \n",
       "20                                                     \n",
       "21  A data contributor should be able to write in ...  \n",
       "22                                                     \n",
       "23                                                     \n",
       "24                                                     \n",
       "25                                                     \n",
       "26                                                     \n",
       "27  A data contributor should be able to write in ...  \n",
       "28                                                     \n",
       "29  A data contributor should be able to write in ...  \n",
       "30  A data contributor should be able to write in ...  \n",
       "31                                                     \n",
       "32                                                     \n",
       "33  A data contributor should be able to write in ...  \n",
       "34  A data contributor should be able to write in ...  \n",
       "35                                                     \n",
       "36                                                     \n",
       "37  A data contributor should be able to write in ...  \n",
       "38  A data contributor should be able to write in ...  \n",
       "39                                                     \n",
       "40  A data contributor should be able to write in ...  \n",
       "41  A data contributor should be able to write in ...  \n",
       "42  A data contributor should be able to write in ...  \n",
       "43  A data contributor should be able to write in ...  \n",
       "44  A data contributor should be able to write in ...  \n",
       "45  A data contributor should be able to write in ...  \n",
       "46  A data contributor should be able to write in ...  \n",
       "47  A data contributor should be able to write in ...  \n",
       "48  A data contributor should be able to write in ...  \n",
       "49  A data contributor should be able to write in ...  \n",
       "50                                                     \n",
       "51                                                     \n",
       "52  A data contributor should be able to write in ...  \n",
       "53                                                     \n",
       "54                                                     \n",
       "55                                                     "
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [],
   "source": [
    "# string clean up\n",
    "# Do not need to explicitly add \"other\" type values. Will add back later for list type attributes\n",
    "p_df[\"Valid Values\"] = (\n",
    "    p_df[\"Valid Values\"]\n",
    "    .replace(\"n/a (unique to each data contributor)\", \"\", regex=False)\n",
    "    .str.replace(\n",
    "        \"Other|Unknown|Not collected|Not applicable|Not specified\", \"\", regex=True\n",
    "    )\n",
    "    .str.strip(\",\")\n",
    ")\n",
    "\n",
    "# replace note type values\n",
    "p_df[\"Valid Values\"] = p_df[\"Valid Values\"].apply(\n",
    "    lambda x: \",\".join(\n",
    "        [y.strip() for y in x.split(\",\") if not bool(re.search(\"Possible values\", y))]\n",
    "    ).strip(\",\")\n",
    ")\n",
    "\n",
    "p_df = p_df.replace(r\"\\n\", \",\", regex=True).replace(r\",,\", \",\", regex=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>Module</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>columnType</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>specimenID</td>\n",
       "      <td>Identifying string linked to a particular samp...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Biospecimen,proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>values unique to each data contributor)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sampleType</td>\n",
       "      <td>The type of sample collected or the term used ...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>specifySampleType</td>\n",
       "      <td>If \"other\" is selected list the type of sample</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics,sampleType = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>experimentType</td>\n",
       "      <td>The type of experiment used.,Provide a value O...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomic Data Commons</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The name of the measurement technique describi...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the me...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics,measurementTechnique = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>technologyPlatformVersion</td>\n",
       "      <td>The specific version (application, manufacture...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C45378</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>specifyPlatformVersion</td>\n",
       "      <td>If \"other\" list the name of the platform version</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics,technologyPlatformVersion = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>platformLocation</td>\n",
       "      <td>The name of the laboratory, facility, vendor, ...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>specifyPlatformLocation</td>\n",
       "      <td>If \"other\" list the name of the platform location</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics,platformLocation = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>instrumentModel</td>\n",
       "      <td>The model of the instrument used.,Provide a va...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG,Proposed minimum metada...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>samplePrepProtocol</td>\n",
       "      <td>An internet address that may provide more deta...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>dataFile</td>\n",
       "      <td>Name of additional file(s) accompanying the da...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>databaseSource</td>\n",
       "      <td>The name of the organization, project, or labo...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001012</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>databaseWeblink</td>\n",
       "      <td>An internet address that may provide details o...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>databaseName</td>\n",
       "      <td>The name of the search database (nr, SwissProt...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001013</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>modificationParameters</td>\n",
       "      <td>Modification parameters for a search engine ru...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/terms?...</td>\n",
       "      <td>Possible values are listed under modification ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>quantationStrategy</td>\n",
       "      <td>The general strategy used for differential ana...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>labelQuantitation</td>\n",
       "      <td>Type of labeling used ,Provide a value OR prov...</td>\n",
       "      <td>AminoxyTMT,DiART,DiLeu,Glyco-TMT,Hydrazide TMT...</td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG,https://www.ebi.ac.uk/o...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>labelFreeQuantitation</td>\n",
       "      <td>Type of label-free data analysis strategy ,Pro...</td>\n",
       "      <td>Label-free protein level quantitation,Label-fr...</td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,labelQuantiation = Label Free</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Data Commons</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>specifyLabelFreeQuantitation</td>\n",
       "      <td>If \"other\" is selected please specify</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics,labelFreeQuantitation = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>isobaricLabelingReagent</td>\n",
       "      <td>Reagent set used for isobaric labeling.,Exampl...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Data Commons</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>msTarget</td>\n",
       "      <td>Specifies whether or not a specific molecule(s...</td>\n",
       "      <td>Targeted,Untargeted</td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>msAnalyzerTypeMS1</td>\n",
       "      <td>Describes the mass spectrometer analyzer used ...</td>\n",
       "      <td>LTQ Orbitrap XL,Orbitrap Fusion,Orbitrap Fusio...</td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,Proposed minimum metadata rel...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>specifyMSAnalyzerTypeMS1</td>\n",
       "      <td>If \"other\" is selected list the name of the ma...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics,msAnalyzerTypeMS1 = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>msAnalyzerTypeMS2</td>\n",
       "      <td>Describes the mass spectrometer analyzer used ...</td>\n",
       "      <td>LTQ Orbitrap XL,Orbitrap Fusion,Orbitrap Fusio...</td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,Proposed minimum metadata rel...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>specifyMSAnalyzerTypeMS2</td>\n",
       "      <td>If \"other\" is selected list the name of the ma...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics,msAnalyzerTypeMS2 = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>msAssayTechnique</td>\n",
       "      <td>Specified the mass spectrometer technique used...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>specifyMSAssayTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the ma...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,msAssayTechnique = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>msAnalyteType</td>\n",
       "      <td>The type of biospecimen subjected to analysis ...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C156434,Pr...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>specifyMSAnalyteType</td>\n",
       "      <td>If \"other\" list the type of analyte</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,msAnalyteType = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>specifyProteomicsAssayType</td>\n",
       "      <td>List the name of the type of peptide quantific...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,proteomicsAssayType = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>specifyMSInstrumentModel</td>\n",
       "      <td>List the name of the instrument model</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,msInstrumentModel = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>digestionMethod</td>\n",
       "      <td>The type of digestion (ex. in-solution, in-gel...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>specifyDigestionMethod</td>\n",
       "      <td>If \"other\" list the type of digestion</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,digestionMethod = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>digestionReagent</td>\n",
       "      <td>Enzymes or reagents used for digestion. ,</td>\n",
       "      <td>2-iodobenzoate,Arg-C,Asp-N,Asp-N_ambic,Chymotr...</td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Data Commons</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>specifyDigestionReagent</td>\n",
       "      <td>If \"other\" is selected specify the name of the...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics digestionReagent=other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>fractionIdentifier</td>\n",
       "      <td>Number of fractions in the experimental run,Pr...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,The Proteomics Sample Metadat...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>enrichmentStrategy</td>\n",
       "      <td>Type of peptide or protein enrichment,</td>\n",
       "      <td>Acetylome,Proteome,Phosphoproteome,Glycoproteo...</td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomic Data Commons</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>specifyEnrichmentStrategy</td>\n",
       "      <td>If \"other\" is selected specify the name</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomic Data Commons</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>enrichment</td>\n",
       "      <td>Protocol or reagent used for enrichment (ex. F...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomic Data Commons</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>fractionIdentifier</td>\n",
       "      <td>An identifier string that describes the sample...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>number</td>\n",
       "      <td>Sage Bionetworks,The Proteomics Sample Metadat...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>acquisitionMode</td>\n",
       "      <td>The specific aspect of a mass spectrometer met...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/terms?...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>acquisitionSoftware</td>\n",
       "      <td>The name of the acquisition software used</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>acquistionSoftwareVersion</td>\n",
       "      <td>The version number of software used</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics,acquisitionSoftware</td>\n",
       "      <td>False</td>\n",
       "      <td>number</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>databaseName</td>\n",
       "      <td>The name of the search database (nr, SwissProt...</td>\n",
       "      <td>MassBank of North America (MoNA),NIST17,Metlin...</td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001013</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>specifyDatabaseName</td>\n",
       "      <td>If \"other\" list the name of the search database</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics,databaseName = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>databaseSource</td>\n",
       "      <td>The name of the organization, project, or labo...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001012</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>specifyDatabaseSource</td>\n",
       "      <td>If \"other\" list the name of the database source</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics,databaseSource = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>spectrometerFrequency</td>\n",
       "      <td>The frequency at which a spectrometer causes h...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>number</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>specifyOtherDataFile</td>\n",
       "      <td>If \"other\" is selected, list the name of the d...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics, dataFile = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>specifyOtherModificationParameters</td>\n",
       "      <td>If \"other\" is selected, list the custom modifi...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,modificationParameters = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>cleavageAgents</td>\n",
       "      <td>Name(s) of the enzyme used (possible values ar...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>specifyOtherCleavageAgent</td>\n",
       "      <td>If \"other\" is selected, list the enzyme name.</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,cleavageAgents = other</td>\n",
       "      <td>True</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>fragmentMassTolerance</td>\n",
       "      <td>Units of measure are Da or ppm (ex. \"0.6 Da\" o...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>precursorMassTolerance</td>\n",
       "      <td>Units of measure are Da or ppm (ex. \"0.6 Da\" o...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Attribute  \\\n",
       "0                           specimenID   \n",
       "1                           sampleType   \n",
       "2                    specifySampleType   \n",
       "3                       experimentType   \n",
       "4                 measurementTechnique   \n",
       "5          specifyMeasurementTechnique   \n",
       "6            technologyPlatformVersion   \n",
       "7               specifyPlatformVersion   \n",
       "8                     platformLocation   \n",
       "9              specifyPlatformLocation   \n",
       "10                     instrumentModel   \n",
       "11                  samplePrepProtocol   \n",
       "12                            dataFile   \n",
       "13                      databaseSource   \n",
       "14                     databaseWeblink   \n",
       "15                        databaseName   \n",
       "16              modificationParameters   \n",
       "17                  quantationStrategy   \n",
       "18                   labelQuantitation   \n",
       "19               labelFreeQuantitation   \n",
       "20        specifyLabelFreeQuantitation   \n",
       "21             isobaricLabelingReagent   \n",
       "22                            msTarget   \n",
       "23                   msAnalyzerTypeMS1   \n",
       "24            specifyMSAnalyzerTypeMS1   \n",
       "25                   msAnalyzerTypeMS2   \n",
       "26            specifyMSAnalyzerTypeMS2   \n",
       "27                    msAssayTechnique   \n",
       "28             specifyMSAssayTechnique   \n",
       "29                       msAnalyteType   \n",
       "30                specifyMSAnalyteType   \n",
       "31          specifyProteomicsAssayType   \n",
       "32            specifyMSInstrumentModel   \n",
       "33                     digestionMethod   \n",
       "34              specifyDigestionMethod   \n",
       "35                    digestionReagent   \n",
       "36             specifyDigestionReagent   \n",
       "37                  fractionIdentifier   \n",
       "38                  enrichmentStrategy   \n",
       "39           specifyEnrichmentStrategy   \n",
       "40                          enrichment   \n",
       "41                  fractionIdentifier   \n",
       "42                     acquisitionMode   \n",
       "43                 acquisitionSoftware   \n",
       "44           acquistionSoftwareVersion   \n",
       "45                        databaseName   \n",
       "46                 specifyDatabaseName   \n",
       "47                      databaseSource   \n",
       "48               specifyDatabaseSource   \n",
       "49               spectrometerFrequency   \n",
       "50                specifyOtherDataFile   \n",
       "51  specifyOtherModificationParameters   \n",
       "52                      cleavageAgents   \n",
       "53           specifyOtherCleavageAgent   \n",
       "54               fragmentMassTolerance   \n",
       "55              precursorMassTolerance   \n",
       "\n",
       "                                          Description  \\\n",
       "0   Identifying string linked to a particular samp...   \n",
       "1   The type of sample collected or the term used ...   \n",
       "2      If \"other\" is selected list the type of sample   \n",
       "3   The type of experiment used.,Provide a value O...   \n",
       "4   The name of the measurement technique describi...   \n",
       "5   If \"other\" is selected list the name of the me...   \n",
       "6   The specific version (application, manufacture...   \n",
       "7    If \"other\" list the name of the platform version   \n",
       "8   The name of the laboratory, facility, vendor, ...   \n",
       "9   If \"other\" list the name of the platform location   \n",
       "10  The model of the instrument used.,Provide a va...   \n",
       "11  An internet address that may provide more deta...   \n",
       "12  Name of additional file(s) accompanying the da...   \n",
       "13  The name of the organization, project, or labo...   \n",
       "14  An internet address that may provide details o...   \n",
       "15  The name of the search database (nr, SwissProt...   \n",
       "16  Modification parameters for a search engine ru...   \n",
       "17  The general strategy used for differential ana...   \n",
       "18  Type of labeling used ,Provide a value OR prov...   \n",
       "19  Type of label-free data analysis strategy ,Pro...   \n",
       "20              If \"other\" is selected please specify   \n",
       "21  Reagent set used for isobaric labeling.,Exampl...   \n",
       "22  Specifies whether or not a specific molecule(s...   \n",
       "23  Describes the mass spectrometer analyzer used ...   \n",
       "24  If \"other\" is selected list the name of the ma...   \n",
       "25  Describes the mass spectrometer analyzer used ...   \n",
       "26  If \"other\" is selected list the name of the ma...   \n",
       "27  Specified the mass spectrometer technique used...   \n",
       "28  If \"other\" is selected list the name of the ma...   \n",
       "29  The type of biospecimen subjected to analysis ...   \n",
       "30               If \"other\" list the type of analyte    \n",
       "31  List the name of the type of peptide quantific...   \n",
       "32              List the name of the instrument model   \n",
       "33  The type of digestion (ex. in-solution, in-gel...   \n",
       "34              If \"other\" list the type of digestion   \n",
       "35          Enzymes or reagents used for digestion. ,   \n",
       "36  If \"other\" is selected specify the name of the...   \n",
       "37  Number of fractions in the experimental run,Pr...   \n",
       "38             Type of peptide or protein enrichment,   \n",
       "39            If \"other\" is selected specify the name   \n",
       "40  Protocol or reagent used for enrichment (ex. F...   \n",
       "41  An identifier string that describes the sample...   \n",
       "42  The specific aspect of a mass spectrometer met...   \n",
       "43          The name of the acquisition software used   \n",
       "44                The version number of software used   \n",
       "45  The name of the search database (nr, SwissProt...   \n",
       "46    If \"other\" list the name of the search database   \n",
       "47  The name of the organization, project, or labo...   \n",
       "48    If \"other\" list the name of the database source   \n",
       "49  The frequency at which a spectrometer causes h...   \n",
       "50  If \"other\" is selected, list the name of the d...   \n",
       "51  If \"other\" is selected, list the custom modifi...   \n",
       "52  Name(s) of the enzyme used (possible values ar...   \n",
       "53     If \"other\" is selected, list the enzyme name.    \n",
       "54  Units of measure are Da or ppm (ex. \"0.6 Da\" o...   \n",
       "55  Units of measure are Da or ppm (ex. \"0.6 Da\" o...   \n",
       "\n",
       "                                         Valid Values  Required  \\\n",
       "0                                                          True   \n",
       "1                                                          True   \n",
       "2                                                         False   \n",
       "3                                                          True   \n",
       "4                                                          True   \n",
       "5                                                         False   \n",
       "6                                                          True   \n",
       "7                                                         False   \n",
       "8                                                          True   \n",
       "9                                                         False   \n",
       "10                                                         True   \n",
       "11                                                         True   \n",
       "12                                                         True   \n",
       "13                                                         True   \n",
       "14                                                         True   \n",
       "15                                                         True   \n",
       "16                                                         True   \n",
       "17                                                         True   \n",
       "18  AminoxyTMT,DiART,DiLeu,Glyco-TMT,Hydrazide TMT...      True   \n",
       "19  Label-free protein level quantitation,Label-fr...      True   \n",
       "20                                                        False   \n",
       "21                                                         True   \n",
       "22                                Targeted,Untargeted      True   \n",
       "23  LTQ Orbitrap XL,Orbitrap Fusion,Orbitrap Fusio...      True   \n",
       "24                                                         True   \n",
       "25  LTQ Orbitrap XL,Orbitrap Fusion,Orbitrap Fusio...      True   \n",
       "26                                                         True   \n",
       "27                                                         True   \n",
       "28                                                         True   \n",
       "29                                                         True   \n",
       "30                                                         True   \n",
       "31                                                         True   \n",
       "32                                                         True   \n",
       "33                                                         True   \n",
       "34                                                         True   \n",
       "35  2-iodobenzoate,Arg-C,Asp-N,Asp-N_ambic,Chymotr...      True   \n",
       "36                                                        False   \n",
       "37                                                         True   \n",
       "38  Acetylome,Proteome,Phosphoproteome,Glycoproteo...      True   \n",
       "39                                                        False   \n",
       "40                                                         True   \n",
       "41                                                         True   \n",
       "42                                                         True   \n",
       "43                                                         True   \n",
       "44                                                         True   \n",
       "45  MassBank of North America (MoNA),NIST17,Metlin...      True   \n",
       "46                                                         True   \n",
       "47                                                         True   \n",
       "48                                                         True   \n",
       "49                                                         True   \n",
       "50                                                         True   \n",
       "51                                                         True   \n",
       "52                                                         True   \n",
       "53                                                         True   \n",
       "54                                                         True   \n",
       "55                                                         True   \n",
       "\n",
       "                                            Module  multivalue columnType  \\\n",
       "0                           Biospecimen,proteomics       False     string   \n",
       "1                                       proteomics       False     string   \n",
       "2                    proteomics,sampleType = other       False     string   \n",
       "3                                       proteomics        True     string   \n",
       "4                                       proteomics        True     string   \n",
       "5          proteomics,measurementTechnique = other        True     string   \n",
       "6                                       proteomics        True     string   \n",
       "7     proteomics,technologyPlatformVersion = other        True     string   \n",
       "8                                       proteomics        True     string   \n",
       "9              proteomics,platformLocation = other        True     string   \n",
       "10                                      proteomics        True     string   \n",
       "11                                      proteomics        True     string   \n",
       "12                                      proteomics        True     string   \n",
       "13                                      proteomics        True     string   \n",
       "14                                      proteomics        True     string   \n",
       "15                                      proteomics        True     string   \n",
       "16                                      proteomics        True     string   \n",
       "17                                      proteomics        True     string   \n",
       "18                                      proteomics        True     string   \n",
       "19        proteomics,labelQuantiation = Label Free        True     string   \n",
       "20        proteomics,labelFreeQuantitation = other       False     string   \n",
       "21                                      proteomics        True     string   \n",
       "22                                      proteomics        True     string   \n",
       "23                            mass spec proteomics        True     string   \n",
       "24  mass spec proteomics,msAnalyzerTypeMS1 = other        True     string   \n",
       "25                            mass spec proteomics        True     string   \n",
       "26  mass spec proteomics,msAnalyzerTypeMS2 = other        True     string   \n",
       "27                                      proteomics        True     string   \n",
       "28             proteomics,msAssayTechnique = other        True     string   \n",
       "29                                     proteomics,        True     string   \n",
       "30                proteomics,msAnalyteType = other        True     string   \n",
       "31          proteomics,proteomicsAssayType = other        True     string   \n",
       "32            proteomics,msInstrumentModel = other        True     string   \n",
       "33                                      proteomics        True     string   \n",
       "34              proteomics,digestionMethod = other       False     string   \n",
       "35                                      proteomics        True     string   \n",
       "36               proteomics digestionReagent=other       False     string   \n",
       "37                                      proteomics        True     string   \n",
       "38                                      proteomics        True     string   \n",
       "39                                      proteomics        True     string   \n",
       "40                                      proteomics        True     string   \n",
       "41                                      proteomics       False     number   \n",
       "42                            mass spec proteomics       False     string   \n",
       "43                            mass spec proteomics       False     string   \n",
       "44        mass spec proteomics,acquisitionSoftware       False     number   \n",
       "45                            mass spec proteomics       False     string   \n",
       "46       mass spec proteomics,databaseName = other       False     string   \n",
       "47                            mass spec proteomics       False     string   \n",
       "48     mass spec proteomics,databaseSource = other       False     string   \n",
       "49                            mass spec proteomics       False     number   \n",
       "50                    proteomics, dataFile = other       False     string   \n",
       "51       proteomics,modificationParameters = other       False     string   \n",
       "52                                      proteomics        True     string   \n",
       "53               proteomics,cleavageAgents = other        True     string   \n",
       "54                                      proteomics       False     string   \n",
       "55                                      proteomics       False     string   \n",
       "\n",
       "                                             Ontology  \\\n",
       "0                                    Sage Bionetworks   \n",
       "1                                   Sage Bionetworks,   \n",
       "2                                    Sage Bionetworks   \n",
       "3                              Proteomic Data Commons   \n",
       "4                                    Sage Bionetworks   \n",
       "5                                    Sage Bionetworks   \n",
       "6          http://purl.obolibrary.org/obo/NCIT_C45378   \n",
       "7                                    Sage Bionetworks   \n",
       "8                              Sage Bionetworks,DSLWG   \n",
       "9                                    Sage Bionetworks   \n",
       "10  Sage Bionetworks,DSLWG,Proposed minimum metada...   \n",
       "11                             Sage Bionetworks,DSLWG   \n",
       "12                         Proteomics Sample Metadata   \n",
       "13          http://purl.obolibrary.org/obo/MS_1001012   \n",
       "14                             Sage Bionetworks,DSLWG   \n",
       "15          http://purl.obolibrary.org/obo/MS_1001013   \n",
       "16  https://www.ebi.ac.uk/ols/ontologies/ms/terms?...   \n",
       "17                                   Sage Bionetworks   \n",
       "18  Sage Bionetworks,DSLWG,https://www.ebi.ac.uk/o...   \n",
       "19                            Proteomics Data Commons   \n",
       "20                                   Sage Bionetworks   \n",
       "21                            Proteomics Data Commons   \n",
       "22                             Sage Bionetworks,DSLWG   \n",
       "23  Sage Bionetworks,Proposed minimum metadata rel...   \n",
       "24                                   Sage Bionetworks   \n",
       "25  Sage Bionetworks,Proposed minimum metadata rel...   \n",
       "26                                   Sage Bionetworks   \n",
       "27                             Sage Bionetworks,DSLWG   \n",
       "28                                   Sage Bionetworks   \n",
       "29  http://purl.obolibrary.org/obo/NCIT_C156434,Pr...   \n",
       "30                                   Sage Bionetworks   \n",
       "31                                   Sage Bionetworks   \n",
       "32                                   Sage Bionetworks   \n",
       "33                             Sage Bionetworks,DSLWG   \n",
       "34                                   Sage Bionetworks   \n",
       "35                            Proteomics Data Commons   \n",
       "36                                   Sage Bionetworks   \n",
       "37  Sage Bionetworks,The Proteomics Sample Metadat...   \n",
       "38                             Proteomic Data Commons   \n",
       "39                             Proteomic Data Commons   \n",
       "40                             Proteomic Data Commons   \n",
       "41  Sage Bionetworks,The Proteomics Sample Metadat...   \n",
       "42  https://www.ebi.ac.uk/ols/ontologies/ms/terms?...   \n",
       "43          http://purl.obolibrary.org/obo/MS_1001455   \n",
       "44                             Sage Bionetworks,DSLWG   \n",
       "45          http://purl.obolibrary.org/obo/MS_1001013   \n",
       "46                                   Sage Bionetworks   \n",
       "47          http://purl.obolibrary.org/obo/MS_1001012   \n",
       "48                                   Sage Bionetworks   \n",
       "49                             Sage Bionetworks,DSLWG   \n",
       "50                                   Sage Bionetworks   \n",
       "51                                   Sage Bionetworks   \n",
       "52                         Proteomics Sample Metadata   \n",
       "53                                   Sage Bionetworks   \n",
       "54                         Proteomics Sample Metadata   \n",
       "55                         Proteomics Sample Metadata   \n",
       "\n",
       "                                                Notes  \n",
       "0             values unique to each data contributor)  \n",
       "1   A data contributor should be able to write in ...  \n",
       "2                                                      \n",
       "3   A data contributor should be able to write in ...  \n",
       "4   A data contributor should be able to write in ...  \n",
       "5   A data contributor should be able to write in ...  \n",
       "6   A data contributor should be able to write in ...  \n",
       "7   A data contributor should be able to write in ...  \n",
       "8   A data contributor should be able to write in ...  \n",
       "9                                                      \n",
       "10  A data contributor should be able to write in ...  \n",
       "11  A data contributor should be able to write in ...  \n",
       "12  A data contributor should be able to write in ...  \n",
       "13  A data contributor should be able to write in ...  \n",
       "14  A data contributor should be able to write in ...  \n",
       "15  A data contributor should be able to write in ...  \n",
       "16  Possible values are listed under modification ...  \n",
       "17  A data contributor should be able to write in ...  \n",
       "18                                                     \n",
       "19                                                     \n",
       "20                                                     \n",
       "21  A data contributor should be able to write in ...  \n",
       "22                                                     \n",
       "23                                                     \n",
       "24                                                     \n",
       "25                                                     \n",
       "26                                                     \n",
       "27  A data contributor should be able to write in ...  \n",
       "28                                                     \n",
       "29  A data contributor should be able to write in ...  \n",
       "30  A data contributor should be able to write in ...  \n",
       "31                                                     \n",
       "32                                                     \n",
       "33  A data contributor should be able to write in ...  \n",
       "34  A data contributor should be able to write in ...  \n",
       "35                                                     \n",
       "36                                                     \n",
       "37  A data contributor should be able to write in ...  \n",
       "38  A data contributor should be able to write in ...  \n",
       "39                                                     \n",
       "40  A data contributor should be able to write in ...  \n",
       "41  A data contributor should be able to write in ...  \n",
       "42  A data contributor should be able to write in ...  \n",
       "43  A data contributor should be able to write in ...  \n",
       "44  A data contributor should be able to write in ...  \n",
       "45  A data contributor should be able to write in ...  \n",
       "46  A data contributor should be able to write in ...  \n",
       "47  A data contributor should be able to write in ...  \n",
       "48  A data contributor should be able to write in ...  \n",
       "49  A data contributor should be able to write in ...  \n",
       "50                                                     \n",
       "51                                                     \n",
       "52  A data contributor should be able to write in ...  \n",
       "53                                                     \n",
       "54                                                     \n",
       "55                                                     "
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>Module</th>\n",
       "      <th>columnType</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>specimenID</td>\n",
       "      <td>Identifying string linked to a particular samp...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Biospecimen,proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>values unique to each data contributor)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sampleType</td>\n",
       "      <td>The type of sample collected or the term used ...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>specifySampleType</td>\n",
       "      <td>If \"other\" is selected list the type of sample</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,sampleType = other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>experimentType</td>\n",
       "      <td>The type of experiment used.,Provide a value O...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomic Data Commons</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The name of the measurement technique describi...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the me...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,measurementTechnique = other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>technologyPlatformVersion</td>\n",
       "      <td>The specific version (application, manufacture...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C45378</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>specifyPlatformVersion</td>\n",
       "      <td>If \"other\" list the name of the platform version</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,technologyPlatformVersion = other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>platformLocation</td>\n",
       "      <td>The name of the laboratory, facility, vendor, ...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>specifyPlatformLocation</td>\n",
       "      <td>If \"other\" list the name of the platform location</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,platformLocation = other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>instrumentModel</td>\n",
       "      <td>The model of the instrument used.,Provide a va...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG,Proposed minimum metada...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>samplePrepProtocol</td>\n",
       "      <td>An internet address that may provide more deta...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>dataFile</td>\n",
       "      <td>Name of additional file(s) accompanying the da...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>databaseSource</td>\n",
       "      <td>The name of the organization, project, or labo...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001012</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>databaseWeblink</td>\n",
       "      <td>An internet address that may provide details o...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>databaseName</td>\n",
       "      <td>The name of the search database (nr, SwissProt...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001013</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>modificationParameters</td>\n",
       "      <td>Modification parameters for a search engine ru...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/terms?...</td>\n",
       "      <td>Possible values are listed under modification ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>quantationStrategy</td>\n",
       "      <td>The general strategy used for differential ana...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>labelQuantitation</td>\n",
       "      <td>Type of labeling used ,Provide a value OR prov...</td>\n",
       "      <td>AminoxyTMT,DiART,DiLeu,Glyco-TMT,Hydrazide TMT...</td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG,https://www.ebi.ac.uk/o...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>labelFreeQuantitation</td>\n",
       "      <td>Type of label-free data analysis strategy ,Pro...</td>\n",
       "      <td>Label-free protein level quantitation,Label-fr...</td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,labelQuantiation = Label Free</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Data Commons</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>specifyLabelFreeQuantitation</td>\n",
       "      <td>If \"other\" is selected please specify</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,labelFreeQuantitation = other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>isobaricLabelingReagent</td>\n",
       "      <td>Reagent set used for isobaric labeling.,Exampl...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Data Commons</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>msTarget</td>\n",
       "      <td>Specifies whether or not a specific molecule(s...</td>\n",
       "      <td>Targeted,Untargeted</td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>msAnalyzerTypeMS1</td>\n",
       "      <td>Describes the mass spectrometer analyzer used ...</td>\n",
       "      <td>LTQ Orbitrap XL,Orbitrap Fusion,Orbitrap Fusio...</td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,Proposed minimum metadata rel...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>specifyMSAnalyzerTypeMS1</td>\n",
       "      <td>If \"other\" is selected list the name of the ma...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics,msAnalyzerTypeMS1 = other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>msAnalyzerTypeMS2</td>\n",
       "      <td>Describes the mass spectrometer analyzer used ...</td>\n",
       "      <td>LTQ Orbitrap XL,Orbitrap Fusion,Orbitrap Fusio...</td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,Proposed minimum metadata rel...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>specifyMSAnalyzerTypeMS2</td>\n",
       "      <td>If \"other\" is selected list the name of the ma...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics,msAnalyzerTypeMS2 = other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>msAssayTechnique</td>\n",
       "      <td>Specified the mass spectrometer technique used...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>specifyMSAssayTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the ma...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,msAssayTechnique = other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>msAnalyteType</td>\n",
       "      <td>The type of biospecimen subjected to analysis ...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C156434,Pr...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>specifyMSAnalyteType</td>\n",
       "      <td>If \"other\" list the type of analyte</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,msAnalyteType = other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>specifyProteomicsAssayType</td>\n",
       "      <td>List the name of the type of peptide quantific...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,proteomicsAssayType = other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>specifyMSInstrumentModel</td>\n",
       "      <td>List the name of the instrument model</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,msInstrumentModel = other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>digestionMethod</td>\n",
       "      <td>The type of digestion (ex. in-solution, in-gel...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>specifyDigestionMethod</td>\n",
       "      <td>If \"other\" list the type of digestion</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,digestionMethod = other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>digestionReagent</td>\n",
       "      <td>Enzymes or reagents used for digestion. ,</td>\n",
       "      <td>2-iodobenzoate,Arg-C,Asp-N,Asp-N_ambic,Chymotr...</td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Data Commons</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>specifyDigestionReagent</td>\n",
       "      <td>If \"other\" is selected specify the name of the...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics digestionReagent=other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>fractionIdentifier</td>\n",
       "      <td>Number of fractions in the experimental run,Pr...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,The Proteomics Sample Metadat...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>enrichmentStrategy</td>\n",
       "      <td>Type of peptide or protein enrichment,</td>\n",
       "      <td>Acetylome,Proteome,Phosphoproteome,Glycoproteo...</td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomic Data Commons</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>specifyEnrichmentStrategy</td>\n",
       "      <td>If \"other\" is selected specify the name</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomic Data Commons</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>enrichment</td>\n",
       "      <td>Protocol or reagent used for enrichment (ex. F...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomic Data Commons</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>fractionIdentifier</td>\n",
       "      <td>An identifier string that describes the sample...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>number</td>\n",
       "      <td>Sage Bionetworks,The Proteomics Sample Metadat...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>acquisitionMode</td>\n",
       "      <td>The specific aspect of a mass spectrometer met...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/terms?...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>acquisitionSoftware</td>\n",
       "      <td>The name of the acquisition software used</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>acquistionSoftwareVersion</td>\n",
       "      <td>The version number of software used</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics,acquisitionSoftware</td>\n",
       "      <td>number</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>databaseName</td>\n",
       "      <td>The name of the search database (nr, SwissProt...</td>\n",
       "      <td>MassBank of North America (MoNA),NIST17,Metlin...</td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001013</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>specifyDatabaseName</td>\n",
       "      <td>If \"other\" list the name of the search database</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics,databaseName = other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>databaseSource</td>\n",
       "      <td>The name of the organization, project, or labo...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001012</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>specifyDatabaseSource</td>\n",
       "      <td>If \"other\" list the name of the database source</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics,databaseSource = other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>spectrometerFrequency</td>\n",
       "      <td>The frequency at which a spectrometer causes h...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>mass spec proteomics</td>\n",
       "      <td>number</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>specifyOtherDataFile</td>\n",
       "      <td>If \"other\" is selected, list the name of the d...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics, dataFile = other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>specifyOtherModificationParameters</td>\n",
       "      <td>If \"other\" is selected, list the custom modifi...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,modificationParameters = other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>cleavageAgents</td>\n",
       "      <td>Name(s) of the enzyme used (possible values ar...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>specifyOtherCleavageAgent</td>\n",
       "      <td>If \"other\" is selected, list the enzyme name.</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics,cleavageAgents = other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>fragmentMassTolerance</td>\n",
       "      <td>Units of measure are Da or ppm (ex. \"0.6 Da\" o...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>precursorMassTolerance</td>\n",
       "      <td>Units of measure are Da or ppm (ex. \"0.6 Da\" o...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Attribute  \\\n",
       "0                           specimenID   \n",
       "1                           sampleType   \n",
       "2                    specifySampleType   \n",
       "3                       experimentType   \n",
       "4                 measurementTechnique   \n",
       "5          specifyMeasurementTechnique   \n",
       "6            technologyPlatformVersion   \n",
       "7               specifyPlatformVersion   \n",
       "8                     platformLocation   \n",
       "9              specifyPlatformLocation   \n",
       "10                     instrumentModel   \n",
       "11                  samplePrepProtocol   \n",
       "12                            dataFile   \n",
       "13                      databaseSource   \n",
       "14                     databaseWeblink   \n",
       "15                        databaseName   \n",
       "16              modificationParameters   \n",
       "17                  quantationStrategy   \n",
       "18                   labelQuantitation   \n",
       "19               labelFreeQuantitation   \n",
       "20        specifyLabelFreeQuantitation   \n",
       "21             isobaricLabelingReagent   \n",
       "22                            msTarget   \n",
       "23                   msAnalyzerTypeMS1   \n",
       "24            specifyMSAnalyzerTypeMS1   \n",
       "25                   msAnalyzerTypeMS2   \n",
       "26            specifyMSAnalyzerTypeMS2   \n",
       "27                    msAssayTechnique   \n",
       "28             specifyMSAssayTechnique   \n",
       "29                       msAnalyteType   \n",
       "30                specifyMSAnalyteType   \n",
       "31          specifyProteomicsAssayType   \n",
       "32            specifyMSInstrumentModel   \n",
       "33                     digestionMethod   \n",
       "34              specifyDigestionMethod   \n",
       "35                    digestionReagent   \n",
       "36             specifyDigestionReagent   \n",
       "37                  fractionIdentifier   \n",
       "38                  enrichmentStrategy   \n",
       "39           specifyEnrichmentStrategy   \n",
       "40                          enrichment   \n",
       "41                  fractionIdentifier   \n",
       "42                     acquisitionMode   \n",
       "43                 acquisitionSoftware   \n",
       "44           acquistionSoftwareVersion   \n",
       "45                        databaseName   \n",
       "46                 specifyDatabaseName   \n",
       "47                      databaseSource   \n",
       "48               specifyDatabaseSource   \n",
       "49               spectrometerFrequency   \n",
       "50                specifyOtherDataFile   \n",
       "51  specifyOtherModificationParameters   \n",
       "52                      cleavageAgents   \n",
       "53           specifyOtherCleavageAgent   \n",
       "54               fragmentMassTolerance   \n",
       "55              precursorMassTolerance   \n",
       "\n",
       "                                          Description  \\\n",
       "0   Identifying string linked to a particular samp...   \n",
       "1   The type of sample collected or the term used ...   \n",
       "2      If \"other\" is selected list the type of sample   \n",
       "3   The type of experiment used.,Provide a value O...   \n",
       "4   The name of the measurement technique describi...   \n",
       "5   If \"other\" is selected list the name of the me...   \n",
       "6   The specific version (application, manufacture...   \n",
       "7    If \"other\" list the name of the platform version   \n",
       "8   The name of the laboratory, facility, vendor, ...   \n",
       "9   If \"other\" list the name of the platform location   \n",
       "10  The model of the instrument used.,Provide a va...   \n",
       "11  An internet address that may provide more deta...   \n",
       "12  Name of additional file(s) accompanying the da...   \n",
       "13  The name of the organization, project, or labo...   \n",
       "14  An internet address that may provide details o...   \n",
       "15  The name of the search database (nr, SwissProt...   \n",
       "16  Modification parameters for a search engine ru...   \n",
       "17  The general strategy used for differential ana...   \n",
       "18  Type of labeling used ,Provide a value OR prov...   \n",
       "19  Type of label-free data analysis strategy ,Pro...   \n",
       "20              If \"other\" is selected please specify   \n",
       "21  Reagent set used for isobaric labeling.,Exampl...   \n",
       "22  Specifies whether or not a specific molecule(s...   \n",
       "23  Describes the mass spectrometer analyzer used ...   \n",
       "24  If \"other\" is selected list the name of the ma...   \n",
       "25  Describes the mass spectrometer analyzer used ...   \n",
       "26  If \"other\" is selected list the name of the ma...   \n",
       "27  Specified the mass spectrometer technique used...   \n",
       "28  If \"other\" is selected list the name of the ma...   \n",
       "29  The type of biospecimen subjected to analysis ...   \n",
       "30               If \"other\" list the type of analyte    \n",
       "31  List the name of the type of peptide quantific...   \n",
       "32              List the name of the instrument model   \n",
       "33  The type of digestion (ex. in-solution, in-gel...   \n",
       "34              If \"other\" list the type of digestion   \n",
       "35          Enzymes or reagents used for digestion. ,   \n",
       "36  If \"other\" is selected specify the name of the...   \n",
       "37  Number of fractions in the experimental run,Pr...   \n",
       "38             Type of peptide or protein enrichment,   \n",
       "39            If \"other\" is selected specify the name   \n",
       "40  Protocol or reagent used for enrichment (ex. F...   \n",
       "41  An identifier string that describes the sample...   \n",
       "42  The specific aspect of a mass spectrometer met...   \n",
       "43          The name of the acquisition software used   \n",
       "44                The version number of software used   \n",
       "45  The name of the search database (nr, SwissProt...   \n",
       "46    If \"other\" list the name of the search database   \n",
       "47  The name of the organization, project, or labo...   \n",
       "48    If \"other\" list the name of the database source   \n",
       "49  The frequency at which a spectrometer causes h...   \n",
       "50  If \"other\" is selected, list the name of the d...   \n",
       "51  If \"other\" is selected, list the custom modifi...   \n",
       "52  Name(s) of the enzyme used (possible values ar...   \n",
       "53     If \"other\" is selected, list the enzyme name.    \n",
       "54  Units of measure are Da or ppm (ex. \"0.6 Da\" o...   \n",
       "55  Units of measure are Da or ppm (ex. \"0.6 Da\" o...   \n",
       "\n",
       "                                         Valid Values  Required  \\\n",
       "0                                                          True   \n",
       "1                                                          True   \n",
       "2                                                          True   \n",
       "3                                                          True   \n",
       "4                                                          True   \n",
       "5                                                          True   \n",
       "6                                                          True   \n",
       "7                                                          True   \n",
       "8                                                          True   \n",
       "9                                                          True   \n",
       "10                                                         True   \n",
       "11                                                         True   \n",
       "12                                                         True   \n",
       "13                                                         True   \n",
       "14                                                         True   \n",
       "15                                                         True   \n",
       "16                                                         True   \n",
       "17                                                         True   \n",
       "18  AminoxyTMT,DiART,DiLeu,Glyco-TMT,Hydrazide TMT...      True   \n",
       "19  Label-free protein level quantitation,Label-fr...      True   \n",
       "20                                                         True   \n",
       "21                                                         True   \n",
       "22                                Targeted,Untargeted      True   \n",
       "23  LTQ Orbitrap XL,Orbitrap Fusion,Orbitrap Fusio...      True   \n",
       "24                                                         True   \n",
       "25  LTQ Orbitrap XL,Orbitrap Fusion,Orbitrap Fusio...      True   \n",
       "26                                                         True   \n",
       "27                                                         True   \n",
       "28                                                         True   \n",
       "29                                                         True   \n",
       "30                                                         True   \n",
       "31                                                         True   \n",
       "32                                                         True   \n",
       "33                                                         True   \n",
       "34                                                         True   \n",
       "35  2-iodobenzoate,Arg-C,Asp-N,Asp-N_ambic,Chymotr...      True   \n",
       "36                                                         True   \n",
       "37                                                         True   \n",
       "38  Acetylome,Proteome,Phosphoproteome,Glycoproteo...      True   \n",
       "39                                                         True   \n",
       "40                                                         True   \n",
       "41                                                         True   \n",
       "42                                                         True   \n",
       "43                                                         True   \n",
       "44                                                         True   \n",
       "45  MassBank of North America (MoNA),NIST17,Metlin...      True   \n",
       "46                                                         True   \n",
       "47                                                         True   \n",
       "48                                                         True   \n",
       "49                                                         True   \n",
       "50                                                         True   \n",
       "51                                                         True   \n",
       "52                                                         True   \n",
       "53                                                         True   \n",
       "54                                                         True   \n",
       "55                                                         True   \n",
       "\n",
       "                                            Module columnType  \\\n",
       "0                           Biospecimen,proteomics     string   \n",
       "1                                       proteomics     string   \n",
       "2                    proteomics,sampleType = other     string   \n",
       "3                                       proteomics     string   \n",
       "4                                       proteomics     string   \n",
       "5          proteomics,measurementTechnique = other     string   \n",
       "6                                       proteomics     string   \n",
       "7     proteomics,technologyPlatformVersion = other     string   \n",
       "8                                       proteomics     string   \n",
       "9              proteomics,platformLocation = other     string   \n",
       "10                                      proteomics     string   \n",
       "11                                      proteomics     string   \n",
       "12                                      proteomics     string   \n",
       "13                                      proteomics     string   \n",
       "14                                      proteomics     string   \n",
       "15                                      proteomics     string   \n",
       "16                                      proteomics     string   \n",
       "17                                      proteomics     string   \n",
       "18                                      proteomics     string   \n",
       "19        proteomics,labelQuantiation = Label Free     string   \n",
       "20        proteomics,labelFreeQuantitation = other     string   \n",
       "21                                      proteomics     string   \n",
       "22                                      proteomics     string   \n",
       "23                            mass spec proteomics     string   \n",
       "24  mass spec proteomics,msAnalyzerTypeMS1 = other     string   \n",
       "25                            mass spec proteomics     string   \n",
       "26  mass spec proteomics,msAnalyzerTypeMS2 = other     string   \n",
       "27                                      proteomics     string   \n",
       "28             proteomics,msAssayTechnique = other     string   \n",
       "29                                     proteomics,     string   \n",
       "30                proteomics,msAnalyteType = other     string   \n",
       "31          proteomics,proteomicsAssayType = other     string   \n",
       "32            proteomics,msInstrumentModel = other     string   \n",
       "33                                      proteomics     string   \n",
       "34              proteomics,digestionMethod = other     string   \n",
       "35                                      proteomics     string   \n",
       "36               proteomics digestionReagent=other     string   \n",
       "37                                      proteomics     string   \n",
       "38                                      proteomics     string   \n",
       "39                                      proteomics     string   \n",
       "40                                      proteomics     string   \n",
       "41                                      proteomics     number   \n",
       "42                            mass spec proteomics     string   \n",
       "43                            mass spec proteomics     string   \n",
       "44        mass spec proteomics,acquisitionSoftware     number   \n",
       "45                            mass spec proteomics     string   \n",
       "46       mass spec proteomics,databaseName = other     string   \n",
       "47                            mass spec proteomics     string   \n",
       "48     mass spec proteomics,databaseSource = other     string   \n",
       "49                            mass spec proteomics     number   \n",
       "50                    proteomics, dataFile = other     string   \n",
       "51       proteomics,modificationParameters = other     string   \n",
       "52                                      proteomics     string   \n",
       "53               proteomics,cleavageAgents = other     string   \n",
       "54                                      proteomics     string   \n",
       "55                                      proteomics     string   \n",
       "\n",
       "                                             Ontology  \\\n",
       "0                                    Sage Bionetworks   \n",
       "1                                   Sage Bionetworks,   \n",
       "2                                    Sage Bionetworks   \n",
       "3                              Proteomic Data Commons   \n",
       "4                                    Sage Bionetworks   \n",
       "5                                    Sage Bionetworks   \n",
       "6          http://purl.obolibrary.org/obo/NCIT_C45378   \n",
       "7                                    Sage Bionetworks   \n",
       "8                              Sage Bionetworks,DSLWG   \n",
       "9                                    Sage Bionetworks   \n",
       "10  Sage Bionetworks,DSLWG,Proposed minimum metada...   \n",
       "11                             Sage Bionetworks,DSLWG   \n",
       "12                         Proteomics Sample Metadata   \n",
       "13          http://purl.obolibrary.org/obo/MS_1001012   \n",
       "14                             Sage Bionetworks,DSLWG   \n",
       "15          http://purl.obolibrary.org/obo/MS_1001013   \n",
       "16  https://www.ebi.ac.uk/ols/ontologies/ms/terms?...   \n",
       "17                                   Sage Bionetworks   \n",
       "18  Sage Bionetworks,DSLWG,https://www.ebi.ac.uk/o...   \n",
       "19                            Proteomics Data Commons   \n",
       "20                                   Sage Bionetworks   \n",
       "21                            Proteomics Data Commons   \n",
       "22                             Sage Bionetworks,DSLWG   \n",
       "23  Sage Bionetworks,Proposed minimum metadata rel...   \n",
       "24                                   Sage Bionetworks   \n",
       "25  Sage Bionetworks,Proposed minimum metadata rel...   \n",
       "26                                   Sage Bionetworks   \n",
       "27                             Sage Bionetworks,DSLWG   \n",
       "28                                   Sage Bionetworks   \n",
       "29  http://purl.obolibrary.org/obo/NCIT_C156434,Pr...   \n",
       "30                                   Sage Bionetworks   \n",
       "31                                   Sage Bionetworks   \n",
       "32                                   Sage Bionetworks   \n",
       "33                             Sage Bionetworks,DSLWG   \n",
       "34                                   Sage Bionetworks   \n",
       "35                            Proteomics Data Commons   \n",
       "36                                   Sage Bionetworks   \n",
       "37  Sage Bionetworks,The Proteomics Sample Metadat...   \n",
       "38                             Proteomic Data Commons   \n",
       "39                             Proteomic Data Commons   \n",
       "40                             Proteomic Data Commons   \n",
       "41  Sage Bionetworks,The Proteomics Sample Metadat...   \n",
       "42  https://www.ebi.ac.uk/ols/ontologies/ms/terms?...   \n",
       "43          http://purl.obolibrary.org/obo/MS_1001455   \n",
       "44                             Sage Bionetworks,DSLWG   \n",
       "45          http://purl.obolibrary.org/obo/MS_1001013   \n",
       "46                                   Sage Bionetworks   \n",
       "47          http://purl.obolibrary.org/obo/MS_1001012   \n",
       "48                                   Sage Bionetworks   \n",
       "49                             Sage Bionetworks,DSLWG   \n",
       "50                                   Sage Bionetworks   \n",
       "51                                   Sage Bionetworks   \n",
       "52                         Proteomics Sample Metadata   \n",
       "53                                   Sage Bionetworks   \n",
       "54                         Proteomics Sample Metadata   \n",
       "55                         Proteomics Sample Metadata   \n",
       "\n",
       "                                                Notes  \n",
       "0             values unique to each data contributor)  \n",
       "1   A data contributor should be able to write in ...  \n",
       "2                                                      \n",
       "3   A data contributor should be able to write in ...  \n",
       "4   A data contributor should be able to write in ...  \n",
       "5   A data contributor should be able to write in ...  \n",
       "6   A data contributor should be able to write in ...  \n",
       "7   A data contributor should be able to write in ...  \n",
       "8   A data contributor should be able to write in ...  \n",
       "9                                                      \n",
       "10  A data contributor should be able to write in ...  \n",
       "11  A data contributor should be able to write in ...  \n",
       "12  A data contributor should be able to write in ...  \n",
       "13  A data contributor should be able to write in ...  \n",
       "14  A data contributor should be able to write in ...  \n",
       "15  A data contributor should be able to write in ...  \n",
       "16  Possible values are listed under modification ...  \n",
       "17  A data contributor should be able to write in ...  \n",
       "18                                                     \n",
       "19                                                     \n",
       "20                                                     \n",
       "21  A data contributor should be able to write in ...  \n",
       "22                                                     \n",
       "23                                                     \n",
       "24                                                     \n",
       "25                                                     \n",
       "26                                                     \n",
       "27  A data contributor should be able to write in ...  \n",
       "28                                                     \n",
       "29  A data contributor should be able to write in ...  \n",
       "30  A data contributor should be able to write in ...  \n",
       "31                                                     \n",
       "32                                                     \n",
       "33  A data contributor should be able to write in ...  \n",
       "34  A data contributor should be able to write in ...  \n",
       "35                                                     \n",
       "36                                                     \n",
       "37  A data contributor should be able to write in ...  \n",
       "38  A data contributor should be able to write in ...  \n",
       "39                                                     \n",
       "40  A data contributor should be able to write in ...  \n",
       "41  A data contributor should be able to write in ...  \n",
       "42  A data contributor should be able to write in ...  \n",
       "43  A data contributor should be able to write in ...  \n",
       "44  A data contributor should be able to write in ...  \n",
       "45  A data contributor should be able to write in ...  \n",
       "46  A data contributor should be able to write in ...  \n",
       "47  A data contributor should be able to write in ...  \n",
       "48  A data contributor should be able to write in ...  \n",
       "49  A data contributor should be able to write in ...  \n",
       "50                                                     \n",
       "51                                                     \n",
       "52  A data contributor should be able to write in ...  \n",
       "53                                                     \n",
       "54                                                     \n",
       "55                                                     "
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "p_df[\"Required\"] = (\n",
    "    p_df[\"Required\"].astype(str)\n",
    "    .apply(lambda x: False if bool(re.search(\"f\", x)) else True)\n",
    "    .astype(bool)\n",
    ")\n",
    "\n",
    "p_df = p_df.drop(columns=[\"multivalue\"])\n",
    "\n",
    "p_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>Module</th>\n",
       "      <th>columnType</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Parent</th>\n",
       "      <th>OtherValue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>specifySampleType</td>\n",
       "      <td>If \"other\" is selected list the type of sample</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherSampleType</td>\n",
       "      <td>sampleType</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the me...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>OtherMeasurementTechnique</td>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>specifyPlatformVersion</td>\n",
       "      <td>If \"other\" list the name of the platform version</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>OtherTechnologyPlatformVersion</td>\n",
       "      <td>technologyPlatformVersion</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>specifyPlatformLocation</td>\n",
       "      <td>If \"other\" list the name of the platform location</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherPlatformLocation</td>\n",
       "      <td>platformLocation</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>labelFreeQuantitation</td>\n",
       "      <td>Type of label-free data analysis strategy ,Pro...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Data Commons</td>\n",
       "      <td></td>\n",
       "      <td>Label freeLabelQuantiation</td>\n",
       "      <td>labelQuantiation</td>\n",
       "      <td>Label Free</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>specifyLabelFreeQuantitation</td>\n",
       "      <td>If \"other\" is selected please specify</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherLabelFreeQuantitation</td>\n",
       "      <td>labelFreeQuantitation</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>specifyMSAnalyzerTypeMS1</td>\n",
       "      <td>If \"other\" is selected list the name of the ma...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherMsAnalyzerTypeMS1</td>\n",
       "      <td>msAnalyzerTypeMS1</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>specifyMSAnalyzerTypeMS2</td>\n",
       "      <td>If \"other\" is selected list the name of the ma...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherMsAnalyzerTypeMS2</td>\n",
       "      <td>msAnalyzerTypeMS2</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>specifyMSAssayTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the ma...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherMsAssayTechnique</td>\n",
       "      <td>msAssayTechnique</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>specifyMSAnalyteType</td>\n",
       "      <td>If \"other\" list the type of analyte</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>OtherMsAnalyteType</td>\n",
       "      <td>msAnalyteType</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>specifyProteomicsAssayType</td>\n",
       "      <td>List the name of the type of peptide quantific...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherProteomicsAssayType</td>\n",
       "      <td>proteomicsAssayType</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>specifyMSInstrumentModel</td>\n",
       "      <td>List the name of the instrument model</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherMsInstrumentModel</td>\n",
       "      <td>msInstrumentModel</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>specifyDigestionMethod</td>\n",
       "      <td>If \"other\" list the type of digestion</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>OtherDigestionMethod</td>\n",
       "      <td>digestionMethod</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>specifyDigestionReagent</td>\n",
       "      <td>If \"other\" is selected specify the name of the...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherProteomics digestionReagent</td>\n",
       "      <td>proteomics digestionReagent</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>specifyDatabaseName</td>\n",
       "      <td>If \"other\" list the name of the search database</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>OtherDatabaseName</td>\n",
       "      <td>databaseName</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>specifyDatabaseSource</td>\n",
       "      <td>If \"other\" list the name of the database source</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>OtherDatabaseSource</td>\n",
       "      <td>databaseSource</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>specifyOtherDataFile</td>\n",
       "      <td>If \"other\" is selected, list the name of the d...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherDataFile</td>\n",
       "      <td>dataFile</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>specifyOtherModificationParameters</td>\n",
       "      <td>If \"other\" is selected, list the custom modifi...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherModificationParameters</td>\n",
       "      <td>modificationParameters</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>specifyOtherCleavageAgent</td>\n",
       "      <td>If \"other\" is selected, list the enzyme name.</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherCleavageAgents</td>\n",
       "      <td>cleavageAgents</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             DependsOn  \\\n",
       "2                    specifySampleType   \n",
       "5          specifyMeasurementTechnique   \n",
       "7               specifyPlatformVersion   \n",
       "9              specifyPlatformLocation   \n",
       "19               labelFreeQuantitation   \n",
       "20        specifyLabelFreeQuantitation   \n",
       "24            specifyMSAnalyzerTypeMS1   \n",
       "26            specifyMSAnalyzerTypeMS2   \n",
       "28             specifyMSAssayTechnique   \n",
       "30                specifyMSAnalyteType   \n",
       "31          specifyProteomicsAssayType   \n",
       "32            specifyMSInstrumentModel   \n",
       "34              specifyDigestionMethod   \n",
       "36             specifyDigestionReagent   \n",
       "46                 specifyDatabaseName   \n",
       "48               specifyDatabaseSource   \n",
       "50                specifyOtherDataFile   \n",
       "51  specifyOtherModificationParameters   \n",
       "53           specifyOtherCleavageAgent   \n",
       "\n",
       "                                          Description Valid Values  Required  \\\n",
       "2      If \"other\" is selected list the type of sample                  False   \n",
       "5   If \"other\" is selected list the name of the me...                  False   \n",
       "7    If \"other\" list the name of the platform version                  False   \n",
       "9   If \"other\" list the name of the platform location                  False   \n",
       "19  Type of label-free data analysis strategy ,Pro...                  False   \n",
       "20              If \"other\" is selected please specify                  False   \n",
       "24  If \"other\" is selected list the name of the ma...                  False   \n",
       "26  If \"other\" is selected list the name of the ma...                  False   \n",
       "28  If \"other\" is selected list the name of the ma...                  False   \n",
       "30               If \"other\" list the type of analyte                   False   \n",
       "31  List the name of the type of peptide quantific...                  False   \n",
       "32              List the name of the instrument model                  False   \n",
       "34              If \"other\" list the type of digestion                  False   \n",
       "36  If \"other\" is selected specify the name of the...                  False   \n",
       "46    If \"other\" list the name of the search database                  False   \n",
       "48    If \"other\" list the name of the database source                  False   \n",
       "50  If \"other\" is selected, list the name of the d...                  False   \n",
       "51  If \"other\" is selected, list the custom modifi...                  False   \n",
       "53     If \"other\" is selected, list the enzyme name.                   False   \n",
       "\n",
       "   Module columnType                 Ontology  \\\n",
       "2   Other     string         Sage Bionetworks   \n",
       "5   Other     string         Sage Bionetworks   \n",
       "7   Other     string         Sage Bionetworks   \n",
       "9   Other     string         Sage Bionetworks   \n",
       "19  Other     string  Proteomics Data Commons   \n",
       "20  Other     string         Sage Bionetworks   \n",
       "24  Other     string         Sage Bionetworks   \n",
       "26  Other     string         Sage Bionetworks   \n",
       "28  Other     string         Sage Bionetworks   \n",
       "30  Other     string         Sage Bionetworks   \n",
       "31  Other     string         Sage Bionetworks   \n",
       "32  Other     string         Sage Bionetworks   \n",
       "34  Other     string         Sage Bionetworks   \n",
       "36  Other     string         Sage Bionetworks   \n",
       "46  Other     string         Sage Bionetworks   \n",
       "48  Other     string         Sage Bionetworks   \n",
       "50  Other     string         Sage Bionetworks   \n",
       "51  Other     string         Sage Bionetworks   \n",
       "53  Other     string         Sage Bionetworks   \n",
       "\n",
       "                                                Notes  \\\n",
       "2                                                       \n",
       "5   A data contributor should be able to write in ...   \n",
       "7   A data contributor should be able to write in ...   \n",
       "9                                                       \n",
       "19                                                      \n",
       "20                                                      \n",
       "24                                                      \n",
       "26                                                      \n",
       "28                                                      \n",
       "30  A data contributor should be able to write in ...   \n",
       "31                                                      \n",
       "32                                                      \n",
       "34  A data contributor should be able to write in ...   \n",
       "36                                                      \n",
       "46  A data contributor should be able to write in ...   \n",
       "48  A data contributor should be able to write in ...   \n",
       "50                                                      \n",
       "51                                                      \n",
       "53                                                      \n",
       "\n",
       "                           Attribute                       Parent  OtherValue  \n",
       "2                   OtherSampleType                    sampleType       other  \n",
       "5         OtherMeasurementTechnique          measurementTechnique       other  \n",
       "7    OtherTechnologyPlatformVersion     technologyPlatformVersion       other  \n",
       "9             OtherPlatformLocation              platformLocation       other  \n",
       "19       Label freeLabelQuantiation              labelQuantiation  Label Free  \n",
       "20       OtherLabelFreeQuantitation         labelFreeQuantitation       other  \n",
       "24           OtherMsAnalyzerTypeMS1             msAnalyzerTypeMS1       other  \n",
       "26           OtherMsAnalyzerTypeMS2             msAnalyzerTypeMS2       other  \n",
       "28            OtherMsAssayTechnique              msAssayTechnique       other  \n",
       "30               OtherMsAnalyteType                 msAnalyteType       other  \n",
       "31         OtherProteomicsAssayType           proteomicsAssayType       other  \n",
       "32           OtherMsInstrumentModel             msInstrumentModel       other  \n",
       "34             OtherDigestionMethod               digestionMethod       other  \n",
       "36  OtherProteomics digestionReagent  proteomics digestionReagent       other  \n",
       "46                OtherDatabaseName                  databaseName       other  \n",
       "48              OtherDatabaseSource                databaseSource       other  \n",
       "50                    OtherDataFile                      dataFile       other  \n",
       "51      OtherModificationParameters        modificationParameters       other  \n",
       "53              OtherCleavageAgents                cleavageAgents       other  "
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# split the values out and create new attributes to add to the model.\n",
    "# These will be the intermediate attributes that will allow the original column i.e. SampleType to then have the user fill in specifySampleType\n",
    "p_df_others = p_df.loc[p_df[\"Module\"].str.contains(\"=\", na=False)].copy(deep=True)\n",
    "p_df_others[\"others\"] = (\n",
    "    p_df_others[\"Module\"]\n",
    "    .str.split(\",\")\n",
    "    .apply(lambda x: \"\".join([y.strip() for y in x if bool(re.search(\"=\", y))]))\n",
    ")\n",
    "\n",
    "# in case there are multiple equals signs\n",
    "p_df_others = p_df_others.explode(\"others\")\n",
    "\n",
    "p_df_others[\"Parent\"] = (\n",
    "    p_df_others[\"others\"].str.split(\"=\").apply(lambda x: x[0].strip())\n",
    ")\n",
    "p_df_others[\"OtherValue\"] = (\n",
    "    p_df_others[\"others\"].str.split(\"=\").apply(lambda x: x[1].strip())\n",
    ")\n",
    "\n",
    "p_df_others[\"others\"] = (\n",
    "    p_df_others[\"others\"]\n",
    "    .str.split(\"=\")\n",
    "    .apply(lambda x: x[1].strip().capitalize() + x[0][0].upper() + x[0][1:])\n",
    ")\n",
    "# Swap others -> Attribute and Attribute -> DependsOn\n",
    "p_df_others = p_df_others.rename(\n",
    "    columns={\"Attribute\": \"DependsOn\", \"others\": \"Attribute\"}\n",
    ")\n",
    "p_df_others[\"Required\"] = False\n",
    "p_df_others[\"Module\"] = \"Other\"\n",
    "p_df_others[\"Valid Values\"] = \"\"\n",
    "p_df_others"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign data modeling variables\n",
    "p_df_others = p_df_others.assign(\n",
    "    Description=p_df_others[\"DependsOn\"].apply(\n",
    "        lambda x: f\"When {re.sub('specify|other', '',x, flags=re.IGNORECASE)} = `other`, add your custom value to the cell\"\n",
    "    ),\n",
    "    columnType=\"string\",\n",
    "    Ontology=\"Sage Bionetworks\",\n",
    "    Required=False,\n",
    "    Properties=\"ValidValue\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>Module</th>\n",
       "      <th>columnType</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Parent</th>\n",
       "      <th>OtherValue</th>\n",
       "      <th>Properties</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>specifySampleType</td>\n",
       "      <td>When SampleType = `other`, add your custom val...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherSampleType</td>\n",
       "      <td>sampleType</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>When MeasurementTechnique = `other`, add your ...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>OtherMeasurementTechnique</td>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>specifyPlatformVersion</td>\n",
       "      <td>When PlatformVersion = `other`, add your custo...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>OtherTechnologyPlatformVersion</td>\n",
       "      <td>technologyPlatformVersion</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>specifyPlatformLocation</td>\n",
       "      <td>When PlatformLocation = `other`, add your cust...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherPlatformLocation</td>\n",
       "      <td>platformLocation</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>labelFreeQuantitation</td>\n",
       "      <td>When labelFreeQuantitation = `other`, add your...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>Label freeLabelQuantiation</td>\n",
       "      <td>labelQuantiation</td>\n",
       "      <td>Label Free</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>specifyLabelFreeQuantitation</td>\n",
       "      <td>When LabelFreeQuantitation = `other`, add your...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherLabelFreeQuantitation</td>\n",
       "      <td>labelFreeQuantitation</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>specifyMSAnalyzerTypeMS1</td>\n",
       "      <td>When MSAnalyzerTypeMS1 = `other`, add your cus...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherMsAnalyzerTypeMS1</td>\n",
       "      <td>msAnalyzerTypeMS1</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>specifyMSAnalyzerTypeMS2</td>\n",
       "      <td>When MSAnalyzerTypeMS2 = `other`, add your cus...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherMsAnalyzerTypeMS2</td>\n",
       "      <td>msAnalyzerTypeMS2</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>specifyMSAssayTechnique</td>\n",
       "      <td>When MSAssayTechnique = `other`, add your cust...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherMsAssayTechnique</td>\n",
       "      <td>msAssayTechnique</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>specifyMSAnalyteType</td>\n",
       "      <td>When MSAnalyteType = `other`, add your custom ...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>OtherMsAnalyteType</td>\n",
       "      <td>msAnalyteType</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>specifyProteomicsAssayType</td>\n",
       "      <td>When ProteomicsAssayType = `other`, add your c...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherProteomicsAssayType</td>\n",
       "      <td>proteomicsAssayType</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>specifyMSInstrumentModel</td>\n",
       "      <td>When MSInstrumentModel = `other`, add your cus...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherMsInstrumentModel</td>\n",
       "      <td>msInstrumentModel</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>specifyDigestionMethod</td>\n",
       "      <td>When DigestionMethod = `other`, add your custo...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>OtherDigestionMethod</td>\n",
       "      <td>digestionMethod</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>specifyDigestionReagent</td>\n",
       "      <td>When DigestionReagent = `other`, add your cust...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherProteomics digestionReagent</td>\n",
       "      <td>proteomics digestionReagent</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>specifyDatabaseName</td>\n",
       "      <td>When DatabaseName = `other`, add your custom v...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>OtherDatabaseName</td>\n",
       "      <td>databaseName</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>specifyDatabaseSource</td>\n",
       "      <td>When DatabaseSource = `other`, add your custom...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>OtherDatabaseSource</td>\n",
       "      <td>databaseSource</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>specifyOtherDataFile</td>\n",
       "      <td>When DataFile = `other`, add your custom value...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherDataFile</td>\n",
       "      <td>dataFile</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>specifyOtherModificationParameters</td>\n",
       "      <td>When ModificationParameters = `other`, add you...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherModificationParameters</td>\n",
       "      <td>modificationParameters</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>specifyOtherCleavageAgent</td>\n",
       "      <td>When CleavageAgent = `other`, add your custom ...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>OtherCleavageAgents</td>\n",
       "      <td>cleavageAgents</td>\n",
       "      <td>other</td>\n",
       "      <td>ValidValue</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             DependsOn  \\\n",
       "2                    specifySampleType   \n",
       "5          specifyMeasurementTechnique   \n",
       "7               specifyPlatformVersion   \n",
       "9              specifyPlatformLocation   \n",
       "19               labelFreeQuantitation   \n",
       "20        specifyLabelFreeQuantitation   \n",
       "24            specifyMSAnalyzerTypeMS1   \n",
       "26            specifyMSAnalyzerTypeMS2   \n",
       "28             specifyMSAssayTechnique   \n",
       "30                specifyMSAnalyteType   \n",
       "31          specifyProteomicsAssayType   \n",
       "32            specifyMSInstrumentModel   \n",
       "34              specifyDigestionMethod   \n",
       "36             specifyDigestionReagent   \n",
       "46                 specifyDatabaseName   \n",
       "48               specifyDatabaseSource   \n",
       "50                specifyOtherDataFile   \n",
       "51  specifyOtherModificationParameters   \n",
       "53           specifyOtherCleavageAgent   \n",
       "\n",
       "                                          Description Valid Values  Required  \\\n",
       "2   When SampleType = `other`, add your custom val...                  False   \n",
       "5   When MeasurementTechnique = `other`, add your ...                  False   \n",
       "7   When PlatformVersion = `other`, add your custo...                  False   \n",
       "9   When PlatformLocation = `other`, add your cust...                  False   \n",
       "19  When labelFreeQuantitation = `other`, add your...                  False   \n",
       "20  When LabelFreeQuantitation = `other`, add your...                  False   \n",
       "24  When MSAnalyzerTypeMS1 = `other`, add your cus...                  False   \n",
       "26  When MSAnalyzerTypeMS2 = `other`, add your cus...                  False   \n",
       "28  When MSAssayTechnique = `other`, add your cust...                  False   \n",
       "30  When MSAnalyteType = `other`, add your custom ...                  False   \n",
       "31  When ProteomicsAssayType = `other`, add your c...                  False   \n",
       "32  When MSInstrumentModel = `other`, add your cus...                  False   \n",
       "34  When DigestionMethod = `other`, add your custo...                  False   \n",
       "36  When DigestionReagent = `other`, add your cust...                  False   \n",
       "46  When DatabaseName = `other`, add your custom v...                  False   \n",
       "48  When DatabaseSource = `other`, add your custom...                  False   \n",
       "50  When DataFile = `other`, add your custom value...                  False   \n",
       "51  When ModificationParameters = `other`, add you...                  False   \n",
       "53  When CleavageAgent = `other`, add your custom ...                  False   \n",
       "\n",
       "   Module columnType          Ontology  \\\n",
       "2   Other     string  Sage Bionetworks   \n",
       "5   Other     string  Sage Bionetworks   \n",
       "7   Other     string  Sage Bionetworks   \n",
       "9   Other     string  Sage Bionetworks   \n",
       "19  Other     string  Sage Bionetworks   \n",
       "20  Other     string  Sage Bionetworks   \n",
       "24  Other     string  Sage Bionetworks   \n",
       "26  Other     string  Sage Bionetworks   \n",
       "28  Other     string  Sage Bionetworks   \n",
       "30  Other     string  Sage Bionetworks   \n",
       "31  Other     string  Sage Bionetworks   \n",
       "32  Other     string  Sage Bionetworks   \n",
       "34  Other     string  Sage Bionetworks   \n",
       "36  Other     string  Sage Bionetworks   \n",
       "46  Other     string  Sage Bionetworks   \n",
       "48  Other     string  Sage Bionetworks   \n",
       "50  Other     string  Sage Bionetworks   \n",
       "51  Other     string  Sage Bionetworks   \n",
       "53  Other     string  Sage Bionetworks   \n",
       "\n",
       "                                                Notes  \\\n",
       "2                                                       \n",
       "5   A data contributor should be able to write in ...   \n",
       "7   A data contributor should be able to write in ...   \n",
       "9                                                       \n",
       "19                                                      \n",
       "20                                                      \n",
       "24                                                      \n",
       "26                                                      \n",
       "28                                                      \n",
       "30  A data contributor should be able to write in ...   \n",
       "31                                                      \n",
       "32                                                      \n",
       "34  A data contributor should be able to write in ...   \n",
       "36                                                      \n",
       "46  A data contributor should be able to write in ...   \n",
       "48  A data contributor should be able to write in ...   \n",
       "50                                                      \n",
       "51                                                      \n",
       "53                                                      \n",
       "\n",
       "                           Attribute                       Parent  OtherValue  \\\n",
       "2                   OtherSampleType                    sampleType       other   \n",
       "5         OtherMeasurementTechnique          measurementTechnique       other   \n",
       "7    OtherTechnologyPlatformVersion     technologyPlatformVersion       other   \n",
       "9             OtherPlatformLocation              platformLocation       other   \n",
       "19       Label freeLabelQuantiation              labelQuantiation  Label Free   \n",
       "20       OtherLabelFreeQuantitation         labelFreeQuantitation       other   \n",
       "24           OtherMsAnalyzerTypeMS1             msAnalyzerTypeMS1       other   \n",
       "26           OtherMsAnalyzerTypeMS2             msAnalyzerTypeMS2       other   \n",
       "28            OtherMsAssayTechnique              msAssayTechnique       other   \n",
       "30               OtherMsAnalyteType                 msAnalyteType       other   \n",
       "31         OtherProteomicsAssayType           proteomicsAssayType       other   \n",
       "32           OtherMsInstrumentModel             msInstrumentModel       other   \n",
       "34             OtherDigestionMethod               digestionMethod       other   \n",
       "36  OtherProteomics digestionReagent  proteomics digestionReagent       other   \n",
       "46                OtherDatabaseName                  databaseName       other   \n",
       "48              OtherDatabaseSource                databaseSource       other   \n",
       "50                    OtherDataFile                      dataFile       other   \n",
       "51      OtherModificationParameters        modificationParameters       other   \n",
       "53              OtherCleavageAgents                cleavageAgents       other   \n",
       "\n",
       "    Properties  \n",
       "2   ValidValue  \n",
       "5   ValidValue  \n",
       "7   ValidValue  \n",
       "9   ValidValue  \n",
       "19  ValidValue  \n",
       "20  ValidValue  \n",
       "24  ValidValue  \n",
       "26  ValidValue  \n",
       "28  ValidValue  \n",
       "30  ValidValue  \n",
       "31  ValidValue  \n",
       "32  ValidValue  \n",
       "34  ValidValue  \n",
       "36  ValidValue  \n",
       "46  ValidValue  \n",
       "48  ValidValue  \n",
       "50  ValidValue  \n",
       "51  ValidValue  \n",
       "53  ValidValue  "
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_df_others"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add other attributes to the list of valid values\n",
    "p_df.loc[~p_df.replace(\"\", np.nan)[\"Valid Values\"].isna(), 'Valid Values'] = p_df.loc[~p_df.replace(\"\", np.nan)[\"Valid Values\"].isna()].apply(lambda x: re.sub(',+', ',', ','.join(x['Valid Values'] + ',' + p_df_others.loc[p_df_others['Parent'] == x['Attribute'], 'Attribute'].values)), axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Last bit of cleanup\n",
    "p_df[\"Properties\"] = \"ManifestColumn\"\n",
    "p_df[\"Module\"] = p_df[\"Attribute\"].apply(\n",
    "    lambda x: \"Other\" if bool(re.search(\"specify\", x)) else \"Metadata\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>Module</th>\n",
       "      <th>columnType</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Properties</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>specimenID</td>\n",
       "      <td>Identifying string linked to a particular samp...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>values unique to each data contributor)</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sampleType</td>\n",
       "      <td>The type of sample collected or the term used ...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>specifySampleType</td>\n",
       "      <td>If \"other\" is selected list the type of sample</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>experimentType</td>\n",
       "      <td>The type of experiment used.,Provide a value O...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomic Data Commons</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The name of the measurement technique describi...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the me...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>technologyPlatformVersion</td>\n",
       "      <td>The specific version (application, manufacture...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C45378</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>specifyPlatformVersion</td>\n",
       "      <td>If \"other\" list the name of the platform version</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>platformLocation</td>\n",
       "      <td>The name of the laboratory, facility, vendor, ...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>specifyPlatformLocation</td>\n",
       "      <td>If \"other\" list the name of the platform location</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>instrumentModel</td>\n",
       "      <td>The model of the instrument used.,Provide a va...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG,Proposed minimum metada...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>samplePrepProtocol</td>\n",
       "      <td>An internet address that may provide more deta...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>dataFile</td>\n",
       "      <td>Name of additional file(s) accompanying the da...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>databaseSource</td>\n",
       "      <td>The name of the organization, project, or labo...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001012</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>databaseWeblink</td>\n",
       "      <td>An internet address that may provide details o...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>databaseName</td>\n",
       "      <td>The name of the search database (nr, SwissProt...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001013</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>modificationParameters</td>\n",
       "      <td>Modification parameters for a search engine ru...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/terms?...</td>\n",
       "      <td>Possible values are listed under modification ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>quantationStrategy</td>\n",
       "      <td>The general strategy used for differential ana...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>labelQuantitation</td>\n",
       "      <td>Type of labeling used ,Provide a value OR prov...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG,https://www.ebi.ac.uk/o...</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>labelFreeQuantitation</td>\n",
       "      <td>Type of label-free data analysis strategy ,Pro...</td>\n",
       "      <td>Label-free protein level quantitation,Label-fr...</td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Data Commons</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>specifyLabelFreeQuantitation</td>\n",
       "      <td>If \"other\" is selected please specify</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>isobaricLabelingReagent</td>\n",
       "      <td>Reagent set used for isobaric labeling.,Exampl...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Data Commons</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>msTarget</td>\n",
       "      <td>Specifies whether or not a specific molecule(s...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>msAnalyzerTypeMS1</td>\n",
       "      <td>Describes the mass spectrometer analyzer used ...</td>\n",
       "      <td>LTQ Orbitrap XL,Orbitrap Fusion,Orbitrap Fusio...</td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,Proposed minimum metadata rel...</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>specifyMSAnalyzerTypeMS1</td>\n",
       "      <td>If \"other\" is selected list the name of the ma...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>msAnalyzerTypeMS2</td>\n",
       "      <td>Describes the mass spectrometer analyzer used ...</td>\n",
       "      <td>LTQ Orbitrap XL,Orbitrap Fusion,Orbitrap Fusio...</td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,Proposed minimum metadata rel...</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>specifyMSAnalyzerTypeMS2</td>\n",
       "      <td>If \"other\" is selected list the name of the ma...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>msAssayTechnique</td>\n",
       "      <td>Specified the mass spectrometer technique used...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>specifyMSAssayTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the ma...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>msAnalyteType</td>\n",
       "      <td>The type of biospecimen subjected to analysis ...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C156434,Pr...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>specifyMSAnalyteType</td>\n",
       "      <td>If \"other\" list the type of analyte</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>specifyProteomicsAssayType</td>\n",
       "      <td>List the name of the type of peptide quantific...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>specifyMSInstrumentModel</td>\n",
       "      <td>List the name of the instrument model</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>digestionMethod</td>\n",
       "      <td>The type of digestion (ex. in-solution, in-gel...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>specifyDigestionMethod</td>\n",
       "      <td>If \"other\" list the type of digestion</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>digestionReagent</td>\n",
       "      <td>Enzymes or reagents used for digestion. ,</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Data Commons</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>specifyDigestionReagent</td>\n",
       "      <td>If \"other\" is selected specify the name of the...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>fractionIdentifier</td>\n",
       "      <td>Number of fractions in the experimental run,Pr...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,The Proteomics Sample Metadat...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>enrichmentStrategy</td>\n",
       "      <td>Type of peptide or protein enrichment,</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomic Data Commons</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>specifyEnrichmentStrategy</td>\n",
       "      <td>If \"other\" is selected specify the name</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomic Data Commons</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>enrichment</td>\n",
       "      <td>Protocol or reagent used for enrichment (ex. F...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomic Data Commons</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>fractionIdentifier</td>\n",
       "      <td>An identifier string that describes the sample...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>number</td>\n",
       "      <td>Sage Bionetworks,The Proteomics Sample Metadat...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>acquisitionMode</td>\n",
       "      <td>The specific aspect of a mass spectrometer met...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/terms?...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>acquisitionSoftware</td>\n",
       "      <td>The name of the acquisition software used</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>acquistionSoftwareVersion</td>\n",
       "      <td>The version number of software used</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>number</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>databaseName</td>\n",
       "      <td>The name of the search database (nr, SwissProt...</td>\n",
       "      <td>MassBank of North America (MoNA),NIST17,Metlin...</td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001013</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>specifyDatabaseName</td>\n",
       "      <td>If \"other\" list the name of the search database</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>databaseSource</td>\n",
       "      <td>The name of the organization, project, or labo...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001012</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>specifyDatabaseSource</td>\n",
       "      <td>If \"other\" list the name of the database source</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>spectrometerFrequency</td>\n",
       "      <td>The frequency at which a spectrometer causes h...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>number</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>specifyOtherDataFile</td>\n",
       "      <td>If \"other\" is selected, list the name of the d...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>specifyOtherModificationParameters</td>\n",
       "      <td>If \"other\" is selected, list the custom modifi...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>cleavageAgents</td>\n",
       "      <td>Name(s) of the enzyme used (possible values ar...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>specifyOtherCleavageAgent</td>\n",
       "      <td>If \"other\" is selected, list the enzyme name.</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>fragmentMassTolerance</td>\n",
       "      <td>Units of measure are Da or ppm (ex. \"0.6 Da\" o...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>precursorMassTolerance</td>\n",
       "      <td>Units of measure are Da or ppm (ex. \"0.6 Da\" o...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Attribute  \\\n",
       "0                           specimenID   \n",
       "1                           sampleType   \n",
       "2                    specifySampleType   \n",
       "3                       experimentType   \n",
       "4                 measurementTechnique   \n",
       "5          specifyMeasurementTechnique   \n",
       "6            technologyPlatformVersion   \n",
       "7               specifyPlatformVersion   \n",
       "8                     platformLocation   \n",
       "9              specifyPlatformLocation   \n",
       "10                     instrumentModel   \n",
       "11                  samplePrepProtocol   \n",
       "12                            dataFile   \n",
       "13                      databaseSource   \n",
       "14                     databaseWeblink   \n",
       "15                        databaseName   \n",
       "16              modificationParameters   \n",
       "17                  quantationStrategy   \n",
       "18                   labelQuantitation   \n",
       "19               labelFreeQuantitation   \n",
       "20        specifyLabelFreeQuantitation   \n",
       "21             isobaricLabelingReagent   \n",
       "22                            msTarget   \n",
       "23                   msAnalyzerTypeMS1   \n",
       "24            specifyMSAnalyzerTypeMS1   \n",
       "25                   msAnalyzerTypeMS2   \n",
       "26            specifyMSAnalyzerTypeMS2   \n",
       "27                    msAssayTechnique   \n",
       "28             specifyMSAssayTechnique   \n",
       "29                       msAnalyteType   \n",
       "30                specifyMSAnalyteType   \n",
       "31          specifyProteomicsAssayType   \n",
       "32            specifyMSInstrumentModel   \n",
       "33                     digestionMethod   \n",
       "34              specifyDigestionMethod   \n",
       "35                    digestionReagent   \n",
       "36             specifyDigestionReagent   \n",
       "37                  fractionIdentifier   \n",
       "38                  enrichmentStrategy   \n",
       "39           specifyEnrichmentStrategy   \n",
       "40                          enrichment   \n",
       "41                  fractionIdentifier   \n",
       "42                     acquisitionMode   \n",
       "43                 acquisitionSoftware   \n",
       "44           acquistionSoftwareVersion   \n",
       "45                        databaseName   \n",
       "46                 specifyDatabaseName   \n",
       "47                      databaseSource   \n",
       "48               specifyDatabaseSource   \n",
       "49               spectrometerFrequency   \n",
       "50                specifyOtherDataFile   \n",
       "51  specifyOtherModificationParameters   \n",
       "52                      cleavageAgents   \n",
       "53           specifyOtherCleavageAgent   \n",
       "54               fragmentMassTolerance   \n",
       "55              precursorMassTolerance   \n",
       "\n",
       "                                          Description  \\\n",
       "0   Identifying string linked to a particular samp...   \n",
       "1   The type of sample collected or the term used ...   \n",
       "2      If \"other\" is selected list the type of sample   \n",
       "3   The type of experiment used.,Provide a value O...   \n",
       "4   The name of the measurement technique describi...   \n",
       "5   If \"other\" is selected list the name of the me...   \n",
       "6   The specific version (application, manufacture...   \n",
       "7    If \"other\" list the name of the platform version   \n",
       "8   The name of the laboratory, facility, vendor, ...   \n",
       "9   If \"other\" list the name of the platform location   \n",
       "10  The model of the instrument used.,Provide a va...   \n",
       "11  An internet address that may provide more deta...   \n",
       "12  Name of additional file(s) accompanying the da...   \n",
       "13  The name of the organization, project, or labo...   \n",
       "14  An internet address that may provide details o...   \n",
       "15  The name of the search database (nr, SwissProt...   \n",
       "16  Modification parameters for a search engine ru...   \n",
       "17  The general strategy used for differential ana...   \n",
       "18  Type of labeling used ,Provide a value OR prov...   \n",
       "19  Type of label-free data analysis strategy ,Pro...   \n",
       "20              If \"other\" is selected please specify   \n",
       "21  Reagent set used for isobaric labeling.,Exampl...   \n",
       "22  Specifies whether or not a specific molecule(s...   \n",
       "23  Describes the mass spectrometer analyzer used ...   \n",
       "24  If \"other\" is selected list the name of the ma...   \n",
       "25  Describes the mass spectrometer analyzer used ...   \n",
       "26  If \"other\" is selected list the name of the ma...   \n",
       "27  Specified the mass spectrometer technique used...   \n",
       "28  If \"other\" is selected list the name of the ma...   \n",
       "29  The type of biospecimen subjected to analysis ...   \n",
       "30               If \"other\" list the type of analyte    \n",
       "31  List the name of the type of peptide quantific...   \n",
       "32              List the name of the instrument model   \n",
       "33  The type of digestion (ex. in-solution, in-gel...   \n",
       "34              If \"other\" list the type of digestion   \n",
       "35          Enzymes or reagents used for digestion. ,   \n",
       "36  If \"other\" is selected specify the name of the...   \n",
       "37  Number of fractions in the experimental run,Pr...   \n",
       "38             Type of peptide or protein enrichment,   \n",
       "39            If \"other\" is selected specify the name   \n",
       "40  Protocol or reagent used for enrichment (ex. F...   \n",
       "41  An identifier string that describes the sample...   \n",
       "42  The specific aspect of a mass spectrometer met...   \n",
       "43          The name of the acquisition software used   \n",
       "44                The version number of software used   \n",
       "45  The name of the search database (nr, SwissProt...   \n",
       "46    If \"other\" list the name of the search database   \n",
       "47  The name of the organization, project, or labo...   \n",
       "48    If \"other\" list the name of the database source   \n",
       "49  The frequency at which a spectrometer causes h...   \n",
       "50  If \"other\" is selected, list the name of the d...   \n",
       "51  If \"other\" is selected, list the custom modifi...   \n",
       "52  Name(s) of the enzyme used (possible values ar...   \n",
       "53     If \"other\" is selected, list the enzyme name.    \n",
       "54  Units of measure are Da or ppm (ex. \"0.6 Da\" o...   \n",
       "55  Units of measure are Da or ppm (ex. \"0.6 Da\" o...   \n",
       "\n",
       "                                         Valid Values  Required    Module  \\\n",
       "0                                                          True  Metadata   \n",
       "1                                                          True  Metadata   \n",
       "2                                                          True     Other   \n",
       "3                                                          True  Metadata   \n",
       "4                                                          True  Metadata   \n",
       "5                                                          True     Other   \n",
       "6                                                          True  Metadata   \n",
       "7                                                          True     Other   \n",
       "8                                                          True  Metadata   \n",
       "9                                                          True     Other   \n",
       "10                                                         True  Metadata   \n",
       "11                                                         True  Metadata   \n",
       "12                                                         True  Metadata   \n",
       "13                                                         True  Metadata   \n",
       "14                                                         True  Metadata   \n",
       "15                                                         True  Metadata   \n",
       "16                                                         True  Metadata   \n",
       "17                                                         True  Metadata   \n",
       "18                                                         True  Metadata   \n",
       "19  Label-free protein level quantitation,Label-fr...      True  Metadata   \n",
       "20                                                         True     Other   \n",
       "21                                                         True  Metadata   \n",
       "22                                                         True  Metadata   \n",
       "23  LTQ Orbitrap XL,Orbitrap Fusion,Orbitrap Fusio...      True  Metadata   \n",
       "24                                                         True     Other   \n",
       "25  LTQ Orbitrap XL,Orbitrap Fusion,Orbitrap Fusio...      True  Metadata   \n",
       "26                                                         True     Other   \n",
       "27                                                         True  Metadata   \n",
       "28                                                         True     Other   \n",
       "29                                                         True  Metadata   \n",
       "30                                                         True     Other   \n",
       "31                                                         True     Other   \n",
       "32                                                         True     Other   \n",
       "33                                                         True  Metadata   \n",
       "34                                                         True     Other   \n",
       "35                                                         True  Metadata   \n",
       "36                                                         True     Other   \n",
       "37                                                         True  Metadata   \n",
       "38                                                         True  Metadata   \n",
       "39                                                         True     Other   \n",
       "40                                                         True  Metadata   \n",
       "41                                                         True  Metadata   \n",
       "42                                                         True  Metadata   \n",
       "43                                                         True  Metadata   \n",
       "44                                                         True  Metadata   \n",
       "45  MassBank of North America (MoNA),NIST17,Metlin...      True  Metadata   \n",
       "46                                                         True     Other   \n",
       "47                                                         True  Metadata   \n",
       "48                                                         True     Other   \n",
       "49                                                         True  Metadata   \n",
       "50                                                         True     Other   \n",
       "51                                                         True     Other   \n",
       "52                                                         True  Metadata   \n",
       "53                                                         True     Other   \n",
       "54                                                         True  Metadata   \n",
       "55                                                         True  Metadata   \n",
       "\n",
       "   columnType                                           Ontology  \\\n",
       "0      string                                   Sage Bionetworks   \n",
       "1      string                                  Sage Bionetworks,   \n",
       "2      string                                   Sage Bionetworks   \n",
       "3      string                             Proteomic Data Commons   \n",
       "4      string                                   Sage Bionetworks   \n",
       "5      string                                   Sage Bionetworks   \n",
       "6      string         http://purl.obolibrary.org/obo/NCIT_C45378   \n",
       "7      string                                   Sage Bionetworks   \n",
       "8      string                             Sage Bionetworks,DSLWG   \n",
       "9      string                                   Sage Bionetworks   \n",
       "10     string  Sage Bionetworks,DSLWG,Proposed minimum metada...   \n",
       "11     string                             Sage Bionetworks,DSLWG   \n",
       "12     string                         Proteomics Sample Metadata   \n",
       "13     string          http://purl.obolibrary.org/obo/MS_1001012   \n",
       "14     string                             Sage Bionetworks,DSLWG   \n",
       "15     string          http://purl.obolibrary.org/obo/MS_1001013   \n",
       "16     string  https://www.ebi.ac.uk/ols/ontologies/ms/terms?...   \n",
       "17     string                                   Sage Bionetworks   \n",
       "18     string  Sage Bionetworks,DSLWG,https://www.ebi.ac.uk/o...   \n",
       "19     string                            Proteomics Data Commons   \n",
       "20     string                                   Sage Bionetworks   \n",
       "21     string                            Proteomics Data Commons   \n",
       "22     string                             Sage Bionetworks,DSLWG   \n",
       "23     string  Sage Bionetworks,Proposed minimum metadata rel...   \n",
       "24     string                                   Sage Bionetworks   \n",
       "25     string  Sage Bionetworks,Proposed minimum metadata rel...   \n",
       "26     string                                   Sage Bionetworks   \n",
       "27     string                             Sage Bionetworks,DSLWG   \n",
       "28     string                                   Sage Bionetworks   \n",
       "29     string  http://purl.obolibrary.org/obo/NCIT_C156434,Pr...   \n",
       "30     string                                   Sage Bionetworks   \n",
       "31     string                                   Sage Bionetworks   \n",
       "32     string                                   Sage Bionetworks   \n",
       "33     string                             Sage Bionetworks,DSLWG   \n",
       "34     string                                   Sage Bionetworks   \n",
       "35     string                            Proteomics Data Commons   \n",
       "36     string                                   Sage Bionetworks   \n",
       "37     string  Sage Bionetworks,The Proteomics Sample Metadat...   \n",
       "38     string                             Proteomic Data Commons   \n",
       "39     string                             Proteomic Data Commons   \n",
       "40     string                             Proteomic Data Commons   \n",
       "41     number  Sage Bionetworks,The Proteomics Sample Metadat...   \n",
       "42     string  https://www.ebi.ac.uk/ols/ontologies/ms/terms?...   \n",
       "43     string          http://purl.obolibrary.org/obo/MS_1001455   \n",
       "44     number                             Sage Bionetworks,DSLWG   \n",
       "45     string          http://purl.obolibrary.org/obo/MS_1001013   \n",
       "46     string                                   Sage Bionetworks   \n",
       "47     string          http://purl.obolibrary.org/obo/MS_1001012   \n",
       "48     string                                   Sage Bionetworks   \n",
       "49     number                             Sage Bionetworks,DSLWG   \n",
       "50     string                                   Sage Bionetworks   \n",
       "51     string                                   Sage Bionetworks   \n",
       "52     string                         Proteomics Sample Metadata   \n",
       "53     string                                   Sage Bionetworks   \n",
       "54     string                         Proteomics Sample Metadata   \n",
       "55     string                         Proteomics Sample Metadata   \n",
       "\n",
       "                                                Notes      Properties  \n",
       "0             values unique to each data contributor)  ManifestColumn  \n",
       "1   A data contributor should be able to write in ...  ManifestColumn  \n",
       "2                                                      ManifestColumn  \n",
       "3   A data contributor should be able to write in ...  ManifestColumn  \n",
       "4   A data contributor should be able to write in ...  ManifestColumn  \n",
       "5   A data contributor should be able to write in ...  ManifestColumn  \n",
       "6   A data contributor should be able to write in ...  ManifestColumn  \n",
       "7   A data contributor should be able to write in ...  ManifestColumn  \n",
       "8   A data contributor should be able to write in ...  ManifestColumn  \n",
       "9                                                      ManifestColumn  \n",
       "10  A data contributor should be able to write in ...  ManifestColumn  \n",
       "11  A data contributor should be able to write in ...  ManifestColumn  \n",
       "12  A data contributor should be able to write in ...  ManifestColumn  \n",
       "13  A data contributor should be able to write in ...  ManifestColumn  \n",
       "14  A data contributor should be able to write in ...  ManifestColumn  \n",
       "15  A data contributor should be able to write in ...  ManifestColumn  \n",
       "16  Possible values are listed under modification ...  ManifestColumn  \n",
       "17  A data contributor should be able to write in ...  ManifestColumn  \n",
       "18                                                     ManifestColumn  \n",
       "19                                                     ManifestColumn  \n",
       "20                                                     ManifestColumn  \n",
       "21  A data contributor should be able to write in ...  ManifestColumn  \n",
       "22                                                     ManifestColumn  \n",
       "23                                                     ManifestColumn  \n",
       "24                                                     ManifestColumn  \n",
       "25                                                     ManifestColumn  \n",
       "26                                                     ManifestColumn  \n",
       "27  A data contributor should be able to write in ...  ManifestColumn  \n",
       "28                                                     ManifestColumn  \n",
       "29  A data contributor should be able to write in ...  ManifestColumn  \n",
       "30  A data contributor should be able to write in ...  ManifestColumn  \n",
       "31                                                     ManifestColumn  \n",
       "32                                                     ManifestColumn  \n",
       "33  A data contributor should be able to write in ...  ManifestColumn  \n",
       "34  A data contributor should be able to write in ...  ManifestColumn  \n",
       "35                                                     ManifestColumn  \n",
       "36                                                     ManifestColumn  \n",
       "37  A data contributor should be able to write in ...  ManifestColumn  \n",
       "38  A data contributor should be able to write in ...  ManifestColumn  \n",
       "39                                                     ManifestColumn  \n",
       "40  A data contributor should be able to write in ...  ManifestColumn  \n",
       "41  A data contributor should be able to write in ...  ManifestColumn  \n",
       "42  A data contributor should be able to write in ...  ManifestColumn  \n",
       "43  A data contributor should be able to write in ...  ManifestColumn  \n",
       "44  A data contributor should be able to write in ...  ManifestColumn  \n",
       "45  A data contributor should be able to write in ...  ManifestColumn  \n",
       "46  A data contributor should be able to write in ...  ManifestColumn  \n",
       "47  A data contributor should be able to write in ...  ManifestColumn  \n",
       "48  A data contributor should be able to write in ...  ManifestColumn  \n",
       "49  A data contributor should be able to write in ...  ManifestColumn  \n",
       "50                                                     ManifestColumn  \n",
       "51                                                     ManifestColumn  \n",
       "52  A data contributor should be able to write in ...  ManifestColumn  \n",
       "53                                                     ManifestColumn  \n",
       "54                                                     ManifestColumn  \n",
       "55                                                     ManifestColumn  "
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape of original data frame: (56, 9)\n",
      "Shape of others data frame: (19, 12)\n",
      "Shape of final data frame: (75, 11)\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 75 entries, specimenID to OtherCleavageAgents \n",
      "Data columns (total 11 columns):\n",
      " #   Column        Non-Null Count  Dtype \n",
      "---  ------        --------------  ----- \n",
      " 0   Description   75 non-null     object\n",
      " 1   Valid Values  4 non-null      object\n",
      " 2   Required      75 non-null     bool  \n",
      " 3   Module        75 non-null     object\n",
      " 4   columnType    75 non-null     object\n",
      " 5   Ontology      75 non-null     object\n",
      " 6   Notes         41 non-null     object\n",
      " 7   Properties    75 non-null     object\n",
      " 8   DependsOn     19 non-null     object\n",
      " 9   Parent        19 non-null     object\n",
      " 10  OtherValue    19 non-null     object\n",
      "dtypes: bool(1), object(10)\n",
      "memory usage: 6.5+ KB\n"
     ]
    }
   ],
   "source": [
    "print(\"Shape of original data frame:\", p_df.shape)\n",
    "print(\"Shape of others data frame:\", p_df_others.shape)\n",
    "\n",
    "p_df_final = pd.concat([p_df, p_df_others]).reset_index(drop=True).replace(\"\", np.nan)\n",
    "p_df_final = p_df_final.set_index(\"Attribute\")\n",
    "\n",
    "print(\"Shape of final data frame:\", p_df_final.shape)\n",
    "\n",
    "p_df_final.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_df_final['UsedIn'] = 'proteomics'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Module</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>UsedIn</th>\n",
       "      <th>columnType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>proteomics</td>\n",
       "      <td>Template used for contributing metadata to the...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Component,Filename,specimenID,sampleType,speci...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>Component</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Template</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Attribute                                        Description  \\\n",
       "38  proteomics  Template used for contributing metadata to the...   \n",
       "\n",
       "   Valid Values                                          DependsOn  \\\n",
       "38          NaN  Component,Filename,specimenID,sampleType,speci...   \n",
       "\n",
       "    Properties Required     Parent  DependsOn Component Source  \\\n",
       "38         NaN     True  Component                  NaN    NaN   \n",
       "\n",
       "   Validation Rules    Module Ontology UsedIn columnType  \n",
       "38              NaN  Template      NaN    NaN        NaN  "
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add proteomics template to the data model updated with the attributes in DependsOn\n",
    "dm.loc[dm[\"Attribute\"] == \"proteomics\", \"DependsOn\"] = (\n",
    "    \"Component,\" + \"Filename,\" + \",\".join(p_df[\"Attribute\"])\n",
    ")\n",
    "\n",
    "dm.loc[dm[\"Attribute\"] == \"proteomics\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Instrument', 'Template', 'Analysis', 'ValidValues', 'Assay',\n",
       "       'Other', 'Consortium', 'Ontology', 'Repository', 'SampleType',\n",
       "       'Unspecified', 'Unit', 'Metadata', 'Model Organism',\n",
       "       'BaseAnnotation'], dtype=object)"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm['Module'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ValidValue', 'Component', 'AnalysisType', 'MeasurementUnit',\n",
       "       'ManifestColumn', 'Ontology', 'Repository', 'Demographics',\n",
       "       'BaseAnnotation'], dtype=object)"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm[\"Parent\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm = dm.set_index('Attribute')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(700, 13)"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Module</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>UsedIn</th>\n",
       "      <th>columnType</th>\n",
       "      <th>Notes</th>\n",
       "      <th>OtherValue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Attribute</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10x Visium Spatial Gene Expression</th>\n",
       "      <td>Instrument Model Name</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Instrument</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnVision 2103 Multiplate Reader</th>\n",
       "      <td>Instrument Model Name</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Instrument</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Exploris 240Exploris 480quadrupole time-of-flight</th>\n",
       "      <td>Instrument Model Name</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Instrument</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FIA-MSMS</th>\n",
       "      <td>Instrument Model Name</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Instrument</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Illumina 1M</th>\n",
       "      <td>Instrument Model Name</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Instrument</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OtherDatabaseName</th>\n",
       "      <td>When DatabaseName = `other`, add your custom v...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specifyDatabaseName</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>False</td>\n",
       "      <td>databaseName</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OtherDatabaseSource</th>\n",
       "      <td>When DatabaseSource = `other`, add your custom...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specifyDatabaseSource</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>False</td>\n",
       "      <td>databaseSource</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OtherDataFile</th>\n",
       "      <td>When DataFile = `other`, add your custom value...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specifyOtherDataFile</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>False</td>\n",
       "      <td>dataFile</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OtherModificationParameters</th>\n",
       "      <td>When ModificationParameters = `other`, add you...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specifyOtherModificationParameters</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>False</td>\n",
       "      <td>modificationParameters</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OtherCleavageAgents</th>\n",
       "      <td>When CleavageAgent = `other`, add your custom ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>specifyOtherCleavageAgent</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>False</td>\n",
       "      <td>cleavageAgents</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>775 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                         Description  \\\n",
       "Attribute                                                                                              \n",
       "10x Visium Spatial Gene Expression                                             Instrument Model Name   \n",
       "EnVision 2103 Multiplate Reader                                                Instrument Model Name   \n",
       "Exploris 240Exploris 480quadrupole time-of-flight                              Instrument Model Name   \n",
       "FIA-MSMS                                                                       Instrument Model Name   \n",
       "Illumina 1M                                                                    Instrument Model Name   \n",
       "...                                                                                              ...   \n",
       "OtherDatabaseName                                  When DatabaseName = `other`, add your custom v...   \n",
       "OtherDatabaseSource                                When DatabaseSource = `other`, add your custom...   \n",
       "OtherDataFile                                      When DataFile = `other`, add your custom value...   \n",
       "OtherModificationParameters                        When ModificationParameters = `other`, add you...   \n",
       "OtherCleavageAgents                                When CleavageAgent = `other`, add your custom ...   \n",
       "\n",
       "                                                  Valid Values  \\\n",
       "Attribute                                                        \n",
       "10x Visium Spatial Gene Expression                         NaN   \n",
       "EnVision 2103 Multiplate Reader                            NaN   \n",
       "Exploris 240Exploris 480quadrupole time-of-flight          NaN   \n",
       "FIA-MSMS                                                   NaN   \n",
       "Illumina 1M                                                NaN   \n",
       "...                                                        ...   \n",
       "OtherDatabaseName                                          NaN   \n",
       "OtherDatabaseSource                                        NaN   \n",
       "OtherDataFile                                              NaN   \n",
       "OtherModificationParameters                                NaN   \n",
       "OtherCleavageAgents                                        NaN   \n",
       "\n",
       "                                                                            DependsOn  \\\n",
       "Attribute                                                                               \n",
       "10x Visium Spatial Gene Expression                                                NaN   \n",
       "EnVision 2103 Multiplate Reader                                                   NaN   \n",
       "Exploris 240Exploris 480quadrupole time-of-flight                                 NaN   \n",
       "FIA-MSMS                                                                          NaN   \n",
       "Illumina 1M                                                                       NaN   \n",
       "...                                                                               ...   \n",
       "OtherDatabaseName                                                 specifyDatabaseName   \n",
       "OtherDatabaseSource                                             specifyDatabaseSource   \n",
       "OtherDataFile                                                    specifyOtherDataFile   \n",
       "OtherModificationParameters                        specifyOtherModificationParameters   \n",
       "OtherCleavageAgents                                         specifyOtherCleavageAgent   \n",
       "\n",
       "                                                   Properties Required  \\\n",
       "Attribute                                                                \n",
       "10x Visium Spatial Gene Expression                        NaN    False   \n",
       "EnVision 2103 Multiplate Reader                           NaN    False   \n",
       "Exploris 240Exploris 480quadrupole time-of-flight         NaN    False   \n",
       "FIA-MSMS                                                  NaN    False   \n",
       "Illumina 1M                                               NaN    False   \n",
       "...                                                       ...      ...   \n",
       "OtherDatabaseName                                  ValidValue    False   \n",
       "OtherDatabaseSource                                ValidValue    False   \n",
       "OtherDataFile                                      ValidValue    False   \n",
       "OtherModificationParameters                        ValidValue    False   \n",
       "OtherCleavageAgents                                ValidValue    False   \n",
       "\n",
       "                                                                   Parent  \\\n",
       "Attribute                                                                   \n",
       "10x Visium Spatial Gene Expression                             ValidValue   \n",
       "EnVision 2103 Multiplate Reader                                ValidValue   \n",
       "Exploris 240Exploris 480quadrupole time-of-flight              ValidValue   \n",
       "FIA-MSMS                                                       ValidValue   \n",
       "Illumina 1M                                                    ValidValue   \n",
       "...                                                                   ...   \n",
       "OtherDatabaseName                                            databaseName   \n",
       "OtherDatabaseSource                                        databaseSource   \n",
       "OtherDataFile                                                    dataFile   \n",
       "OtherModificationParameters                        modificationParameters   \n",
       "OtherCleavageAgents                                        cleavageAgents   \n",
       "\n",
       "                                                   DependsOn Component Source  \\\n",
       "Attribute                                                                       \n",
       "10x Visium Spatial Gene Expression                                 NaN    NaN   \n",
       "EnVision 2103 Multiplate Reader                                    NaN    NaN   \n",
       "Exploris 240Exploris 480quadrupole time-of-flight                  NaN    NaN   \n",
       "FIA-MSMS                                                           NaN    NaN   \n",
       "Illumina 1M                                                        NaN    NaN   \n",
       "...                                                                ...    ...   \n",
       "OtherDatabaseName                                                  NaN    NaN   \n",
       "OtherDatabaseSource                                                NaN    NaN   \n",
       "OtherDataFile                                                      NaN    NaN   \n",
       "OtherModificationParameters                                        NaN    NaN   \n",
       "OtherCleavageAgents                                                NaN    NaN   \n",
       "\n",
       "                                                  Validation Rules  \\\n",
       "Attribute                                                            \n",
       "10x Visium Spatial Gene Expression                             NaN   \n",
       "EnVision 2103 Multiplate Reader                                NaN   \n",
       "Exploris 240Exploris 480quadrupole time-of-flight              NaN   \n",
       "FIA-MSMS                                                       NaN   \n",
       "Illumina 1M                                                    NaN   \n",
       "...                                                            ...   \n",
       "OtherDatabaseName                                              NaN   \n",
       "OtherDatabaseSource                                            NaN   \n",
       "OtherDataFile                                                  NaN   \n",
       "OtherModificationParameters                                    NaN   \n",
       "OtherCleavageAgents                                            NaN   \n",
       "\n",
       "                                                       Module  \\\n",
       "Attribute                                                       \n",
       "10x Visium Spatial Gene Expression                 Instrument   \n",
       "EnVision 2103 Multiplate Reader                    Instrument   \n",
       "Exploris 240Exploris 480quadrupole time-of-flight  Instrument   \n",
       "FIA-MSMS                                           Instrument   \n",
       "Illumina 1M                                        Instrument   \n",
       "...                                                       ...   \n",
       "OtherDatabaseName                                       Other   \n",
       "OtherDatabaseSource                                     Other   \n",
       "OtherDataFile                                           Other   \n",
       "OtherModificationParameters                             Other   \n",
       "OtherCleavageAgents                                     Other   \n",
       "\n",
       "                                                           Ontology  \\\n",
       "Attribute                                                             \n",
       "10x Visium Spatial Gene Expression                              NaN   \n",
       "EnVision 2103 Multiplate Reader                                 NaN   \n",
       "Exploris 240Exploris 480quadrupole time-of-flight               NaN   \n",
       "FIA-MSMS                                                        NaN   \n",
       "Illumina 1M                                                     NaN   \n",
       "...                                                             ...   \n",
       "OtherDatabaseName                                  Sage Bionetworks   \n",
       "OtherDatabaseSource                                Sage Bionetworks   \n",
       "OtherDataFile                                      Sage Bionetworks   \n",
       "OtherModificationParameters                        Sage Bionetworks   \n",
       "OtherCleavageAgents                                Sage Bionetworks   \n",
       "\n",
       "                                                       UsedIn columnType  \\\n",
       "Attribute                                                                  \n",
       "10x Visium Spatial Gene Expression                        NaN        NaN   \n",
       "EnVision 2103 Multiplate Reader                           NaN        NaN   \n",
       "Exploris 240Exploris 480quadrupole time-of-flight         NaN        NaN   \n",
       "FIA-MSMS                                                  NaN        NaN   \n",
       "Illumina 1M                                               NaN        NaN   \n",
       "...                                                       ...        ...   \n",
       "OtherDatabaseName                                  proteomics     string   \n",
       "OtherDatabaseSource                                proteomics     string   \n",
       "OtherDataFile                                      proteomics     string   \n",
       "OtherModificationParameters                        proteomics     string   \n",
       "OtherCleavageAgents                                proteomics     string   \n",
       "\n",
       "                                                                                               Notes  \\\n",
       "Attribute                                                                                              \n",
       "10x Visium Spatial Gene Expression                                                               NaN   \n",
       "EnVision 2103 Multiplate Reader                                                                  NaN   \n",
       "Exploris 240Exploris 480quadrupole time-of-flight                                                NaN   \n",
       "FIA-MSMS                                                                                         NaN   \n",
       "Illumina 1M                                                                                      NaN   \n",
       "...                                                                                              ...   \n",
       "OtherDatabaseName                                  A data contributor should be able to write in ...   \n",
       "OtherDatabaseSource                                A data contributor should be able to write in ...   \n",
       "OtherDataFile                                                                                    NaN   \n",
       "OtherModificationParameters                                                                      NaN   \n",
       "OtherCleavageAgents                                                                              NaN   \n",
       "\n",
       "                                                  OtherValue  \n",
       "Attribute                                                     \n",
       "10x Visium Spatial Gene Expression                       NaN  \n",
       "EnVision 2103 Multiplate Reader                          NaN  \n",
       "Exploris 240Exploris 480quadrupole time-of-flight        NaN  \n",
       "FIA-MSMS                                                 NaN  \n",
       "Illumina 1M                                              NaN  \n",
       "...                                                      ...  \n",
       "OtherDatabaseName                                      other  \n",
       "OtherDatabaseSource                                    other  \n",
       "OtherDataFile                                          other  \n",
       "OtherModificationParameters                            other  \n",
       "OtherCleavageAgents                                    other  \n",
       "\n",
       "[775 rows x 15 columns]"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([dm, p_df_final])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Module</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>UsedIn</th>\n",
       "      <th>columnType</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Attribute</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>platformLocation</th>\n",
       "      <td>The name of the laboratory, facility, vendor, ...</td>\n",
       "      <td>OtherPlatformLocation, Unknown, Not collected,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Assay</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,Microbiome,RNAseq,Whole Gen...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>technologyPlatformVersion</th>\n",
       "      <td>The specific version (application, manufacture...</td>\n",
       "      <td>10x Visium Spatial Gene Expression, Affymetrix...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Assay</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,Microbiome,RNAseq,Whole Gen...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyDatabaseName</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyDatabaseSource</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyDigestionMethod</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyMeasurementTechnique</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,Microbiome,RNAseq,Whole Gen...</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyMSAnalyteType</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyMSAnalyzerTypeMS1</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyMSAnalyzerTypeMS2</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyMSAssayTechnique</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyMSInstrumentModel</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyOtherCleavageAgent</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyOtherModificationParameters</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyPlatformLocation</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,Microbiome,RNAseq,Whole Gen...</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyPlatformVersion</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,Microbiome,RNAseq,Whole Gen...</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyProteomicsAssayType</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifySampleType</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,Microbiome,RNAseq,Whole Gen...</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acquisitionMode</th>\n",
       "      <td>The specific aspect of a mass spectrometer met...</td>\n",
       "      <td>Unknown, Not collected, Not applicable, Not Sp...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/termsi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acquisitionSoftware</th>\n",
       "      <td>The name of the acquisition software used</td>\n",
       "      <td>Unknown, Not collected, Not applicable, Not Sp...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acquistionSoftwareVersion</th>\n",
       "      <td>The version number of software used</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cleavageAgents</th>\n",
       "      <td>Name(s) of the enzyme used</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td>str</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>databaseName</th>\n",
       "      <td>The name of the search database (nr, SwissProt...</td>\n",
       "      <td>MassBank of North America (MoNA), NIST17, Metl...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001013</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>databaseSource</th>\n",
       "      <td>The name of the organization, project, or labo...</td>\n",
       "      <td>OtherDatabaseSource, Unknown, Not collected, N...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001012</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>databaseWeblink</th>\n",
       "      <td>An internet address that may provide details o...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DSLWG,Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>digestionMethod</th>\n",
       "      <td>The type of digestion - in-solution, in-gel, e...</td>\n",
       "      <td>in-solution, in-gel, OtherDigestionMethod, Unk...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DSLWG,Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fractionIdentifier</th>\n",
       "      <td>Identifier string that describes the sample fr...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MS Ontology,Sage Bionetworks,The Proteomics Sa...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fragmentMassTolerance</th>\n",
       "      <td>Units of measure are Da or ppm (ex. \"0.6 Da\" o...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>modificationParameters</th>\n",
       "      <td>Modification parameters for the search engine ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/termsi...</td>\n",
       "      <td>str</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>msAnalyteType</th>\n",
       "      <td>the type of biospecimen subjected to analysis</td>\n",
       "      <td>Protein, metabolite, Lipid, Other, Unknown, No...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C156434</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>msAnalyzerTypeMS1</th>\n",
       "      <td>Describes the mass spectrometer analyzer used ...</td>\n",
       "      <td>LTQ Orbitrap XL, Orbitrap Fusion, Orbitrap Fus...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Proposed minimum metadata relative to mass spe...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>msAnalyzerTypeMS2</th>\n",
       "      <td>Describes the mass spectrometer analyzer used ...</td>\n",
       "      <td>LTQ Orbitrap XL, Orbitrap Fusion, Orbitrap Fus...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Proposed minimum metadata relative to mass spe...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>msAssayTechnique</th>\n",
       "      <td>Specified the mass spectrometer technique used</td>\n",
       "      <td>DDA-MS, DIA-MS, FIA-MSMS, HPLC-MSMS, LC-MS, LC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DSLWG,Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>msInstrumentModel</th>\n",
       "      <td>The model of the mass spectrometer  used</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Proposed minimum metadata relative to mass spe...</td>\n",
       "      <td>str</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>msTarget</th>\n",
       "      <td>specifies whether or not a specific molecule(s...</td>\n",
       "      <td>TargetedMsTarget, UnTargetedMsTarget, Unknown,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>precursorMassTolerance</th>\n",
       "      <td>Units of measure are Da or ppm (ex. \"0.6 Da\" o...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protoemicsAssayType</th>\n",
       "      <td>Type of peptide quantification used (ex. Label...</td>\n",
       "      <td>Label Free, TMT, ITRAQ, SILAC, Targeted Absolu...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DSLWG,Sage Bionetworks,https://www.ebi.ac.uk/o...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>samplePrepProtocol</th>\n",
       "      <td>An internet address that may provide more deta...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleType</th>\n",
       "      <td>The type of sample collected</td>\n",
       "      <td>Amniotic Fluid, Appendix, B cell, Basophils, B...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ImmPort,Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,Microbiome,RNAseq,Whole Gen...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>spectrometerFrequency</th>\n",
       "      <td>The frequency at which a spectrometer causes h...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dataFile</th>\n",
       "      <td>The data file provides the name of the raw fil...</td>\n",
       "      <td>Unknown, Not collected, Not applicable, Not Sp...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Proteomics Sample Metadata</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specimenID</th>\n",
       "      <td>Identifying string linked to a particular samp...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Biospecimen human,Biospecimen nonHuman,Metabol...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                          Description  \\\n",
       "Attribute                                                                               \n",
       "platformLocation                    The name of the laboratory, facility, vendor, ...   \n",
       "technologyPlatformVersion           The specific version (application, manufacture...   \n",
       "specifyDatabaseName                       Value is determined by the data contributor   \n",
       "specifyDatabaseSource                     Value is determined by the data contributor   \n",
       "specifyDigestionMethod                    Value is determined by the data contributor   \n",
       "specifyMeasurementTechnique               Value is determined by the data contributor   \n",
       "specifyMSAnalyteType                      Value is determined by the data contributor   \n",
       "specifyMSAnalyzerTypeMS1                  Value is determined by the data contributor   \n",
       "specifyMSAnalyzerTypeMS2                  Value is determined by the data contributor   \n",
       "specifyMSAssayTechnique                   Value is determined by the data contributor   \n",
       "specifyMSInstrumentModel                  Value is determined by the data contributor   \n",
       "specifyOtherCleavageAgent                 Value is determined by the data contributor   \n",
       "specifyOtherModificationParameters        Value is determined by the data contributor   \n",
       "specifyPlatformLocation                   Value is determined by the data contributor   \n",
       "specifyPlatformVersion                    Value is determined by the data contributor   \n",
       "specifyProteomicsAssayType                Value is determined by the data contributor   \n",
       "specifySampleType                         Value is determined by the data contributor   \n",
       "acquisitionMode                     The specific aspect of a mass spectrometer met...   \n",
       "acquisitionSoftware                         The name of the acquisition software used   \n",
       "acquistionSoftwareVersion                         The version number of software used   \n",
       "cleavageAgents                                             Name(s) of the enzyme used   \n",
       "databaseName                        The name of the search database (nr, SwissProt...   \n",
       "databaseSource                      The name of the organization, project, or labo...   \n",
       "databaseWeblink                     An internet address that may provide details o...   \n",
       "digestionMethod                     The type of digestion - in-solution, in-gel, e...   \n",
       "fractionIdentifier                  Identifier string that describes the sample fr...   \n",
       "fragmentMassTolerance               Units of measure are Da or ppm (ex. \"0.6 Da\" o...   \n",
       "modificationParameters              Modification parameters for the search engine ...   \n",
       "msAnalyteType                           the type of biospecimen subjected to analysis   \n",
       "msAnalyzerTypeMS1                   Describes the mass spectrometer analyzer used ...   \n",
       "msAnalyzerTypeMS2                   Describes the mass spectrometer analyzer used ...   \n",
       "msAssayTechnique                       Specified the mass spectrometer technique used   \n",
       "msInstrumentModel                            The model of the mass spectrometer  used   \n",
       "msTarget                            specifies whether or not a specific molecule(s...   \n",
       "precursorMassTolerance              Units of measure are Da or ppm (ex. \"0.6 Da\" o...   \n",
       "protoemicsAssayType                 Type of peptide quantification used (ex. Label...   \n",
       "samplePrepProtocol                  An internet address that may provide more deta...   \n",
       "sampleType                                               The type of sample collected   \n",
       "spectrometerFrequency               The frequency at which a spectrometer causes h...   \n",
       "dataFile                            The data file provides the name of the raw fil...   \n",
       "specimenID                          Identifying string linked to a particular samp...   \n",
       "\n",
       "                                                                         Valid Values  \\\n",
       "Attribute                                                                               \n",
       "platformLocation                    OtherPlatformLocation, Unknown, Not collected,...   \n",
       "technologyPlatformVersion           10x Visium Spatial Gene Expression, Affymetrix...   \n",
       "specifyDatabaseName                                                               NaN   \n",
       "specifyDatabaseSource                                                             NaN   \n",
       "specifyDigestionMethod                                                            NaN   \n",
       "specifyMeasurementTechnique                                                       NaN   \n",
       "specifyMSAnalyteType                                                              NaN   \n",
       "specifyMSAnalyzerTypeMS1                                                          NaN   \n",
       "specifyMSAnalyzerTypeMS2                                                          NaN   \n",
       "specifyMSAssayTechnique                                                           NaN   \n",
       "specifyMSInstrumentModel                                                          NaN   \n",
       "specifyOtherCleavageAgent                                                         NaN   \n",
       "specifyOtherModificationParameters                                                NaN   \n",
       "specifyPlatformLocation                                                           NaN   \n",
       "specifyPlatformVersion                                                            NaN   \n",
       "specifyProteomicsAssayType                                                        NaN   \n",
       "specifySampleType                                                                 NaN   \n",
       "acquisitionMode                     Unknown, Not collected, Not applicable, Not Sp...   \n",
       "acquisitionSoftware                 Unknown, Not collected, Not applicable, Not Sp...   \n",
       "acquistionSoftwareVersion                                                         NaN   \n",
       "cleavageAgents                                                                    NaN   \n",
       "databaseName                        MassBank of North America (MoNA), NIST17, Metl...   \n",
       "databaseSource                      OtherDatabaseSource, Unknown, Not collected, N...   \n",
       "databaseWeblink                                                                   NaN   \n",
       "digestionMethod                     in-solution, in-gel, OtherDigestionMethod, Unk...   \n",
       "fractionIdentifier                                                                NaN   \n",
       "fragmentMassTolerance                                                             NaN   \n",
       "modificationParameters                                                            NaN   \n",
       "msAnalyteType                       Protein, metabolite, Lipid, Other, Unknown, No...   \n",
       "msAnalyzerTypeMS1                   LTQ Orbitrap XL, Orbitrap Fusion, Orbitrap Fus...   \n",
       "msAnalyzerTypeMS2                   LTQ Orbitrap XL, Orbitrap Fusion, Orbitrap Fus...   \n",
       "msAssayTechnique                    DDA-MS, DIA-MS, FIA-MSMS, HPLC-MSMS, LC-MS, LC...   \n",
       "msInstrumentModel                                                                 NaN   \n",
       "msTarget                            TargetedMsTarget, UnTargetedMsTarget, Unknown,...   \n",
       "precursorMassTolerance                                                            NaN   \n",
       "protoemicsAssayType                 Label Free, TMT, ITRAQ, SILAC, Targeted Absolu...   \n",
       "samplePrepProtocol                                                                NaN   \n",
       "sampleType                          Amniotic Fluid, Appendix, B cell, Basophils, B...   \n",
       "spectrometerFrequency                                                             NaN   \n",
       "dataFile                            Unknown, Not collected, Not applicable, Not Sp...   \n",
       "specimenID                                                                        NaN   \n",
       "\n",
       "                                   DependsOn  Properties Required  \\\n",
       "Attribute                                                           \n",
       "platformLocation                         NaN         NaN     True   \n",
       "technologyPlatformVersion                NaN         NaN     True   \n",
       "specifyDatabaseName                      NaN         NaN    False   \n",
       "specifyDatabaseSource                    NaN         NaN    False   \n",
       "specifyDigestionMethod                   NaN         NaN    False   \n",
       "specifyMeasurementTechnique              NaN         NaN    False   \n",
       "specifyMSAnalyteType                     NaN         NaN    False   \n",
       "specifyMSAnalyzerTypeMS1                 NaN         NaN    False   \n",
       "specifyMSAnalyzerTypeMS2                 NaN         NaN    False   \n",
       "specifyMSAssayTechnique                  NaN         NaN    False   \n",
       "specifyMSInstrumentModel                 NaN         NaN    False   \n",
       "specifyOtherCleavageAgent                NaN         NaN    False   \n",
       "specifyOtherModificationParameters       NaN         NaN    False   \n",
       "specifyPlatformLocation                  NaN         NaN    False   \n",
       "specifyPlatformVersion                   NaN         NaN    False   \n",
       "specifyProteomicsAssayType               NaN         NaN    False   \n",
       "specifySampleType                        NaN         NaN    False   \n",
       "acquisitionMode                          NaN         NaN     True   \n",
       "acquisitionSoftware                      NaN         NaN     True   \n",
       "acquistionSoftwareVersion                NaN         NaN     True   \n",
       "cleavageAgents                           NaN         NaN     True   \n",
       "databaseName                             NaN         NaN     True   \n",
       "databaseSource                           NaN         NaN    False   \n",
       "databaseWeblink                          NaN         NaN    False   \n",
       "digestionMethod                          NaN         NaN     True   \n",
       "fractionIdentifier                       NaN         NaN     True   \n",
       "fragmentMassTolerance                    NaN         NaN     True   \n",
       "modificationParameters                   NaN         NaN     True   \n",
       "msAnalyteType                            NaN         NaN    False   \n",
       "msAnalyzerTypeMS1                        NaN         NaN     True   \n",
       "msAnalyzerTypeMS2                        NaN         NaN     True   \n",
       "msAssayTechnique                         NaN         NaN     True   \n",
       "msInstrumentModel                        NaN         NaN     True   \n",
       "msTarget                                 NaN         NaN     True   \n",
       "precursorMassTolerance                   NaN         NaN     True   \n",
       "protoemicsAssayType                      NaN         NaN     True   \n",
       "samplePrepProtocol                       NaN         NaN    False   \n",
       "sampleType                               NaN         NaN     True   \n",
       "spectrometerFrequency                    NaN         NaN     True   \n",
       "dataFile                                 NaN         NaN     True   \n",
       "specimenID                               NaN         NaN     True   \n",
       "\n",
       "                                            Parent  DependsOn Component  \\\n",
       "Attribute                                                                 \n",
       "platformLocation                    ManifestColumn                  NaN   \n",
       "technologyPlatformVersion           ManifestColumn                  NaN   \n",
       "specifyDatabaseName                 ManifestColumn                  NaN   \n",
       "specifyDatabaseSource               ManifestColumn                  NaN   \n",
       "specifyDigestionMethod              ManifestColumn                  NaN   \n",
       "specifyMeasurementTechnique         ManifestColumn                  NaN   \n",
       "specifyMSAnalyteType                ManifestColumn                  NaN   \n",
       "specifyMSAnalyzerTypeMS1            ManifestColumn                  NaN   \n",
       "specifyMSAnalyzerTypeMS2            ManifestColumn                  NaN   \n",
       "specifyMSAssayTechnique             ManifestColumn                  NaN   \n",
       "specifyMSInstrumentModel            ManifestColumn                  NaN   \n",
       "specifyOtherCleavageAgent           ManifestColumn                  NaN   \n",
       "specifyOtherModificationParameters  ManifestColumn                  NaN   \n",
       "specifyPlatformLocation             ManifestColumn                  NaN   \n",
       "specifyPlatformVersion              ManifestColumn                  NaN   \n",
       "specifyProteomicsAssayType          ManifestColumn                  NaN   \n",
       "specifySampleType                   ManifestColumn                  NaN   \n",
       "acquisitionMode                     ManifestColumn                  NaN   \n",
       "acquisitionSoftware                 ManifestColumn                  NaN   \n",
       "acquistionSoftwareVersion           ManifestColumn                  NaN   \n",
       "cleavageAgents                      ManifestColumn                  NaN   \n",
       "databaseName                        ManifestColumn                  NaN   \n",
       "databaseSource                      ManifestColumn                  NaN   \n",
       "databaseWeblink                     ManifestColumn                  NaN   \n",
       "digestionMethod                     ManifestColumn                  NaN   \n",
       "fractionIdentifier                  ManifestColumn                  NaN   \n",
       "fragmentMassTolerance               ManifestColumn                  NaN   \n",
       "modificationParameters              ManifestColumn                  NaN   \n",
       "msAnalyteType                       ManifestColumn                  NaN   \n",
       "msAnalyzerTypeMS1                   ManifestColumn                  NaN   \n",
       "msAnalyzerTypeMS2                   ManifestColumn                  NaN   \n",
       "msAssayTechnique                    ManifestColumn                  NaN   \n",
       "msInstrumentModel                   ManifestColumn                  NaN   \n",
       "msTarget                            ManifestColumn                  NaN   \n",
       "precursorMassTolerance              ManifestColumn                  NaN   \n",
       "protoemicsAssayType                 ManifestColumn                  NaN   \n",
       "samplePrepProtocol                  ManifestColumn                  NaN   \n",
       "sampleType                          ManifestColumn                  NaN   \n",
       "spectrometerFrequency               ManifestColumn                  NaN   \n",
       "dataFile                            ManifestColumn                  NaN   \n",
       "specimenID                          ManifestColumn                  NaN   \n",
       "\n",
       "                                                                               Source  \\\n",
       "Attribute                                                                               \n",
       "platformLocation                                                     Sage Bionetworks   \n",
       "technologyPlatformVersion                                            Sage Bionetworks   \n",
       "specifyDatabaseName                                                  Sage Bionetworks   \n",
       "specifyDatabaseSource                                                Sage Bionetworks   \n",
       "specifyDigestionMethod                                               Sage Bionetworks   \n",
       "specifyMeasurementTechnique                                          Sage Bionetworks   \n",
       "specifyMSAnalyteType                                                 Sage Bionetworks   \n",
       "specifyMSAnalyzerTypeMS1                                             Sage Bionetworks   \n",
       "specifyMSAnalyzerTypeMS2                                             Sage Bionetworks   \n",
       "specifyMSAssayTechnique                                              Sage Bionetworks   \n",
       "specifyMSInstrumentModel                                             Sage Bionetworks   \n",
       "specifyOtherCleavageAgent                                            Sage Bionetworks   \n",
       "specifyOtherModificationParameters                                   Sage Bionetworks   \n",
       "specifyPlatformLocation                                              Sage Bionetworks   \n",
       "specifyPlatformVersion                                               Sage Bionetworks   \n",
       "specifyProteomicsAssayType                                           Sage Bionetworks   \n",
       "specifySampleType                                                    Sage Bionetworks   \n",
       "acquisitionMode                     https://www.ebi.ac.uk/ols/ontologies/ms/termsi...   \n",
       "acquisitionSoftware                         http://purl.obolibrary.org/obo/MS_1001455   \n",
       "acquistionSoftwareVersion                                            Sage Bionetworks   \n",
       "cleavageAgents                                             Proteomics Sample Metadata   \n",
       "databaseName                                http://purl.obolibrary.org/obo/MS_1001013   \n",
       "databaseSource                              http://purl.obolibrary.org/obo/MS_1001012   \n",
       "databaseWeblink                                                DSLWG,Sage Bionetworks   \n",
       "digestionMethod                                                DSLWG,Sage Bionetworks   \n",
       "fractionIdentifier                  MS Ontology,Sage Bionetworks,The Proteomics Sa...   \n",
       "fragmentMassTolerance                                      Proteomics Sample Metadata   \n",
       "modificationParameters              https://www.ebi.ac.uk/ols/ontologies/ms/termsi...   \n",
       "msAnalyteType                             http://purl.obolibrary.org/obo/NCIT_C156434   \n",
       "msAnalyzerTypeMS1                   Proposed minimum metadata relative to mass spe...   \n",
       "msAnalyzerTypeMS2                   Proposed minimum metadata relative to mass spe...   \n",
       "msAssayTechnique                                               DSLWG,Sage Bionetworks   \n",
       "msInstrumentModel                   Proposed minimum metadata relative to mass spe...   \n",
       "msTarget                                                             Sage Bionetworks   \n",
       "precursorMassTolerance                                     Proteomics Sample Metadata   \n",
       "protoemicsAssayType                 DSLWG,Sage Bionetworks,https://www.ebi.ac.uk/o...   \n",
       "samplePrepProtocol                                                   Sage Bionetworks   \n",
       "sampleType                                                   ImmPort,Sage Bionetworks   \n",
       "spectrometerFrequency                                                Sage Bionetworks   \n",
       "dataFile                                                   Proteomics Sample Metadata   \n",
       "specimenID                                                           Sage Bionetworks   \n",
       "\n",
       "                                   Validation Rules       Module Ontology  \\\n",
       "Attribute                                                                   \n",
       "platformLocation                                NaN        Assay      NaN   \n",
       "technologyPlatformVersion                       NaN        Assay      NaN   \n",
       "specifyDatabaseName                             NaN        Other      NaN   \n",
       "specifyDatabaseSource                           NaN        Other      NaN   \n",
       "specifyDigestionMethod                          NaN        Other      NaN   \n",
       "specifyMeasurementTechnique                     NaN        Other      NaN   \n",
       "specifyMSAnalyteType                            NaN        Other      NaN   \n",
       "specifyMSAnalyzerTypeMS1                        NaN        Other      NaN   \n",
       "specifyMSAnalyzerTypeMS2                        NaN        Other      NaN   \n",
       "specifyMSAssayTechnique                         NaN        Other      NaN   \n",
       "specifyMSInstrumentModel                        NaN        Other      NaN   \n",
       "specifyOtherCleavageAgent                       NaN        Other      NaN   \n",
       "specifyOtherModificationParameters              NaN        Other      NaN   \n",
       "specifyPlatformLocation                         NaN        Other      NaN   \n",
       "specifyPlatformVersion                          NaN        Other      NaN   \n",
       "specifyProteomicsAssayType                      NaN        Other      NaN   \n",
       "specifySampleType                               NaN        Other      NaN   \n",
       "acquisitionMode                                 NaN  Unspecified      NaN   \n",
       "acquisitionSoftware                             NaN  Unspecified      NaN   \n",
       "acquistionSoftwareVersion                       NaN  Unspecified      NaN   \n",
       "cleavageAgents                                  str  Unspecified      NaN   \n",
       "databaseName                                    NaN  Unspecified      NaN   \n",
       "databaseSource                                  NaN  Unspecified      NaN   \n",
       "databaseWeblink                                 NaN  Unspecified      NaN   \n",
       "digestionMethod                                 NaN  Unspecified      NaN   \n",
       "fractionIdentifier                              NaN  Unspecified      NaN   \n",
       "fragmentMassTolerance                           NaN  Unspecified      NaN   \n",
       "modificationParameters                          str  Unspecified      NaN   \n",
       "msAnalyteType                                   NaN  Unspecified      NaN   \n",
       "msAnalyzerTypeMS1                               NaN  Unspecified      NaN   \n",
       "msAnalyzerTypeMS2                               NaN  Unspecified      NaN   \n",
       "msAssayTechnique                                NaN  Unspecified      NaN   \n",
       "msInstrumentModel                               str  Unspecified      NaN   \n",
       "msTarget                                        NaN  Unspecified      NaN   \n",
       "precursorMassTolerance                          NaN  Unspecified      NaN   \n",
       "protoemicsAssayType                             NaN  Unspecified      NaN   \n",
       "samplePrepProtocol                              NaN  Unspecified      NaN   \n",
       "sampleType                                      NaN  Unspecified      NaN   \n",
       "spectrometerFrequency                           NaN  Unspecified      NaN   \n",
       "dataFile                                        NaN     Metadata      NaN   \n",
       "specimenID                                      NaN     Metadata      NaN   \n",
       "\n",
       "                                                                               UsedIn  \\\n",
       "Attribute                                                                               \n",
       "platformLocation                    Metabolomics Human,Microbiome,RNAseq,Whole Gen...   \n",
       "technologyPlatformVersion           Metabolomics Human,Microbiome,RNAseq,Whole Gen...   \n",
       "specifyDatabaseName                                     Metabolomics Human,proteomics   \n",
       "specifyDatabaseSource                                   Metabolomics Human,proteomics   \n",
       "specifyDigestionMethod                                                     proteomics   \n",
       "specifyMeasurementTechnique         Metabolomics Human,Microbiome,RNAseq,Whole Gen...   \n",
       "specifyMSAnalyteType                                    Metabolomics Human,proteomics   \n",
       "specifyMSAnalyzerTypeMS1                                Metabolomics Human,proteomics   \n",
       "specifyMSAnalyzerTypeMS2                                Metabolomics Human,proteomics   \n",
       "specifyMSAssayTechnique                                 Metabolomics Human,proteomics   \n",
       "specifyMSInstrumentModel                                Metabolomics Human,proteomics   \n",
       "specifyOtherCleavageAgent                                                  proteomics   \n",
       "specifyOtherModificationParameters                                         proteomics   \n",
       "specifyPlatformLocation             Metabolomics Human,Microbiome,RNAseq,Whole Gen...   \n",
       "specifyPlatformVersion              Metabolomics Human,Microbiome,RNAseq,Whole Gen...   \n",
       "specifyProteomicsAssayType                                                 proteomics   \n",
       "specifySampleType                   Metabolomics Human,Microbiome,RNAseq,Whole Gen...   \n",
       "acquisitionMode                                         Metabolomics Human,proteomics   \n",
       "acquisitionSoftware                                     Metabolomics Human,proteomics   \n",
       "acquistionSoftwareVersion                               Metabolomics Human,proteomics   \n",
       "cleavageAgents                                                             proteomics   \n",
       "databaseName                                            Metabolomics Human,proteomics   \n",
       "databaseSource                                          Metabolomics Human,proteomics   \n",
       "databaseWeblink                                         Metabolomics Human,proteomics   \n",
       "digestionMethod                                                            proteomics   \n",
       "fractionIdentifier                                                         proteomics   \n",
       "fragmentMassTolerance                                                      proteomics   \n",
       "modificationParameters                                                     proteomics   \n",
       "msAnalyteType                                           Metabolomics Human,proteomics   \n",
       "msAnalyzerTypeMS1                                       Metabolomics Human,proteomics   \n",
       "msAnalyzerTypeMS2                                       Metabolomics Human,proteomics   \n",
       "msAssayTechnique                                        Metabolomics Human,proteomics   \n",
       "msInstrumentModel                                       Metabolomics Human,proteomics   \n",
       "msTarget                                                Metabolomics Human,proteomics   \n",
       "precursorMassTolerance                                                     proteomics   \n",
       "protoemicsAssayType                                                        proteomics   \n",
       "samplePrepProtocol                                      Metabolomics Human,proteomics   \n",
       "sampleType                          Metabolomics Human,Microbiome,RNAseq,Whole Gen...   \n",
       "spectrometerFrequency                                   Metabolomics Human,proteomics   \n",
       "dataFile                                                                   proteomics   \n",
       "specimenID                          Biospecimen human,Biospecimen nonHuman,Metabol...   \n",
       "\n",
       "                                   columnType  \n",
       "Attribute                                      \n",
       "platformLocation                          NaN  \n",
       "technologyPlatformVersion                 NaN  \n",
       "specifyDatabaseName                    String  \n",
       "specifyDatabaseSource                  String  \n",
       "specifyDigestionMethod                 String  \n",
       "specifyMeasurementTechnique            String  \n",
       "specifyMSAnalyteType                   String  \n",
       "specifyMSAnalyzerTypeMS1               String  \n",
       "specifyMSAnalyzerTypeMS2               String  \n",
       "specifyMSAssayTechnique                String  \n",
       "specifyMSInstrumentModel               String  \n",
       "specifyOtherCleavageAgent              String  \n",
       "specifyOtherModificationParameters     String  \n",
       "specifyPlatformLocation                String  \n",
       "specifyPlatformVersion                 String  \n",
       "specifyProteomicsAssayType             String  \n",
       "specifySampleType                      String  \n",
       "acquisitionMode                           NaN  \n",
       "acquisitionSoftware                       NaN  \n",
       "acquistionSoftwareVersion                 NaN  \n",
       "cleavageAgents                            NaN  \n",
       "databaseName                              NaN  \n",
       "databaseSource                            NaN  \n",
       "databaseWeblink                           NaN  \n",
       "digestionMethod                           NaN  \n",
       "fractionIdentifier                        NaN  \n",
       "fragmentMassTolerance                     NaN  \n",
       "modificationParameters                    NaN  \n",
       "msAnalyteType                             NaN  \n",
       "msAnalyzerTypeMS1                         NaN  \n",
       "msAnalyzerTypeMS2                         NaN  \n",
       "msAssayTechnique                          NaN  \n",
       "msInstrumentModel                         NaN  \n",
       "msTarget                                  NaN  \n",
       "precursorMassTolerance                    NaN  \n",
       "protoemicsAssayType                       NaN  \n",
       "samplePrepProtocol                        NaN  \n",
       "sampleType                                NaN  \n",
       "spectrometerFrequency                     NaN  \n",
       "dataFile                                  NaN  \n",
       "specimenID                                NaN  "
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm.loc[dm['UsedIn'].str.contains('proteomics', na = False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description_x</th>\n",
       "      <th>Valid Values_x</th>\n",
       "      <th>DependsOn_x</th>\n",
       "      <th>Properties_x</th>\n",
       "      <th>Required_x</th>\n",
       "      <th>Parent_x</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Module_x</th>\n",
       "      <th>...</th>\n",
       "      <th>Required_y</th>\n",
       "      <th>Module_y</th>\n",
       "      <th>columnType_y</th>\n",
       "      <th>Ontology_y</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Properties_y</th>\n",
       "      <th>DependsOn_y</th>\n",
       "      <th>Parent_y</th>\n",
       "      <th>OtherValue</th>\n",
       "      <th>UsedIn_y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Attribute</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>platformLocation</th>\n",
       "      <td>The name of the laboratory, facility, vendor, ...</td>\n",
       "      <td>OtherPlatformLocation, Unknown, Not collected,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Assay</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>technologyPlatformVersion</th>\n",
       "      <td>The specific version (application, manufacture...</td>\n",
       "      <td>10x Visium Spatial Gene Expression, Affymetrix...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Assay</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C45378</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyDatabaseName</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyDatabaseSource</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyDigestionMethod</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OtherDatabaseName</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>specifyDatabaseName</td>\n",
       "      <td>databaseName</td>\n",
       "      <td>other</td>\n",
       "      <td>proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OtherDatabaseSource</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>specifyDatabaseSource</td>\n",
       "      <td>databaseSource</td>\n",
       "      <td>other</td>\n",
       "      <td>proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OtherDataFile</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>specifyOtherDataFile</td>\n",
       "      <td>dataFile</td>\n",
       "      <td>other</td>\n",
       "      <td>proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OtherModificationParameters</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>specifyOtherModificationParameters</td>\n",
       "      <td>modificationParameters</td>\n",
       "      <td>other</td>\n",
       "      <td>proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OtherCleavageAgents</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Other</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>specifyOtherCleavageAgent</td>\n",
       "      <td>cleavageAgents</td>\n",
       "      <td>other</td>\n",
       "      <td>proteomics</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  Description_x  \\\n",
       "Attribute                                                                         \n",
       "platformLocation              The name of the laboratory, facility, vendor, ...   \n",
       "technologyPlatformVersion     The specific version (application, manufacture...   \n",
       "specifyDatabaseName                 Value is determined by the data contributor   \n",
       "specifyDatabaseSource               Value is determined by the data contributor   \n",
       "specifyDigestionMethod              Value is determined by the data contributor   \n",
       "...                                                                         ...   \n",
       "OtherDatabaseName                                                           NaN   \n",
       "OtherDatabaseSource                                                         NaN   \n",
       "OtherDataFile                                                               NaN   \n",
       "OtherModificationParameters                                                 NaN   \n",
       "OtherCleavageAgents                                                         NaN   \n",
       "\n",
       "                                                                 Valid Values_x  \\\n",
       "Attribute                                                                         \n",
       "platformLocation              OtherPlatformLocation, Unknown, Not collected,...   \n",
       "technologyPlatformVersion     10x Visium Spatial Gene Expression, Affymetrix...   \n",
       "specifyDatabaseName                                                         NaN   \n",
       "specifyDatabaseSource                                                       NaN   \n",
       "specifyDigestionMethod                                                      NaN   \n",
       "...                                                                         ...   \n",
       "OtherDatabaseName                                                           NaN   \n",
       "OtherDatabaseSource                                                         NaN   \n",
       "OtherDataFile                                                               NaN   \n",
       "OtherModificationParameters                                                 NaN   \n",
       "OtherCleavageAgents                                                         NaN   \n",
       "\n",
       "                             DependsOn_x  Properties_x Required_x  \\\n",
       "Attribute                                                           \n",
       "platformLocation                     NaN           NaN       True   \n",
       "technologyPlatformVersion            NaN           NaN       True   \n",
       "specifyDatabaseName                  NaN           NaN      False   \n",
       "specifyDatabaseSource                NaN           NaN      False   \n",
       "specifyDigestionMethod               NaN           NaN      False   \n",
       "...                                  ...           ...        ...   \n",
       "OtherDatabaseName                    NaN           NaN        NaN   \n",
       "OtherDatabaseSource                  NaN           NaN        NaN   \n",
       "OtherDataFile                        NaN           NaN        NaN   \n",
       "OtherModificationParameters          NaN           NaN        NaN   \n",
       "OtherCleavageAgents                  NaN           NaN        NaN   \n",
       "\n",
       "                                    Parent_x  DependsOn Component  \\\n",
       "Attribute                                                           \n",
       "platformLocation              ManifestColumn                  NaN   \n",
       "technologyPlatformVersion     ManifestColumn                  NaN   \n",
       "specifyDatabaseName           ManifestColumn                  NaN   \n",
       "specifyDatabaseSource         ManifestColumn                  NaN   \n",
       "specifyDigestionMethod        ManifestColumn                  NaN   \n",
       "...                                      ...                  ...   \n",
       "OtherDatabaseName                        NaN                  NaN   \n",
       "OtherDatabaseSource                      NaN                  NaN   \n",
       "OtherDataFile                            NaN                  NaN   \n",
       "OtherModificationParameters              NaN                  NaN   \n",
       "OtherCleavageAgents                      NaN                  NaN   \n",
       "\n",
       "                                        Source Validation Rules Module_x  ...  \\\n",
       "Attribute                                                                 ...   \n",
       "platformLocation              Sage Bionetworks              NaN    Assay  ...   \n",
       "technologyPlatformVersion     Sage Bionetworks              NaN    Assay  ...   \n",
       "specifyDatabaseName           Sage Bionetworks              NaN    Other  ...   \n",
       "specifyDatabaseSource         Sage Bionetworks              NaN    Other  ...   \n",
       "specifyDigestionMethod        Sage Bionetworks              NaN    Other  ...   \n",
       "...                                        ...              ...      ...  ...   \n",
       "OtherDatabaseName                          NaN              NaN      NaN  ...   \n",
       "OtherDatabaseSource                        NaN              NaN      NaN  ...   \n",
       "OtherDataFile                              NaN              NaN      NaN  ...   \n",
       "OtherModificationParameters                NaN              NaN      NaN  ...   \n",
       "OtherCleavageAgents                        NaN              NaN      NaN  ...   \n",
       "\n",
       "                             Required_y  Module_y columnType_y  \\\n",
       "Attribute                                                        \n",
       "platformLocation                   True  Metadata       string   \n",
       "technologyPlatformVersion          True  Metadata       string   \n",
       "specifyDatabaseName                True     Other       string   \n",
       "specifyDatabaseSource              True     Other       string   \n",
       "specifyDigestionMethod             True     Other       string   \n",
       "...                                 ...       ...          ...   \n",
       "OtherDatabaseName                 False     Other       string   \n",
       "OtherDatabaseSource               False     Other       string   \n",
       "OtherDataFile                     False     Other       string   \n",
       "OtherModificationParameters       False     Other       string   \n",
       "OtherCleavageAgents               False     Other       string   \n",
       "\n",
       "                                                              Ontology_y  \\\n",
       "Attribute                                                                  \n",
       "platformLocation                                  Sage Bionetworks,DSLWG   \n",
       "technologyPlatformVersion     http://purl.obolibrary.org/obo/NCIT_C45378   \n",
       "specifyDatabaseName                                     Sage Bionetworks   \n",
       "specifyDatabaseSource                                   Sage Bionetworks   \n",
       "specifyDigestionMethod                                  Sage Bionetworks   \n",
       "...                                                                  ...   \n",
       "OtherDatabaseName                                       Sage Bionetworks   \n",
       "OtherDatabaseSource                                     Sage Bionetworks   \n",
       "OtherDataFile                                           Sage Bionetworks   \n",
       "OtherModificationParameters                             Sage Bionetworks   \n",
       "OtherCleavageAgents                                     Sage Bionetworks   \n",
       "\n",
       "                                                                          Notes  \\\n",
       "Attribute                                                                         \n",
       "platformLocation              A data contributor should be able to write in ...   \n",
       "technologyPlatformVersion     A data contributor should be able to write in ...   \n",
       "specifyDatabaseName           A data contributor should be able to write in ...   \n",
       "specifyDatabaseSource         A data contributor should be able to write in ...   \n",
       "specifyDigestionMethod        A data contributor should be able to write in ...   \n",
       "...                                                                         ...   \n",
       "OtherDatabaseName             A data contributor should be able to write in ...   \n",
       "OtherDatabaseSource           A data contributor should be able to write in ...   \n",
       "OtherDataFile                                                               NaN   \n",
       "OtherModificationParameters                                                 NaN   \n",
       "OtherCleavageAgents                                                         NaN   \n",
       "\n",
       "                                Properties_y  \\\n",
       "Attribute                                      \n",
       "platformLocation              ManifestColumn   \n",
       "technologyPlatformVersion     ManifestColumn   \n",
       "specifyDatabaseName           ManifestColumn   \n",
       "specifyDatabaseSource         ManifestColumn   \n",
       "specifyDigestionMethod        ManifestColumn   \n",
       "...                                      ...   \n",
       "OtherDatabaseName                 ValidValue   \n",
       "OtherDatabaseSource               ValidValue   \n",
       "OtherDataFile                     ValidValue   \n",
       "OtherModificationParameters       ValidValue   \n",
       "OtherCleavageAgents               ValidValue   \n",
       "\n",
       "                                                     DependsOn_y  \\\n",
       "Attribute                                                          \n",
       "platformLocation                                             NaN   \n",
       "technologyPlatformVersion                                    NaN   \n",
       "specifyDatabaseName                                          NaN   \n",
       "specifyDatabaseSource                                        NaN   \n",
       "specifyDigestionMethod                                       NaN   \n",
       "...                                                          ...   \n",
       "OtherDatabaseName                            specifyDatabaseName   \n",
       "OtherDatabaseSource                        specifyDatabaseSource   \n",
       "OtherDataFile                               specifyOtherDataFile   \n",
       "OtherModificationParameters   specifyOtherModificationParameters   \n",
       "OtherCleavageAgents                    specifyOtherCleavageAgent   \n",
       "\n",
       "                                            Parent_y OtherValue    UsedIn_y  \n",
       "Attribute                                                                    \n",
       "platformLocation                                 NaN        NaN  proteomics  \n",
       "technologyPlatformVersion                        NaN        NaN  proteomics  \n",
       "specifyDatabaseName                              NaN        NaN  proteomics  \n",
       "specifyDatabaseSource                            NaN        NaN  proteomics  \n",
       "specifyDigestionMethod                           NaN        NaN  proteomics  \n",
       "...                                              ...        ...         ...  \n",
       "OtherDatabaseName                       databaseName      other  proteomics  \n",
       "OtherDatabaseSource                   databaseSource      other  proteomics  \n",
       "OtherDataFile                               dataFile      other  proteomics  \n",
       "OtherModificationParameters   modificationParameters      other  proteomics  \n",
       "OtherCleavageAgents                   cleavageAgents      other  proteomics  \n",
       "\n",
       "[77 rows x 25 columns]"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm.loc[dm[\"UsedIn\"].str.contains(\"proteomics\", na=False)].merge(p_df_final, how = 'outer', on = 'Attribute')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine duplicated rows\n",
    "p_df_final = p_df_final.fillna(\"\").astype(str).groupby(level=0).apply(\n",
    "    lambda x: x.apply(lambda y: \",\".join(y))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_df_final.index.is_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update empty values with the new dataframe\n",
    "dm.update(p_df_final, overwrite=False, errors = 'ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(41, 13)"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm.loc[dm['UsedIn'].str.contains('proteomics', na = False)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(733, 15)"
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# join in the new template\n",
    "dm_final = pd.concat([dm, p_df_final.loc[~p_df_final.index.isin(dm.index)]])\n",
    "dm_final.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace any other values to leave open for contributors to add\n",
    "dm_final.loc[\"platformLocation\", \"Valid Values\"] = np.nan\n",
    "dm_final.loc[\"dataFile\", \"Valid Values\"] = np.nan\n",
    "dm_final.loc[\"acquisitionMode\", \"Valid Values\"] = np.nan\n",
    "dm_final.loc[\"acquisitionSoftware\", \"Valid Values\"] = np.nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Module</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>UsedIn</th>\n",
       "      <th>columnType</th>\n",
       "      <th>Notes</th>\n",
       "      <th>OtherValue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Attribute</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Label freeLabelQuantiation</th>\n",
       "      <td>When labelFreeQuantitation = `other`, add your...</td>\n",
       "      <td></td>\n",
       "      <td>labelFreeQuantitation</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>False</td>\n",
       "      <td>labelQuantiation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>Label Free</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OtherCleavageAgents</th>\n",
       "      <td>When CleavageAgent = `other`, add your custom ...</td>\n",
       "      <td></td>\n",
       "      <td>specifyOtherCleavageAgent</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>False</td>\n",
       "      <td>cleavageAgents</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OtherDataFile</th>\n",
       "      <td>When DataFile = `other`, add your custom value...</td>\n",
       "      <td></td>\n",
       "      <td>specifyOtherDataFile</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>False</td>\n",
       "      <td>dataFile</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td></td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OtherDatabaseName</th>\n",
       "      <td>When DatabaseName = `other`, add your custom v...</td>\n",
       "      <td></td>\n",
       "      <td>specifyDatabaseName</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>False</td>\n",
       "      <td>databaseName</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OtherDatabaseSource</th>\n",
       "      <td>When DatabaseSource = `other`, add your custom...</td>\n",
       "      <td></td>\n",
       "      <td>specifyDatabaseSource</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>False</td>\n",
       "      <td>databaseSource</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>string</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifyProteomicsAssayType</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specifySampleType</th>\n",
       "      <td>Value is determined by the data contributor</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>Metabolomics Human,Microbiome,RNAseq,Whole Gen...</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specimenID</th>\n",
       "      <td>Identifying string linked to a particular samp...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metadata</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>Biospecimen human,Biospecimen nonHuman,Metabol...</td>\n",
       "      <td>string</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>spectrometerFrequency</th>\n",
       "      <td>The frequency at which a spectrometer causes h...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>Sage Bionetworks,DSLWG</td>\n",
       "      <td>Metabolomics Human,proteomics</td>\n",
       "      <td>number</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>technologyPlatformVersion</th>\n",
       "      <td>The specific version (application, manufacture...</td>\n",
       "      <td>10x Visium Spatial Gene Expression, Affymetrix...</td>\n",
       "      <td></td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Assay</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C45378</td>\n",
       "      <td>Metabolomics Human,Microbiome,RNAseq,Whole Gen...</td>\n",
       "      <td>string</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                   Description  \\\n",
       "Attribute                                                                        \n",
       "Label freeLabelQuantiation   When labelFreeQuantitation = `other`, add your...   \n",
       "OtherCleavageAgents          When CleavageAgent = `other`, add your custom ...   \n",
       "OtherDataFile                When DataFile = `other`, add your custom value...   \n",
       "OtherDatabaseName            When DatabaseName = `other`, add your custom v...   \n",
       "OtherDatabaseSource          When DatabaseSource = `other`, add your custom...   \n",
       "...                                                                        ...   \n",
       "specifyProteomicsAssayType         Value is determined by the data contributor   \n",
       "specifySampleType                  Value is determined by the data contributor   \n",
       "specimenID                   Identifying string linked to a particular samp...   \n",
       "spectrometerFrequency        The frequency at which a spectrometer causes h...   \n",
       "technologyPlatformVersion    The specific version (application, manufacture...   \n",
       "\n",
       "                                                                  Valid Values  \\\n",
       "Attribute                                                                        \n",
       "Label freeLabelQuantiation                                                       \n",
       "OtherCleavageAgents                                                              \n",
       "OtherDataFile                                                                    \n",
       "OtherDatabaseName                                                                \n",
       "OtherDatabaseSource                                                              \n",
       "...                                                                        ...   \n",
       "specifyProteomicsAssayType                                                       \n",
       "specifySampleType                                                                \n",
       "specimenID                                                                       \n",
       "spectrometerFrequency                                                            \n",
       "technologyPlatformVersion    10x Visium Spatial Gene Expression, Affymetrix...   \n",
       "\n",
       "                                             DependsOn      Properties  \\\n",
       "Attribute                                                                \n",
       "Label freeLabelQuantiation       labelFreeQuantitation      ValidValue   \n",
       "OtherCleavageAgents          specifyOtherCleavageAgent      ValidValue   \n",
       "OtherDataFile                     specifyOtherDataFile      ValidValue   \n",
       "OtherDatabaseName                  specifyDatabaseName      ValidValue   \n",
       "OtherDatabaseSource              specifyDatabaseSource      ValidValue   \n",
       "...                                                ...             ...   \n",
       "specifyProteomicsAssayType                              ManifestColumn   \n",
       "specifySampleType                                       ManifestColumn   \n",
       "specimenID                                              ManifestColumn   \n",
       "spectrometerFrequency                                   ManifestColumn   \n",
       "technologyPlatformVersion                               ManifestColumn   \n",
       "\n",
       "                            Required            Parent DependsOn Component  \\\n",
       "Attribute                                                                    \n",
       "Label freeLabelQuantiation     False  labelQuantiation                 NaN   \n",
       "OtherCleavageAgents            False    cleavageAgents                 NaN   \n",
       "OtherDataFile                  False          dataFile                 NaN   \n",
       "OtherDatabaseName              False      databaseName                 NaN   \n",
       "OtherDatabaseSource            False    databaseSource                 NaN   \n",
       "...                              ...               ...                 ...   \n",
       "specifyProteomicsAssayType     False    ManifestColumn                 NaN   \n",
       "specifySampleType              False    ManifestColumn                 NaN   \n",
       "specimenID                      True    ManifestColumn                 NaN   \n",
       "spectrometerFrequency           True    ManifestColumn                 NaN   \n",
       "technologyPlatformVersion       True    ManifestColumn                 NaN   \n",
       "\n",
       "                                       Source Validation Rules       Module  \\\n",
       "Attribute                                                                     \n",
       "Label freeLabelQuantiation                NaN              NaN        Other   \n",
       "OtherCleavageAgents                       NaN              NaN        Other   \n",
       "OtherDataFile                             NaN              NaN        Other   \n",
       "OtherDatabaseName                         NaN              NaN        Other   \n",
       "OtherDatabaseSource                       NaN              NaN        Other   \n",
       "...                                       ...              ...          ...   \n",
       "specifyProteomicsAssayType   Sage Bionetworks              NaN        Other   \n",
       "specifySampleType            Sage Bionetworks              NaN        Other   \n",
       "specimenID                   Sage Bionetworks              NaN     Metadata   \n",
       "spectrometerFrequency        Sage Bionetworks              NaN  Unspecified   \n",
       "technologyPlatformVersion    Sage Bionetworks              NaN        Assay   \n",
       "\n",
       "                                                               Ontology  \\\n",
       "Attribute                                                                 \n",
       "Label freeLabelQuantiation                             Sage Bionetworks   \n",
       "OtherCleavageAgents                                    Sage Bionetworks   \n",
       "OtherDataFile                                          Sage Bionetworks   \n",
       "OtherDatabaseName                                      Sage Bionetworks   \n",
       "OtherDatabaseSource                                    Sage Bionetworks   \n",
       "...                                                                 ...   \n",
       "specifyProteomicsAssayType                             Sage Bionetworks   \n",
       "specifySampleType                                      Sage Bionetworks   \n",
       "specimenID                                             Sage Bionetworks   \n",
       "spectrometerFrequency                            Sage Bionetworks,DSLWG   \n",
       "technologyPlatformVersion    http://purl.obolibrary.org/obo/NCIT_C45378   \n",
       "\n",
       "                                                                        UsedIn  \\\n",
       "Attribute                                                                        \n",
       "Label freeLabelQuantiation                                          proteomics   \n",
       "OtherCleavageAgents                                                 proteomics   \n",
       "OtherDataFile                                                       proteomics   \n",
       "OtherDatabaseName                                                   proteomics   \n",
       "OtherDatabaseSource                                                 proteomics   \n",
       "...                                                                        ...   \n",
       "specifyProteomicsAssayType                                          proteomics   \n",
       "specifySampleType            Metabolomics Human,Microbiome,RNAseq,Whole Gen...   \n",
       "specimenID                   Biospecimen human,Biospecimen nonHuman,Metabol...   \n",
       "spectrometerFrequency                            Metabolomics Human,proteomics   \n",
       "technologyPlatformVersion    Metabolomics Human,Microbiome,RNAseq,Whole Gen...   \n",
       "\n",
       "                            columnType  \\\n",
       "Attribute                                \n",
       "Label freeLabelQuantiation      string   \n",
       "OtherCleavageAgents             string   \n",
       "OtherDataFile                   string   \n",
       "OtherDatabaseName               string   \n",
       "OtherDatabaseSource             string   \n",
       "...                                ...   \n",
       "specifyProteomicsAssayType      String   \n",
       "specifySampleType               String   \n",
       "specimenID                      string   \n",
       "spectrometerFrequency           number   \n",
       "technologyPlatformVersion       string   \n",
       "\n",
       "                                                                         Notes  \\\n",
       "Attribute                                                                        \n",
       "Label freeLabelQuantiation                                                       \n",
       "OtherCleavageAgents                                                              \n",
       "OtherDataFile                                                                    \n",
       "OtherDatabaseName            A data contributor should be able to write in ...   \n",
       "OtherDatabaseSource          A data contributor should be able to write in ...   \n",
       "...                                                                        ...   \n",
       "specifyProteomicsAssayType                                                 NaN   \n",
       "specifySampleType                                                          NaN   \n",
       "specimenID                                                                 NaN   \n",
       "spectrometerFrequency                                                      NaN   \n",
       "technologyPlatformVersion                                                  NaN   \n",
       "\n",
       "                             OtherValue  \n",
       "Attribute                                \n",
       "Label freeLabelQuantiation   Label Free  \n",
       "OtherCleavageAgents               other  \n",
       "OtherDataFile                     other  \n",
       "OtherDatabaseName                 other  \n",
       "OtherDatabaseSource               other  \n",
       "...                                 ...  \n",
       "specifyProteomicsAssayType          NaN  \n",
       "specifySampleType                   NaN  \n",
       "specimenID                          NaN  \n",
       "spectrometerFrequency               NaN  \n",
       "technologyPlatformVersion           NaN  \n",
       "\n",
       "[72 rows x 15 columns]"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm_final.loc[p_df_final.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Module</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>UsedIn</th>\n",
       "      <th>columnType</th>\n",
       "      <th>Notes</th>\n",
       "      <th>OtherValue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Attribute</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Description, Valid Values, DependsOn, Properties, Required, Parent, DependsOn Component, Source, Validation Rules, Module, Ontology, UsedIn, columnType, Notes, OtherValue]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# qa check\n",
    "with pd.option_context('display.max_rows', None):\n",
    "    display(dm_final[dm_final.index.duplicated(keep=False)].sort_index())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check there are unique attributes\n",
    "dm_final.index.is_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [],
   "source": [
    "# no longer need column\n",
    "dm_final = dm_final.drop(columns=[\"OtherValue\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write out data model\n",
    "dm_final.to_csv('../EL.data.model.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "elite-data-models-LjmgyG4K-py3.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
