{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Fix inconsistencies in parent column\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "from glob import glob\n",
    "import yaml\n",
    "import re\n",
    "import copy\n",
    "import json\n",
    "from thefuzz import fuzz\n",
    "\n",
    "# Custom package\n",
    "from utils import utils, synapse_login"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Schematic config:  ./config.yml \n",
      " CSV model:  EL.data.model.csv \n",
      " JSON LD Model:  EL.data.model.jsonld\n"
     ]
    }
   ],
   "source": [
    "with open(\"./local_configs/notebook_config.yaml\", \"r\") as f:\n",
    "    config = yaml.safe_load(f)\n",
    "\n",
    "# paths to import files\n",
    "schematic_config = config[\"paths\"][\"schematic\"]\n",
    "csv_model = config[\"file_names\"][\"csv_model\"]\n",
    "json_model = config[\"file_names\"][\"json_model\"]\n",
    "\n",
    "print(\n",
    "    \"Schematic config: \",\n",
    "    schematic_config,\n",
    "    \"\\n\",\n",
    "    \"CSV model: \",\n",
    "    csv_model,\n",
    "    \"\\n\",\n",
    "    \"JSON LD Model: \",\n",
    "    json_model,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hard Coded stuff\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "validation_coder = {\n",
    "    \"number\": \"regex search ([0-9]+\\.[0-9]*.?)|([0-9]+)\",\n",
    "    \"integer\": \"regex search ([0-9]+)\",\n",
    "    \"string\": \"\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_cols = [\n",
    "    \"Attribute\",\n",
    "    \"Description\",\n",
    "    \"Valid Values\",\n",
    "    \"DependsOn\",\n",
    "    \"Properties\",\n",
    "    \"Required\",\n",
    "    \"Template\",\n",
    "    \"Parent\",\n",
    "    \"DependsOn Component\",\n",
    "    \"Source\",\n",
    "    \"Validation Rules\",\n",
    "]\n",
    "\n",
    "keep_cols = [\n",
    "    \"Attribute\",\n",
    "    \"Description\",\n",
    "    \"Valid Values\",\n",
    "    \"Required\",\n",
    "    \"DependsOn\",\n",
    "    \"DependsOn Component\",\n",
    "    \"Properties\",\n",
    "    \"Validation Rules\",\n",
    "    \"Template\",\n",
    "    \"Parent\",\n",
    "    \"Source\",\n",
    "    \"Type\",\n",
    "    \"Ontology\",\n",
    "    \"multivalue\",\n",
    "]\n",
    "\n",
    "list_cols = [\n",
    "    \"UsedIn\",\n",
    "    \"DependsOn\",\n",
    "    \"Properties\",\n",
    "    \"Validation Rules\",\n",
    "    \"Template\",\n",
    "    \"Parent\",\n",
    "    \"Source\",\n",
    "    \"Type\",\n",
    "    \"Ontology\",\n",
    "    \"Required\",\n",
    "    \"multivalue\",\n",
    "]\n",
    "\n",
    "template_hard_coded_attrs = {\n",
    "    \"Attribute\": \"\",\n",
    "    \"Description\": \"\",\n",
    "    \"Valid Values\": \"\",\n",
    "    \"Required\": \"False\",\n",
    "    \"Validation Rules\": \"\",\n",
    "    \"Template\": \"\",\n",
    "    \"Parent\": \"\",\n",
    "    \"Source\": \"\",\n",
    "    \"Type\": \"\",\n",
    "    \"Ontology\": \"\",\n",
    "    \"UsedIn\": \"\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "# hard coded dictionary\n",
    "recoder_valid_values = {\n",
    "    re.compile(\"Not Specified\", flags=re.IGNORECASE): \"Not Specified\",\n",
    "    re.compile(\"(Other$)\", flags=re.IGNORECASE): \"Other\",\n",
    "    re.compile(\"lipid\", re.IGNORECASE): \"Lipid\",\n",
    "    re.compile(\"plasma\", re.IGNORECASE): \"Plasma\",\n",
    "    re.compile(\"protein\", re.IGNORECASE): \"Protein\",\n",
    "    re.compile(\"saliva\", re.IGNORECASE): \"Saliva\",\n",
    "    re.compile(\"serum\", re.IGNORECASE): \"Serum\",\n",
    "    re.compile(\"sputum\", re.IGNORECASE): \"Sputum\",\n",
    "    re.compile(\"urine\", re.IGNORECASE): \"Urine\",\n",
    "    re.compile(\n",
    "        \"^0x Visium Spatial Gene Expression\"\n",
    "    ): \"10x Visium Spatial Gene Expression\",\n",
    "    re.compile(\"falseFalseFALSEtrueTrueTRUE\t\"): \"True, False\",\n",
    "    re.compile(\"TRUE|TRUEDiagnosisStatus\", re.IGNORECASE): \"True\",\n",
    "    re.compile(\"FALSE|FASLSE\", re.IGNORECASE): \"False\",\n",
    "    re.compile(\"UnknownNot collected\"): \"Unknown, Not collected\",\n",
    "    re.compile(r\"\\u200b\\u200b\"): \"\",\n",
    "    re.compile(\n",
    "        \"The Health,Aging,and Body Composition Study \\(HealthABC\\)\"\n",
    "    ): \"The Health and Aging and Body Composition Study (HealthABC)\",\n",
    "    re.compile(\"Not Hispanic or latinoEthnicity\"): \"Not Hispanic or latino\",\n",
    "    re.compile(\"Hispanic or latinoEthnicity\"): \"Hispanic or latino\",\n",
    "    re.compile(\n",
    "        \"HPO, MONDO, MAXO codes or labels \\(not listed for purposes of this RFC\\)\"\n",
    "    ): \"HPO and MONDO and MAXO codes or labels (not listed for purposes of this RFC)\",\n",
    "    re.escape(\n",
    "        r\"Possible values are listed under the instrument model term.OtherMsInstrumentModel\"\n",
    "    ): \"OtherMsInstrumentModel\",\n",
    "    re.compile(\n",
    "        r\"Possible values are listed under the cleavage agent nameOtherCleavageAgents\"\n",
    "    ): \"OtherCleavageAgents\",\n",
    "    re.compile(r\"Possible values are listed under modification parameters\"): \"\",\n",
    "    re.compile(\"Uknown\"): \"Unknown\",\n",
    "    re.compile(\"OtherControlType\", re.IGNORECASE): \"OtherControlType\",\n",
    "    re.compile(\"OtherMsAnalyteType\", re.IGNORECASE): \"OtherMsAnalyteType\",\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean list columns into single string\n",
    "def join_strings(string):\n",
    "    try:\n",
    "        return \",\".join(string)\n",
    "    except:\n",
    "        return \"\"\n",
    "\n",
    "\n",
    "def clean_list(string):\n",
    "    \"\"\"Takes a list represented as a string and returns only unique values found\n",
    "\n",
    "    Args:\n",
    "        string (str): list represented as string\n",
    "\n",
    "    Returns:\n",
    "        string: list as string of unique values\n",
    "    \"\"\"\n",
    "\n",
    "    new_list = string.split(\",\")\n",
    "    new_list = [n.strip() for n in new_list if n != \"nan\"]\n",
    "    new_list = \",\".join(sorted(list(np.unique(new_list)))).strip(\",\")\n",
    "    return new_list\n",
    "\n",
    "\n",
    "def search_df(df, pattern):\n",
    "    mask = np.column_stack(\n",
    "        [df[col].str.contains(pattern, na=False, flags=re.IGNORECASE) for col in df]\n",
    "    )\n",
    "\n",
    "    df = df.loc[mask.any(axis=1)]\n",
    "\n",
    "    with pd.option_context(\"display.max_colwidth\", None):\n",
    "        display(df)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def find_row(df, attribute):\n",
    "    \"\"\"Get indexes of the dataframe\"\"\"\n",
    "    indexes = df.index[\n",
    "        df[\"Attribute\"].str.contains(\n",
    "            \"(^\" + re.escape(attribute) + \"$)\", flags=re.IGNORECASE\n",
    "        )\n",
    "    ].tolist()\n",
    "    if len(indexes) != 0:\n",
    "        return indexes\n",
    "    else:\n",
    "        print(attribute)\n",
    "        return None\n",
    "\n",
    "\n",
    "def replace_valid_value(df, indexes, regex_dict, attribute):\n",
    "    \"\"\"Alter the dataframe valid values with the replacement value\"\"\"\n",
    "    if indexes == None:\n",
    "        return df\n",
    "\n",
    "    elif len(indexes) > 0:\n",
    "        regex_dict = regex_dict[attribute]\n",
    "\n",
    "        for index in indexes:\n",
    "            df.loc[index, \"Valid Values\"] = re.sub(\n",
    "                **regex_dict, string=df.loc[index, \"Valid Values\"]\n",
    "            )\n",
    "\n",
    "            # print(df.loc[index, 'Valid Values'])\n",
    "        return df\n",
    "    else:\n",
    "        return df\n",
    "\n",
    "\n",
    "def code_equals_values(df, regex_dict, attribute):\n",
    "    print(\"attribute: \", attribute)\n",
    "\n",
    "    indexes = find_row(df, attribute)\n",
    "\n",
    "    print(\"Index: \", indexes)\n",
    "\n",
    "    if indexes == None:\n",
    "        return df\n",
    "    else:\n",
    "        df = replace_valid_value(df, indexes, regex_dict, attribute)\n",
    "        return df\n",
    "\n",
    "\n",
    "def rewrite_df_value(df, col_name, search_term, col_value, new_value):\n",
    "    try:\n",
    "        df.loc[df[df[col_name] == search_term].index[0], col_value] = new_value\n",
    "        return df\n",
    "    except:\n",
    "        return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Unzip compressed folder if downloaded from Google Drive\n",
    "# %unzip 'RFC Tables-20230620T181152Z-001.zip'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Collect RFCs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get all the RFC file paths\n",
    "file_paths = glob(\"../_data/RFC Tables/*\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create Data Model for Schematic\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm = pd.DataFrame()\n",
    "\n",
    "# parse through files to create complete data model\n",
    "for fp in file_paths:\n",
    "    file_name = os.path.basename(fp)\n",
    "\n",
    "    temp = pd.read_excel(fp)\n",
    "\n",
    "    # Create file_name column to check\n",
    "    temp.insert(loc=0, column=\"file_name\", value=file_name)\n",
    "\n",
    "    # Create new columnn for data model name\n",
    "    temp.insert(\n",
    "        loc=1,\n",
    "        column=\"dm\",\n",
    "        value=re.sub(\n",
    "            \"\\s\\s+\",\n",
    "            \" \",\n",
    "            re.sub(\n",
    "                \"_\",\n",
    "                \" \",\n",
    "                re.sub(\n",
    "                    \"(EL)|(RFC)|(\\.xlsx)|([Aa]ssay)|([Dd]ata [Mm]odel)\", \"\", file_name\n",
    "                ),\n",
    "            ).strip(),\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    dm = pd.concat([dm, temp])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>dm</th>\n",
       "      <th>key</th>\n",
       "      <th>description</th>\n",
       "      <th>valid values</th>\n",
       "      <th>required</th>\n",
       "      <th>requires</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>type</th>\n",
       "      <th>concept source ontology</th>\n",
       "      <th>note</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "      <th>ontology</th>\n",
       "      <th>term id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EL Assay_ scRNAseq data model.xlsx</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>specimenID</td>\n",
       "      <td>Identifying string linked to a particular samp...</td>\n",
       "      <td>n/a (unique to each data contributor)</td>\n",
       "      <td>True</td>\n",
       "      <td>scRNAseq, Biospecimen</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EL Assay_ scRNAseq data model.xlsx</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>sampleType</td>\n",
       "      <td>The type of sample collected</td>\n",
       "      <td>Amniotic Fluid,\\nAppendix,\\nB cell,\\nBasophils...</td>\n",
       "      <td>True</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,\\nImmPort</td>\n",
       "      <td>The sample types are adopted from Uberon, Cell...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EL Assay_ scRNAseq data model.xlsx</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>specifySampleType</td>\n",
       "      <td>If \"other\" is selected list the type of sample</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>scRNAseq,\\nsampleType = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EL Assay_ scRNAseq data model.xlsx</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The measurement technique describing the assay...</td>\n",
       "      <td>16S rRNA gene sequencing,\\n1D Gel,\\n2D Gel,\\nA...</td>\n",
       "      <td>True</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EL Assay_ scRNAseq data model.xlsx</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the me...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>scRNAseq,\\nmeasurementTechnique = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            file_name        dm                          key  \\\n",
       "0  EL Assay_ scRNAseq data model.xlsx  scRNAseq                   specimenID   \n",
       "1  EL Assay_ scRNAseq data model.xlsx  scRNAseq                   sampleType   \n",
       "2  EL Assay_ scRNAseq data model.xlsx  scRNAseq            specifySampleType   \n",
       "3  EL Assay_ scRNAseq data model.xlsx  scRNAseq         measurementTechnique   \n",
       "4  EL Assay_ scRNAseq data model.xlsx  scRNAseq  specifyMeasurementTechnique   \n",
       "\n",
       "                                         description  \\\n",
       "0  Identifying string linked to a particular samp...   \n",
       "1                       The type of sample collected   \n",
       "2     If \"other\" is selected list the type of sample   \n",
       "3  The measurement technique describing the assay...   \n",
       "4  If \"other\" is selected list the name of the me...   \n",
       "\n",
       "                                        valid values required  \\\n",
       "0              n/a (unique to each data contributor)     True   \n",
       "1  Amniotic Fluid,\\nAppendix,\\nB cell,\\nBasophils...     True   \n",
       "2                                                NaN    False   \n",
       "3  16S rRNA gene sequencing,\\n1D Gel,\\n2D Gel,\\nA...     True   \n",
       "4                                                NaN    False   \n",
       "\n",
       "                                  requires multivalue    type  \\\n",
       "0                    scRNAseq, Biospecimen      False  string   \n",
       "1                                 scRNAseq      False  string   \n",
       "2            scRNAseq,\\nsampleType = other      False  string   \n",
       "3                                 scRNAseq      False  string   \n",
       "4  scRNAseq,\\nmeasurementTechnique = other      False  string   \n",
       "\n",
       "      concept source ontology  \\\n",
       "0            Sage Bionetworks   \n",
       "1  Sage Bionetworks,\\nImmPort   \n",
       "2            Sage Bionetworks   \n",
       "3            Sage Bionetworks   \n",
       "4            Sage Bionetworks   \n",
       "\n",
       "                                                note  Unnamed: 9  \\\n",
       "0                                                NaN         NaN   \n",
       "1  The sample types are adopted from Uberon, Cell...         NaN   \n",
       "2                                                NaN         NaN   \n",
       "3                                                NaN         NaN   \n",
       "4                                                NaN         NaN   \n",
       "\n",
       "        Unnamed: 10 ontology term id  \n",
       "0  Sage Bionetworks      NaN     NaN  \n",
       "1               NaN      NaN     NaN  \n",
       "2               NaN      NaN     NaN  \n",
       "3               NaN      NaN     NaN  \n",
       "4               NaN      NaN     NaN  "
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# initial cleaning\n",
    "dm[[\"required\", \"multivalue\"]] = (\n",
    "    dm[[\"required\", \"multivalue\"]]\n",
    "    .fillna(False)\n",
    "    .astype(str)\n",
    "    .replace({\"1.0\": True, \"0.0\": False})\n",
    ")\n",
    "\n",
    "dm.fillna(\"\")\n",
    "dm.reset_index(drop=True, inplace=True)\n",
    "dm.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data model clean up\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/p0/5m4pdsm55jn_d5nzbjv6bjf40000gq/T/ipykernel_11826/3476214067.py:14: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.\n",
      "  dm[\"valid values\"].str.contains(\n"
     ]
    }
   ],
   "source": [
    "# collapse presumed ontology columns and join with existing\n",
    "dm.loc[:, \"ontology\"] = dm.iloc[:, 11:].bfill(axis=1).iloc[:, 0]\n",
    "\n",
    "dm[\"ontology\"] = (\n",
    "    dm[[\"concept source ontology\", \"ontology\"]]\n",
    "    .fillna(\"\")\n",
    "    .apply(lambda x: \",\".join([y.strip() for y in x.unique() if len(y) > 0]), axis=1)\n",
    ")\n",
    "\n",
    "# if unique values are provided by data contributor then add this note in the ontology\n",
    "\n",
    "dm[\"ontology\"] = (\n",
    "    dm.loc[\n",
    "        dm[\"valid values\"].str.contains(\n",
    "            \"(n/a \\(unique to each data contributor\\))\", na=False\n",
    "        ),\n",
    "        \"ontology\",\n",
    "    ]\n",
    "    + \",\"\n",
    "    + \"Data Contributor\"\n",
    ")\n",
    "\n",
    "# new ontologies:\n",
    "ontology_list = []\n",
    "for i, v in dm[[\"concept source ontology\", \"ontology\"]].fillna(\"\").iterrows():\n",
    "    ontology_list.append(\",\".join(v))\n",
    "\n",
    "dm[\"ontology\"] = ontology_list\n",
    "dm[\"ontology\"] = dm[\"ontology\"].str.strip(\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>dm</th>\n",
       "      <th>key</th>\n",
       "      <th>description</th>\n",
       "      <th>valid values</th>\n",
       "      <th>required</th>\n",
       "      <th>requires</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>type</th>\n",
       "      <th>concept source ontology</th>\n",
       "      <th>note</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "      <th>ontology</th>\n",
       "      <th>term id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[EL Assay_ scRNAseq data model.xlsx]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[specimenID]</td>\n",
       "      <td>[Identifying string linked to a particular sam...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[scRNAseq,  Biospecimen]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[string]</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>[Sage Bionetworks, Sage Bionetworks, Data Cont...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[EL Assay_ scRNAseq data model.xlsx]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[sampleType]</td>\n",
       "      <td>[The type of sample collected]</td>\n",
       "      <td>[Amniotic Fluid, Appendix, B cell, Basophils, ...</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[string]</td>\n",
       "      <td>[Sage Bionetworks, ImmPort]</td>\n",
       "      <td>[The sample types are adopted from Uberon,  Ce...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Sage Bionetworks, ImmPort]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[EL Assay_ scRNAseq data model.xlsx]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[specifySampleType]</td>\n",
       "      <td>[If \"other\" is selected list the type of sample]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[scRNAseq, sampleType = other]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[string]</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[EL Assay_ scRNAseq data model.xlsx]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[measurementTechnique]</td>\n",
       "      <td>[The measurement technique describing the assa...</td>\n",
       "      <td>[16S rRNA gene sequencing, 1D Gel, 2D Gel, Arr...</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[string]</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[EL Assay_ scRNAseq data model.xlsx]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[specifyMeasurementTechnique]</td>\n",
       "      <td>[If \"other\" is selected list the name of the m...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[scRNAseq, measurementTechnique = other]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[string]</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              file_name          dm  \\\n",
       "0  [EL Assay_ scRNAseq data model.xlsx]  [scRNAseq]   \n",
       "1  [EL Assay_ scRNAseq data model.xlsx]  [scRNAseq]   \n",
       "2  [EL Assay_ scRNAseq data model.xlsx]  [scRNAseq]   \n",
       "3  [EL Assay_ scRNAseq data model.xlsx]  [scRNAseq]   \n",
       "4  [EL Assay_ scRNAseq data model.xlsx]  [scRNAseq]   \n",
       "\n",
       "                             key  \\\n",
       "0                   [specimenID]   \n",
       "1                   [sampleType]   \n",
       "2            [specifySampleType]   \n",
       "3         [measurementTechnique]   \n",
       "4  [specifyMeasurementTechnique]   \n",
       "\n",
       "                                         description  \\\n",
       "0  [Identifying string linked to a particular sam...   \n",
       "1                     [The type of sample collected]   \n",
       "2   [If \"other\" is selected list the type of sample]   \n",
       "3  [The measurement technique describing the assa...   \n",
       "4  [If \"other\" is selected list the name of the m...   \n",
       "\n",
       "                                        valid values required  \\\n",
       "0                                                 []   [True]   \n",
       "1  [Amniotic Fluid, Appendix, B cell, Basophils, ...   [True]   \n",
       "2                                                NaN  [False]   \n",
       "3  [16S rRNA gene sequencing, 1D Gel, 2D Gel, Arr...   [True]   \n",
       "4                                                NaN  [False]   \n",
       "\n",
       "                                   requires multivalue      type  \\\n",
       "0                  [scRNAseq,  Biospecimen]    [False]  [string]   \n",
       "1                                [scRNAseq]    [False]  [string]   \n",
       "2            [scRNAseq, sampleType = other]    [False]  [string]   \n",
       "3                                [scRNAseq]    [False]  [string]   \n",
       "4  [scRNAseq, measurementTechnique = other]    [False]  [string]   \n",
       "\n",
       "       concept source ontology  \\\n",
       "0           [Sage Bionetworks]   \n",
       "1  [Sage Bionetworks, ImmPort]   \n",
       "2           [Sage Bionetworks]   \n",
       "3           [Sage Bionetworks]   \n",
       "4           [Sage Bionetworks]   \n",
       "\n",
       "                                                note  Unnamed: 9  \\\n",
       "0                                                NaN         NaN   \n",
       "1  [The sample types are adopted from Uberon,  Ce...         NaN   \n",
       "2                                                NaN         NaN   \n",
       "3                                                NaN         NaN   \n",
       "4                                                NaN         NaN   \n",
       "\n",
       "          Unnamed: 10                                           ontology  \\\n",
       "0  [Sage Bionetworks]  [Sage Bionetworks, Sage Bionetworks, Data Cont...   \n",
       "1                 NaN                        [Sage Bionetworks, ImmPort]   \n",
       "2                 NaN                                 [Sage Bionetworks]   \n",
       "3                 NaN                                 [Sage Bionetworks]   \n",
       "4                 NaN                                 [Sage Bionetworks]   \n",
       "\n",
       "  term id  \n",
       "0     NaN  \n",
       "1     NaN  \n",
       "2     NaN  \n",
       "3     NaN  \n",
       "4     NaN  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm = dm.apply(\n",
    "    lambda x: x.str.replace(\n",
    "        pat=\"\\n|(n/a \\(unique to each data contributor\\))\", repl=\"\", regex=True\n",
    "    ).str.split(\",\"),\n",
    "    axis=1,\n",
    ")\n",
    "\n",
    "dm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>specimenID</td>\n",
       "      <td>Identifying string linked to a particular samp...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq, Biospecimen</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,Sage Bionetworks,Data Contrib...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sampleType</td>\n",
       "      <td>The type of sample collected</td>\n",
       "      <td>Amniotic Fluid,Appendix,B cell,Basophils,Bone,...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks,ImmPort</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,ImmPort</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>specifySampleType</td>\n",
       "      <td>If \"other\" is selected list the type of sample</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq,sampleType = other</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The measurement technique describing the assay...</td>\n",
       "      <td>16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the me...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq,measurementTechnique = other</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Attribute  \\\n",
       "0                   specimenID   \n",
       "1                   sampleType   \n",
       "2            specifySampleType   \n",
       "3         measurementTechnique   \n",
       "4  specifyMeasurementTechnique   \n",
       "\n",
       "                                         Description  \\\n",
       "0  Identifying string linked to a particular samp...   \n",
       "1                       The type of sample collected   \n",
       "2     If \"other\" is selected list the type of sample   \n",
       "3  The measurement technique describing the assay...   \n",
       "4  If \"other\" is selected list the name of the me...   \n",
       "\n",
       "                                        Valid Values Required DependsOn  \\\n",
       "0                                                        True             \n",
       "1  Amniotic Fluid,Appendix,B cell,Basophils,Bone,...     True             \n",
       "2                                                       False             \n",
       "3  16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...     True             \n",
       "4                                                       False             \n",
       "\n",
       "                     DependsOn Component Properties Validation Rules  \\\n",
       "0                  scRNAseq, Biospecimen                               \n",
       "1                               scRNAseq                               \n",
       "2            scRNAseq,sampleType = other                               \n",
       "3                               scRNAseq                               \n",
       "4  scRNAseq,measurementTechnique = other                               \n",
       "\n",
       "   Template Parent                    Source    Type  \\\n",
       "0  scRNAseq                 Sage Bionetworks  string   \n",
       "1  scRNAseq         Sage Bionetworks,ImmPort  string   \n",
       "2  scRNAseq                 Sage Bionetworks  string   \n",
       "3  scRNAseq                 Sage Bionetworks  string   \n",
       "4  scRNAseq                 Sage Bionetworks  string   \n",
       "\n",
       "                                            Ontology multivalue  \n",
       "0  Sage Bionetworks,Sage Bionetworks,Data Contrib...      False  \n",
       "1                           Sage Bionetworks,ImmPort      False  \n",
       "2                                   Sage Bionetworks      False  \n",
       "3                                   Sage Bionetworks      False  \n",
       "4                                   Sage Bionetworks      False  "
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# revert lists back to strings\n",
    "dm = dm.applymap(lambda x: join_strings(x))\n",
    "\n",
    "# Rename columns with DCA standards\n",
    "dm_schema_cols = {\n",
    "    \"dm\": \"Template\",\n",
    "    \"key\": \"Attribute\",\n",
    "    \"description\": \"Description\",\n",
    "    \"valid values\": \"Valid Values\",\n",
    "    \"required\": \"Required\",\n",
    "    \"requires\": \"DependsOn Component\",\n",
    "    \"concept source ontology\": \"Source\",\n",
    "    \"ontology\": \"Ontology\",\n",
    "    \"type\": \"Type\",\n",
    "}\n",
    "\n",
    "dm = dm.rename(dm_schema_cols, axis=1)\n",
    "\n",
    "# drop unimportant columns\n",
    "r = re.compile(\"Unnamed*\", re.IGNORECASE)\n",
    "\n",
    "# Add additional required columns for DCA\n",
    "dm[\"Properties\"] = \"\"\n",
    "dm[\"Validation Rules\"] = dm[\"Type\"].map(validation_coder)\n",
    "dm[\"DependsOn\"] = \"\"\n",
    "dm[\"Parent\"] = \"\"\n",
    "# dm['DependsOn Component'] = \"\"\n",
    "\n",
    "dm = dm[keep_cols]\n",
    "\n",
    "dm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add templates as attributes to generate templates in DCA\n",
    "# Create Manifests in data model\n",
    "unique_templates = np.unique(\n",
    "    \",\".join(dm[\"Template\"].dropna().values.tolist()).split(\",\")\n",
    ")\n",
    "templates = pd.DataFrame({\"Attribute\": unique_templates})\n",
    "templates[\"Required\"] = \"True\"\n",
    "templates[\"Properties\"] = \"dataType\"\n",
    "templates[\"Parent\"] = \"Template\"\n",
    "templates[\"Description\"] = templates[\"Attribute\"].apply(\n",
    "    lambda x: f\"Metadata template for {x}\"\n",
    ")\n",
    "\n",
    "# Adjust bseq\n",
    "templates.loc[\n",
    "    templates[\"Attribute\"] == \"bsSeq (bisulfite-seq WGBS methylseq methylomics)\",\n",
    "    \"Description\",\n",
    "] = \"Metadata template for bisulfite-seq WGBS methylseq methylomics\"\n",
    "templates.loc[\n",
    "    templates[\"Attribute\"] == \"bsSeq (bisulfite-seq WGBS methylseq methylomics)\",\n",
    "    \"Attribute\",\n",
    "] = \"bsSeq\"\n",
    "\n",
    "# update templates column for other attributes\n",
    "dm[\"Template\"] = dm[\"Template\"].str.replace(\n",
    "    \"bsSeq (bisulfite-seq WGBS methylseq methylomics)\", \"bsSeq\", regex=False\n",
    ")\n",
    "\n",
    "# create depends On column for templates\n",
    "for i, m in enumerate(templates[\"Attribute\"]):\n",
    "    dm[\"Template\"].str.contains(m)\n",
    "    templates.loc[i, \"DependsOn\"] = \",\".join(\n",
    "        dm.loc[\n",
    "            dm[\"Template\"].str.contains(m, regex=False, na=False), \"Attribute\"\n",
    "        ].tolist()\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: Valid Values, dtype: object)"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fixing \"not listed\" columns\n",
    "dm[\"Valid Values\"] = dm[\"Valid Values\"].str.replace(\n",
    "    \"Genbank common names (not listed for purposes of this RFC)Unknown\",\n",
    "    \"Genbank common names (not listed for purposes of this RFC),Unknown\",\n",
    "    regex=False,\n",
    ")\n",
    "\n",
    "dm.loc[dm[\"Valid Values\"].str.contains(\"not listed\"), \"Valid Values\"] = (\n",
    "    dm.loc[dm[\"Valid Values\"].str.contains(\"not listed\"), \"Valid Values\"]\n",
    "    .str.split(\")\")\n",
    "    .apply(\n",
    "        lambda x: \",\".join(\n",
    "            [y.strip(\",\") for y in x if not bool(re.search(\"not listed\", y))]\n",
    "        )\n",
    "    )\n",
    ")\n",
    "\n",
    "# QC\n",
    "dm.loc[dm[\"Valid Values\"].str.contains(\"not listed\", na=False), \"Valid Values\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dropping measurement technique\n",
    "dm = dm.drop(\n",
    "    index=dm.query('Attribute == \"measurementTechnique\"').index.values\n",
    ").reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine duplicated attributes\n",
    "dm = dm.groupby(\"Attribute\").agg(lambda x: \",\".join(set(x.astype(str)))).reset_index()\n",
    "\n",
    "# found extra commas in strings at beginning and end\n",
    "dm = dm.applymap(lambda x: x.strip(\",\"))\n",
    "\n",
    "template_recoder = {\n",
    "    re.compile(\"genotyping\", re.IGNORECASE): \"Genotyping\",\n",
    "    re.compile(\"proteomics\", re.IGNORECASE): \"Proteomics\",\n",
    "}\n",
    "\n",
    "dm = dm.replace(template_recoder, regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Attribute              object\n",
       "Description            object\n",
       "Valid Values           object\n",
       "Required               object\n",
       "DependsOn              object\n",
       "DependsOn Component    object\n",
       "Properties             object\n",
       "Validation Rules       object\n",
       "Template               object\n",
       "Parent                 object\n",
       "Source                 object\n",
       "Type                   object\n",
       "Ontology               object\n",
       "multivalue             object\n",
       "dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>acquisitionBatchID</td>\n",
       "      <td>Acquisition batch identifier, provided by the ...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>acquisitionBatchSize</td>\n",
       "      <td>The number of samples</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>acquisitionBatchSizeUnit</td>\n",
       "      <td>The unit of measurement for number of samples ...</td>\n",
       "      <td>AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolmics,acquisitionBatchSize</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>acquisitionMode</td>\n",
       "      <td>The specific aspect of a mass spectrometer met...</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not speci...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec Proteomics,mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Proteomics,Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/terms?...</td>\n",
       "      <td>string</td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/terms?...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>acquisitionSoftware</td>\n",
       "      <td>The name of the acquisition software used</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not speci...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec Proteomics,mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Proteomics,Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Attribute  \\\n",
       "0        acquisitionBatchID   \n",
       "1      acquisitionBatchSize   \n",
       "2  acquisitionBatchSizeUnit   \n",
       "3           acquisitionMode   \n",
       "4       acquisitionSoftware   \n",
       "\n",
       "                                         Description  \\\n",
       "0  Acquisition batch identifier, provided by the ...   \n",
       "1                              The number of samples   \n",
       "2  The unit of measurement for number of samples ...   \n",
       "3  The specific aspect of a mass spectrometer met...   \n",
       "4          The name of the acquisition software used   \n",
       "\n",
       "                                        Valid Values Required DependsOn  \\\n",
       "0                                                       False             \n",
       "1                                                       False             \n",
       "2  AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...    False             \n",
       "3  Unknown,Not collected,Not applicable,Not speci...     True             \n",
       "4  Unknown,Not collected,Not applicable,Not speci...     True             \n",
       "\n",
       "                          DependsOn Component Properties Validation Rules  \\\n",
       "0                       mass spec metabolmics                               \n",
       "1                       mass spec metabolmics                               \n",
       "2  mass spec metabolmics,acquisitionBatchSize                               \n",
       "3  mass spec Proteomics,mass spec metabolmics                               \n",
       "4  mass spec Proteomics,mass spec metabolmics                               \n",
       "\n",
       "                        Template Parent  \\\n",
       "0             Metabolomics Human          \n",
       "1             Metabolomics Human          \n",
       "2             Metabolomics Human          \n",
       "3  Proteomics,Metabolomics Human          \n",
       "4  Proteomics,Metabolomics Human          \n",
       "\n",
       "                                              Source    Type  \\\n",
       "0                                   Sage Bionetworks  string   \n",
       "1                                   Sage Bionetworks  string   \n",
       "2                                   Sage Bionetworks  string   \n",
       "3  https://www.ebi.ac.uk/ols/ontologies/ms/terms?...  string   \n",
       "4          http://purl.obolibrary.org/obo/MS_1001455  string   \n",
       "\n",
       "                                            Ontology multivalue  \n",
       "0                                   Sage Bionetworks             \n",
       "1                                   Sage Bionetworks             \n",
       "2                                   Sage Bionetworks             \n",
       "3  https://www.ebi.ac.uk/ols/ontologies/ms/terms?...      False  \n",
       "4          http://purl.obolibrary.org/obo/MS_1001455      False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "array(['Metabolomics Human', 'Proteomics,Metabolomics Human',\n",
       "       'Individual Human', 'Biospecimen nonHuman', 'Individual nonHuman',\n",
       "       'Biospecimen nonHuman,Biospecimen human', 'Proteomics',\n",
       "       'Individual nonHuman,Individual Human', 'bsSeq',\n",
       "       'bsSeq,Microbiome', 'Microbiome,Metabolomics Human',\n",
       "       'Individual nonHuman,Individual Human,Biospecimen nonHuman,Biospecimen human',\n",
       "       'bsSeq,RNAseq,Microbiome,Whole Genome Sequencing,scRNAseq',\n",
       "       'bsSeq,Microbiome,scRNAseq,Whole Genome Sequencing',\n",
       "       'bsSeq,RNAseq,Microbiome,Whole Genome Sequencing,Proteomics,scRNAseq,Metabolomics Human',\n",
       "       'Genotyping', 'RNAseq,bsSeq,scRNAseq,Whole Genome Sequencing',\n",
       "       'RNAseq,scRNAseq,Whole Genome Sequencing',\n",
       "       'scRNAseq,Whole Genome Sequencing', 'scRNAseq',\n",
       "       'bsSeq,Genotyping,RNAseq,Microbiome,Whole Genome Sequencing,Proteomics,scRNAseq,Metabolomics Human',\n",
       "       'bsSeq,Microbiome,scRNAseq', 'Microbiome',\n",
       "       'bsSeq,Genotyping,RNAseq,Microbiome,Biospecimen nonHuman,Whole Genome Sequencing,Proteomics,scRNAseq,Metabolomics Human,Biospecimen human'],\n",
       "      dtype=object)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(dm.dtypes)\n",
    "display(dm.head())\n",
    "display(dm.Template.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reorder Columns based on DCA Standards\n",
    "dm = dm.loc[:, keep_cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clean up\n",
    "\n",
    "## DependsOn Component\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "recoder = {\n",
    "    \"metabolmics\": \"metabolomics\",\n",
    "    \"(mass spec proteomics)\": \"Proteomics\",\n",
    "    \"(mass spec metabolomics)\": \"Metabolomics Human\",\n",
    "    \"(assay_otheruseTreatment? = Yes)\": \"assay_other, useTreatment? = Yes\",\n",
    "    \"OtherUnknown\": \"Other, Unknown\",\n",
    "    \"falseFalseFALSEtrueTrueTRUE\": \"TRUE, FALSE\",\n",
    "    \"protoemicsAssayType\": \"proteomicsAssayType\",\n",
    "    \"Hispanic or latinoEthnicity\": \"Hispanic or Latino\",\n",
    "    re.compile(\"Forwardreverse\", flags=re.IGNORECASE): \"forward,reverse\",\n",
    "    re.compile(\"singleEndpairedEnd\"): \"singleEnd, pairedEnd\",\n",
    "    re.compile(\"(WGS)\"): \"Whole Genome Sequencing\",\n",
    "    re.compile(\"\\?\"): \"\",\n",
    "    \"Zeiss LSM 980Other\": \"Zeiss LSM 980,Other\",\n",
    "    \"bsSeqsampleType = other\": \"bsSeq, sampleType = other\",\n",
    "    re.compile(\n",
    "        \"HPO, MONDO, MAXO codes or labels \\(not listed for purposes of this RFC\\)\"\n",
    "    ): \"HPO and MONDO and MAXO codes or labels (not listed for purposes of this RFC)\",\n",
    "    \"The Health, Aging, and Body Composition Study \\(HealthABC\\)\": \"The Health and Aging and Body Composition Study (HealthABC)\",\n",
    "}\n",
    "\n",
    "# 'mass spec metabolomics,measurementTechnique = other'\n",
    "# falseFalseFALSEtrueTrueTRUE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm = dm.replace(recoder, regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>cohort</td>\n",
       "      <td>Name of the cohort the individual belongs to</td>\n",
       "      <td>Other, Unknown,Not collected,Not applicable,Centenarian, US Family, Denmark Family, The Osteoporotic Fractures in Men (MrOS) Study, Study of Osteoporotic Fractures (SOF), The Health and Aging and Body Composition Study (HealthABC), Cardiovascular Health Study (CHS),Other,Unknown,Not collected,Not applicable</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>Individual</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Individual nonHuman,Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Attribute                                   Description  \\\n",
       "16    cohort  Name of the cohort the individual belongs to   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                            Valid Values  \\\n",
       "16  Other, Unknown,Not collected,Not applicable,Centenarian, US Family, Denmark Family, The Osteoporotic Fractures in Men (MrOS) Study, Study of Osteoporotic Fractures (SOF), The Health and Aging and Body Composition Study (HealthABC), Cardiovascular Health Study (CHS),Other,Unknown,Not collected,Not applicable   \n",
       "\n",
       "   Required DependsOn DependsOn Component Properties Validation Rules  \\\n",
       "16     True                    Individual                               \n",
       "\n",
       "                                Template Parent            Source    Type  \\\n",
       "16  Individual nonHuman,Individual Human         Sage Bionetworks  string   \n",
       "\n",
       "            Ontology multivalue  \n",
       "16  Sage Bionetworks       True  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with pd.option_context(\"display.max_colwidth\", None):\n",
    "    display(dm[dm[\"Valid Values\"].str.contains(\"and Body Composition Study\")])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ethnicity</td>\n",
       "      <td>Ethnicity of individual</td>\n",
       "      <td>Not Hispanic or Latino, Hispanic or Latino, Prefer not to answer,Other,Unknown,Not collected,Not applicable</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>Individual</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks,https://www.synapse.org/#!Synapse:syn25878249</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,https://www.synapse.org/#!Synapse:syn25878249</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Attribute              Description  \\\n",
       "37  ethnicity  Ethnicity of individual   \n",
       "\n",
       "                                                                                                   Valid Values  \\\n",
       "37  Not Hispanic or Latino, Hispanic or Latino, Prefer not to answer,Other,Unknown,Not collected,Not applicable   \n",
       "\n",
       "   Required DependsOn DependsOn Component Properties Validation Rules  \\\n",
       "37     True                    Individual                               \n",
       "\n",
       "            Template Parent  \\\n",
       "37  Individual Human          \n",
       "\n",
       "                                                            Source    Type  \\\n",
       "37  Sage Bionetworks,https://www.synapse.org/#!Synapse:syn25878249  string   \n",
       "\n",
       "                                                          Ontology multivalue  \n",
       "37  Sage Bionetworks,https://www.synapse.org/#!Synapse:syn25878249      False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with pd.option_context(\"display.max_colwidth\", None):\n",
    "    display(dm.query('Attribute == \"ethnicity\"'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>acquisitionBatchID</td>\n",
       "      <td>Acquisition batch identifier, provided by the ...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolomics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>acquisitionBatchSize</td>\n",
       "      <td>The number of samples</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolomics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>acquisitionBatchSizeUnit</td>\n",
       "      <td>The unit of measurement for number of samples ...</td>\n",
       "      <td>AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolomics,acquisitionBatchSize</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>acquisitionMode</td>\n",
       "      <td>The specific aspect of a mass spectrometer met...</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not speci...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec Proteomics,mass spec metabolomics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Proteomics,Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/termsi...</td>\n",
       "      <td>string</td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/termsi...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>acquisitionSoftware</td>\n",
       "      <td>The name of the acquisition software used</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not speci...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec Proteomics,mass spec metabolomics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Proteomics,Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>specimenID</td>\n",
       "      <td>Identifying string linked to a particular samp...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>Biospecimen,bsSeq,Biospecimen,Biospecimen,mass...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>bsSeq,Genotyping,RNAseq,Microbiome,Biospecimen...</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,Sage Bionetworks,Data Contrib...</td>\n",
       "      <td>false,False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>spectrometerFrequency</td>\n",
       "      <td>The frequency at which a spectrometer causes h...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec Proteomics,mass spec metabolomics</td>\n",
       "      <td></td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.)|([0-9]+)</td>\n",
       "      <td>Proteomics,Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks,DSLWG,Sage Bionetworks</td>\n",
       "      <td>number</td>\n",
       "      <td>Sage Bionetworks,DSLWG,Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>technologyPlatformVersion</td>\n",
       "      <td>The specific version (application, manufacture...</td>\n",
       "      <td>Other, Unknown,Not collected,Not applicable,No...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>bsSeq,mass spec Proteomics,mass spec metabolom...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>bsSeq,RNAseq,Microbiome,Whole Genome Sequencin...</td>\n",
       "      <td></td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C45378,Sag...</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C45378,Sag...</td>\n",
       "      <td>true,False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>vacuumPressure</td>\n",
       "      <td>The recorded vacuum pressure value</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolomics,hasIonizationSource = Yes</td>\n",
       "      <td></td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.)|([0-9]+)</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks,Proposed minimum metadata rel...</td>\n",
       "      <td>number</td>\n",
       "      <td>Sage Bionetworks,Proposed minimum metadata rel...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>vacuumPressureUnit</td>\n",
       "      <td>Unit of vacuum pressure value</td>\n",
       "      <td>AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolomics,hasIonizationSource = Y...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks,Proposed minimum metadata rel...</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,Proposed minimum metadata rel...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Attribute  \\\n",
       "0           acquisitionBatchID   \n",
       "1         acquisitionBatchSize   \n",
       "2     acquisitionBatchSizeUnit   \n",
       "3              acquisitionMode   \n",
       "4          acquisitionSoftware   \n",
       "..                         ...   \n",
       "152                 specimenID   \n",
       "156      spectrometerFrequency   \n",
       "159  technologyPlatformVersion   \n",
       "175             vacuumPressure   \n",
       "176         vacuumPressureUnit   \n",
       "\n",
       "                                           Description  \\\n",
       "0    Acquisition batch identifier, provided by the ...   \n",
       "1                                The number of samples   \n",
       "2    The unit of measurement for number of samples ...   \n",
       "3    The specific aspect of a mass spectrometer met...   \n",
       "4            The name of the acquisition software used   \n",
       "..                                                 ...   \n",
       "152  Identifying string linked to a particular samp...   \n",
       "156  The frequency at which a spectrometer causes h...   \n",
       "159  The specific version (application, manufacture...   \n",
       "175                 The recorded vacuum pressure value   \n",
       "176                      Unit of vacuum pressure value   \n",
       "\n",
       "                                          Valid Values Required DependsOn  \\\n",
       "0                                                         False             \n",
       "1                                                         False             \n",
       "2    AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...    False             \n",
       "3    Unknown,Not collected,Not applicable,Not speci...     True             \n",
       "4    Unknown,Not collected,Not applicable,Not speci...     True             \n",
       "..                                                 ...      ...       ...   \n",
       "152                                                        True             \n",
       "156                                                        True             \n",
       "159  Other, Unknown,Not collected,Not applicable,No...     True             \n",
       "175                                                       False             \n",
       "176  AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...    False             \n",
       "\n",
       "                                   DependsOn Component Properties  \\\n",
       "0                               mass spec metabolomics              \n",
       "1                               mass spec metabolomics              \n",
       "2          mass spec metabolomics,acquisitionBatchSize              \n",
       "3          mass spec Proteomics,mass spec metabolomics              \n",
       "4          mass spec Proteomics,mass spec metabolomics              \n",
       "..                                                 ...        ...   \n",
       "152  Biospecimen,bsSeq,Biospecimen,Biospecimen,mass...              \n",
       "156        mass spec Proteomics,mass spec metabolomics              \n",
       "159  bsSeq,mass spec Proteomics,mass spec metabolom...              \n",
       "175   mass spec metabolomics,hasIonizationSource = Yes              \n",
       "176  mass spec metabolomics,hasIonizationSource = Y...              \n",
       "\n",
       "                            Validation Rules  \\\n",
       "0                                              \n",
       "1                                              \n",
       "2                                              \n",
       "3                                              \n",
       "4                                              \n",
       "..                                       ...   \n",
       "152                                            \n",
       "156  regex search ([0-9]+\\.[0-9]*.)|([0-9]+)   \n",
       "159                                            \n",
       "175  regex search ([0-9]+\\.[0-9]*.)|([0-9]+)   \n",
       "176                                            \n",
       "\n",
       "                                              Template Parent  \\\n",
       "0                                   Metabolomics Human          \n",
       "1                                   Metabolomics Human          \n",
       "2                                   Metabolomics Human          \n",
       "3                        Proteomics,Metabolomics Human          \n",
       "4                        Proteomics,Metabolomics Human          \n",
       "..                                                 ...    ...   \n",
       "152  bsSeq,Genotyping,RNAseq,Microbiome,Biospecimen...          \n",
       "156                      Proteomics,Metabolomics Human          \n",
       "159  bsSeq,RNAseq,Microbiome,Whole Genome Sequencin...          \n",
       "175                                 Metabolomics Human          \n",
       "176                                 Metabolomics Human          \n",
       "\n",
       "                                                Source    Type  \\\n",
       "0                                     Sage Bionetworks  string   \n",
       "1                                     Sage Bionetworks  string   \n",
       "2                                     Sage Bionetworks  string   \n",
       "3    https://www.ebi.ac.uk/ols/ontologies/ms/termsi...  string   \n",
       "4            http://purl.obolibrary.org/obo/MS_1001455  string   \n",
       "..                                                 ...     ...   \n",
       "152                                   Sage Bionetworks  string   \n",
       "156            Sage Bionetworks,DSLWG,Sage Bionetworks  number   \n",
       "159  http://purl.obolibrary.org/obo/NCIT_C45378,Sag...  string   \n",
       "175  Sage Bionetworks,Proposed minimum metadata rel...  number   \n",
       "176  Sage Bionetworks,Proposed minimum metadata rel...  string   \n",
       "\n",
       "                                              Ontology   multivalue  \n",
       "0                                     Sage Bionetworks               \n",
       "1                                     Sage Bionetworks               \n",
       "2                                     Sage Bionetworks               \n",
       "3    https://www.ebi.ac.uk/ols/ontologies/ms/termsi...        False  \n",
       "4            http://purl.obolibrary.org/obo/MS_1001455        False  \n",
       "..                                                 ...          ...  \n",
       "152  Sage Bionetworks,Sage Bionetworks,Data Contrib...  false,False  \n",
       "156            Sage Bionetworks,DSLWG,Sage Bionetworks        False  \n",
       "159  http://purl.obolibrary.org/obo/NCIT_C45378,Sag...   true,False  \n",
       "175  Sage Bionetworks,Proposed minimum metadata rel...               \n",
       "176  Sage Bionetworks,Proposed minimum metadata rel...               \n",
       "\n",
       "[65 rows x 14 columns]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# QA check\n",
    "dm[dm[\"DependsOn Component\"].str.contains(\n",
    "    \"metabolomics\", case=False, na=False)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cleaning other values and equal values\n",
    "\n",
    "Removing illegal characters\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove any special characters\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm[\"Attribute\"] = dm[\"Attribute\"].str.replace(\"\\(|\\)|\\?\", \"\", regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm[\"Valid Values\"] = (\n",
    "    dm[\"Valid Values\"]\n",
    "    .apply(\n",
    "        lambda x: re.sub(\n",
    "            \"^0x Visium Spatial Gene Expression\",\n",
    "            \"10x Visium Spatial Gene Expression\",\n",
    "            x,\n",
    "        )\n",
    "    )\n",
    "    .apply(clean_list)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clean up equals in depends on\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_new_value(old_value):\n",
    "    new_vals = old_value.split(\"=\")\n",
    "    new_vals = [nv.strip() for nv in new_vals]\n",
    "    # convert to camel case\n",
    "    nv = new_vals[1].capitalize() + new_vals[0][0].upper() + new_vals[0][1:]\n",
    "    return nv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "def recode_yes_no(v):\n",
    "    if v.lower() == \"yes\":\n",
    "        return \"TRUE\"\n",
    "    elif v.lower() == \"no\":\n",
    "        return \"FALSE\"\n",
    "    else:\n",
    "        return v\n",
    "\n",
    "\n",
    "dm[\"Valid Values\"] = (\n",
    "    dm[\"Valid Values\"]\n",
    "    .apply(lambda x: x.split(\",\"))\n",
    "    .apply(lambda x: \",\".join([recode_yes_no(y) for y in x]))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split list to process other values\n",
    "# Find the other columns in the data model\n",
    "others = dm[dm[\"DependsOn Component\"].str.contains(\"=\", na=False)].copy()\n",
    "\n",
    "others[\"DependsOn Component Original\"] = others[\"DependsOn Component\"].str.split(\",\")\n",
    "\n",
    "# Create series of equals values to use for new attributes/ valid values relationship\n",
    "others[\"equals_series\"] = others[\"DependsOn Component Original\"].apply(\n",
    "    lambda x: [y for y in x if bool(re.search(\"=\", y))][0]\n",
    ")\n",
    "\n",
    "# others[\"equals_attribute\"] = others[\"equals_series\"].apply(create_new_value)\n",
    "\n",
    "others[[\"baseAttribute\", \"equalsValue\"]] = (\n",
    "    others[\"equals_series\"]\n",
    "    .str.split(\"=\", expand=True)\n",
    "    .apply(lambda x: [y.strip() for y in x])\n",
    "    .rename({0: \"base_attribute\", 1: \"equalsValue\"}, axis=1)\n",
    ")\n",
    "\n",
    "# Deciding to use true and false for all yes/no values\n",
    "recoder = {\n",
    "    re.compile(\"^[Yy]es\", flags=re.IGNORECASE): \"TRUE\",\n",
    "    re.compile(\"true\", flags=re.IGNORECASE): \"TRUE\",\n",
    "    re.compile(\"^[Nn]o\", flags=re.IGNORECASE): \"FALSE\",\n",
    "}\n",
    "\n",
    "others[\"equalsValue\"] = others[\"equalsValue\"].replace(recoder, regex=True)\n",
    "\n",
    "others[\"newDescription\"] = others[[\"baseAttribute\", \"equalsValue\"]].apply(\n",
    "    lambda x: f\"When {x[0].strip()} = {x[1].strip()}\", axis=1\n",
    ")\n",
    "\n",
    "others[\"equalsAttribute\"] = others[[\"baseAttribute\", \"equalsValue\"]].apply(\n",
    "    lambda x: f\"{x[1].strip()}{x[0].strip()[0].upper()+x[0].strip()[1:]}\", axis=1\n",
    ")\n",
    "others[\"DependsOn Component\"] = \"\"\n",
    "others[\"Properties\"] = \"dataProperty\"\n",
    "others.loc[others[\"equalsValue\"] == \"other\", \"Parent\"] = \"Specification\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "equals_df = others.copy()\n",
    "equals_df = equals_df.drop(columns=[\"DependsOn\", \"Description\"])\n",
    "\n",
    "equals_df = equals_df.rename(\n",
    "    {\n",
    "        \"Attribute\": \"DependsOn\",\n",
    "        \"newDescription\": \"Description\",\n",
    "        \"equalsAttribute\": \"Attribute\",\n",
    "    },\n",
    "    axis=1,\n",
    ")[base_cols]\n",
    "\n",
    "equals_df[\"DependsOn Component\"] = \"\"\n",
    "\n",
    "equals_df[\"Valid Values\"] = \"\"\n",
    "\n",
    "equals_df[\"Properties\"] = \"ValidValue\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Update base attribute equals values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create mapping\n",
    "temp = others[\"equals_series\"].str.split(\"=\", expand=True)\n",
    "temp = temp.apply(lambda x: x.str.strip(), axis=1)\n",
    "temp = temp.rename({0: \"base_attribute\", 1: \"value_to_replace\"}, axis=1)\n",
    "temp[\"value_to_replace\"] = temp[\"value_to_replace\"].str.capitalize()\n",
    "\n",
    "# new value\n",
    "temp[\"new_value\"] = others[\"equalsAttribute\"]\n",
    "temp.reset_index(drop=True, inplace=True)\n",
    "\n",
    "temp = temp.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'captivityStatus': {'pattern': re.compile(r'(Captive)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'captiveCaptivityStatus'},\n",
       " 'hasAssayControl': {'pattern': re.compile(r'(Yes)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'TRUEHasAssayControl'},\n",
       " 'diagnosisStatus': {'pattern': re.compile(r'(True)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'TRUEDiagnosisStatus'},\n",
       " 'ethnicity': {'pattern': re.compile(r'(Hispanic or latino)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'Hispanic or LatinoEthnicity'},\n",
       " 'hasIonizationSource': {'pattern': re.compile(r'(Yes)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'TRUEHasIonizationSource'},\n",
       " 'msTarget': {'pattern': re.compile(r'(Targeted)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'TargetedMsTarget'},\n",
       " 'useReagent': {'pattern': re.compile(r'(Yes)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'TRUEUseReagent'},\n",
       " 'acquisitionBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherAcquisitionBatchSizeUnit'},\n",
       " 'batchSizeUnit': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherBatchSizeUnit'},\n",
       " 'controlType': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherControlType'},\n",
       " 'dnaBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherDnaBatchSizeUnit'},\n",
       " 'databaseName': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherDatabaseName'},\n",
       " 'databaseSource': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherDatabaseSource'},\n",
       " 'digestionMethod': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherDigestionMethod'},\n",
       " 'experimentalBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherExperimentalBatchSizeUnit'},\n",
       " 'extractionMethod': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherExtractionMethod'},\n",
       " 'lensVoltageUnits': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherLensVoltageUnits'},\n",
       " 'libraryPrep': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherLibraryPrep'},\n",
       " 'libraryPreparationMethod': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherLibraryPreparationMethod'},\n",
       " 'libraryVersion': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherLibraryVersion'},\n",
       " 'msAnalyteType': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherMsAnalyteType'},\n",
       " 'msAnalyzerTypeMS1': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherMsAnalyzerTypeMS1'},\n",
       " 'msAnalyzerTypeMS2': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherMsAnalyzerTypeMS2'},\n",
       " 'msAssayTechnique': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherMsAssayTechnique'},\n",
       " 'msInstrumentModel': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherMsInstrumentModel'},\n",
       " 'measurementTechnique': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherMeasurementTechnique'},\n",
       " 'cleavageAgents': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherCleavageAgents'},\n",
       " 'modificationParameters': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherModificationParameters'},\n",
       " 'platformLocation': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherPlatformLocation'},\n",
       " 'technologyPlatformVersion': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherTechnologyPlatformVersion'},\n",
       " 'ProcessingBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherProcessingBatchSizeUnit'},\n",
       " 'ProteomicsAssayType': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherProteomicsAssayType'},\n",
       " 'rnaBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherRnaBatchSizeUnit'},\n",
       " 'readLengthUnits': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherReadLengthUnits'},\n",
       " 'repositoryName': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherRepositoryName'},\n",
       " 'sampleBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherSampleBatchSizeUnit'},\n",
       " 'sampleIntdocution': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherSampleIntdocution'},\n",
       " 'sampleType': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherSampleType'},\n",
       " 'sequencingBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherSequencingBatchSizeUnit'},\n",
       " 'transcriptType': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherTranscriptType'},\n",
       " 'vaccumPressureUnits': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherVaccumPressureUnits'},\n",
       " 'useTreatment': {'pattern': re.compile(r'(Yes)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'TRUEUseTreatment'}}"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "replacements = {}\n",
    "\n",
    "for i, x in temp.iterrows():\n",
    "    ba, vtr, nv = x\n",
    "    replacements[ba] = {\n",
    "        \"pattern\": re.compile(\"(\" + vtr + \")\", flags=re.IGNORECASE),\n",
    "        \"repl\": nv,\n",
    "    }\n",
    "\n",
    "replacements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "attribute:  captivityStatus\n",
      "Index:  [13]\n",
      "--------------------\n",
      "attribute:  hasAssayControl\n",
      "Index:  [49]\n",
      "--------------------\n",
      "attribute:  diagnosisStatus\n",
      "Index:  [29]\n",
      "--------------------\n",
      "attribute:  ethnicity\n",
      "Index:  [37]\n",
      "--------------------\n",
      "attribute:  hasIonizationSource\n",
      "Index:  [50]\n",
      "--------------------\n",
      "attribute:  msTarget\n",
      "Index:  [71]\n",
      "--------------------\n",
      "attribute:  useReagent\n",
      "Index:  [173]\n",
      "--------------------\n",
      "attribute:  acquisitionBatchSizeUnit\n",
      "Index:  [2]\n",
      "--------------------\n",
      "attribute:  batchSizeUnit\n",
      "Index:  [11]\n",
      "--------------------\n",
      "attribute:  controlType\n",
      "Index:  [19]\n",
      "--------------------\n",
      "attribute:  dnaBatchSizeUnit\n",
      "Index:  [34]\n",
      "--------------------\n",
      "attribute:  databaseName\n",
      "Index:  [25]\n",
      "--------------------\n",
      "attribute:  databaseSource\n",
      "Index:  [26]\n",
      "--------------------\n",
      "attribute:  digestionMethod\n",
      "Index:  [30]\n",
      "--------------------\n",
      "attribute:  experimentalBatchSizeUnit\n",
      "experimentalBatchSizeUnit\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  extractionMethod\n",
      "Index:  [40]\n",
      "--------------------\n",
      "attribute:  lensVoltageUnits\n",
      "lensVoltageUnits\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  libraryPrep\n",
      "Index:  [60]\n",
      "--------------------\n",
      "attribute:  libraryPreparationMethod\n",
      "Index:  [61]\n",
      "--------------------\n",
      "attribute:  libraryVersion\n",
      "Index:  [62]\n",
      "--------------------\n",
      "attribute:  msAnalyteType\n",
      "Index:  [66]\n",
      "--------------------\n",
      "attribute:  msAnalyzerTypeMS1\n",
      "Index:  [67]\n",
      "--------------------\n",
      "attribute:  msAnalyzerTypeMS2\n",
      "Index:  [68]\n",
      "--------------------\n",
      "attribute:  msAssayTechnique\n",
      "Index:  [69]\n",
      "--------------------\n",
      "attribute:  msInstrumentModel\n",
      "Index:  [70]\n",
      "--------------------\n",
      "attribute:  measurementTechnique\n",
      "measurementTechnique\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  cleavageAgents\n",
      "Index:  [15]\n",
      "--------------------\n",
      "attribute:  modificationParameters\n",
      "Index:  [65]\n",
      "--------------------\n",
      "attribute:  platformLocation\n",
      "Index:  [77]\n",
      "--------------------\n",
      "attribute:  technologyPlatformVersion\n",
      "Index:  [159]\n",
      "--------------------\n",
      "attribute:  ProcessingBatchSizeUnit\n",
      "Index:  [81]\n",
      "--------------------\n",
      "attribute:  ProteomicsAssayType\n",
      "ProteomicsAssayType\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  rnaBatchSizeUnit\n",
      "Index:  [99]\n",
      "--------------------\n",
      "attribute:  readLengthUnits\n",
      "Index:  [85]\n",
      "--------------------\n",
      "attribute:  repositoryName\n",
      "Index:  [95]\n",
      "--------------------\n",
      "attribute:  sampleBatchSizeUnit\n",
      "Index:  [103]\n",
      "--------------------\n",
      "attribute:  sampleIntdocution\n",
      "sampleIntdocution\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  sampleType\n",
      "Index:  [106]\n",
      "--------------------\n",
      "attribute:  sequencingBatchSizeUnit\n",
      "Index:  [110]\n",
      "--------------------\n",
      "attribute:  transcriptType\n",
      "Index:  [164]\n",
      "--------------------\n",
      "attribute:  vaccumPressureUnits\n",
      "vaccumPressureUnits\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  useTreatment\n",
      "Index:  [174]\n",
      "--------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/p0/5m4pdsm55jn_d5nzbjv6bjf40000gq/T/ipykernel_11826/3370338766.py:42: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.\n",
      "  df[\"Attribute\"].str.contains(\n"
     ]
    }
   ],
   "source": [
    "for attribute in replacements.keys():\n",
    "    dm = code_equals_values(dm, replacements, attribute)\n",
    "    print(\"-\" * 20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add derived attributes from \"=\" valid values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dm[\"Properties\"] = \"dataProperty\"\n",
    "dm.update(others[base_cols])\n",
    "dm = pd.concat([dm, equals_df], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Do not need\n",
    "dm[\"DependsOn Component\"] = \"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Valid Values Work\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Not sure what happened here in the RFCs\n",
    "dm[\"Valid Values\"] = dm[\"Valid Values\"].replace(recoder_valid_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "attribute:  acquisitionBatchSizeUnit\n",
      "Index:  [2]\n",
      "--------------------\n",
      "attribute:  batchSizeUnit\n",
      "Index:  [11]\n",
      "--------------------\n",
      "attribute:  captivityStatus\n",
      "Index:  [13]\n",
      "--------------------\n",
      "attribute:  cellType\n",
      "Index:  [14]\n",
      "--------------------\n",
      "attribute:  cleavageAgents\n",
      "Index:  [15]\n",
      "--------------------\n",
      "attribute:  cohort\n",
      "Index:  [16]\n",
      "--------------------\n",
      "attribute:  consentGroupID\n",
      "Index:  [18]\n",
      "--------------------\n",
      "attribute:  controlType\n",
      "Index:  [19]\n",
      "--------------------\n",
      "attribute:  conversionRatioUnits\n",
      "Index:  [21]\n",
      "--------------------\n",
      "attribute:  countryCode\n",
      "Index:  [22]\n",
      "--------------------\n",
      "attribute:  databaseName\n",
      "Index:  [25]\n",
      "--------------------\n",
      "attribute:  databaseSource\n",
      "Index:  [26]\n",
      "--------------------\n",
      "attribute:  diagnosis\n",
      "Index:  [28]\n",
      "--------------------\n",
      "attribute:  digestionMethod\n",
      "Index:  [30]\n",
      "--------------------\n",
      "attribute:  dnaBatchID\n",
      "Index:  [32]\n",
      "--------------------\n",
      "attribute:  dnaBatchSizeUnit\n",
      "Index:  [34]\n",
      "--------------------\n",
      "attribute:  enrichmentMethod\n",
      "Index:  [35]\n",
      "--------------------\n",
      "attribute:  ethnicity\n",
      "Index:  [37]\n",
      "--------------------\n",
      "attribute:  experiementalBatchSizeUnit\n",
      "Index:  [39]\n",
      "--------------------\n",
      "attribute:  extractionMethod\n",
      "Index:  [40]\n",
      "--------------------\n",
      "attribute:  fastingState\n",
      "Index:  [42]\n",
      "--------------------\n",
      "attribute:  fieldCenterCode\n",
      "Index:  [43]\n",
      "--------------------\n",
      "attribute:  isPostMortem\n",
      "Index:  [53]\n",
      "--------------------\n",
      "attribute:  lensVoltagesUnit \n",
      "Index:  [58]\n",
      "--------------------\n",
      "attribute:  libraryPrep\n",
      "Index:  [60]\n",
      "--------------------\n",
      "attribute:  libraryPreparationMethod\n",
      "Index:  [61]\n",
      "--------------------\n",
      "attribute:  libraryVersion\n",
      "Index:  [62]\n",
      "--------------------\n",
      "attribute:  lifeStage\n",
      "Index:  [63]\n",
      "--------------------\n",
      "attribute:  modificationParameters\n",
      "Index:  [65]\n",
      "--------------------\n",
      "attribute:  msAnalyteType\n",
      "Index:  [66]\n",
      "--------------------\n",
      "attribute:  msAnalyzerTypeMS1\n",
      "Index:  [67]\n",
      "--------------------\n",
      "attribute:  msAnalyzerTypeMS2\n",
      "Index:  [68]\n",
      "--------------------\n",
      "attribute:  msAssayTechnique\n",
      "Index:  [69]\n",
      "--------------------\n",
      "attribute:  msInstrumentModel\n",
      "Index:  [70]\n",
      "--------------------\n",
      "attribute:  nucleicAcidSource\n",
      "Index:  [72]\n",
      "--------------------\n",
      "attribute:  organ\n",
      "Index:  [73]\n",
      "--------------------\n",
      "attribute:  platformLocation\n",
      "Index:  [77]\n",
      "--------------------\n",
      "attribute:  processingBatchSizeUnit\n",
      "Index:  [81]\n",
      "--------------------\n",
      "attribute:  protoemicsAssayType\n",
      "Index:  [82]\n",
      "--------------------\n",
      "attribute:  race\n",
      "Index:  [83]\n",
      "--------------------\n",
      "attribute:  readLengthUnits\n",
      "Index:  [85]\n",
      "--------------------\n",
      "attribute:  repositoryName\n",
      "Index:  [95]\n",
      "--------------------\n",
      "attribute:  rnaBatchSizeUnit\n",
      "Index:  [99]\n",
      "--------------------\n",
      "attribute:  sampleBatchSizeUnit\n",
      "Index:  [103]\n",
      "--------------------\n",
      "attribute:  sampleIntroduction\n",
      "Index:  [104]\n",
      "--------------------\n",
      "attribute:  sampleType\n",
      "Index:  [106]\n",
      "--------------------\n",
      "attribute:  samplingAge\n",
      "Index:  [107]\n",
      "--------------------\n",
      "attribute:  sequencingBatchSizeUnit\n",
      "Index:  [110]\n",
      "--------------------\n",
      "attribute:  sex\n",
      "Index:  [111]\n",
      "--------------------\n",
      "attribute:  speciesGroup\n",
      "Index:  [113]\n",
      "--------------------\n",
      "attribute:  specimenAge\n",
      "Index:  [151]\n",
      "--------------------\n",
      "attribute:  specimenSource\n",
      "Index:  [154]\n",
      "--------------------\n",
      "attribute:  specimenType\n",
      "Index:  [155]\n",
      "--------------------\n",
      "attribute:  technologyPlatformVersion\n",
      "Index:  [159]\n",
      "--------------------\n",
      "attribute:  tissue\n",
      "Index:  [160]\n",
      "--------------------\n",
      "attribute:  transcriptType\n",
      "Index:  [164]\n",
      "--------------------\n",
      "attribute:  vacuumPressureUnit\n",
      "Index:  [176]\n",
      "--------------------\n",
      "attribute:  visitCode\n",
      "Index:  [178]\n",
      "--------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/p0/5m4pdsm55jn_d5nzbjv6bjf40000gq/T/ipykernel_11826/519273947.py:5: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.\n",
      "  dm[\"Valid Values\"].str.contains(pattern, flags=re.IGNORECASE, regex=True)\n",
      "/var/folders/p0/5m4pdsm55jn_d5nzbjv6bjf40000gq/T/ipykernel_11826/3370338766.py:42: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.\n",
      "  df[\"Attribute\"].str.contains(\n"
     ]
    }
   ],
   "source": [
    "# valid values that contain other\n",
    "pattern = \"([Oo]ther)\"\n",
    "\n",
    "pure_others = dm[\n",
    "    dm[\"Valid Values\"].str.contains(pattern, flags=re.IGNORECASE, regex=True)\n",
    "].copy()\n",
    "\n",
    "pure_others.loc[:, \"replacement_value\"] = pure_others.loc[:, \"Attribute\"].apply(\n",
    "    lambda x: \"Other\" + (x[0].upper() + x[1:])\n",
    ")\n",
    "\n",
    "regex_dict = {}\n",
    "\n",
    "for i, r in pure_others.iterrows():\n",
    "    regex_dict[r[\"Attribute\"]] = {\n",
    "        \"pattern\": re.compile(pattern, flags=re.IGNORECASE),\n",
    "        \"repl\": r[\"replacement_value\"],\n",
    "    }\n",
    "\n",
    "# json_formatted = json.dumps(regex_dict, indent=4)\n",
    "# print(json_formatted)\n",
    "\n",
    "for attribute in regex_dict.keys():\n",
    "    dm = code_equals_values(dm, regex_dict, attribute)\n",
    "    print(\"-\" * 20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cleanup valid values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fuzzy matching to find misspellings\n",
    "# Fuzzy matching\n",
    "\n",
    "valid_values = \",\".join(dm[\"Valid Values\"])\n",
    "valid_values = valid_values.split(\",\")\n",
    "valid_values = list(np.unique(valid_values))\n",
    "valid_values = [v.strip() for v in valid_values if len(v) > 0]\n",
    "\n",
    "\n",
    "scores = {}\n",
    "for v in valid_values:\n",
    "    scores[v] = {}\n",
    "    for v2 in valid_values:\n",
    "        if v == v2:\n",
    "            next\n",
    "        else:\n",
    "            score = fuzz.ratio(v.lower(), v2.lower())\n",
    "            if score == 100:\n",
    "                scores[v][v2] = score\n",
    "    if len(scores[v]) == 0:\n",
    "        scores.pop(v)\n",
    "\n",
    "# create recoding variables off fuzzy matching\n",
    "new_values_recoded = []\n",
    "\n",
    "for v in scores.values():\n",
    "    new_values_recoded.append(list(v.keys())[0].title())\n",
    "\n",
    "new_values_recoded = np.unique(new_values_recoded)\n",
    "\n",
    "# recoder_valid_values = {}\n",
    "for nv in new_values_recoded:\n",
    "    recoder_valid_values[re.compile(nv, flags=re.IGNORECASE)] = nv\n",
    "    # recoder_valid_values.append(value_add)\n",
    "\n",
    "dm[[\"Valid Values\", \"multivalue\"]] = (\n",
    "    dm[[\"Valid Values\", \"multivalue\"]]\n",
    "    .replace(recoder_valid_values, regex=True)\n",
    "    .fillna(\"\")\n",
    "    .applymap(lambda x: clean_list(x))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>UsedIn</th>\n",
       "      <th>Description</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10x</td>\n",
       "      <td>libraryPreparationMethod</td>\n",
       "      <td>Method by which library was prepared</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ValidValue</td>\n",
       "      <td></td>\n",
       "      <td>Microbiome,RNAseq,Whole Genome Sequencing,bsSe...</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>STRING</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10x Visium Spatial Gene Expression</td>\n",
       "      <td>platformLocation,technologyPlatformVersion</td>\n",
       "      <td>The name of the laboratory, facility, vendor, ...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ValidValue</td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human,Microbiome,Proteomics,RNAse...</td>\n",
       "      <td></td>\n",
       "      <td>DSLWG,Sage Bionetworks,http://purl.obolibrary....</td>\n",
       "      <td>STRING</td>\n",
       "      <td>DSLWG,Sage Bionetworks,http://purl.obolibrary....</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AFU</td>\n",
       "      <td>acquisitionBatchSizeUnit,batchSizeUnit,convers...</td>\n",
       "      <td>Unit of treatment amount,Unit of lens voltages...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.)|([0-9]+)</td>\n",
       "      <td>Genotyping,Metabolomics Human,Microbiome,RNAse...</td>\n",
       "      <td></td>\n",
       "      <td>ImmPort,Proposed minimum metadata relative to ...</td>\n",
       "      <td>STRING</td>\n",
       "      <td>ImmPort,Proposed minimum metadata relative to ...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AI</td>\n",
       "      <td>acquisitionBatchSizeUnit,batchSizeUnit,convers...</td>\n",
       "      <td>Unit of treatment amount,Unit of lens voltages...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.)|([0-9]+)</td>\n",
       "      <td>Genotyping,Metabolomics Human,Microbiome,RNAse...</td>\n",
       "      <td></td>\n",
       "      <td>ImmPort,Proposed minimum metadata relative to ...</td>\n",
       "      <td>STRING</td>\n",
       "      <td>ImmPort,Proposed minimum metadata relative to ...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AIBL pool</td>\n",
       "      <td>controlType</td>\n",
       "      <td>Control samples suitable for normalization and...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ValidValue</td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>https://doi.org/10.1101/2020.05.19.105197https...</td>\n",
       "      <td>STRING</td>\n",
       "      <td>https://doi.org/10.1101/2020.05.19.105197https...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Attribute  \\\n",
       "0                                 10x   \n",
       "1  10x Visium Spatial Gene Expression   \n",
       "2                                 AFU   \n",
       "3                                  AI   \n",
       "4                           AIBL pool   \n",
       "\n",
       "                                              UsedIn  \\\n",
       "0                           libraryPreparationMethod   \n",
       "1         platformLocation,technologyPlatformVersion   \n",
       "2  acquisitionBatchSizeUnit,batchSizeUnit,convers...   \n",
       "3  acquisitionBatchSizeUnit,batchSizeUnit,convers...   \n",
       "4                                        controlType   \n",
       "\n",
       "                                         Description Required DependsOn  \\\n",
       "0               Method by which library was prepared    False             \n",
       "1  The name of the laboratory, facility, vendor, ...    False             \n",
       "2  Unit of treatment amount,Unit of lens voltages...    False             \n",
       "3  Unit of treatment amount,Unit of lens voltages...    False             \n",
       "4  Control samples suitable for normalization and...    False             \n",
       "\n",
       "  DependsOn Component  Properties                         Validation Rules  \\\n",
       "0                      ValidValue                                            \n",
       "1                      ValidValue                                            \n",
       "2                      ValidValue  regex search ([0-9]+\\.[0-9]*.)|([0-9]+)   \n",
       "3                      ValidValue  regex search ([0-9]+\\.[0-9]*.)|([0-9]+)   \n",
       "4                      ValidValue                                            \n",
       "\n",
       "                                            Template Parent  \\\n",
       "0  Microbiome,RNAseq,Whole Genome Sequencing,bsSe...          \n",
       "1  Metabolomics Human,Microbiome,Proteomics,RNAse...          \n",
       "2  Genotyping,Metabolomics Human,Microbiome,RNAse...          \n",
       "3  Genotyping,Metabolomics Human,Microbiome,RNAse...          \n",
       "4                                 Metabolomics Human          \n",
       "\n",
       "                                              Source    Type  \\\n",
       "0                                   Sage Bionetworks  STRING   \n",
       "1  DSLWG,Sage Bionetworks,http://purl.obolibrary....  STRING   \n",
       "2  ImmPort,Proposed minimum metadata relative to ...  STRING   \n",
       "3  ImmPort,Proposed minimum metadata relative to ...  STRING   \n",
       "4  https://doi.org/10.1101/2020.05.19.105197https...  STRING   \n",
       "\n",
       "                                            Ontology multivalue  \n",
       "0                                   Sage Bionetworks      False  \n",
       "1  DSLWG,Sage Bionetworks,http://purl.obolibrary....      False  \n",
       "2  ImmPort,Proposed minimum metadata relative to ...      False  \n",
       "3  ImmPort,Proposed minimum metadata relative to ...      False  \n",
       "4  https://doi.org/10.1101/2020.05.19.105197https...      False  "
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm[\"Valid Values\"] = (\n",
    "    dm[\"Valid Values\"].apply(lambda x: clean_list(x)\n",
    "                             ).apply(lambda x: x.split(\",\"))\n",
    ")\n",
    "\n",
    "# Expand each valid value into its own row\n",
    "dm_vv = dm.explode(\"Valid Values\")\n",
    "\n",
    "# join valid values back together in data model as string\n",
    "dm[\"Valid Values\"] = dm[\"Valid Values\"].apply(lambda x: \",\".join(x).strip(\",\"))\n",
    "\n",
    "# Group valid values to create unique attribute and trace where value is used in another attribute and template\n",
    "dm_vv = (\n",
    "    dm_vv.dropna(subset=\"Valid Values\")\n",
    "    .groupby(\"Valid Values\")\n",
    "    .agg(lambda x: \",\".join(set(x.astype(str))).strip(\",\"))\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "dm_vv[\"Properties\"] = \"ValidValue\"\n",
    "dm_vv[\"Required\"] = \"False\"\n",
    "\n",
    "# rename for concatenating with data model\n",
    "dm_vv = dm_vv.rename({\"Valid Values\": \"Attribute\",\n",
    "                     \"Attribute\": \"UsedIn\"}, axis=1)\n",
    "\n",
    "dm_vv[list_cols] = dm_vv[list_cols].applymap(clean_list)\n",
    "\n",
    "dm_vv = dm_vv[dm_vv[\"Attribute\"] != \"\"].reset_index(drop=True)\n",
    "\n",
    "# clean up type\n",
    "dm_vv[\"Type\"] = \"STRING\"\n",
    "\n",
    "# cleanup multivalue\n",
    "dm_vv[\"multivalue\"] = \"False\"\n",
    "\n",
    "dm_vv.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(435, 14)"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm_vv.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dm[\"Valid Values\"] = (\n",
    "#     dm[\"Valid Values\"]\n",
    "#     .replace(recoder_valid_values, regex=True)\n",
    "#     .apply(lambda x: clean_list(x))\n",
    "# )\n",
    "\n",
    "# valid_values = list(np.unique(\",\".join(dm[\"Valid Values\"]).split(\",\")))\n",
    "\n",
    "# valid_values = [v.strip() for v in valid_values if len(v) > 0]\n",
    "\n",
    "# valid_values_df = pd.DataFrame({\"Attribute\": pd.Series(valid_values)})\n",
    "\n",
    "# valid_values_df[\"Properties\"] = \"validValue\"\n",
    "# valid_values_df[\"Required\"] = \"False\"\n",
    "\n",
    "# valid_values_df = valid_values_df[\n",
    "#     ~valid_values_df[\"Attribute\"].isin(dm[\"Attribute\"].tolist())\n",
    "# ]\n",
    "\n",
    "# valid_values_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(239, 14)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(674, 15)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# adding valid values found in attribute columns\n",
    "print(dm.shape)\n",
    "\n",
    "dm2 = pd.concat([dm, dm_vv], axis=0, ignore_index=True)\n",
    "\n",
    "display(dm2.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29\n"
     ]
    }
   ],
   "source": [
    "print(sum(dm2.duplicated(subset=\"Attribute\", keep=False)))\n",
    "\n",
    "# dm2.loc[dm2.duplicated(subset=\"Attribute\", keep=False),].sort_values(by=\"Attribute\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "tags": [
     "new attributes"
    ]
   },
   "outputs": [],
   "source": [
    "# Create measurement unit attributes\n",
    "# Separate out measurement units\n",
    "r = re.compile(\"(^Other)\")\n",
    "\n",
    "measurement_units = np.unique(\n",
    "    \",\".join(\n",
    "        dm.loc[\n",
    "            dm[\"Valid Values\"].str.contains(\"units\", regex=True), \"Valid Values\"\n",
    "        ].values.tolist()\n",
    "    ).split(\",\")\n",
    ")\n",
    "\n",
    "measurement_units = [\n",
    "    x\n",
    "    for x in measurement_units\n",
    "    if x not in [\"Not Specified\", \"Other\", \"Unknown\", \"Not Available\"]\n",
    "    and not bool(r.search(x))\n",
    "]\n",
    "\n",
    "dm2.loc[dm2[\"Attribute\"].isin(measurement_units), \"Parent\"] = \"MeasurementUnit\"\n",
    "dm2.loc[dm2[\"Attribute\"].isin(measurement_units), \"Description\"] = \"Measurement unit\"\n",
    "dm2.loc[dm2[\"Attribute\"].isin(measurement_units), \"Type\"] = \"STRING\"\n",
    "dm2.loc[dm2[\"Attribute\"].isin(measurement_units), \"multivalue\"] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {
    "tags": [
     "cleanup"
    ]
   },
   "outputs": [],
   "source": [
    "# Nonsense attributes\n",
    "dm2 = dm2.drop(\n",
    "    index=dm2.loc[dm2[\"Attribute\"].str.contains(\n",
    "        \"Possible values\"),].index.tolist()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'f': {'F': 100}, 'F': {'f': 100}}"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fuzzy matching to find misspellings\n",
    "# Fuzzy matching\n",
    "\n",
    "valid_values = dm2[\"Attribute\"].replace(\n",
    "    recoder_valid_values, regex=True).tolist()\n",
    "\n",
    "scores = {}\n",
    "for v in valid_values:\n",
    "    scores[v] = {}\n",
    "    for v2 in valid_values:\n",
    "        if v == v2:\n",
    "            next\n",
    "        else:\n",
    "            score = fuzz.ratio(v.lower(), v2.lower())\n",
    "            if score == 100:\n",
    "                scores[v][v2] = score\n",
    "    if len(scores[v]) == 0:\n",
    "        scores.pop(v)\n",
    "\n",
    "scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm2[\"Attribute\"] = dm2[\"Attribute\"].replace(recoder_valid_values, regex=True)\n",
    "\n",
    "dm2[[\"Valid Values\", \"multivalue\"]] = (\n",
    "    dm2[[\"Valid Values\", \"multivalue\"]]\n",
    "    .fillna(\"\")\n",
    "    .astype(str)\n",
    "    .replace(recoder_valid_values, regex=True)\n",
    "    .applymap(lambda x: clean_list(x))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dm2 shape: (648, 15)\n",
      "Duplicates: 0\n"
     ]
    }
   ],
   "source": [
    "dm2 = (\n",
    "    dm2.dropna(subset=\"Attribute\")\n",
    "    .groupby(\"Attribute\")\n",
    "    .agg(lambda x: \",\".join(set(x.astype(str))).strip(\",\"))\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "dm2 = dm2.drop(index=dm2[dm2[\"Attribute\"] == \"f\"].index.tolist()).reset_index(drop=True)\n",
    "\n",
    "dm2[list_cols] = dm2[list_cols].applymap(clean_list)\n",
    "\n",
    "print(f\"dm2 shape: {dm2.shape}\")\n",
    "print(f'Duplicates: {sum(dm2.duplicated(subset=\"Attribute\", keep=False))}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check columns for Speical Characters\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Cardiovascular Health Study (CHS)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Cerebrospinal Fluid (CSF)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>Illumina Infinium MethylationEPIC BeadChip v1.0 (850k)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>Illumina Infinium MethylationEPIC BeadChip v2.0 (935k)Illumina MiSeq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>MassBank of North America (MoNA)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>Study of Osteoporotic Fractures (SOF)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>The Health and Aging and Body Composition Study (HealthABC)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>The Osteoporotic Fractures in Men (MrOS) Study</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                Attribute\n",
       "39                                      Cardiovascular Health Study (CHS)\n",
       "43                                              Cerebrospinal Fluid (CSF)\n",
       "111                Illumina Infinium MethylationEPIC BeadChip v1.0 (850k)\n",
       "112  Illumina Infinium MethylationEPIC BeadChip v2.0 (935k)Illumina MiSeq\n",
       "159                                      MassBank of North America (MoNA)\n",
       "309                                 Study of Osteoporotic Fractures (SOF)\n",
       "320           The Health and Aging and Body Composition Study (HealthABC)\n",
       "321                        The Osteoporotic Fractures in Men (MrOS) Study"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "check_cols = [\"Attribute\"]\n",
    "\n",
    "mask = np.column_stack(\n",
    "    [dm2[col].str.contains(\"\\(|\\)\", na=False) for col in dm2[check_cols]]\n",
    ")\n",
    "\n",
    "with pd.option_context(\"display.max_colwidth\", None):\n",
    "    display(dm2[check_cols].loc[mask.any(axis=1)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(660, 15)\n"
     ]
    }
   ],
   "source": [
    "dm2 = pd.concat([dm2, templates], axis=0, ignore_index=True)\n",
    "\n",
    "print(dm2.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm2[[\"Valid Values\", \"Template\", \"Ontology\", \"Source\"]] = (\n",
    "    dm2[[\"Valid Values\", \"Template\", \"Ontology\", \"Source\"]]\n",
    "    .fillna(\"\")\n",
    "    .applymap(clean_list)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Recode required columns and fix spelling mistakes\n",
    "required_recoder = {\"0.0\": \"False\", \"1.0\": \"True\", \"FASLSE\": \"False\"}\n",
    "\n",
    "dm2[\"Required\"] = dm2[\"Required\"].replace(required_recoder)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Last bit of cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove measurement technique dependency from biospecimen human\n",
    "dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"] = (\n",
    "    dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"]\n",
    "    .values[0]\n",
    "    .replace(\"measurementTechnique,\", \"\")\n",
    ")\n",
    "bio_measure_technique_index = dm2.query(\n",
    "    'Attribute == \"measurementTechnique\" and Parent == \"Biospecimen human\"'\n",
    ")\n",
    "\n",
    "if len(bio_measure_technique_index.index) > 0:\n",
    "    dm2 = dm2.drop(\n",
    "        index=bio_measure_technique_index.index[0]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove measurement technique dependency from biospecimen human\n",
    "dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"] = (\n",
    "    dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"]\n",
    "    .values[0]\n",
    "    .replace(\"specifyMeasurementTechnique,\", \"\")\n",
    ")\n",
    "bio_measure_technique_index = dm2.query(\n",
    "    'Attribute == \"specifyMeasurementTechnique\" and Parent == \"Biospecimen human\"'\n",
    ")\n",
    "\n",
    "\n",
    "if len(bio_measure_technique_index.index) > 0:\n",
    "    dm2 = dm2.drop(\n",
    "        index=bio_measure_technique_index.index[0]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove measurement technique dependency from biospecimen human\n",
    "dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"] = (\n",
    "    dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"]\n",
    "    .values[0]\n",
    "    .replace(\"OtherMeasurementTechnique,\", \"\")\n",
    ")\n",
    "bio_measure_technique_index = dm2.query(\n",
    "    'Attribute == \"OtherMeasurementTechnique\"')\n",
    "\n",
    "\n",
    "if len(bio_measure_technique_index.index) > 0:\n",
    "    dm2 = dm2.drop(\n",
    "        index=bio_measure_technique_index.index[0]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm2.loc[dm2[dm2[\"Attribute\"] == \"visitCode\"].index.values[0], \"Valid Values\"] = \"\"\n",
    "\n",
    "dm2.loc[dm2[dm2[\"Attribute\"] == \"visitCode\"].index.values[0],\n",
    "        \"Validation Rules\"] = \"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extra comma at beginning of valid values\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validation Rules\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "mixed_attrs = [\n",
    "    {\"attribute\": \"tissueWeight\", \"val_type\": \"mixed float\", \"regex\": \"regex search\"},\n",
    "    {\"attribute\": \"tissueVolume\", \"val_type\": \"mixed float\", \"regex\": \"regex search\"},\n",
    "    {\"attribute\": \"specimenAge\", \"val_type\": \"mixed integer\", \"regex\": \"regex search\"},\n",
    "    {\"attribute\": \"samplingAge\", \"val_type\": \"mixed integer\", \"regex\": \"regex search\"},\n",
    "    {\"attribute\": \"age\", \"val_type\": \"mixed integer\", \"regex\": \"regex search\"},\n",
    "]\n",
    "\n",
    "for ma in mixed_attrs:\n",
    "    attribute = ma[\"attribute\"]\n",
    "    val_type = ma[\"val_type\"]\n",
    "\n",
    "    # get indexes for new validation rules based on attribute\n",
    "    indexes = dm2[dm2[\"Attribute\"] == attribute].index.tolist()\n",
    "\n",
    "    for i in indexes:\n",
    "        if val_type == \"integer\":\n",
    "            first_part = \"[0-9]+\"\n",
    "        elif val_type == \"float\":\n",
    "            first_part = \"^\\d*?\\.?\\d$\"\n",
    "        elif val_type == \"mixed integer\":\n",
    "            regex = \"regex search\"\n",
    "            num_match = \"^\\d*?\"\n",
    "        elif val_type == \"mixed float\":\n",
    "            regex = \"regex search\"\n",
    "            num_match = \"^\\d*?\\.?\\d$\"\n",
    "            # All valid values are applicable\n",
    "            new_string = (\n",
    "                regex\n",
    "                + num_match\n",
    "                + \"|\"\n",
    "                + \"\"\n",
    "                + \"|\".join(dm2.loc[i, \"Valid Values\"].split(\",\"))\n",
    "            )\n",
    "\n",
    "        dm2.loc[i, \"Validation Rules\"] = new_string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['',\n",
       " 'regex search ([0-9]+\\\\.[0-9]*.)|([0-9]+)',\n",
       " 'regex search ([0-9]+),regex search ([0-9]+\\\\.[0-9]*.)|([0-9]+)',\n",
       " 'regex search ([0-9]+)',\n",
       " 'regex search^\\\\d*?\\\\.?\\\\d$|Not applicable|Not collected|Unknown',\n",
       " nan]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm2[\"Validation Rules\"].unique().tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Building Dependencies\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "dependencies = {\n",
    "    \"specimenID\": \"matchAtLeastOne Biospecimenhuman.specimenID value\",\n",
    "    \"individualID\": \"matchExactlyOne IndividualHuman.individualID set\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "for k, v in dependencies.items():\n",
    "    indexes = dm2[dm2[\"Attribute\"] == k][\"Validation Rules\"].index.values\n",
    "    dm2.loc[indexes, \"Validation Rules\"] = v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm2.loc[3, \"Valid Values\"] = dm2.loc[3, \"Valid Values\"].replace(\n",
    "    \",Whole Genome Sequencing\", \"\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# More Adjustments\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>UsedIn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>visitCode</td>\n",
       "      <td>Indicate which longitudinal visit for the individual the data comes from, provided by the data contributor's data dictionary</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>number</td>\n",
       "      <td>Data Contributor,Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Attribute  \\\n",
       "646  visitCode   \n",
       "\n",
       "                                                                                                                      Description  \\\n",
       "646  Indicate which longitudinal visit for the individual the data comes from, provided by the data contributor's data dictionary   \n",
       "\n",
       "    Valid Values Required DependsOn DependsOn Component Properties  \\\n",
       "646                  True                                            \n",
       "\n",
       "    Validation Rules          Template Parent            Source    Type  \\\n",
       "646                   Individual Human         Sage Bionetworks  number   \n",
       "\n",
       "                              Ontology multivalue UsedIn  \n",
       "646  Data Contributor,Sage Bionetworks      False         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with pd.option_context(\"display.max_colwidth\", None):\n",
    "    display(dm2.query('Attribute == \"visitCode\"'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "hard_coded_valid_values = [\n",
    "    {\n",
    "        \"attribute\": \"visitCode\",\n",
    "        \"valid_value\": \"1,2,3,4,Other,Unknown,Not collected,Not applicable\",\n",
    "    },\n",
    "    {\"attribute\": \"tissueWeight\", \"valid_value\": \"\"},\n",
    "    {\"attribute\": \"consentGroupID\", \"valid_value\": \"1,2,3\"},\n",
    "    {\"attribute\": \"samplingAge\", \"valid_value\": \"\"},\n",
    "    {\"attribute\": \"specimenAge\", \"valid_value\": \"\"},\n",
    "    {\"attribute\": \"age\", \"valid_value\": \"\"},\n",
    "]\n",
    "\n",
    "for h in hard_coded_valid_values:\n",
    "    dm2.loc[\n",
    "        dm2[dm2[\"Attribute\"] == h[\"attribute\"]].index.values[0], \"Valid Values\"\n",
    "    ] = h[\"valid_value\"]\n",
    "\n",
    "# # Only needed for the first time since the Valid values were TBD\n",
    "dm2 = rewrite_df_value(dm2, \"Attribute\", \"assay\", \"Valid Values\", np.nan)\n",
    "dm2 = dm2.replace(\"\", np.nan)\n",
    "# Add valid values from the AD model\n",
    "dm2[\"Valid Values\"] = dm2[\"Valid Values\"].fillna(\n",
    "    dm2[\"Attribute\"].map(df_new_attrs.set_index(\"Attribute\")[\"Valid Values\"])\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>UsedIn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>age</td>\n",
       "      <td>Age of the individual (age in years of the individual at first recorded study event (enrollment, visit, observation, sample collection, survey completion, etc.)</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>regex search^\\d*?\\.?\\d$|Not applicable|Not collected|Unknown</td>\n",
       "      <td>Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383</th>\n",
       "      <td>consentGroupID</td>\n",
       "      <td>Indicate the consent group for the individual, provided by the data contributor's data dictionary</td>\n",
       "      <td>1,2,3</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Data Contributor,Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552</th>\n",
       "      <td>samplingAge</td>\n",
       "      <td>The calculated age of the sample, measurement is determined or coded by the data contributor.</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>regex search^\\d*?\\.?\\d$|Not applicable|Not collected|Unknown</td>\n",
       "      <td>Biospecimen human,Biospecimen nonHuman</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>603</th>\n",
       "      <td>specimenAge</td>\n",
       "      <td>The subject's age at the time of specimen extraction is the individual's age (since birth) at the time a given specimen was extracted. Measured as age in years.</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>regex search^\\d*?\\.?\\d$|Not applicable|Not collected|Unknown</td>\n",
       "      <td>Biospecimen human,Biospecimen nonHuman</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>618</th>\n",
       "      <td>tissueWeight</td>\n",
       "      <td>The mass of the tissue specimen. Measured in mg.</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>regex search^\\d*?\\.?\\d$|Not applicable|Not collected|Unknown</td>\n",
       "      <td>Biospecimen human,Biospecimen nonHuman</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>visitCode</td>\n",
       "      <td>Indicate which longitudinal visit for the individual the data comes from, provided by the data contributor's data dictionary</td>\n",
       "      <td>1,2,3,4,Other,Unknown,Not collected,Not applicable</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>number</td>\n",
       "      <td>Data Contributor,Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Attribute  \\\n",
       "362             age   \n",
       "383  consentGroupID   \n",
       "552     samplingAge   \n",
       "603     specimenAge   \n",
       "618    tissueWeight   \n",
       "646       visitCode   \n",
       "\n",
       "                                                                                                                                                          Description  \\\n",
       "362  Age of the individual (age in years of the individual at first recorded study event (enrollment, visit, observation, sample collection, survey completion, etc.)   \n",
       "383                                                                 Indicate the consent group for the individual, provided by the data contributor's data dictionary   \n",
       "552                                                                    The calculated age of the sample, measurement is determined or coded by the data contributor.    \n",
       "603  The subject's age at the time of specimen extraction is the individual's age (since birth) at the time a given specimen was extracted. Measured as age in years.   \n",
       "618                                                                                                                  The mass of the tissue specimen. Measured in mg.   \n",
       "646                                      Indicate which longitudinal visit for the individual the data comes from, provided by the data contributor's data dictionary   \n",
       "\n",
       "                                           Valid Values Required DependsOn  \\\n",
       "362                                                         True             \n",
       "383                                               1,2,3     True             \n",
       "552                                                         True             \n",
       "603                                                         True             \n",
       "618                                                         True             \n",
       "646  1,2,3,4,Other,Unknown,Not collected,Not applicable     True             \n",
       "\n",
       "    DependsOn Component Properties  \\\n",
       "362                                  \n",
       "383                                  \n",
       "552                                  \n",
       "603                                  \n",
       "618                                  \n",
       "646                                  \n",
       "\n",
       "                                                 Validation Rules  \\\n",
       "362  regex search^\\d*?\\.?\\d$|Not applicable|Not collected|Unknown   \n",
       "383                                                                 \n",
       "552  regex search^\\d*?\\.?\\d$|Not applicable|Not collected|Unknown   \n",
       "603  regex search^\\d*?\\.?\\d$|Not applicable|Not collected|Unknown   \n",
       "618  regex search^\\d*?\\.?\\d$|Not applicable|Not collected|Unknown   \n",
       "646                                                                 \n",
       "\n",
       "                                   Template Parent            Source    Type  \\\n",
       "362                        Individual Human         Sage Bionetworks  string   \n",
       "383                        Individual Human         Sage Bionetworks  string   \n",
       "552  Biospecimen human,Biospecimen nonHuman         Sage Bionetworks  string   \n",
       "603  Biospecimen human,Biospecimen nonHuman         Sage Bionetworks  string   \n",
       "618  Biospecimen human,Biospecimen nonHuman         Sage Bionetworks  string   \n",
       "646                        Individual Human         Sage Bionetworks  number   \n",
       "\n",
       "                              Ontology multivalue UsedIn  \n",
       "362                   Sage Bionetworks      False         \n",
       "383  Data Contributor,Sage Bionetworks      False         \n",
       "552                   Sage Bionetworks      False         \n",
       "603                   Sage Bionetworks      False         \n",
       "618                   Sage Bionetworks      False         \n",
       "646  Data Contributor,Sage Bionetworks      False         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "checks = [c[\"attribute\"] for c in hard_coded_valid_values]\n",
    "\n",
    "with pd.option_context(\"display.max_colwidth\", None):\n",
    "    display(dm2.query(f\"Attribute in @checks\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(660, 15)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(660, 15)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(dm2.shape)\n",
    "\n",
    "dm2 = (\n",
    "    dm2.fillna(\"\").drop_duplicates(subset=[\"Attribute\"]).reset_index(drop=True)\n",
    ")  # 'DependsOn', 'Properties',\n",
    "\n",
    "display(dm2.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add ADKP attrs to dm\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\n",
    "    \"https://raw.githubusercontent.com/adknowledgeportal/data-models/main/AD.model.csv\"\n",
    ")\n",
    "\n",
    "# preprocess AD data model to remove duplicates\n",
    "df = df.sort_values(by=[\"Attribute\", \"Valid Values\"]).reset_index(drop=True)\n",
    "df = df.drop_duplicates(keep=\"first\", subset=[\"Attribute\"])\n",
    "\n",
    "df.loc[\n",
    "    df.query('Attribute.str.contains(\"template\")',\n",
    "             engine=\"python\").index.tolist(),\n",
    "    \"Properties\",\n",
    "] = \"template\"\n",
    "\n",
    "attrs_interest = [\n",
    "    \"analysisType\",\n",
    "    \"analysisType\",\n",
    "    \"analytical covariates\",\n",
    "    \"assay\",\n",
    "    \"assay\",\n",
    "    \"biospecimen\",\n",
    "    \"consortium\",\n",
    "    \"data dictionary\",\n",
    "    \"dataSubtype\",\n",
    "    \"dataType\",\n",
    "    \"fileFormat\",\n",
    "    \"grant\",\n",
    "    \"ID mapping\",\n",
    "    \"individual\",\n",
    "    \"isConsortiumAnalysis\",\n",
    "    \"isModelSystem\",\n",
    "    \"isMultiSpecimen\",\n",
    "    \"libraryPrep\",\n",
    "    \"libraryType\",\n",
    "    \"manifest\",\n",
    "    \"manifest\",\n",
    "    \"metadata\",\n",
    "    \"metadataType\",\n",
    "    \"modelSystemName\",\n",
    "    \"modelSystemType\",\n",
    "    \"platform\",\n",
    "    \"project\",\n",
    "    \"protocol\",\n",
    "    \"protocol\",\n",
    "    \"resourceType\",\n",
    "    \"type\",\n",
    "]\n",
    "\n",
    "metadataTypes = [\n",
    "    \"analytical covariates\",\n",
    "    \"assay\",\n",
    "    \"biospecimen\",\n",
    "    \"data dictionary\",\n",
    "    \"ID mapping\",\n",
    "    \"individual\",\n",
    "    \"manifest\",\n",
    "    \"protocol\",\n",
    "]\n",
    "\n",
    "df_new_attrs = df.query(\"Attribute in @attrs_interest\").copy()\n",
    "\n",
    "df_new_attrs[\"Properties\"] = \"BaseAnnotation\"\n",
    "\n",
    "# Cleanup data model attributes to fit ELITE data model\n",
    "df_new_attrs = rewrite_df_value(\n",
    "    df_new_attrs, \"Attribute\", \"study\", \"Valid Values\", \"LLFS,ILO,LG,LC\"\n",
    ")\n",
    "\n",
    "df_new_attrs = rewrite_df_value(\n",
    "    df_new_attrs, \"Attribute\", \"consortium\", \"Valid Values\", \"ELITE\"\n",
    ")\n",
    "\n",
    "df_new_attrs = rewrite_df_value(\n",
    "    df_new_attrs,\n",
    "    \"Attribute\",\n",
    "    \"metadataType\",\n",
    "    \"Valid Values\",\n",
    "    \"analytical covariates, assay, biospecimen, data dictionary, ID mapping, individual, manifest, protocol\",\n",
    ")\n",
    "\n",
    "# recode Parent\n",
    "recoder = {\n",
    "    \"metadataType\": \"MetadataType\",\n",
    "    \"dataProperty\": \"DataProperty\",\n",
    "    \"dataType\": \"DataType\",\n",
    "    \"dataSubtype\": \"DataSubtype\",\n",
    "}\n",
    "\n",
    "df_new_attrs = df_new_attrs.replace(recoder)\n",
    "\n",
    "# Rename Columns\n",
    "\n",
    "df_new_attrs = df_new_attrs.rename(\n",
    "    columns={\"Parent\": \"Properties\", \"Properties\": \"Parent\"}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge new attributes with existiing data model\n",
    "dm_new = pd.concat([dm2, df_new_attrs])\n",
    "\n",
    "dm_new.drop(columns=\"columnType\", inplace=True)\n",
    "\n",
    "print(f\"Duplicates: {sum(dm_new.duplicated(subset=\"Attribute\"))}\")\n",
    "# dm_new[dm_new.duplicated(subset=\"Attribute\", keep=False)]\n",
    "\n",
    "dm_new.sample(5)\n",
    "dm_new[\"Validation Rules\"] = dm_new[\"Validation Rules\"].replace(\n",
    "    \"regex search ([0-9]+),regex search ([0-9]+\\\\.[0-9]*.)|([0-9]+)\",\n",
    "    \"regex search ([0-9]+\\\\.[0-9]*.)|([0-9]+)\",\n",
    ")\n",
    "dm_new[\n",
    "    dm_new[\"Validation Rules\"]\n",
    "    == \"regex search ([0-9]+),regex search ([0-9]+\\\\.[0-9]*.)|([0-9]+)\"\n",
    "]\n",
    "\n",
    "# recode Parent\n",
    "recoder = {\n",
    "    \"metadataType\": \"MetadataType\",\n",
    "    \"dataProperty\": \"DataProperty\",\n",
    "    \"dataType\": \"DataType\",\n",
    "    \"dataSubtype\": \"DataSubtype\",\n",
    "}\n",
    "\n",
    "dm_new = dm_new.replace(recoder)\n",
    "\n",
    "# drop duplicates\n",
    "print(dm_new.shape)\n",
    "\n",
    "dm_new = dm_new.drop_duplicates(subset=[\"Attribute\"], keep=\"first\")\n",
    "\n",
    "print(dm_new.shape)\n",
    "\n",
    "# Fill in blanks with unspecified to update later and categorize for site\n",
    "dm_new[[\"Parent\", \"Properties\", \"Type\"]] = dm_new[\n",
    "    [\"Parent\", \"Properties\", \"Type\"]\n",
    "].fillna(\"unspecified\")\n",
    "\n",
    "dm_new[\"Required\"] = dm_new[\"Required\"].fillna(\"False\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clean up template depends on\n",
    "dm_new[\"DependsOn\"] = (\n",
    "    dm_new[\"DependsOn\"].fillna(\"\").apply(clean_list).replace(\"\", np.nan)\n",
    ")\n",
    "\n",
    "remove_values = [\"False\", \"Not applicable\", \"Not collected\", \"Unknown\", \"Other\"]\n",
    "\n",
    "dm_new.loc[dm_new[\"Parent\"] == \"Template\", \"DependsOn\"] = dm_new.loc[\n",
    "    dm_new[\"Parent\"] == \"Template\", \"DependsOn\"\n",
    "].apply(lambda x: \",\".join([y for y in x.split(\",\") if y not in remove_values]))\n",
    "\n",
    "dm_new.loc[dm_new[\"Properties\"] == \"Valid Value\", \"Validation Rules\"] = np.nan"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pull Valid Values from Sources\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hard Coded Attributes Fixed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add study name attribute\n",
    "# Later get from synapse\n",
    "new_attribute = {\n",
    "    \"Attribute\": [\"studyName\"],\n",
    "    \"Description\": [\n",
    "        \"Name of studies found in project. Values include: MRGWAS,ELPSCRNA,Aging-PheWAS,Organoid scRNAseq\"\n",
    "    ],\n",
    "    \"Valid Values\": [np.nan],\n",
    "    \"DependsOn\": [np.nan],\n",
    "    \"Properties\": [np.nan],\n",
    "    \"Required\": [np.nan],\n",
    "    \"Parent\": [\"validValues\"],\n",
    "    \"DependsOn Component\": [np.nan],\n",
    "    \"Source\": [np.nan],\n",
    "    \"Validation Rules\": [\"str\"],\n",
    "    \"Module\": [np.nan],\n",
    "    \"Type\": [np.nan],\n",
    "    \"Ontology\": [np.nan],\n",
    "}\n",
    "\n",
    "dm = pd.concat([dm, pd.DataFrame.from_dict(new_attribute)])\n",
    "\n",
    "atrs_to_fix = [\n",
    "    \"sequencingBatchID\",  # needs to accept numerical values\n",
    "    \"libraryVersion\",  # needs to accept numerical values\n",
    "]\n",
    "vv = \"\"\n",
    "vr = \"regex search ^[0-9]+|(Unknown)|(Not collected)|(Not applicable)|(Not Specified)\"\n",
    "for a in atrs_to_fix:\n",
    "    fix_index = dm_elite.query(\"Attribute == @a\").index[0]\n",
    "    dm_elite.loc[fix_index, \"Valid Values\"] = vv\n",
    "    dm_elite.loc[dm_elite.query(\"Attribute == @a\").index[0], \"Validation Rules\"] = vr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Update Attributes with Synapse Data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "syn = synapse_login.main()\n",
    "\n",
    "# Pull grant information\n",
    "grants = syn.tableQuery(\"SELECT * FROM syn51209786\").asDataFrame()[\"grantNumber\"]\n",
    "\n",
    "grantIds = []\n",
    "for i in grants:\n",
    "    grantIds += i\n",
    "\n",
    "grantIds = \",\".join([i.strip() for i in grantIds])\n",
    "\n",
    "dm.loc[\n",
    "    dm[\"Attribute\"].str.contains(\"grant\", flags=re.IGNORECASE), \"Valid Values\"\n",
    "] = grantIds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write Out Data Model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm_new[\"Type\"] = dm_new[\"Type\"].str.upper()\n",
    "\n",
    "dm_new = dm_new.reset_index(drop=True)\n",
    "dm_new.to_csv(\"../\" + csv_model, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Convert Data Model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "schematic schema convert ../EL.data.model.csv --output_jsonld EL.data.model.jsonld\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting schematic...\n",
      "Done adding requirements and value ranges to attributes\n",
      "The Data Model was created and saved to '../EL.data.model.jsonld' location.\n"
     ]
    }
   ],
   "source": [
    "# Convert CSV to JSON LD\n",
    "print(f'schematic schema convert {\"../\" + csv_model} --output_jsonld {json_model}')\n",
    "\n",
    "!schematic schema convert {\"../\" + csv_model} --output_jsonld {\"../\" + json_model}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
