{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Fix inconsistencies in parent column\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "from glob import glob\n",
    "import yaml\n",
    "import re\n",
    "import copy\n",
    "import json\n",
    "from thefuzz import fuzz\n",
    "\n",
    "# Custom package\n",
    "from utils import utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Schematic config:  ./config.yml \n",
      " CSV model:  EL.data.model.csv \n",
      " JSON LD Model:  EL.data.model.jsonld\n"
     ]
    }
   ],
   "source": [
    "with open(\"./local_configs/notebook_config.yaml\", \"r\") as f:\n",
    "    config = yaml.safe_load(f)\n",
    "\n",
    "# paths to import files\n",
    "schematic_config = config[\"paths\"][\"schematic\"]\n",
    "csv_model = config[\"file_names\"][\"csv_model\"]\n",
    "json_model = config[\"file_names\"][\"json_model\"]\n",
    "\n",
    "print(\n",
    "    \"Schematic config: \",\n",
    "    schematic_config,\n",
    "    \"\\n\",\n",
    "    \"CSV model: \",\n",
    "    csv_model,\n",
    "    \"\\n\",\n",
    "    \"JSON LD Model: \",\n",
    "    json_model,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hard Coded stuff\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "validation_coder = {\n",
    "    \"number\": \"regex search ([0-9]+\\.[0-9]*.?)|([0-9]+)\",\n",
    "    \"integer\": \"regex search ([0-9]+)\",\n",
    "    \"string\": \"\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_cols = [\n",
    "    \"Attribute\",\n",
    "    \"Description\",\n",
    "    \"Valid Values\",\n",
    "    \"DependsOn\",\n",
    "    \"Properties\",\n",
    "    \"Required\",\n",
    "    \"Template\",\n",
    "    \"Parent\",\n",
    "    \"DependsOn Component\",\n",
    "    \"Source\",\n",
    "    \"Validation Rules\",\n",
    "]\n",
    "\n",
    "keep_cols = [\n",
    "    \"Attribute\",\n",
    "    \"Description\",\n",
    "    \"Valid Values\",\n",
    "    \"Required\",\n",
    "    \"DependsOn\",\n",
    "    \"DependsOn Component\",\n",
    "    \"Properties\",\n",
    "    \"Validation Rules\",\n",
    "    \"Template\",\n",
    "    \"Parent\",\n",
    "    \"Source\",\n",
    "    \"Type\",\n",
    "    \"Ontology\",\n",
    "    \"multivalue\",\n",
    "]\n",
    "\n",
    "list_cols = [\n",
    "    \"UsedIn\",\n",
    "    \"DependsOn\",\n",
    "    \"Properties\",\n",
    "    \"Validation Rules\",\n",
    "    \"Template\",\n",
    "    \"Parent\",\n",
    "    \"Source\",\n",
    "    \"Type\",\n",
    "    \"Ontology\",\n",
    "    \"Required\",\n",
    "    \"multivalue\",\n",
    "]\n",
    "\n",
    "hard_coded_valid_values = [\n",
    "    {\n",
    "        \"attribute\": \"visitCode\",\n",
    "        \"valid_value\": \"1,2,3,4,Other,Unknown,Not collected,Not applicable\",\n",
    "    },\n",
    "    {\"attribute\": \"tissueWeight\", \"valid_value\": \"\"},\n",
    "    {\"attribute\": \"consentGroupID\", \"valid_value\": \"1,2,3\"},\n",
    "    {\"attribute\": \"samplingAge\", \"valid_value\": \"\"},\n",
    "    {\"attribute\": \"specimenAge\", \"valid_value\": \"\"},\n",
    "    {\"attribute\": \"age\", \"valid_value\": \"\"},\n",
    "]\n",
    "\n",
    "template_hard_coded_attrs = {\n",
    "    \"Attribute\": \"\",\n",
    "    \"Description\": \"\",\n",
    "    \"Valid Values\": \"\",\n",
    "    \"Required\": \"False\",\n",
    "    \"Validation Rules\": \"\",\n",
    "    \"Template\": \"\",\n",
    "    \"Parent\": \"\",\n",
    "    \"Source\": \"\",\n",
    "    \"Type\": \"\",\n",
    "    \"Ontology\": \"\",\n",
    "    \"UsedIn\": \"\",\n",
    "}\n",
    "\n",
    "hard_coded_attrs = [\n",
    "    {\n",
    "        \"Attribute\": \"Other\",\n",
    "        \"Description\": \"When value is not apart of the list\",\n",
    "        \"Valid Values\": \"\",\n",
    "        \"Required\": \"False\",\n",
    "        \"Validation Rules\": \"\",\n",
    "        \"Template\": \"\",\n",
    "        \"Parent\": \"\",\n",
    "        \"Source\": \"\",\n",
    "        \"Type\": \"\",\n",
    "        \"Ontology\": \"\",\n",
    "        \"UsedIn\": \"\",\n",
    "    },\n",
    "    {\n",
    "        \"Attribute\": \"\",\n",
    "        \"Description\": \"\",\n",
    "        \"Valid Values\": \"\",\n",
    "        \"Required\": \"False\",\n",
    "        \"Validation Rules\": \"\",\n",
    "        \"Template\": \"\",\n",
    "        \"Parent\": \"\",\n",
    "        \"Source\": \"\",\n",
    "        \"Type\": \"\",\n",
    "        \"Ontology\": \"\",\n",
    "        \"UsedIn\": \"\",\n",
    "    },\n",
    "    {\n",
    "        \"Attribute\": \"\",\n",
    "        \"Description\": \"\",\n",
    "        \"Valid Values\": \"\",\n",
    "        \"Required\": \"False\",\n",
    "        \"Validation Rules\": \"\",\n",
    "        \"Template\": \"\",\n",
    "        \"Parent\": \"\",\n",
    "        \"Source\": \"\",\n",
    "        \"Type\": \"\",\n",
    "        \"Ontology\": \"\",\n",
    "        \"UsedIn\": \"\",\n",
    "    },\n",
    "    {\n",
    "        \"Attribute\": \"\",\n",
    "        \"Description\": \"\",\n",
    "        \"Valid Values\": \"\",\n",
    "        \"Required\": \"False\",\n",
    "        \"Validation Rules\": \"\",\n",
    "        \"Template\": \"\",\n",
    "        \"Parent\": \"\",\n",
    "        \"Source\": \"\",\n",
    "        \"Type\": \"\",\n",
    "        \"Ontology\": \"\",\n",
    "        \"UsedIn\": \"\",\n",
    "    },\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# hard coded dictionary\n",
    "recoder_valid_values = {\n",
    "    re.compile(\"Not Specified\", flags=re.IGNORECASE): \"Not Specified\",\n",
    "    re.compile(\"(Other$)\", flags=re.IGNORECASE): \"Other\",\n",
    "    re.compile(\"lipid\", re.IGNORECASE): \"Lipid\",\n",
    "    re.compile(\"plasma\", re.IGNORECASE): \"Plasma\",\n",
    "    re.compile(\"protein\", re.IGNORECASE): \"Protein\",\n",
    "    re.compile(\"saliva\", re.IGNORECASE): \"Saliva\",\n",
    "    re.compile(\"serum\", re.IGNORECASE): \"Serum\",\n",
    "    re.compile(\"sputum\", re.IGNORECASE): \"Sputum\",\n",
    "    re.compile(\"urine\", re.IGNORECASE): \"Urine\",\n",
    "    re.compile(\n",
    "        \"^0x Visium Spatial Gene Expression\"\n",
    "    ): \"10x Visium Spatial Gene Expression\",\n",
    "    re.compile(\"falseFalseFALSEtrueTrueTRUE\t\"): \"True, False\",\n",
    "    re.compile(\"TRUE|TRUEDiagnosisStatus\", re.IGNORECASE): \"True\",\n",
    "    re.compile(\"FALSE|FASLSE\", re.IGNORECASE): \"False\",\n",
    "    re.compile(\"UnknownNot collected\"): \"Unknown, Not collected\",\n",
    "    re.compile(r\"\\u200b\\u200b\"): \"\",\n",
    "    re.compile(\n",
    "        \"The Health,Aging,and Body Composition Study \\(HealthABC\\)\"\n",
    "    ): \"The Health and Aging and Body Composition Study (HealthABC)\",\n",
    "    re.compile(\"Not Hispanic or latinoEthnicity\"): \"Not Hispanic or latino\",\n",
    "    re.compile(\"Hispanic or latinoEthnicity\"): \"Hispanic or latino\",\n",
    "    re.compile(\n",
    "        \"HPO, MONDO, MAXO codes or labels \\(not listed for purposes of this RFC\\)\"\n",
    "    ): \"HPO and MONDO and MAXO codes or labels (not listed for purposes of this RFC)\",\n",
    "    re.escape(\n",
    "        r\"Possible values are listed under the instrument model term.OtherMsInstrumentModel\"\n",
    "    ): \"OtherMsInstrumentModel\",\n",
    "    re.compile(\n",
    "        r\"Possible values are listed under the cleavage agent nameOtherCleavageAgents\"\n",
    "    ): \"OtherCleavageAgents\",\n",
    "    re.compile(r\"Possible values are listed under modification parameters\"): \"\",\n",
    "    re.compile(\"Uknown\"): \"Unknown\",\n",
    "    re.compile(\"OtherControlType\", re.IGNORECASE): \"OtherControlType\",\n",
    "    re.compile(\"OtherMsAnalyteType\", re.IGNORECASE): \"OtherMsAnalyteType\",\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean list columns into single string\n",
    "def join_strings(string):\n",
    "    try:\n",
    "        return \",\".join(string)\n",
    "    except:\n",
    "        return \"\"\n",
    "\n",
    "\n",
    "def clean_list(string):\n",
    "    \"\"\"Takes a list represented as a string and returns only unique values found\n",
    "\n",
    "    Args:\n",
    "        string (str): list represented as string\n",
    "\n",
    "    Returns:\n",
    "        string: list as string of unique values\n",
    "    \"\"\"\n",
    "\n",
    "    new_list = string.split(\",\")\n",
    "    new_list = [n.strip() for n in new_list if n != \"nan\"]\n",
    "    new_list = \",\".join(sorted(list(np.unique(new_list)))).strip(\",\")\n",
    "    return new_list\n",
    "\n",
    "\n",
    "def search_df(df, pattern):\n",
    "    mask = np.column_stack(\n",
    "        [df[col].str.contains(pattern, na=False, flags=re.IGNORECASE) for col in df]\n",
    "    )\n",
    "\n",
    "    df = df.loc[mask.any(axis=1)]\n",
    "\n",
    "    with pd.option_context(\"display.max_colwidth\", None):\n",
    "        display(df)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def find_row(df, attribute):\n",
    "    \"\"\"Get indexes of the dataframe\"\"\"\n",
    "    indexes = df.index[\n",
    "        df[\"Attribute\"].str.contains(\n",
    "            \"(^\" + re.escape(attribute) + \"$)\", flags=re.IGNORECASE\n",
    "        )\n",
    "    ].tolist()\n",
    "    if len(indexes) != 0:\n",
    "        return indexes\n",
    "    else:\n",
    "        print(attribute)\n",
    "        return None\n",
    "\n",
    "\n",
    "def replace_valid_value(df, indexes, regex_dict, attribute):\n",
    "    \"\"\"Alter the dataframe valid values with the replacement value\"\"\"\n",
    "    if indexes == None:\n",
    "        return df\n",
    "\n",
    "    elif len(indexes) > 0:\n",
    "        regex_dict = regex_dict[attribute]\n",
    "\n",
    "        for index in indexes:\n",
    "            df.loc[index, \"Valid Values\"] = re.sub(\n",
    "                **regex_dict, string=df.loc[index, \"Valid Values\"]\n",
    "            )\n",
    "\n",
    "            # print(df.loc[index, 'Valid Values'])\n",
    "        return df\n",
    "    else:\n",
    "        return df\n",
    "\n",
    "\n",
    "def code_equals_values(df, regex_dict, attribute):\n",
    "    print(\"attribute: \", attribute)\n",
    "\n",
    "    indexes = find_row(df, attribute)\n",
    "\n",
    "    print(\"Index: \", indexes)\n",
    "\n",
    "    if indexes == None:\n",
    "        return df\n",
    "    else:\n",
    "        df = replace_valid_value(df, indexes, regex_dict, attribute)\n",
    "        return df\n",
    "\n",
    "\n",
    "def rewrite_df_value(df, col_name, search_term, col_value, new_value):\n",
    "    try:\n",
    "        df.loc[df[df[col_name] == search_term].index[0], col_value] = new_value\n",
    "        return df\n",
    "    except:\n",
    "        return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Unzip compressed folder if downloaded from Google Drive\n",
    "# %unzip 'RFC Tables-20230620T181152Z-001.zip'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Collect RFCs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get all the RFC file paths\n",
    "file_paths = glob(\"../_data/RFC Tables/*\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create Data Model for Schematic\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm = pd.DataFrame()\n",
    "\n",
    "# parse through files to create complete data model\n",
    "for fp in file_paths:\n",
    "    file_name = os.path.basename(fp)\n",
    "\n",
    "    temp = pd.read_excel(fp)\n",
    "\n",
    "    # Create file_name column to check\n",
    "    temp.insert(loc=0, column=\"file_name\", value=file_name)\n",
    "\n",
    "    # Create new columnn for data model name\n",
    "    temp.insert(\n",
    "        loc=1,\n",
    "        column=\"dm\",\n",
    "        value=re.sub(\n",
    "            \"\\s\\s+\",\n",
    "            \" \",\n",
    "            re.sub(\n",
    "                \"_\",\n",
    "                \" \",\n",
    "                re.sub(\n",
    "                    \"(EL)|(RFC)|(\\.xlsx)|([Aa]ssay)|([Dd]ata [Mm]odel)\", \"\", file_name\n",
    "                ),\n",
    "            ).strip(),\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    dm = pd.concat([dm, temp])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>dm</th>\n",
       "      <th>key</th>\n",
       "      <th>description</th>\n",
       "      <th>valid values</th>\n",
       "      <th>required</th>\n",
       "      <th>requires</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>type</th>\n",
       "      <th>concept source ontology</th>\n",
       "      <th>note</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "      <th>ontology</th>\n",
       "      <th>term id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EL Assay_ scRNAseq data model.xlsx</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>specimenID</td>\n",
       "      <td>Identifying string linked to a particular samp...</td>\n",
       "      <td>n/a (unique to each data contributor)</td>\n",
       "      <td>True</td>\n",
       "      <td>scRNAseq, Biospecimen</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EL Assay_ scRNAseq data model.xlsx</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>sampleType</td>\n",
       "      <td>The type of sample collected</td>\n",
       "      <td>Amniotic Fluid,\\nAppendix,\\nB cell,\\nBasophils...</td>\n",
       "      <td>True</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,\\nImmPort</td>\n",
       "      <td>The sample types are adopted from Uberon, Cell...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EL Assay_ scRNAseq data model.xlsx</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>specifySampleType</td>\n",
       "      <td>If \"other\" is selected list the type of sample</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>scRNAseq,\\nsampleType = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EL Assay_ scRNAseq data model.xlsx</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The measurement technique describing the assay...</td>\n",
       "      <td>16S rRNA gene sequencing,\\n1D Gel,\\n2D Gel,\\nA...</td>\n",
       "      <td>True</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EL Assay_ scRNAseq data model.xlsx</td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the me...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>scRNAseq,\\nmeasurementTechnique = other</td>\n",
       "      <td>False</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            file_name        dm                          key  \\\n",
       "0  EL Assay_ scRNAseq data model.xlsx  scRNAseq                   specimenID   \n",
       "1  EL Assay_ scRNAseq data model.xlsx  scRNAseq                   sampleType   \n",
       "2  EL Assay_ scRNAseq data model.xlsx  scRNAseq            specifySampleType   \n",
       "3  EL Assay_ scRNAseq data model.xlsx  scRNAseq         measurementTechnique   \n",
       "4  EL Assay_ scRNAseq data model.xlsx  scRNAseq  specifyMeasurementTechnique   \n",
       "\n",
       "                                         description  \\\n",
       "0  Identifying string linked to a particular samp...   \n",
       "1                       The type of sample collected   \n",
       "2     If \"other\" is selected list the type of sample   \n",
       "3  The measurement technique describing the assay...   \n",
       "4  If \"other\" is selected list the name of the me...   \n",
       "\n",
       "                                        valid values required  \\\n",
       "0              n/a (unique to each data contributor)     True   \n",
       "1  Amniotic Fluid,\\nAppendix,\\nB cell,\\nBasophils...     True   \n",
       "2                                                NaN    False   \n",
       "3  16S rRNA gene sequencing,\\n1D Gel,\\n2D Gel,\\nA...     True   \n",
       "4                                                NaN    False   \n",
       "\n",
       "                                  requires multivalue    type  \\\n",
       "0                    scRNAseq, Biospecimen      False  string   \n",
       "1                                 scRNAseq      False  string   \n",
       "2            scRNAseq,\\nsampleType = other      False  string   \n",
       "3                                 scRNAseq      False  string   \n",
       "4  scRNAseq,\\nmeasurementTechnique = other      False  string   \n",
       "\n",
       "      concept source ontology  \\\n",
       "0            Sage Bionetworks   \n",
       "1  Sage Bionetworks,\\nImmPort   \n",
       "2            Sage Bionetworks   \n",
       "3            Sage Bionetworks   \n",
       "4            Sage Bionetworks   \n",
       "\n",
       "                                                note  Unnamed: 9  \\\n",
       "0                                                NaN         NaN   \n",
       "1  The sample types are adopted from Uberon, Cell...         NaN   \n",
       "2                                                NaN         NaN   \n",
       "3                                                NaN         NaN   \n",
       "4                                                NaN         NaN   \n",
       "\n",
       "        Unnamed: 10 ontology term id  \n",
       "0  Sage Bionetworks      NaN     NaN  \n",
       "1               NaN      NaN     NaN  \n",
       "2               NaN      NaN     NaN  \n",
       "3               NaN      NaN     NaN  \n",
       "4               NaN      NaN     NaN  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# initial cleaning\n",
    "dm[[\"required\", \"multivalue\"]] = (\n",
    "    dm[[\"required\", \"multivalue\"]]\n",
    "    .fillna(False)\n",
    "    .astype(str)\n",
    "    .replace({\"1.0\": True, \"0.0\": False})\n",
    ")\n",
    "\n",
    "dm.fillna(\"\")\n",
    "dm.reset_index(drop=True, inplace=True)\n",
    "dm.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data model clean up\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/p0/5m4pdsm55jn_d5nzbjv6bjf40000gq/T/ipykernel_15102/2830742079.py:14: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.\n",
      "  dm[\"valid values\"].str.contains(\n"
     ]
    }
   ],
   "source": [
    "# collapse presumed ontology columns and join with existing\n",
    "dm.loc[:, \"ontology\"] = dm.iloc[:, 11:].bfill(axis=1).iloc[:, 0]\n",
    "\n",
    "dm[\"ontology\"] = (\n",
    "    dm[[\"concept source ontology\", \"ontology\"]]\n",
    "    .fillna(\"\")\n",
    "    .apply(lambda x: \",\".join([y.strip() for y in x.unique() if len(y) > 0]), axis=1)\n",
    ")\n",
    "\n",
    "# if unique values are provided by data contributor then add this note in the ontology\n",
    "\n",
    "dm[\"ontology\"] = (\n",
    "    dm.loc[\n",
    "        dm[\"valid values\"].str.contains(\n",
    "            \"(n/a \\(unique to each data contributor\\))\", na=False\n",
    "        ),\n",
    "        \"ontology\",\n",
    "    ]\n",
    "    + \",\"\n",
    "    + \"Data Contributor\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# new ontologies:\n",
    "ontology_list = []\n",
    "for i, v in dm[[\"concept source ontology\", \"ontology\"]].fillna(\"\").iterrows():\n",
    "    ontology_list.append(\",\".join(v))\n",
    "\n",
    "dm[\"ontology\"] = ontology_list\n",
    "dm[\"ontology\"] = dm[\"ontology\"].str.strip(\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>dm</th>\n",
       "      <th>key</th>\n",
       "      <th>description</th>\n",
       "      <th>valid values</th>\n",
       "      <th>required</th>\n",
       "      <th>requires</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>type</th>\n",
       "      <th>concept source ontology</th>\n",
       "      <th>note</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "      <th>ontology</th>\n",
       "      <th>term id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[EL Assay_ scRNAseq data model.xlsx]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[specimenID]</td>\n",
       "      <td>[Identifying string linked to a particular sam...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[scRNAseq,  Biospecimen]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[string]</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>[Sage Bionetworks, Sage Bionetworks, Data Cont...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[EL Assay_ scRNAseq data model.xlsx]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[sampleType]</td>\n",
       "      <td>[The type of sample collected]</td>\n",
       "      <td>[Amniotic Fluid, Appendix, B cell, Basophils, ...</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[string]</td>\n",
       "      <td>[Sage Bionetworks, ImmPort]</td>\n",
       "      <td>[The sample types are adopted from Uberon,  Ce...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Sage Bionetworks, ImmPort]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[EL Assay_ scRNAseq data model.xlsx]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[specifySampleType]</td>\n",
       "      <td>[If \"other\" is selected list the type of sample]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[scRNAseq, sampleType = other]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[string]</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[EL Assay_ scRNAseq data model.xlsx]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[measurementTechnique]</td>\n",
       "      <td>[The measurement technique describing the assa...</td>\n",
       "      <td>[16S rRNA gene sequencing, 1D Gel, 2D Gel, Arr...</td>\n",
       "      <td>[True]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[string]</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[EL Assay_ scRNAseq data model.xlsx]</td>\n",
       "      <td>[scRNAseq]</td>\n",
       "      <td>[specifyMeasurementTechnique]</td>\n",
       "      <td>[If \"other\" is selected list the name of the m...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[scRNAseq, measurementTechnique = other]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[string]</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Sage Bionetworks]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              file_name          dm  \\\n",
       "0  [EL Assay_ scRNAseq data model.xlsx]  [scRNAseq]   \n",
       "1  [EL Assay_ scRNAseq data model.xlsx]  [scRNAseq]   \n",
       "2  [EL Assay_ scRNAseq data model.xlsx]  [scRNAseq]   \n",
       "3  [EL Assay_ scRNAseq data model.xlsx]  [scRNAseq]   \n",
       "4  [EL Assay_ scRNAseq data model.xlsx]  [scRNAseq]   \n",
       "\n",
       "                             key  \\\n",
       "0                   [specimenID]   \n",
       "1                   [sampleType]   \n",
       "2            [specifySampleType]   \n",
       "3         [measurementTechnique]   \n",
       "4  [specifyMeasurementTechnique]   \n",
       "\n",
       "                                         description  \\\n",
       "0  [Identifying string linked to a particular sam...   \n",
       "1                     [The type of sample collected]   \n",
       "2   [If \"other\" is selected list the type of sample]   \n",
       "3  [The measurement technique describing the assa...   \n",
       "4  [If \"other\" is selected list the name of the m...   \n",
       "\n",
       "                                        valid values required  \\\n",
       "0                                                 []   [True]   \n",
       "1  [Amniotic Fluid, Appendix, B cell, Basophils, ...   [True]   \n",
       "2                                                NaN  [False]   \n",
       "3  [16S rRNA gene sequencing, 1D Gel, 2D Gel, Arr...   [True]   \n",
       "4                                                NaN  [False]   \n",
       "\n",
       "                                   requires multivalue      type  \\\n",
       "0                  [scRNAseq,  Biospecimen]    [False]  [string]   \n",
       "1                                [scRNAseq]    [False]  [string]   \n",
       "2            [scRNAseq, sampleType = other]    [False]  [string]   \n",
       "3                                [scRNAseq]    [False]  [string]   \n",
       "4  [scRNAseq, measurementTechnique = other]    [False]  [string]   \n",
       "\n",
       "       concept source ontology  \\\n",
       "0           [Sage Bionetworks]   \n",
       "1  [Sage Bionetworks, ImmPort]   \n",
       "2           [Sage Bionetworks]   \n",
       "3           [Sage Bionetworks]   \n",
       "4           [Sage Bionetworks]   \n",
       "\n",
       "                                                note  Unnamed: 9  \\\n",
       "0                                                NaN         NaN   \n",
       "1  [The sample types are adopted from Uberon,  Ce...         NaN   \n",
       "2                                                NaN         NaN   \n",
       "3                                                NaN         NaN   \n",
       "4                                                NaN         NaN   \n",
       "\n",
       "          Unnamed: 10                                           ontology  \\\n",
       "0  [Sage Bionetworks]  [Sage Bionetworks, Sage Bionetworks, Data Cont...   \n",
       "1                 NaN                        [Sage Bionetworks, ImmPort]   \n",
       "2                 NaN                                 [Sage Bionetworks]   \n",
       "3                 NaN                                 [Sage Bionetworks]   \n",
       "4                 NaN                                 [Sage Bionetworks]   \n",
       "\n",
       "  term id  \n",
       "0     NaN  \n",
       "1     NaN  \n",
       "2     NaN  \n",
       "3     NaN  \n",
       "4     NaN  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm = dm.apply(\n",
    "    lambda x: x.str.replace(\n",
    "        pat=\"\\n|(n/a \\(unique to each data contributor\\))\", repl=\"\", regex=True\n",
    "    ).str.split(\",\"),\n",
    "    axis=1,\n",
    ")\n",
    "\n",
    "dm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>specimenID</td>\n",
       "      <td>Identifying string linked to a particular samp...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq, Biospecimen</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,Sage Bionetworks,Data Contrib...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sampleType</td>\n",
       "      <td>The type of sample collected</td>\n",
       "      <td>Amniotic Fluid,Appendix,B cell,Basophils,Bone,...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks,ImmPort</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,ImmPort</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>specifySampleType</td>\n",
       "      <td>If \"other\" is selected list the type of sample</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq,sampleType = other</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>measurementTechnique</td>\n",
       "      <td>The measurement technique describing the assay...</td>\n",
       "      <td>16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>specifyMeasurementTechnique</td>\n",
       "      <td>If \"other\" is selected list the name of the me...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq,measurementTechnique = other</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Attribute  \\\n",
       "0                   specimenID   \n",
       "1                   sampleType   \n",
       "2            specifySampleType   \n",
       "3         measurementTechnique   \n",
       "4  specifyMeasurementTechnique   \n",
       "\n",
       "                                         Description  \\\n",
       "0  Identifying string linked to a particular samp...   \n",
       "1                       The type of sample collected   \n",
       "2     If \"other\" is selected list the type of sample   \n",
       "3  The measurement technique describing the assay...   \n",
       "4  If \"other\" is selected list the name of the me...   \n",
       "\n",
       "                                        Valid Values Required DependsOn  \\\n",
       "0                                                        True             \n",
       "1  Amniotic Fluid,Appendix,B cell,Basophils,Bone,...     True             \n",
       "2                                                       False             \n",
       "3  16S rRNA gene sequencing,1D Gel,2D Gel,Array,B...     True             \n",
       "4                                                       False             \n",
       "\n",
       "                     DependsOn Component Properties Validation Rules  \\\n",
       "0                  scRNAseq, Biospecimen                               \n",
       "1                               scRNAseq                               \n",
       "2            scRNAseq,sampleType = other                               \n",
       "3                               scRNAseq                               \n",
       "4  scRNAseq,measurementTechnique = other                               \n",
       "\n",
       "   Template Parent                    Source    Type  \\\n",
       "0  scRNAseq                 Sage Bionetworks  string   \n",
       "1  scRNAseq         Sage Bionetworks,ImmPort  string   \n",
       "2  scRNAseq                 Sage Bionetworks  string   \n",
       "3  scRNAseq                 Sage Bionetworks  string   \n",
       "4  scRNAseq                 Sage Bionetworks  string   \n",
       "\n",
       "                                            Ontology multivalue  \n",
       "0  Sage Bionetworks,Sage Bionetworks,Data Contrib...      False  \n",
       "1                           Sage Bionetworks,ImmPort      False  \n",
       "2                                   Sage Bionetworks      False  \n",
       "3                                   Sage Bionetworks      False  \n",
       "4                                   Sage Bionetworks      False  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# revert lists back to strings\n",
    "dm = dm.applymap(lambda x: join_strings(x))\n",
    "\n",
    "# Rename columns with DCA standards\n",
    "dm_schema_cols = {\n",
    "    \"dm\": \"Template\",\n",
    "    \"key\": \"Attribute\",\n",
    "    \"description\": \"Description\",\n",
    "    \"valid values\": \"Valid Values\",\n",
    "    \"required\": \"Required\",\n",
    "    \"requires\": \"DependsOn Component\",\n",
    "    \"concept source ontology\": \"Source\",\n",
    "    \"ontology\": \"Ontology\",\n",
    "    \"type\": \"Type\",\n",
    "}\n",
    "\n",
    "dm = dm.rename(dm_schema_cols, axis=1)\n",
    "\n",
    "# drop unimportant columns\n",
    "r = re.compile(\"Unnamed*\", re.IGNORECASE)\n",
    "\n",
    "# Add additional required columns for DCA\n",
    "dm[\"Properties\"] = \"\"\n",
    "dm[\"Validation Rules\"] = dm[\"Type\"].map(validation_coder)\n",
    "dm[\"DependsOn\"] = \"\"\n",
    "dm[\"Parent\"] = \"\"\n",
    "# dm['DependsOn Component'] = \"\"\n",
    "\n",
    "dm = dm[keep_cols]\n",
    "\n",
    "dm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: Valid Values, dtype: object)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fixing \"not listed\" columns\n",
    "dm[\"Valid Values\"] = dm[\"Valid Values\"].str.replace(\n",
    "    \"Genbank common names (not listed for purposes of this RFC)Unknown\",\n",
    "    \"Genbank common names (not listed for purposes of this RFC),Unknown\",\n",
    "    regex=False,\n",
    ")\n",
    "\n",
    "dm.loc[dm[\"Valid Values\"].str.contains(\"not listed\"), \"Valid Values\"] = (\n",
    "    dm.loc[dm[\"Valid Values\"].str.contains(\"not listed\"), \"Valid Values\"]\n",
    "    .str.split(\")\")\n",
    "    .apply(\n",
    "        lambda x: \",\".join(\n",
    "            [y.strip(\",\") for y in x if not bool(re.search(\"not listed\", y))]\n",
    "        )\n",
    "    )\n",
    ")\n",
    "\n",
    "# QC\n",
    "dm.loc[dm[\"Valid Values\"].str.contains(\"not listed\", na=False), \"Valid Values\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dropping measurement technique\n",
    "dm = dm.drop(\n",
    "    index=dm.query('Attribute == \"measurementTechnique\"').index.values\n",
    ").reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine duplicated attributes\n",
    "dm = dm.groupby(\"Attribute\").agg(lambda x: \",\".join(set(x.astype(str)))).reset_index()\n",
    "\n",
    "# found extra commas in strings at beginning and end\n",
    "dm = dm.applymap(lambda x: x.strip(\",\"))\n",
    "\n",
    "template_recoder = {\n",
    "    re.compile(\"genotyping\", re.IGNORECASE): \"Genotyping\",\n",
    "    re.compile(\"proteomics\", re.IGNORECASE): \"Proteomics\",\n",
    "}\n",
    "\n",
    "dm = dm.replace(template_recoder, regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Metabolomics Human', 'Metabolomics Human,Proteomics',\n",
       "       'Individual Human', 'Biospecimen nonHuman', 'Individual nonHuman',\n",
       "       'Biospecimen nonHuman,Biospecimen human', 'Proteomics',\n",
       "       'Individual Human,Individual nonHuman',\n",
       "       'bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'Microbiome,bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'Microbiome,Metabolomics Human',\n",
       "       'Individual Human,Biospecimen nonHuman,Individual nonHuman,Biospecimen human',\n",
       "       'Microbiome,RNAseq,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),Whole Genome Sequencing',\n",
       "       'scRNAseq,Whole Genome Sequencing,Microbiome,bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'Microbiome,Metabolomics Human,RNAseq,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),Whole Genome Sequencing,Proteomics',\n",
       "       'Genotyping',\n",
       "       'scRNAseq,Whole Genome Sequencing,RNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'scRNAseq,Whole Genome Sequencing,RNAseq',\n",
       "       'scRNAseq,Whole Genome Sequencing', 'scRNAseq',\n",
       "       'Microbiome,Metabolomics Human,Genotyping,RNAseq,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),Whole Genome Sequencing,Proteomics',\n",
       "       'scRNAseq,Microbiome,bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'Microbiome',\n",
       "       'Microbiome,Metabolomics Human,Biospecimen nonHuman,Genotyping,RNAseq,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),Whole Genome Sequencing,Proteomics,Biospecimen human'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# QA Check\n",
    "dm[\"Template\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Attribute              object\n",
       "Description            object\n",
       "Valid Values           object\n",
       "Required               object\n",
       "DependsOn              object\n",
       "DependsOn Component    object\n",
       "Properties             object\n",
       "Validation Rules       object\n",
       "Template               object\n",
       "Parent                 object\n",
       "Source                 object\n",
       "Type                   object\n",
       "Ontology               object\n",
       "multivalue             object\n",
       "dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>acquisitionBatchID</td>\n",
       "      <td>Acquisition batch identifier, provided by the ...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>acquisitionBatchSize</td>\n",
       "      <td>The number of samples</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>acquisitionBatchSizeUnit</td>\n",
       "      <td>The unit of measurement for number of samples ...</td>\n",
       "      <td>AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolmics,acquisitionBatchSize</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>acquisitionMode</td>\n",
       "      <td>The specific aspect of a mass spectrometer met...</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not speci...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec Proteomics,mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human,Proteomics</td>\n",
       "      <td></td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/terms?...</td>\n",
       "      <td>string</td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/terms?...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>acquisitionSoftware</td>\n",
       "      <td>The name of the acquisition software used</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not speci...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec Proteomics,mass spec metabolmics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human,Proteomics</td>\n",
       "      <td></td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Attribute  \\\n",
       "0        acquisitionBatchID   \n",
       "1      acquisitionBatchSize   \n",
       "2  acquisitionBatchSizeUnit   \n",
       "3           acquisitionMode   \n",
       "4       acquisitionSoftware   \n",
       "\n",
       "                                         Description  \\\n",
       "0  Acquisition batch identifier, provided by the ...   \n",
       "1                              The number of samples   \n",
       "2  The unit of measurement for number of samples ...   \n",
       "3  The specific aspect of a mass spectrometer met...   \n",
       "4          The name of the acquisition software used   \n",
       "\n",
       "                                        Valid Values Required DependsOn  \\\n",
       "0                                                       False             \n",
       "1                                                       False             \n",
       "2  AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...    False             \n",
       "3  Unknown,Not collected,Not applicable,Not speci...     True             \n",
       "4  Unknown,Not collected,Not applicable,Not speci...     True             \n",
       "\n",
       "                          DependsOn Component Properties Validation Rules  \\\n",
       "0                       mass spec metabolmics                               \n",
       "1                       mass spec metabolmics                               \n",
       "2  mass spec metabolmics,acquisitionBatchSize                               \n",
       "3  mass spec Proteomics,mass spec metabolmics                               \n",
       "4  mass spec Proteomics,mass spec metabolmics                               \n",
       "\n",
       "                        Template Parent  \\\n",
       "0             Metabolomics Human          \n",
       "1             Metabolomics Human          \n",
       "2             Metabolomics Human          \n",
       "3  Metabolomics Human,Proteomics          \n",
       "4  Metabolomics Human,Proteomics          \n",
       "\n",
       "                                              Source    Type  \\\n",
       "0                                   Sage Bionetworks  string   \n",
       "1                                   Sage Bionetworks  string   \n",
       "2                                   Sage Bionetworks  string   \n",
       "3  https://www.ebi.ac.uk/ols/ontologies/ms/terms?...  string   \n",
       "4          http://purl.obolibrary.org/obo/MS_1001455  string   \n",
       "\n",
       "                                            Ontology multivalue  \n",
       "0                                   Sage Bionetworks             \n",
       "1                                   Sage Bionetworks             \n",
       "2                                   Sage Bionetworks             \n",
       "3  https://www.ebi.ac.uk/ols/ontologies/ms/terms?...      False  \n",
       "4          http://purl.obolibrary.org/obo/MS_1001455      False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "array(['Metabolomics Human', 'Metabolomics Human,Proteomics',\n",
       "       'Individual Human', 'Biospecimen nonHuman', 'Individual nonHuman',\n",
       "       'Biospecimen nonHuman,Biospecimen human', 'Proteomics',\n",
       "       'Individual Human,Individual nonHuman',\n",
       "       'bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'Microbiome,bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'Microbiome,Metabolomics Human',\n",
       "       'Individual Human,Biospecimen nonHuman,Individual nonHuman,Biospecimen human',\n",
       "       'Microbiome,RNAseq,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),Whole Genome Sequencing',\n",
       "       'scRNAseq,Whole Genome Sequencing,Microbiome,bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'Microbiome,Metabolomics Human,RNAseq,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),Whole Genome Sequencing,Proteomics',\n",
       "       'Genotyping',\n",
       "       'scRNAseq,Whole Genome Sequencing,RNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'scRNAseq,Whole Genome Sequencing,RNAseq',\n",
       "       'scRNAseq,Whole Genome Sequencing', 'scRNAseq',\n",
       "       'Microbiome,Metabolomics Human,Genotyping,RNAseq,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),Whole Genome Sequencing,Proteomics',\n",
       "       'scRNAseq,Microbiome,bsSeq (bisulfite-seq WGBS methylseq methylomics)',\n",
       "       'Microbiome',\n",
       "       'Microbiome,Metabolomics Human,Biospecimen nonHuman,Genotyping,RNAseq,scRNAseq,bsSeq (bisulfite-seq WGBS methylseq methylomics),Whole Genome Sequencing,Proteomics,Biospecimen human'],\n",
       "      dtype=object)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(dm.dtypes)\n",
    "display(dm.head())\n",
    "display(dm.Template.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reorder Columns based on DCA Standards\n",
    "dm = dm.loc[:, keep_cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clean up\n",
    "\n",
    "## DependsOn Component\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "recoder = {\n",
    "    \"metabolmics\": \"metabolomics\",\n",
    "    \"(mass spec proteomics)\": \"Proteomics\",\n",
    "    \"(mass spec metabolomics)\": \"Metabolomics Human\",\n",
    "    \"(assay_otheruseTreatment? = Yes)\": \"assay_other, useTreatment? = Yes\",\n",
    "    \"OtherUnknown\": \"Other, Unknown\",\n",
    "    \"falseFalseFALSEtrueTrueTRUE\": \"TRUE, FALSE\",\n",
    "    \"Hispanic or latinoEthnicity\": \"Hispanic or Latino\",\n",
    "    re.compile(\"Forwardreverse\", flags=re.IGNORECASE): \"forward,reverse\",\n",
    "    re.compile(\"singleEndpairedEnd\"): \"singleEnd, pairedEnd\",\n",
    "    re.compile(\"(WGS)\"): \"Whole Genome Sequencing\",\n",
    "    re.compile(\"\\?\"): \"\",\n",
    "    \"Zeiss LSM 980Other\": \"Zeiss LSM 980,Other\",\n",
    "    \"bsSeqsampleType = other\": \"bsSeq, sampleType = other\",\n",
    "    re.compile(\n",
    "        \"HPO, MONDO, MAXO codes or labels \\(not listed for purposes of this RFC\\)\"\n",
    "    ): \"HPO and MONDO and MAXO codes or labels (not listed for purposes of this RFC)\",\n",
    "    \"The Health, Aging, and Body Composition Study \\(HealthABC\\)\": \"The Health and Aging and Body Composition Study (HealthABC)\",\n",
    "}\n",
    "\n",
    "# 'mass spec metabolomics,measurementTechnique = other'\n",
    "# falseFalseFALSEtrueTrueTRUE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm = dm.apply(lambda x: x.replace(recoder, regex=True), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>cohort</td>\n",
       "      <td>Name of the cohort the individual belongs to</td>\n",
       "      <td>Other, Unknown,Not collected,Not applicable,Centenarian, US Family, Denmark Family, The Osteoporotic Fractures in Men (MrOS) Study, Study of Osteoporotic Fractures (SOF), The Health and Aging and Body Composition Study (HealthABC), Cardiovascular Health Study (CHS),Other,Unknown,Not collected,Not applicable</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>Individual</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Individual Human,Individual nonHuman</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Attribute                                   Description  \\\n",
       "16    cohort  Name of the cohort the individual belongs to   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                            Valid Values  \\\n",
       "16  Other, Unknown,Not collected,Not applicable,Centenarian, US Family, Denmark Family, The Osteoporotic Fractures in Men (MrOS) Study, Study of Osteoporotic Fractures (SOF), The Health and Aging and Body Composition Study (HealthABC), Cardiovascular Health Study (CHS),Other,Unknown,Not collected,Not applicable   \n",
       "\n",
       "   Required DependsOn DependsOn Component Properties Validation Rules  \\\n",
       "16     True                    Individual                               \n",
       "\n",
       "                                Template Parent            Source    Type  \\\n",
       "16  Individual Human,Individual nonHuman         Sage Bionetworks  string   \n",
       "\n",
       "            Ontology multivalue  \n",
       "16  Sage Bionetworks       True  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with pd.option_context(\"display.max_colwidth\", None):\n",
    "    display(dm[dm[\"Valid Values\"].str.contains(\"and Body Composition Study\")])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ethnicity</td>\n",
       "      <td>Ethnicity of individual</td>\n",
       "      <td>Not Hispanic or Latino, Hispanic or Latino, Prefer not to answer,Other,Unknown,Not collected,Not applicable</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>Individual</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks,https://www.synapse.org/#!Synapse:syn25878249</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,https://www.synapse.org/#!Synapse:syn25878249</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Attribute              Description  \\\n",
       "37  ethnicity  Ethnicity of individual   \n",
       "\n",
       "                                                                                                   Valid Values  \\\n",
       "37  Not Hispanic or Latino, Hispanic or Latino, Prefer not to answer,Other,Unknown,Not collected,Not applicable   \n",
       "\n",
       "   Required DependsOn DependsOn Component Properties Validation Rules  \\\n",
       "37     True                    Individual                               \n",
       "\n",
       "            Template Parent  \\\n",
       "37  Individual Human          \n",
       "\n",
       "                                                            Source    Type  \\\n",
       "37  Sage Bionetworks,https://www.synapse.org/#!Synapse:syn25878249  string   \n",
       "\n",
       "                                                          Ontology multivalue  \n",
       "37  Sage Bionetworks,https://www.synapse.org/#!Synapse:syn25878249      False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with pd.option_context(\"display.max_colwidth\", None):\n",
    "    display(dm.query('Attribute == \"ethnicity\"'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>acquisitionBatchID</td>\n",
       "      <td>Acquisition batch identifier, provided by the ...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolomics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>acquisitionBatchSize</td>\n",
       "      <td>The number of samples</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolomics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>acquisitionBatchSizeUnit</td>\n",
       "      <td>The unit of measurement for number of samples ...</td>\n",
       "      <td>AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolomics,acquisitionBatchSize</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>acquisitionMode</td>\n",
       "      <td>The specific aspect of a mass spectrometer met...</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not speci...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec Proteomics,mass spec metabolomics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human,Proteomics</td>\n",
       "      <td></td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/termsi...</td>\n",
       "      <td>string</td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/termsi...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>acquisitionSoftware</td>\n",
       "      <td>The name of the acquisition software used</td>\n",
       "      <td>Unknown,Not collected,Not applicable,Not speci...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec Proteomics,mass spec metabolomics</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human,Proteomics</td>\n",
       "      <td></td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>specimenID</td>\n",
       "      <td>Identifying string linked to a particular samp...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>Biospecimen,Genotyping ,Biospecimen,Biospecime...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Microbiome,Metabolomics Human,Biospecimen nonH...</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,Sage Bionetworks,Data Contrib...</td>\n",
       "      <td>false,False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>spectrometerFrequency</td>\n",
       "      <td>The frequency at which a spectrometer causes h...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec Proteomics,mass spec metabolomics</td>\n",
       "      <td></td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.)|([0-9]+)</td>\n",
       "      <td>Metabolomics Human,Proteomics</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks,DSLWG,Sage Bionetworks</td>\n",
       "      <td>number</td>\n",
       "      <td>Sage Bionetworks,DSLWG,Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>technologyPlatformVersion</td>\n",
       "      <td>The specific version (application, manufacture...</td>\n",
       "      <td>0x Visium Spatial Gene Expression,​​Affymetrix...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolomics,RNAseq,scRNAseq,bsSeq,m...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Microbiome,Metabolomics Human,RNAseq,scRNAseq,...</td>\n",
       "      <td></td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C45378,Sag...</td>\n",
       "      <td>string</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C45378,Sag...</td>\n",
       "      <td>true,False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>vacuumPressure</td>\n",
       "      <td>The recorded vacuum pressure value</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolomics,hasIonizationSource = Yes</td>\n",
       "      <td></td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.)|([0-9]+)</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks,Proposed minimum metadata rel...</td>\n",
       "      <td>number</td>\n",
       "      <td>Sage Bionetworks,Proposed minimum metadata rel...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>vacuumPressureUnit</td>\n",
       "      <td>Unit of vacuum pressure value</td>\n",
       "      <td>AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>mass spec metabolomics,hasIonizationSource = Y...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks,Proposed minimum metadata rel...</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks,Proposed minimum metadata rel...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Attribute  \\\n",
       "0           acquisitionBatchID   \n",
       "1         acquisitionBatchSize   \n",
       "2     acquisitionBatchSizeUnit   \n",
       "3              acquisitionMode   \n",
       "4          acquisitionSoftware   \n",
       "..                         ...   \n",
       "152                 specimenID   \n",
       "156      spectrometerFrequency   \n",
       "159  technologyPlatformVersion   \n",
       "175             vacuumPressure   \n",
       "176         vacuumPressureUnit   \n",
       "\n",
       "                                           Description  \\\n",
       "0    Acquisition batch identifier, provided by the ...   \n",
       "1                                The number of samples   \n",
       "2    The unit of measurement for number of samples ...   \n",
       "3    The specific aspect of a mass spectrometer met...   \n",
       "4            The name of the acquisition software used   \n",
       "..                                                 ...   \n",
       "152  Identifying string linked to a particular samp...   \n",
       "156  The frequency at which a spectrometer causes h...   \n",
       "159  The specific version (application, manufacture...   \n",
       "175                 The recorded vacuum pressure value   \n",
       "176                      Unit of vacuum pressure value   \n",
       "\n",
       "                                          Valid Values Required DependsOn  \\\n",
       "0                                                         False             \n",
       "1                                                         False             \n",
       "2    AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...    False             \n",
       "3    Unknown,Not collected,Not applicable,Not speci...     True             \n",
       "4    Unknown,Not collected,Not applicable,Not speci...     True             \n",
       "..                                                 ...      ...       ...   \n",
       "152                                                        True             \n",
       "156                                                        True             \n",
       "159  0x Visium Spatial Gene Expression,​​Affymetrix...     True             \n",
       "175                                                       False             \n",
       "176  AFU,AI,AU/ml,DK units/ml,bpg/dl,g/l,gm,HAU,IU,...    False             \n",
       "\n",
       "                                   DependsOn Component Properties  \\\n",
       "0                               mass spec metabolomics              \n",
       "1                               mass spec metabolomics              \n",
       "2          mass spec metabolomics,acquisitionBatchSize              \n",
       "3          mass spec Proteomics,mass spec metabolomics              \n",
       "4          mass spec Proteomics,mass spec metabolomics              \n",
       "..                                                 ...        ...   \n",
       "152  Biospecimen,Genotyping ,Biospecimen,Biospecime...              \n",
       "156        mass spec Proteomics,mass spec metabolomics              \n",
       "159  mass spec metabolomics,RNAseq,scRNAseq,bsSeq,m...              \n",
       "175   mass spec metabolomics,hasIonizationSource = Yes              \n",
       "176  mass spec metabolomics,hasIonizationSource = Y...              \n",
       "\n",
       "                            Validation Rules  \\\n",
       "0                                              \n",
       "1                                              \n",
       "2                                              \n",
       "3                                              \n",
       "4                                              \n",
       "..                                       ...   \n",
       "152                                            \n",
       "156  regex search ([0-9]+\\.[0-9]*.)|([0-9]+)   \n",
       "159                                            \n",
       "175  regex search ([0-9]+\\.[0-9]*.)|([0-9]+)   \n",
       "176                                            \n",
       "\n",
       "                                              Template Parent  \\\n",
       "0                                   Metabolomics Human          \n",
       "1                                   Metabolomics Human          \n",
       "2                                   Metabolomics Human          \n",
       "3                        Metabolomics Human,Proteomics          \n",
       "4                        Metabolomics Human,Proteomics          \n",
       "..                                                 ...    ...   \n",
       "152  Microbiome,Metabolomics Human,Biospecimen nonH...          \n",
       "156                      Metabolomics Human,Proteomics          \n",
       "159  Microbiome,Metabolomics Human,RNAseq,scRNAseq,...          \n",
       "175                                 Metabolomics Human          \n",
       "176                                 Metabolomics Human          \n",
       "\n",
       "                                                Source    Type  \\\n",
       "0                                     Sage Bionetworks  string   \n",
       "1                                     Sage Bionetworks  string   \n",
       "2                                     Sage Bionetworks  string   \n",
       "3    https://www.ebi.ac.uk/ols/ontologies/ms/termsi...  string   \n",
       "4            http://purl.obolibrary.org/obo/MS_1001455  string   \n",
       "..                                                 ...     ...   \n",
       "152                                   Sage Bionetworks  string   \n",
       "156            Sage Bionetworks,DSLWG,Sage Bionetworks  number   \n",
       "159  http://purl.obolibrary.org/obo/NCIT_C45378,Sag...  string   \n",
       "175  Sage Bionetworks,Proposed minimum metadata rel...  number   \n",
       "176  Sage Bionetworks,Proposed minimum metadata rel...  string   \n",
       "\n",
       "                                              Ontology   multivalue  \n",
       "0                                     Sage Bionetworks               \n",
       "1                                     Sage Bionetworks               \n",
       "2                                     Sage Bionetworks               \n",
       "3    https://www.ebi.ac.uk/ols/ontologies/ms/termsi...        False  \n",
       "4            http://purl.obolibrary.org/obo/MS_1001455        False  \n",
       "..                                                 ...          ...  \n",
       "152  Sage Bionetworks,Sage Bionetworks,Data Contrib...  false,False  \n",
       "156            Sage Bionetworks,DSLWG,Sage Bionetworks        False  \n",
       "159  http://purl.obolibrary.org/obo/NCIT_C45378,Sag...   true,False  \n",
       "175  Sage Bionetworks,Proposed minimum metadata rel...               \n",
       "176  Sage Bionetworks,Proposed minimum metadata rel...               \n",
       "\n",
       "[65 rows x 14 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# QA check\n",
    "dm[dm[\"DependsOn Component\"].str.contains(\n",
    "    \"metabolomics\", case=False, na=False)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cleaning other values and equal values\n",
    "\n",
    "Removing illegal characters\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove any special characters\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm[\"Attribute\"] = dm[\"Attribute\"].str.replace(\"\\(|\\)|\\?\", \"\", regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm[\"Valid Values\"] = (\n",
    "    dm[\"Valid Values\"]\n",
    "    .apply(\n",
    "        lambda x: re.sub(\n",
    "            \"^0x Visium Spatial Gene Expression\",\n",
    "            \"10x Visium Spatial Gene Expression\",\n",
    "            x,\n",
    "        )\n",
    "    )\n",
    "    .apply(clean_list)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clean up equals in depends on\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_new_value(old_value):\n",
    "    new_vals = old_value.split(\"=\")\n",
    "    new_vals = [nv.strip() for nv in new_vals]\n",
    "    # convert to camel case\n",
    "    nv = new_vals[1].capitalize() + new_vals[0][0].upper() + new_vals[0][1:]\n",
    "    return nv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def recode_yes_no(v):\n",
    "    if v.lower() == \"yes\":\n",
    "        return \"TRUE\"\n",
    "    elif v.lower() == \"no\":\n",
    "        return \"FALSE\"\n",
    "    else:\n",
    "        return v\n",
    "\n",
    "\n",
    "dm[\"Valid Values\"] = (\n",
    "    dm[\"Valid Values\"]\n",
    "    .apply(lambda x: x.split(\",\"))\n",
    "    .apply(lambda x: \",\".join([recode_yes_no(y) for y in x]))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split list to process other values\n",
    "# Find the other columns in the data model\n",
    "others = dm[dm[\"DependsOn Component\"].str.contains(\"=\", na=False)].copy()\n",
    "\n",
    "others[\"DependsOn Component Original\"] = others[\"DependsOn Component\"].str.split(\",\")\n",
    "\n",
    "# Create series of equals values to use for new attributes/ valid values relationship\n",
    "others[\"equals_series\"] = others[\"DependsOn Component Original\"].apply(\n",
    "    lambda x: [y for y in x if bool(re.search(\"=\", y))][0]\n",
    ")\n",
    "\n",
    "# others[\"equals_attribute\"] = others[\"equals_series\"].apply(create_new_value)\n",
    "\n",
    "others[[\"baseAttribute\", \"equalsValue\"]] = (\n",
    "    others[\"equals_series\"]\n",
    "    .str.split(\"=\", expand=True)\n",
    "    .apply(lambda x: [y.strip() for y in x])\n",
    "    .rename({0: \"base_attribute\", 1: \"equalsValue\"}, axis=1)\n",
    ")\n",
    "\n",
    "# Deciding to use true and false for all yes/no values\n",
    "recoder = {\n",
    "    re.compile(\"^[Yy]es\", flags=re.IGNORECASE): \"TRUE\",\n",
    "    re.compile(\"true\", flags=re.IGNORECASE): \"TRUE\",\n",
    "    re.compile(\"^[Nn]o\", flags=re.IGNORECASE): \"FALSE\",\n",
    "}\n",
    "\n",
    "others[\"equalsValue\"] = others[\"equalsValue\"].replace(recoder, regex=True)\n",
    "\n",
    "others[\"newDescription\"] = others[[\"baseAttribute\", \"equalsValue\"]].apply(\n",
    "    lambda x: f\"When {x[0].strip()} = {x[1].strip()}\", axis=1\n",
    ")\n",
    "\n",
    "others[\"equalsAttribute\"] = others[[\"baseAttribute\", \"equalsValue\"]].apply(\n",
    "    lambda x: f\"{x[1].strip()}{x[0].strip()[0].upper()+x[0].strip()[1:]}\", axis=1\n",
    ")\n",
    "others[\"DependsOn Component\"] = \"\"\n",
    "others[\"Properties\"] = \"dataProperty\"\n",
    "others.loc[others[\"equalsValue\"] == \"other\", \"Parent\"] = \"Specification\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "equals_df = others.copy()\n",
    "equals_df = equals_df.drop(columns=[\"DependsOn\", \"Description\"])\n",
    "\n",
    "equals_df = equals_df.rename(\n",
    "    {\n",
    "        \"Attribute\": \"DependsOn\",\n",
    "        \"newDescription\": \"Description\",\n",
    "        \"equalsAttribute\": \"Attribute\",\n",
    "    },\n",
    "    axis=1,\n",
    ")[base_cols]\n",
    "\n",
    "equals_df[\"DependsOn Component\"] = \"\"\n",
    "\n",
    "equals_df[\"Valid Values\"] = \"\"\n",
    "\n",
    "equals_df[\"Properties\"] = \"ValidValue\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Update base attribute equals values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create mapping\n",
    "temp = others[\"equals_series\"].str.split(\"=\", expand=True)\n",
    "temp = temp.apply(lambda x: x.str.strip(), axis=1)\n",
    "temp = temp.rename({0: \"base_attribute\", 1: \"value_to_replace\"}, axis=1)\n",
    "temp[\"value_to_replace\"] = temp[\"value_to_replace\"].str.capitalize()\n",
    "\n",
    "# new value\n",
    "temp[\"new_value\"] = others[\"equalsAttribute\"]\n",
    "temp.reset_index(drop=True, inplace=True)\n",
    "\n",
    "temp = temp.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'captivityStatus': {'pattern': re.compile(r'(Captive)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'captiveCaptivityStatus'},\n",
       " 'hasAssayControl': {'pattern': re.compile(r'(Yes)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'TRUEHasAssayControl'},\n",
       " 'diagnosisStatus': {'pattern': re.compile(r'(True)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'TRUEDiagnosisStatus'},\n",
       " 'ethnicity': {'pattern': re.compile(r'(Hispanic or latino)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'Hispanic or LatinoEthnicity'},\n",
       " 'hasIonizationSource': {'pattern': re.compile(r'(Yes)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'TRUEHasIonizationSource'},\n",
       " 'msTarget': {'pattern': re.compile(r'(Targeted)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'TargetedMsTarget'},\n",
       " 'useReagent': {'pattern': re.compile(r'(Yes)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'TRUEUseReagent'},\n",
       " 'acquisitionBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherAcquisitionBatchSizeUnit'},\n",
       " 'batchSizeUnit': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherBatchSizeUnit'},\n",
       " 'controlType': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherControlType'},\n",
       " 'dnaBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherDnaBatchSizeUnit'},\n",
       " 'databaseName': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherDatabaseName'},\n",
       " 'databaseSource': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherDatabaseSource'},\n",
       " 'digestionMethod': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherDigestionMethod'},\n",
       " 'experimentalBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherExperimentalBatchSizeUnit'},\n",
       " 'extractionMethod': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherExtractionMethod'},\n",
       " 'lensVoltageUnits': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherLensVoltageUnits'},\n",
       " 'libraryPrep': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherLibraryPrep'},\n",
       " 'libraryPreparationMethod': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherLibraryPreparationMethod'},\n",
       " 'libraryVersion': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherLibraryVersion'},\n",
       " 'msAnalyteType': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherMsAnalyteType'},\n",
       " 'msAnalyzerTypeMS1': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherMsAnalyzerTypeMS1'},\n",
       " 'msAnalyzerTypeMS2': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherMsAnalyzerTypeMS2'},\n",
       " 'msAssayTechnique': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherMsAssayTechnique'},\n",
       " 'msInstrumentModel': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherMsInstrumentModel'},\n",
       " 'measurementTechnique': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherMeasurementTechnique'},\n",
       " 'cleavageAgents': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherCleavageAgents'},\n",
       " 'modificationParameters': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherModificationParameters'},\n",
       " 'platformLocation': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherPlatformLocation'},\n",
       " 'technologyPlatformVersion': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherTechnologyPlatformVersion'},\n",
       " 'ProcessingBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherProcessingBatchSizeUnit'},\n",
       " 'ProteomicsAssayType': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherProteomicsAssayType'},\n",
       " 'rnaBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherRnaBatchSizeUnit'},\n",
       " 'readLengthUnits': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherReadLengthUnits'},\n",
       " 'repositoryName': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherRepositoryName'},\n",
       " 'sampleBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherSampleBatchSizeUnit'},\n",
       " 'sampleIntdocution': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherSampleIntdocution'},\n",
       " 'sampleType': {'pattern': re.compile(r'(Other)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherSampleType'},\n",
       " 'sequencingBatchSizeUnit': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherSequencingBatchSizeUnit'},\n",
       " 'transcriptType': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherTranscriptType'},\n",
       " 'vaccumPressureUnits': {'pattern': re.compile(r'(Other)',\n",
       "             re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'otherVaccumPressureUnits'},\n",
       " 'useTreatment': {'pattern': re.compile(r'(Yes)', re.IGNORECASE|re.UNICODE),\n",
       "  'repl': 'TRUEUseTreatment'}}"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "replacements = {}\n",
    "\n",
    "for i, x in temp.iterrows():\n",
    "    ba, vtr, nv = x\n",
    "    replacements[ba] = {\n",
    "        \"pattern\": re.compile(\"(\" + vtr + \")\", flags=re.IGNORECASE),\n",
    "        \"repl\": nv,\n",
    "    }\n",
    "\n",
    "replacements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "attribute:  captivityStatus\n",
      "Index:  [13]\n",
      "--------------------\n",
      "attribute:  hasAssayControl\n",
      "Index:  [49]\n",
      "--------------------\n",
      "attribute:  diagnosisStatus\n",
      "Index:  [29]\n",
      "--------------------\n",
      "attribute:  ethnicity\n",
      "Index:  [37]\n",
      "--------------------\n",
      "attribute:  hasIonizationSource\n",
      "Index:  [50]\n",
      "--------------------\n",
      "attribute:  msTarget\n",
      "Index:  [71]\n",
      "--------------------\n",
      "attribute:  useReagent\n",
      "Index:  [173]\n",
      "--------------------\n",
      "attribute:  acquisitionBatchSizeUnit\n",
      "Index:  [2]\n",
      "--------------------\n",
      "attribute:  batchSizeUnit\n",
      "Index:  [11]\n",
      "--------------------\n",
      "attribute:  controlType\n",
      "Index:  [19]\n",
      "--------------------\n",
      "attribute:  dnaBatchSizeUnit\n",
      "Index:  [34]\n",
      "--------------------\n",
      "attribute:  databaseName\n",
      "Index:  [25]\n",
      "--------------------\n",
      "attribute:  databaseSource\n",
      "Index:  [26]\n",
      "--------------------\n",
      "attribute:  digestionMethod\n",
      "Index:  [30]\n",
      "--------------------\n",
      "attribute:  experimentalBatchSizeUnit\n",
      "experimentalBatchSizeUnit\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  extractionMethod\n",
      "Index:  [40]\n",
      "--------------------\n",
      "attribute:  lensVoltageUnits\n",
      "lensVoltageUnits\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  libraryPrep\n",
      "Index:  [60]\n",
      "--------------------\n",
      "attribute:  libraryPreparationMethod\n",
      "Index:  [61]\n",
      "--------------------\n",
      "attribute:  libraryVersion\n",
      "Index:  [62]\n",
      "--------------------\n",
      "attribute:  msAnalyteType\n",
      "Index:  [66]\n",
      "--------------------\n",
      "attribute:  msAnalyzerTypeMS1\n",
      "Index:  [67]\n",
      "--------------------\n",
      "attribute:  msAnalyzerTypeMS2\n",
      "Index:  [68]\n",
      "--------------------\n",
      "attribute:  msAssayTechnique\n",
      "Index:  [69]\n",
      "--------------------\n",
      "attribute:  msInstrumentModel\n",
      "Index:  [70]\n",
      "--------------------\n",
      "attribute:  measurementTechnique\n",
      "measurementTechnique\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  cleavageAgents\n",
      "Index:  [15]\n",
      "--------------------\n",
      "attribute:  modificationParameters\n",
      "Index:  [65]\n",
      "--------------------\n",
      "attribute:  platformLocation\n",
      "Index:  [77]\n",
      "--------------------\n",
      "attribute:  technologyPlatformVersion\n",
      "Index:  [159]\n",
      "--------------------\n",
      "attribute:  ProcessingBatchSizeUnit\n",
      "Index:  [81]\n",
      "--------------------\n",
      "attribute:  ProteomicsAssayType\n",
      "ProteomicsAssayType\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  rnaBatchSizeUnit\n",
      "Index:  [99]\n",
      "--------------------\n",
      "attribute:  readLengthUnits\n",
      "Index:  [85]\n",
      "--------------------\n",
      "attribute:  repositoryName\n",
      "Index:  [95]\n",
      "--------------------\n",
      "attribute:  sampleBatchSizeUnit\n",
      "Index:  [103]\n",
      "--------------------\n",
      "attribute:  sampleIntdocution\n",
      "sampleIntdocution\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  sampleType\n",
      "Index:  [106]\n",
      "--------------------\n",
      "attribute:  sequencingBatchSizeUnit\n",
      "Index:  [110]\n",
      "--------------------\n",
      "attribute:  transcriptType\n",
      "Index:  [164]\n",
      "--------------------\n",
      "attribute:  vaccumPressureUnits\n",
      "vaccumPressureUnits\n",
      "Index:  None\n",
      "--------------------\n",
      "attribute:  useTreatment\n",
      "Index:  [174]\n",
      "--------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/p0/5m4pdsm55jn_d5nzbjv6bjf40000gq/T/ipykernel_15102/1305637007.py:41: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.\n",
      "  df[\"Attribute\"].str.contains(\n"
     ]
    }
   ],
   "source": [
    "for attribute in replacements.keys():\n",
    "    dm = code_equals_values(dm, replacements, attribute)\n",
    "    print(\"-\" * 20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add derived attributes from \"=\" valid values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dm[\"Properties\"] = \"dataProperty\"\n",
    "dm.update(others[base_cols])\n",
    "dm = pd.concat([dm, equals_df], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Do not need\n",
    "dm[\"DependsOn Component\"] = \"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Valid Values Work\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Not sure what happened here in the RFCs\n",
    "dm[\"Valid Values\"] = dm[\"Valid Values\"].replace(recoder_valid_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "attribute:  acquisitionBatchSizeUnit\n",
      "Index:  [2]\n",
      "--------------------\n",
      "attribute:  batchSizeUnit\n",
      "Index:  [11]\n",
      "--------------------\n",
      "attribute:  captivityStatus\n",
      "Index:  [13]\n",
      "--------------------\n",
      "attribute:  cellType\n",
      "Index:  [14]\n",
      "--------------------\n",
      "attribute:  cleavageAgents\n",
      "Index:  [15]\n",
      "--------------------\n",
      "attribute:  cohort\n",
      "Index:  [16]\n",
      "--------------------\n",
      "attribute:  consentGroupID\n",
      "Index:  [18]\n",
      "--------------------\n",
      "attribute:  controlType\n",
      "Index:  [19]\n",
      "--------------------\n",
      "attribute:  conversionRatioUnits\n",
      "Index:  [21]\n",
      "--------------------\n",
      "attribute:  countryCode\n",
      "Index:  [22]\n",
      "--------------------\n",
      "attribute:  databaseName\n",
      "Index:  [25]\n",
      "--------------------\n",
      "attribute:  databaseSource\n",
      "Index:  [26]\n",
      "--------------------\n",
      "attribute:  diagnosis\n",
      "Index:  [28]\n",
      "--------------------\n",
      "attribute:  digestionMethod\n",
      "Index:  [30]\n",
      "--------------------\n",
      "attribute:  dnaBatchID\n",
      "Index:  [32]\n",
      "--------------------\n",
      "attribute:  dnaBatchSizeUnit\n",
      "Index:  [34]\n",
      "--------------------\n",
      "attribute:  enrichmentMethod\n",
      "Index:  [35]\n",
      "--------------------\n",
      "attribute:  ethnicity\n",
      "Index:  [37]\n",
      "--------------------\n",
      "attribute:  experiementalBatchSizeUnit\n",
      "Index:  [39]\n",
      "--------------------\n",
      "attribute:  extractionMethod\n",
      "Index:  [40]\n",
      "--------------------\n",
      "attribute:  fastingState\n",
      "Index:  [42]\n",
      "--------------------\n",
      "attribute:  fieldCenterCode\n",
      "Index:  [43]\n",
      "--------------------\n",
      "attribute:  isPostMortem\n",
      "Index:  [53]\n",
      "--------------------\n",
      "attribute:  lensVoltagesUnit \n",
      "Index:  [58]\n",
      "--------------------\n",
      "attribute:  libraryPrep\n",
      "Index:  [60]\n",
      "--------------------\n",
      "attribute:  libraryPreparationMethod\n",
      "Index:  [61]\n",
      "--------------------\n",
      "attribute:  libraryVersion\n",
      "Index:  [62]\n",
      "--------------------\n",
      "attribute:  lifeStage\n",
      "Index:  [63]\n",
      "--------------------\n",
      "attribute:  modificationParameters\n",
      "Index:  [65]\n",
      "--------------------\n",
      "attribute:  msAnalyteType\n",
      "Index:  [66]\n",
      "--------------------\n",
      "attribute:  msAnalyzerTypeMS1\n",
      "Index:  [67]\n",
      "--------------------\n",
      "attribute:  msAnalyzerTypeMS2\n",
      "Index:  [68]\n",
      "--------------------\n",
      "attribute:  msAssayTechnique\n",
      "Index:  [69]\n",
      "--------------------\n",
      "attribute:  msInstrumentModel\n",
      "Index:  [70]\n",
      "--------------------\n",
      "attribute:  nucleicAcidSource\n",
      "Index:  [72]\n",
      "--------------------\n",
      "attribute:  organ\n",
      "Index:  [73]\n",
      "--------------------\n",
      "attribute:  platformLocation\n",
      "Index:  [77]\n",
      "--------------------\n",
      "attribute:  processingBatchSizeUnit\n",
      "Index:  [81]\n",
      "--------------------\n",
      "attribute:  protoemicsAssayType\n",
      "Index:  [82]\n",
      "--------------------\n",
      "attribute:  race\n",
      "Index:  [83]\n",
      "--------------------\n",
      "attribute:  readLengthUnits\n",
      "Index:  [85]\n",
      "--------------------\n",
      "attribute:  repositoryName\n",
      "Index:  [95]\n",
      "--------------------\n",
      "attribute:  rnaBatchSizeUnit\n",
      "Index:  [99]\n",
      "--------------------\n",
      "attribute:  sampleBatchSizeUnit\n",
      "Index:  [103]\n",
      "--------------------\n",
      "attribute:  sampleIntroduction\n",
      "Index:  [104]\n",
      "--------------------\n",
      "attribute:  sampleType\n",
      "Index:  [106]\n",
      "--------------------\n",
      "attribute:  samplingAge\n",
      "Index:  [107]\n",
      "--------------------\n",
      "attribute:  sequencingBatchSizeUnit\n",
      "Index:  [110]\n",
      "--------------------\n",
      "attribute:  sex\n",
      "Index:  [111]\n",
      "--------------------\n",
      "attribute:  speciesGroup\n",
      "Index:  [113]\n",
      "--------------------\n",
      "attribute:  specimenAge\n",
      "Index:  [151]\n",
      "--------------------\n",
      "attribute:  specimenSource\n",
      "Index:  [154]\n",
      "--------------------\n",
      "attribute:  specimenType\n",
      "Index:  [155]\n",
      "--------------------\n",
      "attribute:  technologyPlatformVersion\n",
      "Index:  [159]\n",
      "--------------------\n",
      "attribute:  tissue\n",
      "Index:  [160]\n",
      "--------------------\n",
      "attribute:  transcriptType\n",
      "Index:  [164]\n",
      "--------------------\n",
      "attribute:  vacuumPressureUnit\n",
      "Index:  [176]\n",
      "--------------------\n",
      "attribute:  visitCode\n",
      "Index:  [178]\n",
      "--------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/p0/5m4pdsm55jn_d5nzbjv6bjf40000gq/T/ipykernel_15102/519273947.py:5: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.\n",
      "  dm[\"Valid Values\"].str.contains(pattern, flags=re.IGNORECASE, regex=True)\n",
      "/var/folders/p0/5m4pdsm55jn_d5nzbjv6bjf40000gq/T/ipykernel_15102/1305637007.py:41: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.\n",
      "  df[\"Attribute\"].str.contains(\n"
     ]
    }
   ],
   "source": [
    "# valid values that contain other\n",
    "pattern = \"([Oo]ther)\"\n",
    "\n",
    "pure_others = dm[\n",
    "    dm[\"Valid Values\"].str.contains(pattern, flags=re.IGNORECASE, regex=True)\n",
    "].copy()\n",
    "\n",
    "pure_others.loc[:, \"replacement_value\"] = pure_others.loc[:, \"Attribute\"].apply(\n",
    "    lambda x: \"Other\" + (x[0].upper() + x[1:])\n",
    ")\n",
    "\n",
    "regex_dict = {}\n",
    "\n",
    "for i, r in pure_others.iterrows():\n",
    "    regex_dict[r[\"Attribute\"]] = {\n",
    "        \"pattern\": re.compile(pattern, flags=re.IGNORECASE),\n",
    "        \"repl\": r[\"replacement_value\"],\n",
    "    }\n",
    "\n",
    "# json_formatted = json.dumps(regex_dict, indent=4)\n",
    "# print(json_formatted)\n",
    "\n",
    "for attribute in regex_dict.keys():\n",
    "    dm = code_equals_values(dm, regex_dict, attribute)\n",
    "    print(\"-\" * 20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cleanup valid values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fuzzy matching to find misspellings\n",
    "# Fuzzy matching\n",
    "\n",
    "valid_values = \",\".join(dm[\"Valid Values\"])\n",
    "valid_values = valid_values.split(\",\")\n",
    "valid_values = list(np.unique(valid_values))\n",
    "valid_values = [v.strip() for v in valid_values if len(v) > 0]\n",
    "\n",
    "\n",
    "scores = {}\n",
    "for v in valid_values:\n",
    "    scores[v] = {}\n",
    "    for v2 in valid_values:\n",
    "        if v == v2:\n",
    "            next\n",
    "        else:\n",
    "            score = fuzz.ratio(v.lower(), v2.lower())\n",
    "            if score == 100:\n",
    "                scores[v][v2] = score\n",
    "    if len(scores[v]) == 0:\n",
    "        scores.pop(v)\n",
    "\n",
    "# create recoding variables off fuzzy matching\n",
    "new_values_recoded = []\n",
    "\n",
    "for v in scores.values():\n",
    "    new_values_recoded.append(list(v.keys())[0].title())\n",
    "\n",
    "new_values_recoded = np.unique(new_values_recoded)\n",
    "\n",
    "# recoder_valid_values = {}\n",
    "for nv in new_values_recoded:\n",
    "    recoder_valid_values[re.compile(nv, flags=re.IGNORECASE)] = nv\n",
    "    # recoder_valid_values.append(value_add)\n",
    "\n",
    "dm[[\"Valid Values\", \"multivalue\"]] = (\n",
    "    dm[[\"Valid Values\", \"multivalue\"]]\n",
    "    .replace(recoder_valid_values, regex=True)\n",
    "    .fillna(\"\")\n",
    "    .applymap(lambda x: clean_list(x))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>UsedIn</th>\n",
       "      <th>Description</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10x</td>\n",
       "      <td>libraryPreparationMethod</td>\n",
       "      <td>Method by which library was prepared</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ValidValue</td>\n",
       "      <td></td>\n",
       "      <td>Microbiome,RNAseq,Whole Genome Sequencing,bsSe...</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>STRING</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10x Visium Spatial Gene Expression</td>\n",
       "      <td>platformLocation,technologyPlatformVersion</td>\n",
       "      <td>The specific version (application, manufacture...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ValidValue</td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human,Microbiome,Proteomics,RNAse...</td>\n",
       "      <td></td>\n",
       "      <td>DSLWG,Sage Bionetworks,http://purl.obolibrary....</td>\n",
       "      <td>STRING</td>\n",
       "      <td>DSLWG,Sage Bionetworks,http://purl.obolibrary....</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AFU</td>\n",
       "      <td>acquisitionBatchSizeUnit,batchSizeUnit,convers...</td>\n",
       "      <td>Unit of treatment amount,Unit of lens voltages...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.)|([0-9]+)</td>\n",
       "      <td>Genotyping,Metabolomics Human,Microbiome,RNAse...</td>\n",
       "      <td></td>\n",
       "      <td>ImmPort,Proposed minimum metadata relative to ...</td>\n",
       "      <td>STRING</td>\n",
       "      <td>ImmPort,Proposed minimum metadata relative to ...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AI</td>\n",
       "      <td>acquisitionBatchSizeUnit,batchSizeUnit,convers...</td>\n",
       "      <td>Unit of treatment amount,Unit of lens voltages...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.)|([0-9]+)</td>\n",
       "      <td>Genotyping,Metabolomics Human,Microbiome,RNAse...</td>\n",
       "      <td></td>\n",
       "      <td>ImmPort,Proposed minimum metadata relative to ...</td>\n",
       "      <td>STRING</td>\n",
       "      <td>ImmPort,Proposed minimum metadata relative to ...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AIBL pool</td>\n",
       "      <td>controlType</td>\n",
       "      <td>Control samples suitable for normalization and...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ValidValue</td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>https://doi.org/10.1101/2020.05.19.105197https...</td>\n",
       "      <td>STRING</td>\n",
       "      <td>https://doi.org/10.1101/2020.05.19.105197https...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Attribute  \\\n",
       "0                                 10x   \n",
       "1  10x Visium Spatial Gene Expression   \n",
       "2                                 AFU   \n",
       "3                                  AI   \n",
       "4                           AIBL pool   \n",
       "\n",
       "                                              UsedIn  \\\n",
       "0                           libraryPreparationMethod   \n",
       "1         platformLocation,technologyPlatformVersion   \n",
       "2  acquisitionBatchSizeUnit,batchSizeUnit,convers...   \n",
       "3  acquisitionBatchSizeUnit,batchSizeUnit,convers...   \n",
       "4                                        controlType   \n",
       "\n",
       "                                         Description Required DependsOn  \\\n",
       "0               Method by which library was prepared    False             \n",
       "1  The specific version (application, manufacture...    False             \n",
       "2  Unit of treatment amount,Unit of lens voltages...    False             \n",
       "3  Unit of treatment amount,Unit of lens voltages...    False             \n",
       "4  Control samples suitable for normalization and...    False             \n",
       "\n",
       "  DependsOn Component  Properties                         Validation Rules  \\\n",
       "0                      ValidValue                                            \n",
       "1                      ValidValue                                            \n",
       "2                      ValidValue  regex search ([0-9]+\\.[0-9]*.)|([0-9]+)   \n",
       "3                      ValidValue  regex search ([0-9]+\\.[0-9]*.)|([0-9]+)   \n",
       "4                      ValidValue                                            \n",
       "\n",
       "                                            Template Parent  \\\n",
       "0  Microbiome,RNAseq,Whole Genome Sequencing,bsSe...          \n",
       "1  Metabolomics Human,Microbiome,Proteomics,RNAse...          \n",
       "2  Genotyping,Metabolomics Human,Microbiome,RNAse...          \n",
       "3  Genotyping,Metabolomics Human,Microbiome,RNAse...          \n",
       "4                                 Metabolomics Human          \n",
       "\n",
       "                                              Source    Type  \\\n",
       "0                                   Sage Bionetworks  STRING   \n",
       "1  DSLWG,Sage Bionetworks,http://purl.obolibrary....  STRING   \n",
       "2  ImmPort,Proposed minimum metadata relative to ...  STRING   \n",
       "3  ImmPort,Proposed minimum metadata relative to ...  STRING   \n",
       "4  https://doi.org/10.1101/2020.05.19.105197https...  STRING   \n",
       "\n",
       "                                            Ontology multivalue  \n",
       "0                                   Sage Bionetworks      False  \n",
       "1  DSLWG,Sage Bionetworks,http://purl.obolibrary....      False  \n",
       "2  ImmPort,Proposed minimum metadata relative to ...      False  \n",
       "3  ImmPort,Proposed minimum metadata relative to ...      False  \n",
       "4  https://doi.org/10.1101/2020.05.19.105197https...      False  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm[\"Valid Values\"] = (\n",
    "    dm[\"Valid Values\"].apply(lambda x: clean_list(x)\n",
    "                             ).apply(lambda x: x.split(\",\"))\n",
    ")\n",
    "\n",
    "# Expand each valid value into its own row\n",
    "dm_vv = dm.explode(\"Valid Values\")\n",
    "\n",
    "# join valid values back together in data model as string\n",
    "dm[\"Valid Values\"] = dm[\"Valid Values\"].apply(lambda x: \",\".join(x).strip(\",\"))\n",
    "\n",
    "# Group valid values to create unique attribute and trace where value is used in another attribute and template\n",
    "dm_vv = (\n",
    "    dm_vv.dropna(subset=\"Valid Values\")\n",
    "    .groupby(\"Valid Values\")\n",
    "    .agg(lambda x: \",\".join(set(x.astype(str))).strip(\",\"))\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "dm_vv[\"Properties\"] = \"ValidValue\"\n",
    "dm_vv[\"Required\"] = \"False\"\n",
    "\n",
    "# rename for concatenating with data model\n",
    "dm_vv = dm_vv.rename({\"Valid Values\": \"Attribute\",\n",
    "                     \"Attribute\": \"UsedIn\"}, axis=1)\n",
    "\n",
    "dm_vv[list_cols] = dm_vv[list_cols].applymap(clean_list)\n",
    "\n",
    "dm_vv = dm_vv[dm_vv[\"Attribute\"] != \"\"].reset_index(drop=True)\n",
    "\n",
    "# clean up type\n",
    "dm_vv[\"Type\"] = \"STRING\"\n",
    "\n",
    "# cleanup multivalue\n",
    "dm_vv[\"multivalue\"] = \"False\"\n",
    "\n",
    "dm_vv.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(435, 14)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm_vv.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dm[\"Valid Values\"] = (\n",
    "#     dm[\"Valid Values\"]\n",
    "#     .replace(recoder_valid_values, regex=True)\n",
    "#     .apply(lambda x: clean_list(x))\n",
    "# )\n",
    "\n",
    "# valid_values = list(np.unique(\",\".join(dm[\"Valid Values\"]).split(\",\")))\n",
    "\n",
    "# valid_values = [v.strip() for v in valid_values if len(v) > 0]\n",
    "\n",
    "# valid_values_df = pd.DataFrame({\"Attribute\": pd.Series(valid_values)})\n",
    "\n",
    "# valid_values_df[\"Properties\"] = \"validValue\"\n",
    "# valid_values_df[\"Required\"] = \"False\"\n",
    "\n",
    "# valid_values_df = valid_values_df[\n",
    "#     ~valid_values_df[\"Attribute\"].isin(dm[\"Attribute\"].tolist())\n",
    "# ]\n",
    "\n",
    "# valid_values_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(239, 14)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(674, 15)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# adding valid values found in attribute columns\n",
    "print(dm.shape)\n",
    "\n",
    "dm2 = pd.concat([dm, dm_vv], axis=0, ignore_index=True)\n",
    "\n",
    "display(dm2.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29\n"
     ]
    }
   ],
   "source": [
    "print(sum(dm2.duplicated(subset=\"Attribute\", keep=False)))\n",
    "\n",
    "# dm2.loc[dm2.duplicated(subset=\"Attribute\", keep=False),].sort_values(by=\"Attribute\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "tags": [
     "new attributes"
    ]
   },
   "outputs": [],
   "source": [
    "# Create measurement unit attributes\n",
    "# Separate out measurement units\n",
    "r = re.compile(\"(^Other)\")\n",
    "\n",
    "measurement_units = np.unique(\n",
    "    \",\".join(\n",
    "        dm.loc[\n",
    "            dm[\"Valid Values\"].str.contains(\"units\", regex=True), \"Valid Values\"\n",
    "        ].values.tolist()\n",
    "    ).split(\",\")\n",
    ")\n",
    "\n",
    "measurement_units = [\n",
    "    x\n",
    "    for x in measurement_units\n",
    "    if x not in [\"Not Specified\", \"Other\", \"Unknown\", \"Not Available\"]\n",
    "    and not bool(r.search(x))\n",
    "]\n",
    "\n",
    "dm2.loc[dm2[\"Attribute\"].isin(measurement_units), \"Parent\"] = \"MeasurementUnit\"\n",
    "dm2.loc[dm2[\"Attribute\"].isin(measurement_units), \"Description\"] = \"Measurement unit\"\n",
    "dm2.loc[dm2[\"Attribute\"].isin(measurement_units), \"Type\"] = \"STRING\"\n",
    "dm2.loc[dm2[\"Attribute\"].isin(measurement_units), \"multivalue\"] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "tags": [
     "cleanup"
    ]
   },
   "outputs": [],
   "source": [
    "# Nonsense attributes\n",
    "dm2 = dm2.drop(\n",
    "    index=dm2.loc[dm2[\"Attribute\"].str.contains(\n",
    "        \"Possible values\"),].index.tolist()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'f': {'F': 100}, 'F': {'f': 100}}"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fuzzy matching to find misspellings\n",
    "# Fuzzy matching\n",
    "\n",
    "valid_values = dm2[\"Attribute\"].replace(\n",
    "    recoder_valid_values, regex=True).tolist()\n",
    "\n",
    "scores = {}\n",
    "for v in valid_values:\n",
    "    scores[v] = {}\n",
    "    for v2 in valid_values:\n",
    "        if v == v2:\n",
    "            next\n",
    "        else:\n",
    "            score = fuzz.ratio(v.lower(), v2.lower())\n",
    "            if score == 100:\n",
    "                scores[v][v2] = score\n",
    "    if len(scores[v]) == 0:\n",
    "        scores.pop(v)\n",
    "\n",
    "scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm2[\"Attribute\"] = dm2[\"Attribute\"].replace(recoder_valid_values, regex=True)\n",
    "\n",
    "dm2[[\"Valid Values\", \"multivalue\"]] = (\n",
    "    dm2[[\"Valid Values\", \"multivalue\"]]\n",
    "    .fillna(\"\")\n",
    "    .astype(str)\n",
    "    .replace(recoder_valid_values, regex=True)\n",
    "    .applymap(lambda x: clean_list(x))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dm2 shape: (648, 15)\n",
      "Duplicates: 0\n"
     ]
    }
   ],
   "source": [
    "dm2 = (\n",
    "    dm2.dropna(subset=\"Attribute\")\n",
    "    .groupby(\"Attribute\")\n",
    "    .agg(lambda x: \",\".join(set(x.astype(str))).strip(\",\"))\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "dm2 = dm2.drop(index=dm2[dm2[\"Attribute\"] == \"f\"].index.tolist()).reset_index(drop=True)\n",
    "\n",
    "dm2[list_cols] = dm2[list_cols].applymap(clean_list)\n",
    "\n",
    "print(f\"dm2 shape: {dm2.shape}\")\n",
    "print(f'Duplicates: {sum(dm2.duplicated(subset=\"Attribute\", keep=False))}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check columns for Speical Characters\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Cardiovascular Health Study (CHS)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Cerebrospinal Fluid (CSF)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>Illumina Infinium MethylationEPIC BeadChip v1.0 (850k)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>Illumina Infinium MethylationEPIC BeadChip v2.0 (935k)Illumina MiSeq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>MassBank of North America (MoNA)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>Study of Osteoporotic Fractures (SOF)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>The Health and Aging and Body Composition Study (HealthABC)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>The Osteoporotic Fractures in Men (MrOS) Study</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                Attribute\n",
       "39                                      Cardiovascular Health Study (CHS)\n",
       "43                                              Cerebrospinal Fluid (CSF)\n",
       "111                Illumina Infinium MethylationEPIC BeadChip v1.0 (850k)\n",
       "112  Illumina Infinium MethylationEPIC BeadChip v2.0 (935k)Illumina MiSeq\n",
       "159                                      MassBank of North America (MoNA)\n",
       "309                                 Study of Osteoporotic Fractures (SOF)\n",
       "320           The Health and Aging and Body Composition Study (HealthABC)\n",
       "321                        The Osteoporotic Fractures in Men (MrOS) Study"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "check_cols = [\"Attribute\"]\n",
    "\n",
    "mask = np.column_stack(\n",
    "    [dm2[col].str.contains(\"\\(|\\)\", na=False) for col in dm2[check_cols]]\n",
    ")\n",
    "\n",
    "with pd.option_context(\"display.max_colwidth\", None):\n",
    "    display(dm2[check_cols].loc[mask.any(axis=1)])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create Manifests in data model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_templates = np.unique(\n",
    "    \",\".join(dm2[\"Template\"].dropna().values.tolist()).split(\",\")\n",
    ")\n",
    "templates = pd.DataFrame({\"Attribute\": unique_templates})\n",
    "templates[\"Required\"] = \"True\"\n",
    "templates[\"Properties\"] = \"dataType\"\n",
    "templates[\"Parent\"] = \"Template\"\n",
    "templates[\"Description\"] = templates[\"Attribute\"].apply(\n",
    "    lambda x: f\"Metadata template for {x}\"\n",
    ")\n",
    "\n",
    "# Adjust bseq\n",
    "templates.loc[\n",
    "    templates[\"Attribute\"] == \"bsSeq (bisulfite-seq WGBS methylseq methylomics)\",\n",
    "    \"Description\",\n",
    "] = \"Template for bisulfite-seq WGBS methylseq methylomics\"\n",
    "templates.loc[\n",
    "    templates[\"Attribute\"] == \"bsSeq (bisulfite-seq WGBS methylseq methylomics)\",\n",
    "    \"Attribute\",\n",
    "] = \"bsSeq\"\n",
    "\n",
    "# update templates column for other attributes\n",
    "dm2[\"Template\"] = dm2[\"Template\"].str.replace(\n",
    "    \"bsSeq (bisulfite-seq WGBS methylseq methylomics)\", \"bsSeq\", regex=False\n",
    ")\n",
    "\n",
    "# create depends On column for templates\n",
    "for i, m in enumerate(templates[\"Attribute\"]):\n",
    "    dm2[\"Template\"].str.contains(m)\n",
    "    templates.loc[i, \"DependsOn\"] = \",\".join(\n",
    "        dm2.loc[\n",
    "            dm2[\"Template\"].str.contains(m, regex=False, na=False), \"Attribute\"\n",
    "        ].tolist()\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(660, 15)\n"
     ]
    }
   ],
   "source": [
    "dm2 = pd.concat([dm2, templates], axis=0, ignore_index=True)\n",
    "\n",
    "print(dm2.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm2[[\"Valid Values\", \"Template\", \"Ontology\", \"Source\"]] = (\n",
    "    dm2[[\"Valid Values\", \"Template\", \"Ontology\", \"Source\"]]\n",
    "    .fillna(\"\")\n",
    "    .applymap(clean_list)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Recode required columns and fix spelling mistakes\n",
    "required_recoder = {\"0.0\": \"False\", \"1.0\": \"True\", \"FASLSE\": \"False\"}\n",
    "\n",
    "dm2[\"Required\"] = dm2[\"Required\"].replace(required_recoder)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Last bit of cleanup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove measurement technique dependency from biospecimen human\n",
    "dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"] = (\n",
    "    dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"]\n",
    "    .values[0]\n",
    "    .replace(\"measurementTechnique,\", \"\")\n",
    ")\n",
    "bio_measure_technique_index = dm2.query(\n",
    "    'Attribute == \"measurementTechnique\" and Parent == \"Biospecimen human\"'\n",
    ")\n",
    "\n",
    "if len(bio_measure_technique_index.index) > 0:\n",
    "    dm2 = dm2.drop(\n",
    "        index=bio_measure_technique_index.index[0]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove measurement technique dependency from biospecimen human\n",
    "dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"] = (\n",
    "    dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"]\n",
    "    .values[0]\n",
    "    .replace(\"specifyMeasurementTechnique,\", \"\")\n",
    ")\n",
    "bio_measure_technique_index = dm2.query(\n",
    "    'Attribute == \"specifyMeasurementTechnique\" and Parent == \"Biospecimen human\"'\n",
    ")\n",
    "\n",
    "\n",
    "if len(bio_measure_technique_index.index) > 0:\n",
    "    dm2 = dm2.drop(\n",
    "        index=bio_measure_technique_index.index[0]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove measurement technique dependency from biospecimen human\n",
    "dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"] = (\n",
    "    dm2.loc[dm2[\"Attribute\"] == \"Biospecimen human\", \"DependsOn\"]\n",
    "    .values[0]\n",
    "    .replace(\"OtherMeasurementTechnique,\", \"\")\n",
    ")\n",
    "bio_measure_technique_index = dm2.query(\n",
    "    'Attribute == \"OtherMeasurementTechnique\"')\n",
    "\n",
    "\n",
    "if len(bio_measure_technique_index.index) > 0:\n",
    "    dm2 = dm2.drop(\n",
    "        index=bio_measure_technique_index.index[0]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm2.loc[dm2[dm2[\"Attribute\"] == \"visitCode\"].index.values[0], \"Valid Values\"] = \"\"\n",
    "\n",
    "dm2.loc[dm2[dm2[\"Attribute\"] == \"visitCode\"].index.values[0],\n",
    "        \"Validation Rules\"] = \"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extra comma at beginning of valid values\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validation Rules\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "mixed_attrs = [\n",
    "    {\"attribute\": \"tissueWeight\", \"val_type\": \"mixed float\", \"regex\": \"regex search\"},\n",
    "    {\"attribute\": \"tissueVolume\", \"val_type\": \"mixed float\", \"regex\": \"regex search\"},\n",
    "    {\"attribute\": \"specimenAge\", \"val_type\": \"mixed integer\", \"regex\": \"regex search\"},\n",
    "    {\"attribute\": \"samplingAge\", \"val_type\": \"mixed integer\", \"regex\": \"regex search\"},\n",
    "    {\"attribute\": \"age\", \"val_type\": \"mixed integer\", \"regex\": \"regex search\"},\n",
    "]\n",
    "\n",
    "for ma in mixed_attrs:\n",
    "    attribute = ma[\"attribute\"]\n",
    "    val_type = ma[\"val_type\"]\n",
    "\n",
    "    # get indexes for new validation rules based on attribute\n",
    "    indexes = dm2[dm2[\"Attribute\"] == attribute].index.tolist()\n",
    "\n",
    "    for i in indexes:\n",
    "        if val_type == \"integer\":\n",
    "            first_part = \"[0-9]+\"\n",
    "        elif val_type == \"float\":\n",
    "            first_part = \"^\\d*?\\.?\\d$\"\n",
    "        elif val_type == \"mixed integer\":\n",
    "            regex = \"regex search\"\n",
    "            num_match = \"^\\d*?\"\n",
    "        elif val_type == \"mixed float\":\n",
    "            regex = \"regex search\"\n",
    "            num_match = \"^\\d*?\\.?\\d$\"\n",
    "            # All valid values are applicable\n",
    "            new_string = (\n",
    "                regex\n",
    "                + num_match\n",
    "                + \"|\"\n",
    "                + \"\"\n",
    "                + \"|\".join(dm2.loc[i, \"Valid Values\"].split(\",\"))\n",
    "            )\n",
    "\n",
    "        dm2.loc[i, \"Validation Rules\"] = new_string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['',\n",
       " 'regex search ([0-9]+\\\\.[0-9]*.)|([0-9]+)',\n",
       " 'regex search ([0-9]+),regex search ([0-9]+\\\\.[0-9]*.)|([0-9]+)',\n",
       " 'regex search ([0-9]+)',\n",
       " 'regex search^\\\\d*?\\\\.?\\\\d$|Not applicable|Not collected|Unknown',\n",
       " nan]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm2[\"Validation Rules\"].unique().tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Building Dependencies\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "dependencies = {\n",
    "    \"specimenID\": \"matchAtLeastOne Biospecimenhuman.specimenID value\",\n",
    "    \"individualID\": \"matchExactlyOne IndividualHuman.individualID set\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "for k, v in dependencies.items():\n",
    "    indexes = dm2[dm2[\"Attribute\"] == k][\"Validation Rules\"].index.values\n",
    "    dm2.loc[indexes, \"Validation Rules\"] = v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm2.loc[3, \"Valid Values\"] = dm2.loc[3, \"Valid Values\"].replace(\n",
    "    \",Whole Genome Sequencing\", \"\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create File Annotations Attributes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "# base file annotations\n",
    "base_file_annotations = {\"resourceType\": \"\",\n",
    "                         \"isReleased\": False, \"fileType\": \"\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>UsedIn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>visitCode</td>\n",
       "      <td>Indicate which longitudinal visit for the individual the data comes from, provided by the data contributor's data dictionary</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>number</td>\n",
       "      <td>Data Contributor,Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Attribute  \\\n",
       "646  visitCode   \n",
       "\n",
       "                                                                                                                      Description  \\\n",
       "646  Indicate which longitudinal visit for the individual the data comes from, provided by the data contributor's data dictionary   \n",
       "\n",
       "    Valid Values Required DependsOn DependsOn Component Properties  \\\n",
       "646                  True                                            \n",
       "\n",
       "    Validation Rules          Template Parent            Source    Type  \\\n",
       "646                   Individual Human         Sage Bionetworks  number   \n",
       "\n",
       "                              Ontology multivalue UsedIn  \n",
       "646  Data Contributor,Sage Bionetworks      False         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with pd.option_context(\"display.max_colwidth\", None):\n",
    "    display(dm2.query('Attribute == \"visitCode\"'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write out new data model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "for h in hard_coded_valid_values:\n",
    "    dm2.loc[\n",
    "        dm2[dm2[\"Attribute\"] == h[\"attribute\"]].index.values[0], \"Valid Values\"\n",
    "    ] = h[\"valid_value\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>UsedIn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>age</td>\n",
       "      <td>Age of the individual (age in years of the individual at first recorded study event (enrollment, visit, observation, sample collection, survey completion, etc.)</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>regex search^\\d*?\\.?\\d$|Not applicable|Not collected|Unknown</td>\n",
       "      <td>Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383</th>\n",
       "      <td>consentGroupID</td>\n",
       "      <td>Indicate the consent group for the individual, provided by the data contributor's data dictionary</td>\n",
       "      <td>1,2,3</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Data Contributor,Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552</th>\n",
       "      <td>samplingAge</td>\n",
       "      <td>The calculated age of the sample, measurement is determined or coded by the data contributor.</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>regex search^\\d*?\\.?\\d$|Not applicable|Not collected|Unknown</td>\n",
       "      <td>Biospecimen human,Biospecimen nonHuman</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>603</th>\n",
       "      <td>specimenAge</td>\n",
       "      <td>The subject's age at the time of specimen extraction is the individual's age (since birth) at the time a given specimen was extracted. Measured as age in years.</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>regex search^\\d*?\\.?\\d$|Not applicable|Not collected|Unknown</td>\n",
       "      <td>Biospecimen human,Biospecimen nonHuman</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>618</th>\n",
       "      <td>tissueWeight</td>\n",
       "      <td>The mass of the tissue specimen. Measured in mg.</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>regex search^\\d*?\\.?\\d$|Not applicable|Not collected|Unknown</td>\n",
       "      <td>Biospecimen human,Biospecimen nonHuman</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>string</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>visitCode</td>\n",
       "      <td>Indicate which longitudinal visit for the individual the data comes from, provided by the data contributor's data dictionary</td>\n",
       "      <td>1,2,3,4,Other,Unknown,Not collected,Not applicable</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>number</td>\n",
       "      <td>Data Contributor,Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Attribute  \\\n",
       "362             age   \n",
       "383  consentGroupID   \n",
       "552     samplingAge   \n",
       "603     specimenAge   \n",
       "618    tissueWeight   \n",
       "646       visitCode   \n",
       "\n",
       "                                                                                                                                                          Description  \\\n",
       "362  Age of the individual (age in years of the individual at first recorded study event (enrollment, visit, observation, sample collection, survey completion, etc.)   \n",
       "383                                                                 Indicate the consent group for the individual, provided by the data contributor's data dictionary   \n",
       "552                                                                    The calculated age of the sample, measurement is determined or coded by the data contributor.    \n",
       "603  The subject's age at the time of specimen extraction is the individual's age (since birth) at the time a given specimen was extracted. Measured as age in years.   \n",
       "618                                                                                                                  The mass of the tissue specimen. Measured in mg.   \n",
       "646                                      Indicate which longitudinal visit for the individual the data comes from, provided by the data contributor's data dictionary   \n",
       "\n",
       "                                           Valid Values Required DependsOn  \\\n",
       "362                                                         True             \n",
       "383                                               1,2,3     True             \n",
       "552                                                         True             \n",
       "603                                                         True             \n",
       "618                                                         True             \n",
       "646  1,2,3,4,Other,Unknown,Not collected,Not applicable     True             \n",
       "\n",
       "    DependsOn Component Properties  \\\n",
       "362                                  \n",
       "383                                  \n",
       "552                                  \n",
       "603                                  \n",
       "618                                  \n",
       "646                                  \n",
       "\n",
       "                                                 Validation Rules  \\\n",
       "362  regex search^\\d*?\\.?\\d$|Not applicable|Not collected|Unknown   \n",
       "383                                                                 \n",
       "552  regex search^\\d*?\\.?\\d$|Not applicable|Not collected|Unknown   \n",
       "603  regex search^\\d*?\\.?\\d$|Not applicable|Not collected|Unknown   \n",
       "618  regex search^\\d*?\\.?\\d$|Not applicable|Not collected|Unknown   \n",
       "646                                                                 \n",
       "\n",
       "                                   Template Parent            Source    Type  \\\n",
       "362                        Individual Human         Sage Bionetworks  string   \n",
       "383                        Individual Human         Sage Bionetworks  string   \n",
       "552  Biospecimen human,Biospecimen nonHuman         Sage Bionetworks  string   \n",
       "603  Biospecimen human,Biospecimen nonHuman         Sage Bionetworks  string   \n",
       "618  Biospecimen human,Biospecimen nonHuman         Sage Bionetworks  string   \n",
       "646                        Individual Human         Sage Bionetworks  number   \n",
       "\n",
       "                              Ontology multivalue UsedIn  \n",
       "362                   Sage Bionetworks      False         \n",
       "383  Data Contributor,Sage Bionetworks      False         \n",
       "552                   Sage Bionetworks      False         \n",
       "603                   Sage Bionetworks      False         \n",
       "618                   Sage Bionetworks      False         \n",
       "646  Data Contributor,Sage Bionetworks      False         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "checks = [c[\"attribute\"] for c in hard_coded_valid_values]\n",
    "\n",
    "with pd.option_context(\"display.max_colwidth\", None):\n",
    "    display(dm2.query(f\"Attribute in @checks\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(660, 15)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(660, 15)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(dm2.shape)\n",
    "\n",
    "dm2 = (\n",
    "    dm2.fillna(\"\").drop_duplicates(subset=[\"Attribute\"]).reset_index(drop=True)\n",
    ")  # 'DependsOn', 'Properties',\n",
    "\n",
    "display(dm2.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm2[\"Type\"] = dm2[\"Type\"].str.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>UsedIn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Attribute, Description, Valid Values, Required, DependsOn, DependsOn Component, Properties, Validation Rules, Template, Parent, Source, Type, Ontology, multivalue, UsedIn]\n",
       "Index: []"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm2[dm2[\"Attribute\"] == \"Other\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>UsedIn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>experiementalBatchSize</td>\n",
       "      <td>The number of samples</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.)|([0-9]+)</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NUMBER</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510</th>\n",
       "      <td>pcrCycles</td>\n",
       "      <td>Number of PCR cycles to amplify transposased D...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.)|([0-9]+)</td>\n",
       "      <td>bsSeq</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NUMBER</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>Leukocytes</td>\n",
       "      <td>The type of sample collected</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ValidValue</td>\n",
       "      <td></td>\n",
       "      <td>Genotyping,Metabolomics Human,Microbiome,Prote...</td>\n",
       "      <td></td>\n",
       "      <td>ImmPort,Sage Bionetworks</td>\n",
       "      <td>STRING</td>\n",
       "      <td>ImmPort,Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td>sampleType</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>621</th>\n",
       "      <td>transcriptType</td>\n",
       "      <td>The type of transcript reported</td>\n",
       "      <td>Not Specified,Not applicable,Not collected,Oth...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>RNAseq,Whole Genome Sequencing,scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td>ImmPort</td>\n",
       "      <td>STRING</td>\n",
       "      <td>ImmPort</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>TruSeq</td>\n",
       "      <td>Method by which library was prepared</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ValidValue</td>\n",
       "      <td></td>\n",
       "      <td>Microbiome,RNAseq,Whole Genome Sequencing,bsSe...</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>STRING</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td>libraryPreparationMethod</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>OtherRepositoryNameRepositoryName</td>\n",
       "      <td>The public repository name for the transcript ...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ValidValue</td>\n",
       "      <td></td>\n",
       "      <td>RNAseq,Whole Genome Sequencing,scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>STRING</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td>repositoryName</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>OtherConsentGroupID</td>\n",
       "      <td>Indicate the consent group for the individual,...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ValidValue</td>\n",
       "      <td></td>\n",
       "      <td>Individual Human</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>STRING</td>\n",
       "      <td>Data Contributor,Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td>consentGroupID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>Tonsil</td>\n",
       "      <td>The type of sample collected</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ValidValue</td>\n",
       "      <td></td>\n",
       "      <td>Genotyping,Metabolomics Human,Microbiome,Prote...</td>\n",
       "      <td></td>\n",
       "      <td>ImmPort,Sage Bionetworks</td>\n",
       "      <td>STRING</td>\n",
       "      <td>ImmPort,Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td>sampleType</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550</th>\n",
       "      <td>samplePrepProtocol</td>\n",
       "      <td>An internet address that may provide more deta...</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics Human,Proteomics</td>\n",
       "      <td></td>\n",
       "      <td>DSLWG,Sage Bionetworks</td>\n",
       "      <td>STRING</td>\n",
       "      <td>DSLWG,Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411</th>\n",
       "      <td>fractionIdentifier</td>\n",
       "      <td>Identifier string that describes the sample fr...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.)|([0-9]+)</td>\n",
       "      <td>Proteomics</td>\n",
       "      <td></td>\n",
       "      <td>MS Ontology,Sage Bionetworks,The Proteomics Sa...</td>\n",
       "      <td>NUMBER</td>\n",
       "      <td>MS Ontology,Sage Bionetworks,The Proteomics Sa...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Attribute  \\\n",
       "405             experiementalBatchSize   \n",
       "510                          pcrCycles   \n",
       "141                         Leukocytes   \n",
       "621                     transcriptType   \n",
       "325                             TruSeq   \n",
       "239  OtherRepositoryNameRepositoryName   \n",
       "207                OtherConsentGroupID   \n",
       "323                             Tonsil   \n",
       "550                 samplePrepProtocol   \n",
       "411                 fractionIdentifier   \n",
       "\n",
       "                                           Description  \\\n",
       "405                             The number of samples    \n",
       "510  Number of PCR cycles to amplify transposased D...   \n",
       "141                       The type of sample collected   \n",
       "621                    The type of transcript reported   \n",
       "325               Method by which library was prepared   \n",
       "239  The public repository name for the transcript ...   \n",
       "207  Indicate the consent group for the individual,...   \n",
       "323                       The type of sample collected   \n",
       "550  An internet address that may provide more deta...   \n",
       "411  Identifier string that describes the sample fr...   \n",
       "\n",
       "                                          Valid Values Required DependsOn  \\\n",
       "405                                                       False             \n",
       "510                                                        True             \n",
       "141                                                       False             \n",
       "621  Not Specified,Not applicable,Not collected,Oth...     True             \n",
       "325                                                       False             \n",
       "239                                                       False             \n",
       "207                                                       False             \n",
       "323                                                       False             \n",
       "550                                                       False             \n",
       "411                                                        True             \n",
       "\n",
       "    DependsOn Component  Properties                         Validation Rules  \\\n",
       "405                                  regex search ([0-9]+\\.[0-9]*.)|([0-9]+)   \n",
       "510                                  regex search ([0-9]+\\.[0-9]*.)|([0-9]+)   \n",
       "141                      ValidValue                                            \n",
       "621                                                                            \n",
       "325                      ValidValue                                            \n",
       "239                      ValidValue                                            \n",
       "207                      ValidValue                                            \n",
       "323                      ValidValue                                            \n",
       "550                                                                            \n",
       "411                                  regex search ([0-9]+\\.[0-9]*.)|([0-9]+)   \n",
       "\n",
       "                                              Template Parent  \\\n",
       "405                                 Metabolomics Human          \n",
       "510                                              bsSeq          \n",
       "141  Genotyping,Metabolomics Human,Microbiome,Prote...          \n",
       "621            RNAseq,Whole Genome Sequencing,scRNAseq          \n",
       "325  Microbiome,RNAseq,Whole Genome Sequencing,bsSe...          \n",
       "239            RNAseq,Whole Genome Sequencing,scRNAseq          \n",
       "207                                   Individual Human          \n",
       "323  Genotyping,Metabolomics Human,Microbiome,Prote...          \n",
       "550                      Metabolomics Human,Proteomics          \n",
       "411                                         Proteomics          \n",
       "\n",
       "                                                Source    Type  \\\n",
       "405                                   Sage Bionetworks  NUMBER   \n",
       "510                                   Sage Bionetworks  NUMBER   \n",
       "141                           ImmPort,Sage Bionetworks  STRING   \n",
       "621                                            ImmPort  STRING   \n",
       "325                                   Sage Bionetworks  STRING   \n",
       "239                                   Sage Bionetworks  STRING   \n",
       "207                                   Sage Bionetworks  STRING   \n",
       "323                           ImmPort,Sage Bionetworks  STRING   \n",
       "550                             DSLWG,Sage Bionetworks  STRING   \n",
       "411  MS Ontology,Sage Bionetworks,The Proteomics Sa...  NUMBER   \n",
       "\n",
       "                                              Ontology multivalue  \\\n",
       "405                                   Sage Bionetworks              \n",
       "510                                   Sage Bionetworks      False   \n",
       "141                           ImmPort,Sage Bionetworks      False   \n",
       "621                                            ImmPort      False   \n",
       "325                                   Sage Bionetworks      False   \n",
       "239                                   Sage Bionetworks      False   \n",
       "207                  Data Contributor,Sage Bionetworks      False   \n",
       "323                           ImmPort,Sage Bionetworks      False   \n",
       "550                             DSLWG,Sage Bionetworks      False   \n",
       "411  MS Ontology,Sage Bionetworks,The Proteomics Sa...      False   \n",
       "\n",
       "                       UsedIn  \n",
       "405                            \n",
       "510                            \n",
       "141                sampleType  \n",
       "621                            \n",
       "325  libraryPreparationMethod  \n",
       "239            repositoryName  \n",
       "207            consentGroupID  \n",
       "323                sampleType  \n",
       "550                            \n",
       "411                            "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm2.sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add ADKP attrs to dm\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\n",
    "    \"https://raw.githubusercontent.com/adknowledgeportal/data-models/main/AD.model.csv\"\n",
    ")\n",
    "\n",
    "# preprocess AD data model to remove duplicates\n",
    "df = df.sort_values(by=[\"Attribute\", \"Valid Values\"]).reset_index(drop=True)\n",
    "df = df.drop_duplicates(keep=\"first\", subset=[\"Attribute\"])\n",
    "\n",
    "df.loc[\n",
    "    df.query('Attribute.str.contains(\"template\")',\n",
    "             engine=\"python\").index.tolist(),\n",
    "    \"Properties\",\n",
    "] = \"template\"\n",
    "\n",
    "attrs_interest = [\n",
    "    \"analysisType\",\n",
    "    \"analysisType\",\n",
    "    \"analytical covariates\",\n",
    "    \"assay\",\n",
    "    \"assay\",\n",
    "    \"biospecimen\",\n",
    "    \"consortium\",\n",
    "    \"data dictionary\",\n",
    "    \"dataSubtype\",\n",
    "    \"dataType\",\n",
    "    \"fileFormat\",\n",
    "    \"grant\",\n",
    "    \"ID mapping\",\n",
    "    \"individual\",\n",
    "    \"isConsortiumAnalysis\",\n",
    "    \"isModelSystem\",\n",
    "    \"isMultiSpecimen\",\n",
    "    \"libraryPrep\",\n",
    "    \"libraryType\",\n",
    "    \"manifest\",\n",
    "    \"manifest\",\n",
    "    \"metadata\",\n",
    "    \"metadataType\",\n",
    "    \"modelSystemName\",\n",
    "    \"modelSystemType\",\n",
    "    \"platform\",\n",
    "    \"project\",\n",
    "    \"protocol\",\n",
    "    \"protocol\",\n",
    "    \"resourceType\",\n",
    "    \"type\",\n",
    "]\n",
    "\n",
    "metadataTypes = [\n",
    "    \"analytical covariates\",\n",
    "    \"assay\",\n",
    "    \"biospecimen\",\n",
    "    \"data dictionary\",\n",
    "    \"ID mapping\",\n",
    "    \"individual\",\n",
    "    \"manifest\",\n",
    "    \"protocol\",\n",
    "]\n",
    "\n",
    "df_new_attrs = df.query(\"Attribute in @attrs_interest\").copy()\n",
    "\n",
    "df_new_attrs[\"Properties\"] = \"BaseAnnotation\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cleanup data model attributes to fit ELITE data model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_new_attrs = rewrite_df_value(\n",
    "    df_new_attrs, \"Attribute\", \"study\", \"Valid Values\", \"LLFS,ILO,LG,LC\"\n",
    ")\n",
    "\n",
    "df_new_attrs = rewrite_df_value(\n",
    "    df_new_attrs, \"Attribute\", \"consortium\", \"Valid Values\", \"ELITE\"\n",
    ")\n",
    "\n",
    "df_new_attrs = rewrite_df_value(\n",
    "    df_new_attrs,\n",
    "    \"Attribute\",\n",
    "    \"metadataType\",\n",
    "    \"Valid Values\",\n",
    "    \"analytical covariates, assay, biospecimen, data dictionary, ID mapping, individual, manifest, protocol\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Required</th>\n",
       "      <th>Properties</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Source</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>columnType</th>\n",
       "      <th>module</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>ID mapping</td>\n",
       "      <td>A file or table that maps data identifiers to ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MetadataType</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://edamontology.org/operation_3282</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>730</th>\n",
       "      <td>analysisType</td>\n",
       "      <td>Type of analysis</td>\n",
       "      <td>ANOVA, assessment, batch effect correction, ch...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>False</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sage.annotations-analysis.analysisType-0.0.14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>analysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731</th>\n",
       "      <td>analytical covariates</td>\n",
       "      <td>A file that contains a combination or subset o...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MetadataType</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>745</th>\n",
       "      <td>assay</td>\n",
       "      <td>The technology used to generate the data in th...</td>\n",
       "      <td>10x multiome, 16SrRNAseq, active avoidance lea...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>True</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sage.annotations-experimentalData.assay-0.0.26</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>794</th>\n",
       "      <td>biospecimen</td>\n",
       "      <td>Metadata describing properties of specimens co...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MetadataType</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Attribute                                        Description  \\\n",
       "295             ID mapping  A file or table that maps data identifiers to ...   \n",
       "730           analysisType                                   Type of analysis   \n",
       "731  analytical covariates  A file that contains a combination or subset o...   \n",
       "745                  assay  The technology used to generate the data in th...   \n",
       "794            biospecimen  Metadata describing properties of specimens co...   \n",
       "\n",
       "                                          Valid Values  DependsOn  \\\n",
       "295                                                NaN        NaN   \n",
       "730  ANOVA, assessment, batch effect correction, ch...        NaN   \n",
       "731                                                NaN        NaN   \n",
       "745  10x multiome, 16SrRNAseq, active avoidance lea...        NaN   \n",
       "794                                                NaN        NaN   \n",
       "\n",
       "             Parent Required    Properties  DependsOn Component  \\\n",
       "295  BaseAnnotation      NaN  MetadataType                  NaN   \n",
       "730  BaseAnnotation    False  DataProperty                  NaN   \n",
       "731  BaseAnnotation      NaN  MetadataType                  NaN   \n",
       "745  BaseAnnotation     True  DataProperty                  NaN   \n",
       "794  BaseAnnotation      NaN  MetadataType                  NaN   \n",
       "\n",
       "                                             Source  Validation Rules  \\\n",
       "295          http://edamontology.org/operation_3282               NaN   \n",
       "730   sage.annotations-analysis.analysisType-0.0.14               NaN   \n",
       "731                                Sage Bionetworks               NaN   \n",
       "745  sage.annotations-experimentalData.assay-0.0.26               NaN   \n",
       "794                                Sage Bionetworks               NaN   \n",
       "\n",
       "    columnType            module  \n",
       "295     string  experimentalData  \n",
       "730     string          analysis  \n",
       "731     string  experimentalData  \n",
       "745     string  experimentalData  \n",
       "794     string  experimentalData  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# recode Parent\n",
    "recoder = {\n",
    "    \"metadataType\": \"MetadataType\",\n",
    "    \"dataProperty\": \"DataProperty\",\n",
    "    \"dataType\": \"DataType\",\n",
    "    \"dataSubtype\": \"DataSubtype\",\n",
    "}\n",
    "\n",
    "df_new_attrs = df_new_attrs.replace(recoder)\n",
    "\n",
    "df_new_attrs = df_new_attrs.rename(\n",
    "    columns={\"Parent\": \"Properties\", \"Properties\": \"Parent\"}\n",
    ")\n",
    "df_new_attrs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merge new attributes with existiing data model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>UsedIn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>assay</td>\n",
       "      <td>The analysis or technology used to generate th...</td>\n",
       "      <td>TBD</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Biospecimen nonHuman</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>STRING</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>libraryPrep</td>\n",
       "      <td>The general strategy by which the library was ...</td>\n",
       "      <td>Chromium Single Cell 3',DNALibraryConstruction...</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Microbiome,Whole Genome Sequencing,bsSeq,scRNAseq</td>\n",
       "      <td></td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>STRING</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Attribute                                        Description  \\\n",
       "364        assay  The analysis or technology used to generate th...   \n",
       "436  libraryPrep  The general strategy by which the library was ...   \n",
       "\n",
       "                                          Valid Values Required DependsOn  \\\n",
       "364                                                TBD                      \n",
       "436  Chromium Single Cell 3',DNALibraryConstruction...     True             \n",
       "\n",
       "    DependsOn Component Properties Validation Rules  \\\n",
       "364                                                   \n",
       "436                                                   \n",
       "\n",
       "                                              Template Parent  \\\n",
       "364                               Biospecimen nonHuman          \n",
       "436  Microbiome,Whole Genome Sequencing,bsSeq,scRNAseq          \n",
       "\n",
       "               Source    Type          Ontology multivalue UsedIn  \n",
       "364  Sage Bionetworks  STRING  Sage Bionetworks                    \n",
       "436  Sage Bionetworks  STRING  Sage Bionetworks      False         "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm2.query(\"Attribute in @attrs_interest\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Only needed for the first time since the Valid values were TBD\n",
    "dm2 = rewrite_df_value(dm2, \"Attribute\", \"assay\", \"Valid Values\", np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm2 = dm2.replace(\"\", np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add valid values from the AD model\n",
    "dm2[\"Valid Values\"] = dm2[\"Valid Values\"].fillna(\n",
    "    dm2[\"Attribute\"].map(df_new_attrs.set_index(\"Attribute\")[\"Valid Values\"])\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add new attributes from AD model\n",
    "dm_new = pd.concat([dm2, df_new_attrs])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>UsedIn</th>\n",
       "      <th>columnType</th>\n",
       "      <th>module</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10x</td>\n",
       "      <td>Method by which library was prepared</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Microbiome,RNAseq,Whole Genome Sequencing,bsSe...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>STRING</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td>libraryPreparationMethod</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10x Visium Spatial Gene Expression</td>\n",
       "      <td>The specific version (application, manufacture...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human,Microbiome,Proteomics,RNAse...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DSLWG,Sage Bionetworks,http://purl.obolibrary....</td>\n",
       "      <td>STRING</td>\n",
       "      <td>DSLWG,Sage Bionetworks,http://purl.obolibrary....</td>\n",
       "      <td>False</td>\n",
       "      <td>platformLocation,technologyPlatformVersion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AFU</td>\n",
       "      <td>Measurement unit</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.)|([0-9]+)</td>\n",
       "      <td>Genotyping,Metabolomics Human,Microbiome,RNAse...</td>\n",
       "      <td>MeasurementUnit</td>\n",
       "      <td>ImmPort,Proposed minimum metadata relative to ...</td>\n",
       "      <td>STRING</td>\n",
       "      <td>ImmPort,Proposed minimum metadata relative to ...</td>\n",
       "      <td>False</td>\n",
       "      <td>acquisitionBatchSizeUnit,batchSizeUnit,convers...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AI</td>\n",
       "      <td>Measurement unit</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>regex search ([0-9]+\\.[0-9]*.)|([0-9]+)</td>\n",
       "      <td>Genotyping,Metabolomics Human,Microbiome,RNAse...</td>\n",
       "      <td>MeasurementUnit</td>\n",
       "      <td>ImmPort,Proposed minimum metadata relative to ...</td>\n",
       "      <td>STRING</td>\n",
       "      <td>ImmPort,Proposed minimum metadata relative to ...</td>\n",
       "      <td>False</td>\n",
       "      <td>acquisitionBatchSizeUnit,batchSizeUnit,convers...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AIBL pool</td>\n",
       "      <td>Control samples suitable for normalization and...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ValidValue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://doi.org/10.1101/2020.05.19.105197https...</td>\n",
       "      <td>STRING</td>\n",
       "      <td>https://doi.org/10.1101/2020.05.19.105197https...</td>\n",
       "      <td>False</td>\n",
       "      <td>controlType</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1123</th>\n",
       "      <td>modelSystemName</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3xTg-AD, 5XFAD, AB42, Abca7A1527GAPOE4Trem2R47...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-neuro.modelSystemName-0.0.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>neuro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1124</th>\n",
       "      <td>modelSystemType</td>\n",
       "      <td>Type of model system.</td>\n",
       "      <td>animal, cerebral organoid, immortalized cell l...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-experimentalData.modelSystemT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207</th>\n",
       "      <td>platform</td>\n",
       "      <td>The specific version (manufacturer, model, etc...</td>\n",
       "      <td>Affy5.0, Affy6.0, Affymetrix Human Gene 1.0 ST...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-experimentalData.platform-0.0.26</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1233</th>\n",
       "      <td>protocol</td>\n",
       "      <td>A plan specification which has sufficient leve...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MetadataType</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OBI_0000272</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258</th>\n",
       "      <td>resourceType</td>\n",
       "      <td>The type of resource being stored and annotated</td>\n",
       "      <td>analysis, curatedData, experimentalData, metad...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-sageCommunity.resourceType-0.0.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>sageCommunity</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>684 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               Attribute  \\\n",
       "0                                    10x   \n",
       "1     10x Visium Spatial Gene Expression   \n",
       "2                                    AFU   \n",
       "3                                     AI   \n",
       "4                              AIBL pool   \n",
       "...                                  ...   \n",
       "1123                     modelSystemName   \n",
       "1124                     modelSystemType   \n",
       "1207                            platform   \n",
       "1233                            protocol   \n",
       "1258                        resourceType   \n",
       "\n",
       "                                            Description  \\\n",
       "0                  Method by which library was prepared   \n",
       "1     The specific version (application, manufacture...   \n",
       "2                                      Measurement unit   \n",
       "3                                      Measurement unit   \n",
       "4     Control samples suitable for normalization and...   \n",
       "...                                                 ...   \n",
       "1123                                                NaN   \n",
       "1124                              Type of model system.   \n",
       "1207  The specific version (manufacturer, model, etc...   \n",
       "1233  A plan specification which has sufficient leve...   \n",
       "1258    The type of resource being stored and annotated   \n",
       "\n",
       "                                           Valid Values Required DependsOn  \\\n",
       "0                                                   NaN    False       NaN   \n",
       "1                                                   NaN    False       NaN   \n",
       "2                                                   NaN    False       NaN   \n",
       "3                                                   NaN    False       NaN   \n",
       "4                                                   NaN    False       NaN   \n",
       "...                                                 ...      ...       ...   \n",
       "1123  3xTg-AD, 5XFAD, AB42, Abca7A1527GAPOE4Trem2R47...    False       NaN   \n",
       "1124  animal, cerebral organoid, immortalized cell l...    False       NaN   \n",
       "1207  Affy5.0, Affy6.0, Affymetrix Human Gene 1.0 ST...     True       NaN   \n",
       "1233                                                NaN      NaN       NaN   \n",
       "1258  analysis, curatedData, experimentalData, metad...     True       NaN   \n",
       "\n",
       "      DependsOn Component    Properties  \\\n",
       "0                     NaN    ValidValue   \n",
       "1                     NaN    ValidValue   \n",
       "2                     NaN    ValidValue   \n",
       "3                     NaN    ValidValue   \n",
       "4                     NaN    ValidValue   \n",
       "...                   ...           ...   \n",
       "1123                  NaN  DataProperty   \n",
       "1124                  NaN  DataProperty   \n",
       "1207                  NaN  DataProperty   \n",
       "1233                  NaN  MetadataType   \n",
       "1258                  NaN  DataProperty   \n",
       "\n",
       "                             Validation Rules  \\\n",
       "0                                         NaN   \n",
       "1                                         NaN   \n",
       "2     regex search ([0-9]+\\.[0-9]*.)|([0-9]+)   \n",
       "3     regex search ([0-9]+\\.[0-9]*.)|([0-9]+)   \n",
       "4                                         NaN   \n",
       "...                                       ...   \n",
       "1123                                      NaN   \n",
       "1124                                      NaN   \n",
       "1207                                      NaN   \n",
       "1233                                      NaN   \n",
       "1258                                      NaN   \n",
       "\n",
       "                                               Template           Parent  \\\n",
       "0     Microbiome,RNAseq,Whole Genome Sequencing,bsSe...              NaN   \n",
       "1     Metabolomics Human,Microbiome,Proteomics,RNAse...              NaN   \n",
       "2     Genotyping,Metabolomics Human,Microbiome,RNAse...  MeasurementUnit   \n",
       "3     Genotyping,Metabolomics Human,Microbiome,RNAse...  MeasurementUnit   \n",
       "4                                    Metabolomics Human              NaN   \n",
       "...                                                 ...              ...   \n",
       "1123                                                NaN   BaseAnnotation   \n",
       "1124                                                NaN   BaseAnnotation   \n",
       "1207                                                NaN   BaseAnnotation   \n",
       "1233                                                NaN   BaseAnnotation   \n",
       "1258                                                NaN   BaseAnnotation   \n",
       "\n",
       "                                                 Source    Type  \\\n",
       "0                                      Sage Bionetworks  STRING   \n",
       "1     DSLWG,Sage Bionetworks,http://purl.obolibrary....  STRING   \n",
       "2     ImmPort,Proposed minimum metadata relative to ...  STRING   \n",
       "3     ImmPort,Proposed minimum metadata relative to ...  STRING   \n",
       "4     https://doi.org/10.1101/2020.05.19.105197https...  STRING   \n",
       "...                                                 ...     ...   \n",
       "1123       sage.annotations-neuro.modelSystemName-0.0.8     NaN   \n",
       "1124  sage.annotations-experimentalData.modelSystemT...     NaN   \n",
       "1207  sage.annotations-experimentalData.platform-0.0.26     NaN   \n",
       "1233         http://purl.obolibrary.org/obo/OBI_0000272     NaN   \n",
       "1258  sage.annotations-sageCommunity.resourceType-0.0.5     NaN   \n",
       "\n",
       "                                               Ontology multivalue  \\\n",
       "0                                      Sage Bionetworks      False   \n",
       "1     DSLWG,Sage Bionetworks,http://purl.obolibrary....      False   \n",
       "2     ImmPort,Proposed minimum metadata relative to ...      False   \n",
       "3     ImmPort,Proposed minimum metadata relative to ...      False   \n",
       "4     https://doi.org/10.1101/2020.05.19.105197https...      False   \n",
       "...                                                 ...        ...   \n",
       "1123                                                NaN        NaN   \n",
       "1124                                                NaN        NaN   \n",
       "1207                                                NaN        NaN   \n",
       "1233                                                NaN        NaN   \n",
       "1258                                                NaN        NaN   \n",
       "\n",
       "                                                 UsedIn columnType  \\\n",
       "0                              libraryPreparationMethod        NaN   \n",
       "1            platformLocation,technologyPlatformVersion        NaN   \n",
       "2     acquisitionBatchSizeUnit,batchSizeUnit,convers...        NaN   \n",
       "3     acquisitionBatchSizeUnit,batchSizeUnit,convers...        NaN   \n",
       "4                                           controlType        NaN   \n",
       "...                                                 ...        ...   \n",
       "1123                                                NaN     string   \n",
       "1124                                                NaN     string   \n",
       "1207                                                NaN     string   \n",
       "1233                                                NaN     string   \n",
       "1258                                                NaN     string   \n",
       "\n",
       "                module  \n",
       "0                  NaN  \n",
       "1                  NaN  \n",
       "2                  NaN  \n",
       "3                  NaN  \n",
       "4                  NaN  \n",
       "...                ...  \n",
       "1123             neuro  \n",
       "1124  experimentalData  \n",
       "1207  experimentalData  \n",
       "1233  experimentalData  \n",
       "1258     sageCommunity  \n",
       "\n",
       "[684 rows x 17 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>UsedIn</th>\n",
       "      <th>columnType</th>\n",
       "      <th>module</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>730</th>\n",
       "      <td>analysisType</td>\n",
       "      <td>Type of analysis</td>\n",
       "      <td>ANOVA, assessment, batch effect correction, ch...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-analysis.analysisType-0.0.14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>analysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>745</th>\n",
       "      <td>assay</td>\n",
       "      <td>The technology used to generate the data in th...</td>\n",
       "      <td>10x multiome, 16SrRNAseq, active avoidance lea...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-experimentalData.assay-0.0.26</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>848</th>\n",
       "      <td>consortium</td>\n",
       "      <td>The name of the consortium</td>\n",
       "      <td>ELITE</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-sageCommunity.consortium-0.0.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>sageCommunity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>870</th>\n",
       "      <td>DataSubtype</td>\n",
       "      <td>Further qualification of dataType, which may b...</td>\n",
       "      <td>dataMatrix, metadata, normalized, processed, r...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-experimentalData.dataSubtype-...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>871</th>\n",
       "      <td>DataType</td>\n",
       "      <td>Types of input/output data in bioinformatics p...</td>\n",
       "      <td>Volume, Weight, Pharmacokinetic Study, genomic...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-experimentalData.dataType-0.0.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>936</th>\n",
       "      <td>fileFormat</td>\n",
       "      <td>Defined format of the data file, typically cor...</td>\n",
       "      <td>7z, ab1, abf, avi, bai, bam, bash script, bcf,...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-sageCommunity.fileFormat-0.0.12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>sageCommunity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972</th>\n",
       "      <td>grant</td>\n",
       "      <td>Grant number including activity code, institut...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-neuro.grant-0.0.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>neuro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1028</th>\n",
       "      <td>isModelSystem</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRUE, FALSE</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-neuro.isModelSystem-0.0.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>boolean</td>\n",
       "      <td>neuro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1029</th>\n",
       "      <td>isMultiSpecimen</td>\n",
       "      <td>Boolean flag indicating whether or not a file ...</td>\n",
       "      <td>TRUE, FALSE</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-experimentalData.isMultiSpeci...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>boolean</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1049</th>\n",
       "      <td>libraryPrep</td>\n",
       "      <td>The general strategy by which the library was ...</td>\n",
       "      <td>amplicon, cellHashing, Chromium Single Cell 3'...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-ngs.libraryPrep-0.0.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>ngs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1051</th>\n",
       "      <td>libraryType</td>\n",
       "      <td>The type of library, in assays where samples a...</td>\n",
       "      <td>ADT, cDNA, CMO, HTO, LMO, plasmid DNA</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-ngs.libraryType-0.0.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>ngs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105</th>\n",
       "      <td>MetadataType</td>\n",
       "      <td>For files of dataSubtype: metadata, a descript...</td>\n",
       "      <td>analytical covariates, assay, biospecimen, dat...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-experimentalData.metadataType...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1123</th>\n",
       "      <td>modelSystemName</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3xTg-AD, 5XFAD, AB42, Abca7A1527GAPOE4Trem2R47...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-neuro.modelSystemName-0.0.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>neuro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1124</th>\n",
       "      <td>modelSystemType</td>\n",
       "      <td>Type of model system.</td>\n",
       "      <td>animal, cerebral organoid, immortalized cell l...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-experimentalData.modelSystemT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207</th>\n",
       "      <td>platform</td>\n",
       "      <td>The specific version (manufacturer, model, etc...</td>\n",
       "      <td>Affy5.0, Affy6.0, Affymetrix Human Gene 1.0 ST...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-experimentalData.platform-0.0.26</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258</th>\n",
       "      <td>resourceType</td>\n",
       "      <td>The type of resource being stored and annotated</td>\n",
       "      <td>analysis, curatedData, experimentalData, metad...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-sageCommunity.resourceType-0.0.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>sageCommunity</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Attribute                                        Description  \\\n",
       "730      analysisType                                   Type of analysis   \n",
       "745             assay  The technology used to generate the data in th...   \n",
       "848        consortium                         The name of the consortium   \n",
       "870       DataSubtype  Further qualification of dataType, which may b...   \n",
       "871          DataType  Types of input/output data in bioinformatics p...   \n",
       "936        fileFormat  Defined format of the data file, typically cor...   \n",
       "972             grant  Grant number including activity code, institut...   \n",
       "1028    isModelSystem                                                NaN   \n",
       "1029  isMultiSpecimen  Boolean flag indicating whether or not a file ...   \n",
       "1049      libraryPrep  The general strategy by which the library was ...   \n",
       "1051      libraryType  The type of library, in assays where samples a...   \n",
       "1105     MetadataType  For files of dataSubtype: metadata, a descript...   \n",
       "1123  modelSystemName                                                NaN   \n",
       "1124  modelSystemType                              Type of model system.   \n",
       "1207         platform  The specific version (manufacturer, model, etc...   \n",
       "1258     resourceType    The type of resource being stored and annotated   \n",
       "\n",
       "                                           Valid Values Required DependsOn  \\\n",
       "730   ANOVA, assessment, batch effect correction, ch...    False       NaN   \n",
       "745   10x multiome, 16SrRNAseq, active avoidance lea...     True       NaN   \n",
       "848                                               ELITE     True       NaN   \n",
       "870   dataMatrix, metadata, normalized, processed, r...    False       NaN   \n",
       "871   Volume, Weight, Pharmacokinetic Study, genomic...    False       NaN   \n",
       "936   7z, ab1, abf, avi, bai, bam, bash script, bcf,...    False       NaN   \n",
       "972                                                 NaN    False       NaN   \n",
       "1028                                        TRUE, FALSE    False       NaN   \n",
       "1029                                        TRUE, FALSE    False       NaN   \n",
       "1049  amplicon, cellHashing, Chromium Single Cell 3'...     True       NaN   \n",
       "1051              ADT, cDNA, CMO, HTO, LMO, plasmid DNA    False       NaN   \n",
       "1105  analytical covariates, assay, biospecimen, dat...    False       NaN   \n",
       "1123  3xTg-AD, 5XFAD, AB42, Abca7A1527GAPOE4Trem2R47...    False       NaN   \n",
       "1124  animal, cerebral organoid, immortalized cell l...    False       NaN   \n",
       "1207  Affy5.0, Affy6.0, Affymetrix Human Gene 1.0 ST...     True       NaN   \n",
       "1258  analysis, curatedData, experimentalData, metad...     True       NaN   \n",
       "\n",
       "      DependsOn Component    Properties Validation Rules Template  \\\n",
       "730                   NaN  DataProperty              NaN      NaN   \n",
       "745                   NaN  DataProperty              NaN      NaN   \n",
       "848                   NaN  DataProperty              NaN      NaN   \n",
       "870                   NaN  DataProperty              NaN      NaN   \n",
       "871                   NaN  DataProperty              NaN      NaN   \n",
       "936                   NaN  DataProperty              NaN      NaN   \n",
       "972                   NaN  DataProperty              NaN      NaN   \n",
       "1028                  NaN  DataProperty              NaN      NaN   \n",
       "1029                  NaN  DataProperty              NaN      NaN   \n",
       "1049                  NaN  DataProperty              NaN      NaN   \n",
       "1051                  NaN  DataProperty              NaN      NaN   \n",
       "1105                  NaN  DataProperty              NaN      NaN   \n",
       "1123                  NaN  DataProperty              NaN      NaN   \n",
       "1124                  NaN  DataProperty              NaN      NaN   \n",
       "1207                  NaN  DataProperty              NaN      NaN   \n",
       "1258                  NaN  DataProperty              NaN      NaN   \n",
       "\n",
       "              Parent                                             Source Type  \\\n",
       "730   BaseAnnotation      sage.annotations-analysis.analysisType-0.0.14  NaN   \n",
       "745   BaseAnnotation     sage.annotations-experimentalData.assay-0.0.26  NaN   \n",
       "848   BaseAnnotation    sage.annotations-sageCommunity.consortium-0.0.5  NaN   \n",
       "870   BaseAnnotation  sage.annotations-experimentalData.dataSubtype-...  NaN   \n",
       "871   BaseAnnotation   sage.annotations-experimentalData.dataType-0.0.3  NaN   \n",
       "936   BaseAnnotation   sage.annotations-sageCommunity.fileFormat-0.0.12  NaN   \n",
       "972   BaseAnnotation                 sage.annotations-neuro.grant-0.0.2  NaN   \n",
       "1028  BaseAnnotation         sage.annotations-neuro.isModelSystem-0.0.2  NaN   \n",
       "1029  BaseAnnotation  sage.annotations-experimentalData.isMultiSpeci...  NaN   \n",
       "1049  BaseAnnotation            sage.annotations-ngs.libraryPrep-0.0.13  NaN   \n",
       "1051  BaseAnnotation             sage.annotations-ngs.libraryType-0.0.3  NaN   \n",
       "1105  BaseAnnotation  sage.annotations-experimentalData.metadataType...  NaN   \n",
       "1123  BaseAnnotation       sage.annotations-neuro.modelSystemName-0.0.8  NaN   \n",
       "1124  BaseAnnotation  sage.annotations-experimentalData.modelSystemT...  NaN   \n",
       "1207  BaseAnnotation  sage.annotations-experimentalData.platform-0.0.26  NaN   \n",
       "1258  BaseAnnotation  sage.annotations-sageCommunity.resourceType-0.0.5  NaN   \n",
       "\n",
       "     Ontology multivalue UsedIn columnType            module  \n",
       "730       NaN        NaN    NaN     string          analysis  \n",
       "745       NaN        NaN    NaN     string  experimentalData  \n",
       "848       NaN        NaN    NaN     string     sageCommunity  \n",
       "870       NaN        NaN    NaN     string  experimentalData  \n",
       "871       NaN        NaN    NaN     string  experimentalData  \n",
       "936       NaN        NaN    NaN     string     sageCommunity  \n",
       "972       NaN        NaN    NaN     string             neuro  \n",
       "1028      NaN        NaN    NaN    boolean             neuro  \n",
       "1029      NaN        NaN    NaN    boolean  experimentalData  \n",
       "1049      NaN        NaN    NaN     string               ngs  \n",
       "1051      NaN        NaN    NaN     string               ngs  \n",
       "1105      NaN        NaN    NaN     string  experimentalData  \n",
       "1123      NaN        NaN    NaN     string             neuro  \n",
       "1124      NaN        NaN    NaN     string  experimentalData  \n",
       "1207      NaN        NaN    NaN     string  experimentalData  \n",
       "1258      NaN        NaN    NaN     string     sageCommunity  "
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm_new.query('Properties == \"DataProperty\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>UsedIn</th>\n",
       "      <th>columnType</th>\n",
       "      <th>module</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>611</th>\n",
       "      <td>studyCode</td>\n",
       "      <td>Unique identifier for the study, assigned by t...</td>\n",
       "      <td>ILO,LC,LG,LLFS</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Individual Human,Individual nonHuman</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>STRING</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Attribute                                        Description  \\\n",
       "611  studyCode  Unique identifier for the study, assigned by t...   \n",
       "\n",
       "       Valid Values Required DependsOn  DependsOn Component Properties  \\\n",
       "611  ILO,LC,LG,LLFS     True       NaN                  NaN        NaN   \n",
       "\n",
       "    Validation Rules                              Template Parent  \\\n",
       "611              NaN  Individual Human,Individual nonHuman    NaN   \n",
       "\n",
       "               Source    Type          Ontology multivalue UsedIn columnType  \\\n",
       "611  Sage Bionetworks  STRING  Sage Bionetworks       True    NaN        NaN   \n",
       "\n",
       "    module  \n",
       "611    NaN  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm_new.query('Attribute == \"studyCode\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>UsedIn</th>\n",
       "      <th>columnType</th>\n",
       "      <th>module</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>assay</td>\n",
       "      <td>The analysis or technology used to generate th...</td>\n",
       "      <td>10x multiome, 16SrRNAseq, active avoidance lea...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Biospecimen nonHuman</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>STRING</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>libraryPrep</td>\n",
       "      <td>The general strategy by which the library was ...</td>\n",
       "      <td>Chromium Single Cell 3',DNALibraryConstruction...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Microbiome,Whole Genome Sequencing,bsSeq,scRNAseq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>STRING</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>745</th>\n",
       "      <td>assay</td>\n",
       "      <td>The technology used to generate the data in th...</td>\n",
       "      <td>10x multiome, 16SrRNAseq, active avoidance lea...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-experimentalData.assay-0.0.26</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1049</th>\n",
       "      <td>libraryPrep</td>\n",
       "      <td>The general strategy by which the library was ...</td>\n",
       "      <td>amplicon, cellHashing, Chromium Single Cell 3'...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataProperty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BaseAnnotation</td>\n",
       "      <td>sage.annotations-ngs.libraryPrep-0.0.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>string</td>\n",
       "      <td>ngs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Attribute                                        Description  \\\n",
       "364         assay  The analysis or technology used to generate th...   \n",
       "436   libraryPrep  The general strategy by which the library was ...   \n",
       "745         assay  The technology used to generate the data in th...   \n",
       "1049  libraryPrep  The general strategy by which the library was ...   \n",
       "\n",
       "                                           Valid Values Required DependsOn  \\\n",
       "364   10x multiome, 16SrRNAseq, active avoidance lea...      NaN       NaN   \n",
       "436   Chromium Single Cell 3',DNALibraryConstruction...     True       NaN   \n",
       "745   10x multiome, 16SrRNAseq, active avoidance lea...     True       NaN   \n",
       "1049  amplicon, cellHashing, Chromium Single Cell 3'...     True       NaN   \n",
       "\n",
       "      DependsOn Component    Properties Validation Rules  \\\n",
       "364                   NaN           NaN              NaN   \n",
       "436                   NaN           NaN              NaN   \n",
       "745                   NaN  DataProperty              NaN   \n",
       "1049                  NaN  DataProperty              NaN   \n",
       "\n",
       "                                               Template          Parent  \\\n",
       "364                                Biospecimen nonHuman             NaN   \n",
       "436   Microbiome,Whole Genome Sequencing,bsSeq,scRNAseq             NaN   \n",
       "745                                                 NaN  BaseAnnotation   \n",
       "1049                                                NaN  BaseAnnotation   \n",
       "\n",
       "                                              Source    Type  \\\n",
       "364                                 Sage Bionetworks  STRING   \n",
       "436                                 Sage Bionetworks  STRING   \n",
       "745   sage.annotations-experimentalData.assay-0.0.26     NaN   \n",
       "1049         sage.annotations-ngs.libraryPrep-0.0.13     NaN   \n",
       "\n",
       "              Ontology multivalue UsedIn columnType            module  \n",
       "364   Sage Bionetworks        NaN    NaN        NaN               NaN  \n",
       "436   Sage Bionetworks      False    NaN        NaN               NaN  \n",
       "745                NaN        NaN    NaN     string  experimentalData  \n",
       "1049               NaN        NaN    NaN     string               ngs  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(sum(dm_new.duplicated(subset=\"Attribute\")))\n",
    "dm_new[dm_new.duplicated(subset=\"Attribute\", keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm_new[\"Validation Rules\"] = dm_new[\"Validation Rules\"].replace(\n",
    "    \"regex search ([0-9]+),regex search ([0-9]+\\\\.[0-9]*.)|([0-9]+)\",\n",
    "    \"regex search ([0-9]+\\\\.[0-9]*.)|([0-9]+)\",\n",
    ")\n",
    "dm_new[\n",
    "    dm_new[\"Validation Rules\"]\n",
    "    == \"regex search ([0-9]+),regex search ([0-9]+\\\\.[0-9]*.)|([0-9]+)\"\n",
    "]\n",
    "\n",
    "# recode Parent\n",
    "recoder = {\n",
    "    \"metadataType\": \"MetadataType\",\n",
    "    \"dataProperty\": \"DataProperty\",\n",
    "    \"dataType\": \"DataType\",\n",
    "    \"dataSubtype\": \"DataSubtype\",\n",
    "}\n",
    "\n",
    "dm_new = dm_new.replace(recoder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm_new.drop(columns=\"columnType\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(684, 16)\n",
      "(682, 16)\n"
     ]
    }
   ],
   "source": [
    "print(dm_new.shape)\n",
    "\n",
    "dm_new = dm_new.drop_duplicates(subset=[\"Attribute\"], keep=\"first\")\n",
    "\n",
    "print(dm_new.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm_new[[\"Parent\", \"Properties\", \"Type\"]] = dm_new[\n",
    "    [\"Parent\", \"Properties\", \"Type\"]\n",
    "].fillna(\"unspecified\")\n",
    "\n",
    "dm_new[\"Required\"] = dm_new[\"Required\"].fillna(\"False\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clean up template depends on\n",
    "dm_new[\"DependsOn\"] = (\n",
    "    dm_new[\"DependsOn\"].fillna(\"\").apply(clean_list).replace(\"\", np.nan)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "remove_values = [\"False\", \"Not applicable\", \"Not collected\", \"Unknown\", \"Other\"]\n",
    "\n",
    "dm_new.loc[dm_new[\"Parent\"] == \"Template\", \"DependsOn\"] = dm_new.loc[\n",
    "    dm_new[\"Parent\"] == \"Template\", \"DependsOn\"\n",
    "].apply(lambda x: \",\".join([y for y in x.split(\",\") if y not in remove_values]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>Required</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Template</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>UsedIn</th>\n",
       "      <th>module</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>Biospecimen human</td>\n",
       "      <td>Metadata template for Biospecimen human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>Cerebrospinal Fluid (CSF),OtherCellType,OtherF...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataType</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Template</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>Biospecimen nonHuman</td>\n",
       "      <td>Metadata template for Biospecimen nonHuman</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>Cerebrospinal Fluid (CSF),OtherCellType,OtherF...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataType</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Template</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>650</th>\n",
       "      <td>Genotyping</td>\n",
       "      <td>Metadata template for Genotyping</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>AFU,AI,AU/ml,Amniotic Fluid,Appendix,B cell,Ba...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataType</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Template</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651</th>\n",
       "      <td>Individual Human</td>\n",
       "      <td>Metadata template for Individual Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>American Indian or Alaska Native,Asian,BU,Bird...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataType</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Template</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652</th>\n",
       "      <td>Individual nonHuman</td>\n",
       "      <td>Metadata template for Individual nonHuman</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>Adult,Bird,Cardiovascular Health Study (CHS),C...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataType</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Template</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>653</th>\n",
       "      <td>Metabolomics Human</td>\n",
       "      <td>Metadata template for Metabolomics Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>10x Visium Spatial Gene Expression,AFU,AI,AIBL...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataType</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Template</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>Microbiome</td>\n",
       "      <td>Metadata template for Microbiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>10x,10x Visium Spatial Gene Expression,AFU,AI,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataType</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Template</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>655</th>\n",
       "      <td>Proteomics</td>\n",
       "      <td>Metadata template for Proteomics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>10x Visium Spatial Gene Expression,Affymetrix ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataType</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Template</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>656</th>\n",
       "      <td>RNAseq</td>\n",
       "      <td>Metadata template for RNAseq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>10x,10x Visium Spatial Gene Expression,AFU,AI,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataType</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Template</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>657</th>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "      <td>Metadata template for Whole Genome Sequencing</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>10x,10x Visium Spatial Gene Expression,AFU,AI,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataType</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Template</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>658</th>\n",
       "      <td>bsSeq</td>\n",
       "      <td>Template for bisulfite-seq WGBS methylseq meth...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>10x,10x Visium Spatial Gene Expression,AFU,AI,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataType</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Template</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>659</th>\n",
       "      <td>scRNAseq</td>\n",
       "      <td>Metadata template for scRNAseq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>10x,10x Visium Spatial Gene Expression,AFU,AI,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DataType</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Template</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unspecified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Attribute  \\\n",
       "648        Biospecimen human   \n",
       "649     Biospecimen nonHuman   \n",
       "650               Genotyping   \n",
       "651         Individual Human   \n",
       "652      Individual nonHuman   \n",
       "653       Metabolomics Human   \n",
       "654               Microbiome   \n",
       "655               Proteomics   \n",
       "656                   RNAseq   \n",
       "657  Whole Genome Sequencing   \n",
       "658                    bsSeq   \n",
       "659                 scRNAseq   \n",
       "\n",
       "                                           Description Valid Values Required  \\\n",
       "648            Metadata template for Biospecimen human          NaN     True   \n",
       "649         Metadata template for Biospecimen nonHuman          NaN     True   \n",
       "650                   Metadata template for Genotyping          NaN     True   \n",
       "651             Metadata template for Individual Human          NaN     True   \n",
       "652          Metadata template for Individual nonHuman          NaN     True   \n",
       "653           Metadata template for Metabolomics Human          NaN     True   \n",
       "654                   Metadata template for Microbiome          NaN     True   \n",
       "655                   Metadata template for Proteomics          NaN     True   \n",
       "656                       Metadata template for RNAseq          NaN     True   \n",
       "657      Metadata template for Whole Genome Sequencing          NaN     True   \n",
       "658  Template for bisulfite-seq WGBS methylseq meth...          NaN     True   \n",
       "659                     Metadata template for scRNAseq          NaN     True   \n",
       "\n",
       "                                             DependsOn  DependsOn Component  \\\n",
       "648  Cerebrospinal Fluid (CSF),OtherCellType,OtherF...                  NaN   \n",
       "649  Cerebrospinal Fluid (CSF),OtherCellType,OtherF...                  NaN   \n",
       "650  AFU,AI,AU/ml,Amniotic Fluid,Appendix,B cell,Ba...                  NaN   \n",
       "651  American Indian or Alaska Native,Asian,BU,Bird...                  NaN   \n",
       "652  Adult,Bird,Cardiovascular Health Study (CHS),C...                  NaN   \n",
       "653  10x Visium Spatial Gene Expression,AFU,AI,AIBL...                  NaN   \n",
       "654  10x,10x Visium Spatial Gene Expression,AFU,AI,...                  NaN   \n",
       "655  10x Visium Spatial Gene Expression,Affymetrix ...                  NaN   \n",
       "656  10x,10x Visium Spatial Gene Expression,AFU,AI,...                  NaN   \n",
       "657  10x,10x Visium Spatial Gene Expression,AFU,AI,...                  NaN   \n",
       "658  10x,10x Visium Spatial Gene Expression,AFU,AI,...                  NaN   \n",
       "659  10x,10x Visium Spatial Gene Expression,AFU,AI,...                  NaN   \n",
       "\n",
       "    Properties Validation Rules Template    Parent Source         Type  \\\n",
       "648   DataType              NaN      NaN  Template    NaN  unspecified   \n",
       "649   DataType              NaN      NaN  Template    NaN  unspecified   \n",
       "650   DataType              NaN      NaN  Template    NaN  unspecified   \n",
       "651   DataType              NaN      NaN  Template    NaN  unspecified   \n",
       "652   DataType              NaN      NaN  Template    NaN  unspecified   \n",
       "653   DataType              NaN      NaN  Template    NaN  unspecified   \n",
       "654   DataType              NaN      NaN  Template    NaN  unspecified   \n",
       "655   DataType              NaN      NaN  Template    NaN  unspecified   \n",
       "656   DataType              NaN      NaN  Template    NaN  unspecified   \n",
       "657   DataType              NaN      NaN  Template    NaN  unspecified   \n",
       "658   DataType              NaN      NaN  Template    NaN  unspecified   \n",
       "659   DataType              NaN      NaN  Template    NaN  unspecified   \n",
       "\n",
       "    Ontology multivalue UsedIn module  \n",
       "648      NaN        NaN    NaN    NaN  \n",
       "649      NaN        NaN    NaN    NaN  \n",
       "650      NaN        NaN    NaN    NaN  \n",
       "651      NaN        NaN    NaN    NaN  \n",
       "652      NaN        NaN    NaN    NaN  \n",
       "653      NaN        NaN    NaN    NaN  \n",
       "654      NaN        NaN    NaN    NaN  \n",
       "655      NaN        NaN    NaN    NaN  \n",
       "656      NaN        NaN    NaN    NaN  \n",
       "657      NaN        NaN    NaN    NaN  \n",
       "658      NaN        NaN    NaN    NaN  \n",
       "659      NaN        NaN    NaN    NaN  "
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm_new.loc[dm_new[\"Parent\"] == \"Template\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "dm_new = dm_new.reset_index(drop=True)\n",
    "dm_new.to_csv(\"../\" + csv_model, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "schematic schema convert ../EL.data.model.csv --output_jsonld EL.data.model.jsonld\n",
      "Starting schematic...\n",
      "Done adding requirements and value ranges to attributes\n",
      "The Data Model was created and saved to '../EL.data.model.jsonld' location.\n"
     ]
    }
   ],
   "source": [
    "# Convert CSV to JSON LD\n",
    "print(f'schematic schema convert {\"../\" + csv_model} --output_jsonld {json_model}')\n",
    "\n",
    "!schematic schema convert {\"../\" + csv_model} --output_jsonld {\"../\" + json_model}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
