{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "UPGRADE AVAILABLE\n",
      "\n",
      "A more recent version of the Synapse Client (4.2.0) is available. Your version (3.2.0) can be upgraded by typing:\n",
      "    pip install --upgrade synapseclient\n",
      "\n",
      "Python Synapse Client version 4.2.0 release notes\n",
      "\n",
      "https://python-docs.synapse.org/news/\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome, Nicholas Lee!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\"\"\"Create partitions from the data model based on the Parent names\n",
    "\n",
    "Args:\n",
    "    data_model (object): dataframe that contains the data model\n",
    "    partition_path (str): directory to store the partitioned data model as CSVs\n",
    "\"\"\"\n",
    "\n",
    "import glob\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import datetime\n",
    "import synapseclient as sc\n",
    "\n",
    "syn = sc.login()\n",
    "\n",
    "dm_path = '../../EL.data.model.csv'\n",
    "output_path = \"../backups\"\n",
    "\n",
    "partition_path = \"../modules\"\n",
    "# data_model = utils.load_and_backup_dm(dm_name, output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>Valid Values</th>\n",
       "      <th>DependsOn</th>\n",
       "      <th>DependsOn Component</th>\n",
       "      <th>Required</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Validation Rules</th>\n",
       "      <th>Properties</th>\n",
       "      <th>Source</th>\n",
       "      <th>module</th>\n",
       "      <th>multivalue</th>\n",
       "      <th>columnType</th>\n",
       "      <th>Ontology</th>\n",
       "      <th>Notes</th>\n",
       "      <th>UsedIn</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Attribute</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>acquisitionBatchID</th>\n",
       "      <td>Acquisition batch identifier, provided by the ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>False</td>\n",
       "      <td>STRING</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acquisitionBatchSize</th>\n",
       "      <td>The number of samples</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>False</td>\n",
       "      <td>STRING</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acquisitionBatchSizeUnit</th>\n",
       "      <td>The unit of measurement for number of samples ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>False</td>\n",
       "      <td>STRING</td>\n",
       "      <td>Sage Bionetworks</td>\n",
       "      <td></td>\n",
       "      <td>Metabolomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acquisitionMode</th>\n",
       "      <td>The specific aspect of a mass spectrometer met...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>False</td>\n",
       "      <td>STRING</td>\n",
       "      <td>https://www.ebi.ac.uk/ols/ontologies/ms/terms?...</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>Metabolomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acquisitionSoftware</th>\n",
       "      <td>The name of the acquisition software used. Pro...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ManifestColumn</td>\n",
       "      <td>False</td>\n",
       "      <td>STRING</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MS_1001455</td>\n",
       "      <td>A data contributor should be able to write in ...</td>\n",
       "      <td>Metabolomics</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                Description  \\\n",
       "Attribute                                                                     \n",
       "acquisitionBatchID        Acquisition batch identifier, provided by the ...   \n",
       "acquisitionBatchSize                                  The number of samples   \n",
       "acquisitionBatchSizeUnit  The unit of measurement for number of samples ...   \n",
       "acquisitionMode           The specific aspect of a mass spectrometer met...   \n",
       "acquisitionSoftware       The name of the acquisition software used. Pro...   \n",
       "\n",
       "                         Valid Values DependsOn  DependsOn Component Required  \\\n",
       "Attribute                                                                       \n",
       "acquisitionBatchID                NaN       NaN                  NaN    False   \n",
       "acquisitionBatchSize              NaN       NaN                  NaN    False   \n",
       "acquisitionBatchSizeUnit          NaN       NaN                  NaN    False   \n",
       "acquisitionMode                   NaN       NaN                  NaN     True   \n",
       "acquisitionSoftware               NaN       NaN                  NaN     True   \n",
       "\n",
       "                                  Parent  Validation Rules      Properties  \\\n",
       "Attribute                                                                    \n",
       "acquisitionBatchID        ManifestColumn               NaN  ManifestColumn   \n",
       "acquisitionBatchSize      ManifestColumn               NaN  ManifestColumn   \n",
       "acquisitionBatchSizeUnit  ManifestColumn               NaN  ManifestColumn   \n",
       "acquisitionMode           ManifestColumn               NaN  ManifestColumn   \n",
       "acquisitionSoftware       ManifestColumn               NaN  ManifestColumn   \n",
       "\n",
       "                         Source          module multivalue columnType  \\\n",
       "Attribute                                                               \n",
       "acquisitionBatchID          NaN  ManifestColumn      False     STRING   \n",
       "acquisitionBatchSize        NaN  ManifestColumn      False     STRING   \n",
       "acquisitionBatchSizeUnit    NaN  ManifestColumn      False     STRING   \n",
       "acquisitionMode             NaN  ManifestColumn      False     STRING   \n",
       "acquisitionSoftware         NaN  ManifestColumn      False     STRING   \n",
       "\n",
       "                                                                   Ontology  \\\n",
       "Attribute                                                                     \n",
       "acquisitionBatchID                                         Sage Bionetworks   \n",
       "acquisitionBatchSize                                       Sage Bionetworks   \n",
       "acquisitionBatchSizeUnit                                   Sage Bionetworks   \n",
       "acquisitionMode           https://www.ebi.ac.uk/ols/ontologies/ms/terms?...   \n",
       "acquisitionSoftware               http://purl.obolibrary.org/obo/MS_1001455   \n",
       "\n",
       "                                                                      Notes  \\\n",
       "Attribute                                                                     \n",
       "acquisitionBatchID                                                            \n",
       "acquisitionBatchSize                                                          \n",
       "acquisitionBatchSizeUnit                                                      \n",
       "acquisitionMode           A data contributor should be able to write in ...   \n",
       "acquisitionSoftware       A data contributor should be able to write in ...   \n",
       "\n",
       "                                UsedIn  \n",
       "Attribute                               \n",
       "acquisitionBatchID        Metabolomics  \n",
       "acquisitionBatchSize      Metabolomics  \n",
       "acquisitionBatchSizeUnit  Metabolomics  \n",
       "acquisitionMode           Metabolomics  \n",
       "acquisitionSoftware       Metabolomics  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dm = pd.read_csv(dm_path)\n",
    "dm['module'] = dm['module'].fillna('Unknown')\n",
    "dm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "partition_path = \"../../modules\"\n",
    "\n",
    "# create_data_model_partitions(data_model, partition_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ManifestColumn', 'individual', 'experimentalData', 'batch',\n",
       "       'template', 'biospecimen', 'clinical', 'sequencing',\n",
       "       'demographics', 'mass spectrometry', 'ADKP'], dtype=object)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Split by Parent\n",
    "modules = dm[\"module\"].dropna().unique()\n",
    "modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "import shutil\n",
    "\n",
    "# clean modules folders\n",
    "existing_dir_mods = [s.strip('/') for s in glob.glob(\"**/\", root_dir = partition_path)]\n",
    "\n",
    "for e in list(set(existing_dir_mods) - set(modules)): \n",
    "    print('Removing directory: ', e)\n",
    "    shutil.rmtree(Path(partition_path, e))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create module directories\n",
    "for m in list(set(existing_dir_mods) - set(modules)):\n",
    "    mod_path = Path(partition_path, m)\n",
    "\n",
    "    print(mod_path)\n",
    "\n",
    "    mod_path.mkdir(parents = True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'sequencing/libraryPrep.csv': 'libraryPrep',\n",
       " 'sequencing/readLength.csv': 'readLength',\n",
       " 'sequencing/libraryPreparationMethod.csv': 'libraryPreparationMethod',\n",
       " 'sequencing/lambdaDNAconc.csv': 'lambdaDNAconc',\n",
       " 'sequencing/isStranded.csv': 'isStranded',\n",
       " 'sequencing/meanCoverage.csv': 'meanCoverage',\n",
       " 'sequencing/totalReads.csv': 'totalReads',\n",
       " 'sequencing/gDNAconc.csv': 'gDNAconc',\n",
       " 'sequencing/runType.csv': 'runType',\n",
       " 'sequencing/directionalBSseqLibrary.csv': 'directionalBSseqLibrary',\n",
       " 'sequencing/libraryVersion.csv': 'libraryVersion',\n",
       " 'sequencing/pcrCycles.csv': 'pcrCycles',\n",
       " 'sequencing/readStrandOrigin.csv': 'readStrandOrigin',\n",
       " 'demographics/ethnicity.csv': 'ethnicity',\n",
       " 'demographics/sex.csv': 'sex',\n",
       " 'demographics/race.csv': 'race',\n",
       " 'template/assay_phenotype_human_template.csv': 'assay_phenotype_human_template',\n",
       " 'template/assay_scRNAseq_template.csv': 'assay_scRNAseq_template',\n",
       " 'template/individual_human_template.csv': 'individual_human_template',\n",
       " 'template/individual_non_human_template.csv': 'individual_non_human_template',\n",
       " 'template/assay_metabolomics_template.csv': 'assay_metabolomics_template',\n",
       " 'template/assay_bsSeq_template.csv': 'assay_bsSeq_template',\n",
       " 'template/biospecimen_human_template.csv': 'biospecimen_human_template',\n",
       " 'template/assay_metagenomics_template.csv': 'assay_metagenomics_template',\n",
       " 'template/assay_whole_genome_sequencing_template.csv': 'assay_whole_genome_sequencing_template',\n",
       " 'template/assay_RNAseq_template.csv': 'assay_RNAseq_template',\n",
       " 'template/file_annotation_template.csv': 'file_annotation_template',\n",
       " 'template/genotyping_human_template.csv': 'genotyping_human_template',\n",
       " 'template/study_folder_template.csv': 'study_folder_template',\n",
       " 'template/assay_proteomics_template.csv': 'assay_proteomics_template',\n",
       " 'template/biospecimen_non_human_template.csv': 'biospecimen_non_human_template',\n",
       " 'template/assay_microbiome_template.csv': 'assay_microbiome_template',\n",
       " 'batch/batchSize.csv': 'batchSize',\n",
       " 'ManifestColumn/lensVoltages.csv': 'lensVoltages',\n",
       " 'ManifestColumn/treatmentTemperatureValue.csv': 'treatmentTemperatureValue',\n",
       " 'ManifestColumn/processingBatchSizeUnit.csv': 'processingBatchSizeUnit',\n",
       " 'ManifestColumn/dataTypeAll.csv': 'dataTypeAll',\n",
       " 'ManifestColumn/consortium.csv': 'consortium',\n",
       " 'ManifestColumn/fractionIdentifier.csv': 'fractionIdentifier',\n",
       " 'ManifestColumn/stage.csv': 'stage',\n",
       " 'ManifestColumn/mid.csv': 'mid',\n",
       " 'ManifestColumn/sequencingBatchSizeUnit.csv': 'sequencingBatchSizeUnit',\n",
       " 'ManifestColumn/rnaBatchID.csv': 'rnaBatchID',\n",
       " 'ManifestColumn/speciesAge.csv': 'speciesAge',\n",
       " 'ManifestColumn/fieldCenterCode.csv': 'fieldCenterCode',\n",
       " 'ManifestColumn/acquisitionBatchSizeUnit.csv': 'acquisitionBatchSizeUnit',\n",
       " 'ManifestColumn/treatmentName.csv': 'treatmentName',\n",
       " 'ManifestColumn/studyMetadata.csv': 'studyMetadata',\n",
       " 'ManifestColumn/program.csv': 'program',\n",
       " 'ManifestColumn/rnaBatchSize.csv': 'rnaBatchSize',\n",
       " 'ManifestColumn/methods.csv': 'methods',\n",
       " 'ManifestColumn/dataType.csv': 'dataType',\n",
       " 'ManifestColumn/batchLabel.csv': 'batchLabel',\n",
       " 'ManifestColumn/arrayManufacturer.csv': 'arrayManufacturer',\n",
       " 'ManifestColumn/genotypeTechnology.csv': 'genotypeTechnology',\n",
       " 'ManifestColumn/fileFormat.csv': 'fileFormat',\n",
       " 'ManifestColumn/sourceMatID.csv': 'sourceMatID',\n",
       " 'ManifestColumn/statisticalModel.csv': 'statisticalModel',\n",
       " 'ManifestColumn/databaseName.csv': 'databaseName',\n",
       " 'ManifestColumn/imputation.csv': 'imputation',\n",
       " 'ManifestColumn/processingBatchID.csv': 'processingBatchID',\n",
       " 'ManifestColumn/reagentCatalogNumber.csv': 'reagentCatalogNumber',\n",
       " 'ManifestColumn/sampleBatchID.csv': 'sampleBatchID',\n",
       " 'ManifestColumn/acquisitionBatchID.csv': 'acquisitionBatchID',\n",
       " 'ManifestColumn/sop.csv': 'sop',\n",
       " 'ManifestColumn/diagnosisStatus.csv': 'diagnosisStatus',\n",
       " 'ManifestColumn/reagentContact.csv': 'reagentContact',\n",
       " 'ManifestColumn/dnaBatchSizeUnit.csv': 'dnaBatchSizeUnit',\n",
       " 'ManifestColumn/transcriptType.csv': 'transcriptType',\n",
       " 'ManifestColumn/extractionMethod.csv': 'extractionMethod',\n",
       " 'ManifestColumn/modificationParameters.csv': 'modificationParameters',\n",
       " 'ManifestColumn/libVector.csv': 'libVector',\n",
       " 'ManifestColumn/hasIonizationSource.csv': 'hasIonizationSource',\n",
       " 'ManifestColumn/visitCode.csv': 'visitCode',\n",
       " 'ManifestColumn/repositoryName.csv': 'repositoryName',\n",
       " 'ManifestColumn/ackContext.csv': 'ackContext',\n",
       " 'ManifestColumn/treatmentIDs.csv': 'treatmentIDs',\n",
       " 'ManifestColumn/experimentType.csv': 'experimentType',\n",
       " 'ManifestColumn/captivityDuration.csv': 'captivityDuration',\n",
       " 'ManifestColumn/databaseLibrary.csv': 'databaseLibrary',\n",
       " 'ManifestColumn/treatmentTemperatureUnit.csv': 'treatmentTemperatureUnit',\n",
       " 'ManifestColumn/analysisType.csv': 'analysisType',\n",
       " 'ManifestColumn/experimentalBatchSize.csv': 'experimentalBatchSize',\n",
       " 'ManifestColumn/variantCount.csv': 'variantCount',\n",
       " 'ManifestColumn/relatedStudies.csv': 'relatedStudies',\n",
       " 'ManifestColumn/reagentManufacturer.csv': 'reagentManufacturer',\n",
       " 'ManifestColumn/assay.csv': 'assay',\n",
       " 'ManifestColumn/taxon.csv': 'taxon',\n",
       " 'ManifestColumn/projectFullName.csv': 'projectFullName',\n",
       " 'ManifestColumn/experimentalFactor.csv': 'experimentalFactor',\n",
       " 'ManifestColumn/readmeFile.csv': 'readmeFile',\n",
       " 'ManifestColumn/studyAbstract.csv': 'studyAbstract',\n",
       " 'ManifestColumn/newsRelease.csv': 'newsRelease',\n",
       " 'ManifestColumn/DNAextraction.csv': 'DNAextraction',\n",
       " 'ManifestColumn/acquisitionMode.csv': 'acquisitionMode',\n",
       " 'ManifestColumn/dataSubtype.csv': 'dataSubtype',\n",
       " 'ManifestColumn/Component.csv': 'Component',\n",
       " 'ManifestColumn/assemblyName.csv': 'assemblyName',\n",
       " 'ManifestColumn/sampleIntroduction.csv': 'sampleIntroduction',\n",
       " 'ManifestColumn/hasAssayControl.csv': 'hasAssayControl',\n",
       " 'ManifestColumn/accessReqs.csv': 'accessReqs',\n",
       " 'ManifestColumn/technologyPlatformVersion.csv': 'technologyPlatformVersion',\n",
       " 'ManifestColumn/adjustedCovariates.csv': 'adjustedCovariates',\n",
       " 'ManifestColumn/lensVoltagesUnit.csv': 'lensVoltagesUnit',\n",
       " 'ManifestColumn/vacuumPressureUnit.csv': 'vacuumPressureUnit',\n",
       " 'ManifestColumn/projectShortName.csv': 'projectShortName',\n",
       " 'ManifestColumn/lifeStage.csv': 'lifeStage',\n",
       " 'ManifestColumn/backgroundTrait.csv': 'backgroundTrait',\n",
       " 'ManifestColumn/studyWiki.csv': 'studyWiki',\n",
       " 'ManifestColumn/studyType.csv': 'studyType',\n",
       " 'ManifestColumn/acquisitionBatchSize.csv': 'acquisitionBatchSize',\n",
       " 'ManifestColumn/speciesGroup.csv': 'speciesGroup',\n",
       " 'ManifestColumn/libraryBatchID.csv': 'libraryBatchID',\n",
       " 'ManifestColumn/sampleBatchSize.csv': 'sampleBatchSize',\n",
       " 'ManifestColumn/studyCode.csv': 'studyCode',\n",
       " 'ManifestColumn/consentGroupID.csv': 'consentGroupID',\n",
       " 'ManifestColumn/captivityStatus.csv': 'captivityStatus',\n",
       " 'ManifestColumn/numberContig.csv': 'numberContig',\n",
       " 'ManifestColumn/arrayInformation.csv': 'arrayInformation',\n",
       " 'ManifestColumn/specimenType.csv': 'specimenType',\n",
       " 'ManifestColumn/specimenAge.csv': 'specimenAge',\n",
       " 'ManifestColumn/summaryStatisticsAssembly.csv': 'summaryStatisticsAssembly',\n",
       " 'ManifestColumn/processingBatchSize.csv': 'processingBatchSize',\n",
       " 'ManifestColumn/isFeatured.csv': 'isFeatured',\n",
       " 'ManifestColumn/dataFile.csv': 'dataFile',\n",
       " 'ManifestColumn/specimenID.csv': 'specimenID',\n",
       " 'ManifestColumn/isMultiSpecimen.csv': 'isMultiSpecimen',\n",
       " 'ManifestColumn/useSummaryFile.csv': 'useSummaryFile',\n",
       " 'ManifestColumn/samplePrepProtocol.csv': 'samplePrepProtocol',\n",
       " 'ManifestColumn/treatmentDurationValue.csv': 'treatmentDurationValue',\n",
       " 'ManifestColumn/grantNumber.csv': 'grantNumber',\n",
       " 'ManifestColumn/measurementTechnique.csv': 'measurementTechnique',\n",
       " 'ManifestColumn/sampleType.csv': 'sampleType',\n",
       " 'ManifestColumn/assemblySoftware.csv': 'assemblySoftware',\n",
       " 'ManifestColumn/seqMethod.csv': 'seqMethod',\n",
       " 'ManifestColumn/acquistionSoftwareVersion.csv': 'acquistionSoftwareVersion',\n",
       " 'ManifestColumn/tissueVolume.csv': 'tissueVolume',\n",
       " 'ManifestColumn/conversionRatioUnits.csv': 'conversionRatioUnits',\n",
       " 'ManifestColumn/databaseWeblink.csv': 'databaseWeblink',\n",
       " 'ManifestColumn/libSize.csv': 'libSize',\n",
       " 'ManifestColumn/gasFlowTemperature.csv': 'gasFlowTemperature',\n",
       " 'ManifestColumn/reagentIDs.csv': 'reagentIDs',\n",
       " 'ManifestColumn/digestionMethod.csv': 'digestionMethod',\n",
       " 'ManifestColumn/imputationPanel.csv': 'imputationPanel',\n",
       " 'ManifestColumn/studyDescription.csv': 'studyDescription',\n",
       " 'ManifestColumn/msAssayTechnique.csv': 'msAssayTechnique',\n",
       " 'ManifestColumn/chromosome.csv': 'chromosome',\n",
       " 'ManifestColumn/adapters.csv': 'adapters',\n",
       " 'ManifestColumn/labelQuantitation.csv': 'labelQuantitation',\n",
       " 'ManifestColumn/sequencingBatchID.csv': 'sequencingBatchID',\n",
       " 'ManifestColumn/dnaBatchSize.csv': 'dnaBatchSize',\n",
       " 'ManifestColumn/imputationSoftware.csv': 'imputationSoftware',\n",
       " 'ManifestColumn/specimenSource.csv': 'specimenSource',\n",
       " 'ManifestColumn/individualID.csv': 'individualID',\n",
       " 'ManifestColumn/reagentName.csv': 'reagentName',\n",
       " 'ManifestColumn/countryCode.csv': 'countryCode',\n",
       " 'ManifestColumn/nucleAcidExt.csv': 'nucleAcidExt',\n",
       " 'ManifestColumn/study.csv': 'study',\n",
       " 'ManifestColumn/metadataType.csv': 'metadataType',\n",
       " 'ManifestColumn/studyFocus.csv': 'studyFocus',\n",
       " 'ManifestColumn/resultUnit.csv': 'resultUnit',\n",
       " 'ManifestColumn/sampleBatchSizeUnit.csv': 'sampleBatchSizeUnit',\n",
       " 'ManifestColumn/msAnalyteType.csv': 'msAnalyteType',\n",
       " 'ManifestColumn/reportedTrait.csv': 'reportedTrait',\n",
       " 'ManifestColumn/studyName.csv': 'studyName',\n",
       " 'ManifestColumn/dnaBatchID.csv': 'dnaBatchID',\n",
       " 'ManifestColumn/labelFreeQuantitation.csv': 'labelFreeQuantitation',\n",
       " 'ManifestColumn/isobaricLabelingReagent.csv': 'isobaricLabelingReagent',\n",
       " 'ManifestColumn/reagentWeblink.csv': 'reagentWeblink',\n",
       " 'ManifestColumn/studyKey.csv': 'studyKey',\n",
       " 'ManifestColumn/treatmentAmountUnit.csv': 'treatmentAmountUnit',\n",
       " 'ManifestColumn/referenceTranscriptID.csv': 'referenceTranscriptID',\n",
       " 'ManifestColumn/useTreatment.csv': 'useTreatment',\n",
       " 'ManifestColumn/valueReported.csv': 'valueReported',\n",
       " 'ManifestColumn/batchID.csv': 'batchID',\n",
       " 'ManifestColumn/batchSizeUnit.csv': 'batchSizeUnit',\n",
       " 'ManifestColumn/conversionRatio.csv': 'conversionRatio',\n",
       " 'ManifestColumn/digestionReagent.csv': 'digestionReagent',\n",
       " 'ManifestColumn/associatedResource.csv': 'associatedResource',\n",
       " 'ManifestColumn/rnaBatchSizeUnit.csv': 'rnaBatchSizeUnit',\n",
       " 'ManifestColumn/libReadsSeqd.csv': 'libReadsSeqd',\n",
       " 'ManifestColumn/ethnicGroupCode.csv': 'ethnicGroupCode',\n",
       " 'ManifestColumn/treatmentAmountValue.csv': 'treatmentAmountValue',\n",
       " 'ManifestColumn/Filename.csv': 'Filename',\n",
       " 'ManifestColumn/treatmentDurationUnit.csv': 'treatmentDurationUnit',\n",
       " 'ManifestColumn/seqPlatform.csv': 'seqPlatform',\n",
       " 'ManifestColumn/ionProperty.csv': 'ionProperty',\n",
       " 'ManifestColumn/instrumentModel.csv': 'instrumentModel',\n",
       " 'ManifestColumn/summaryFile.csv': 'summaryFile',\n",
       " 'ManifestColumn/quantationStrategy.csv': 'quantationStrategy',\n",
       " 'ManifestColumn/commonName.csv': 'commonName',\n",
       " 'ManifestColumn/speciesName.csv': 'speciesName',\n",
       " 'ManifestColumn/readLengthUnits.csv': 'readLengthUnits',\n",
       " 'ManifestColumn/isModelSystem.csv': 'isModelSystem',\n",
       " 'ManifestColumn/enrichment.csv': 'enrichment',\n",
       " 'ManifestColumn/numberIndividuals.csv': 'numberIndividuals',\n",
       " 'ManifestColumn/acknowledgment.csv': 'acknowledgment',\n",
       " 'ManifestColumn/vacuumPressure.csv': 'vacuumPressure',\n",
       " 'ManifestColumn/experimentalBatchSizeUnit.csv': 'experimentalBatchSizeUnit',\n",
       " 'ManifestColumn/msTarget.csv': 'msTarget',\n",
       " 'ManifestColumn/enrichmentMethod.csv': 'enrichmentMethod',\n",
       " 'ManifestColumn/resourceType.csv': 'resourceType',\n",
       " 'ManifestColumn/useReagent.csv': 'useReagent',\n",
       " 'ManifestColumn/dataModelVersion.csv': 'dataModelVersion',\n",
       " 'ManifestColumn/gasFlowTemperatureUnit.csv': 'gasFlowTemperatureUnit',\n",
       " 'ManifestColumn/acquisitionSoftware.csv': 'acquisitionSoftware',\n",
       " 'ManifestColumn/isReleased.csv': 'isReleased',\n",
       " 'ManifestColumn/databaseVersion.csv': 'databaseVersion',\n",
       " 'ManifestColumn/reagentLotNumber.csv': 'reagentLotNumber',\n",
       " 'ManifestColumn/targetGene.csv': 'targetGene',\n",
       " 'ManifestColumn/databaseSource.csv': 'databaseSource',\n",
       " 'ManifestColumn/assemblyQual.csv': 'assemblyQual',\n",
       " 'ManifestColumn/libLayout.csv': 'libLayout',\n",
       " 'ManifestColumn/sequencingBatchSize.csv': 'sequencingBatchSize',\n",
       " 'ManifestColumn/grant.csv': 'grant',\n",
       " 'ManifestColumn/useReadMeFile.csv': 'useReadMeFile',\n",
       " 'ManifestColumn/acquisitionSoftwareVersion.csv': 'acquisitionSoftwareVersion',\n",
       " 'ManifestColumn/enrichmentStrategy.csv': 'enrichmentStrategy',\n",
       " 'mass spectrometry/spectrometerFrequency.csv': 'spectrometerFrequency',\n",
       " 'experimentalData/controlType.csv': 'controlType',\n",
       " 'experimentalData/assay.csv': 'assay',\n",
       " 'experimentalData/platformLocation.csv': 'platformLocation',\n",
       " 'biospecimen/nucleicAcidSource.csv': 'nucleicAcidSource',\n",
       " 'biospecimen/tissueWeight.csv': 'tissueWeight',\n",
       " 'biospecimen/tissue.csv': 'tissue',\n",
       " 'biospecimen/fastingState.csv': 'fastingState',\n",
       " 'biospecimen/cellType.csv': 'cellType',\n",
       " 'biospecimen/specimenID.csv': 'specimenID',\n",
       " 'biospecimen/isPostMortem.csv': 'isPostMortem',\n",
       " 'biospecimen/samplingAge.csv': 'samplingAge',\n",
       " 'biospecimen/organ.csv': 'organ',\n",
       " 'clinical/diagnosis.csv': 'diagnosis',\n",
       " 'ADKP/isModelSystem.csv': 'isModelSystem',\n",
       " 'individual/species.csv': 'species',\n",
       " 'individual/age.csv': 'age',\n",
       " 'individual/cohort.csv': 'cohort',\n",
       " 'individual/individualID.csv': 'individualID'}"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check for existing csv files\n",
    "existing_csvs = glob.glob('**/*.csv', recursive = True, root_dir = partition_path)\n",
    "existing_attrs = [Path(e).stem for e in existing_csvs]\n",
    "existing_df = pd.DataFrame(\n",
    "    {\"csv_paths\": existing_csvs, \"existing_attrs\": existing_attrs}\n",
    ")\n",
    "existing_df[\"module\"] = existing_df[\"csv_paths\"].apply(lambda x: str(Path(x).parent))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Updating module CSVs: 100%|██████████| 238/238 [00:00<00:00, 5499.18it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n",
      "CSV already exists\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "for i in tqdm(range(len(dm)), desc = 'Updating module CSVs'): \n",
    "    try: \n",
    "        attr = dm.loc[i, 'Attribute']\n",
    "        module = dm.loc[i, 'module']\n",
    "        csv_path = Path(module, attr + '.csv')\n",
    "        full_csv_path = Path(partition_path, csv_path)\n",
    "        \n",
    "        checker = existing_df[existing_df['existing_attrs'] == attr]\n",
    "        \n",
    "        if all(checker['module'] == module): \n",
    "            # print('CSV already exists')\n",
    "            pass # need to do comparison later\n",
    "        \n",
    "        elif all(checker['module'] != module):  # if module is different\n",
    "            # remove old csv\n",
    "            print('Removing old CSV file: ', attr)\n",
    "            Path.unlink(full_csv_path)\n",
    "            \n",
    "            # create new csv\n",
    "            print('Creating new CSV: ', csv_path)\n",
    "            dm.loc[i, ].to_csv(full_csv_path, index = 'ignore')\n",
    "        \n",
    "        else:     \n",
    "            # create new csv\n",
    "            dm.loc[i, ].to_csv(full_csv_path, index = 'ignore')\n",
    "    \n",
    "    except Exception as e: \n",
    "        print(attr, e)\n",
    "        pass"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "elite-data-models-LjmgyG4K-py3.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
