{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "import re\n",
    "from datetime import datetime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data model\n",
    "with open('../../EL.data.model.jsonld', 'r', encoding='UTF-8') as f: \n",
    "    dm_json = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get template names\n",
    "for i in dm_json['@graph']: \n",
    "    try: \n",
    "        if bool(re.search('template', i['@id'])): \n",
    "            # print(i)\n",
    "            templates = [v for n in i[\"schema:domainIncludes\"] for _,v in n.items()]\n",
    "    except KeyError: \n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "| display_name                           | schema_name                        | type   |\n",
      "|:---------------------------------------|:-----------------------------------|:-------|\n",
      "| assay_RNAseq_template                  | AssayRNAseqTemplate                | file   |\n",
      "| assay_bsSeq_template                   | AssayBsSeqTemplate                 | file   |\n",
      "| assay_metabolomics_template            | AssayMetabolomicsTemplate          | file   |\n",
      "| assay_metagenomics_template            | AssayMetagenomicsTemplate          | file   |\n",
      "| assay_microbiome_template              | AssayMicrobiomeTemplate            | file   |\n",
      "| assay_phenotype_human_template         | AssayPhenotypeHumanTemplate        | file   |\n",
      "| assay_proteomics_template              | AssayProteomicsTemplate            | file   |\n",
      "| assay_scRNAseq_template                | AssayScRNAseqTemplate              | file   |\n",
      "| assay_whole_genome_sequencing_template | AssayWholeGenomeSequencingTemplate | file   |\n",
      "| file_annotation_template               | FileAnnotationTemplate             | file   |\n",
      "| genotyping_human_template              | GenotypingHumanTemplate            | file   |\n",
      "| study_folder_template                  | StudyFolderTemplate                | file   |\n",
      "| biospecimen_human_template             | BiospecimenHumanTemplate           | record |\n",
      "| biospecimen_non_human_template         | BiospecimenNonHumanTemplate        | record |\n",
      "| individual_human_template              | IndividualHumanTemplate            | record |\n",
      "| individual_non_human_template          | IndividualNonHumanTemplate         | record |\n",
      "\n",
      "Shape:  (16, 3)\n"
     ]
    }
   ],
   "source": [
    "# re-create template\n",
    "# get 'sms:displayName' for \"display_name\" and '@id' for \"schema_name\"\n",
    "\n",
    "manifest_schemas = []\n",
    "\n",
    "for i in dm_json['@graph']: \n",
    "    try: \n",
    "        if i['@id'] in templates: \n",
    "            if bool(re.search('individual|biospecimen', i['@id'], flags=re.IGNORECASE)): \n",
    "                record_type = 'record'\n",
    "            else: \n",
    "                record_type = 'file'\n",
    "                \n",
    "            temp_template = {\n",
    "                'display_name': i['sms:displayName'], \n",
    "                \"schema_name\": i['@id'].strip('bts:'),\n",
    "                \"type\": record_type\n",
    "            }\n",
    "\n",
    "            manifest_schemas.append(temp_template)\n",
    "    except KeyError: \n",
    "        pass\n",
    "\n",
    "manifest_schemas_df = (\n",
    "    pd.DataFrame(manifest_schemas)\n",
    "    .sort_values([\"type\", \"display_name\"], ascending=True)\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "print(manifest_schemas_df.to_markdown(index=False))\n",
    "print()\n",
    "print(\"Shape: \", manifest_schemas_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import Markdown"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "| _merge     | display_name                           | schema_name_new                    | schema_name_previous   | type_new   | type_previous   |\n",
      "|:-----------|:---------------------------------------|:-----------------------------------|:-----------------------|:-----------|:----------------|\n",
      "| right_only | Biospecimen_human                      | nan                                | BiospecimenHuman       | nan        | record          |\n",
      "| right_only | Biospecimen_non_Human                  | nan                                | BiospecimenNonHuman    | nan        | record          |\n",
      "| right_only | Genotyping_Human                       | nan                                | GenotypingHuman        | nan        | file            |\n",
      "| right_only | Individual_Human                       | nan                                | IndividualHuman        | nan        | record          |\n",
      "| right_only | Individual_non_Human                   | nan                                | IndividualNonHuman     | nan        | record          |\n",
      "| right_only | Metabolomics                           | nan                                | Metabolomics           | nan        | file            |\n",
      "| right_only | Metagenomics                           | nan                                | Metagenomics           | nan        | file            |\n",
      "| right_only | Microbiome                             | nan                                | Microbiome             | nan        | file            |\n",
      "| right_only | Proteomics                             | nan                                | Proteomics             | nan        | file            |\n",
      "| right_only | RNAseq                                 | nan                                | RNAseq                 | nan        | file            |\n",
      "| right_only | Whole_Genome_Sequencing                | nan                                | WholeGenomeSequencing  | nan        | file            |\n",
      "| left_only  | assay_RNAseq_template                  | AssayRNAseqTemplate                | nan                    | file       | nan             |\n",
      "| left_only  | assay_bsSeq_template                   | AssayBsSeqTemplate                 | nan                    | file       | nan             |\n",
      "| left_only  | assay_metabolomics_template            | AssayMetabolomicsTemplate          | nan                    | file       | nan             |\n",
      "| left_only  | assay_metagenomics_template            | AssayMetagenomicsTemplate          | nan                    | file       | nan             |\n",
      "| left_only  | assay_microbiome_template              | AssayMicrobiomeTemplate            | nan                    | file       | nan             |\n",
      "| left_only  | assay_phenotype_human_template         | AssayPhenotypeHumanTemplate        | nan                    | file       | nan             |\n",
      "| left_only  | assay_proteomics_template              | AssayProteomicsTemplate            | nan                    | file       | nan             |\n",
      "| left_only  | assay_scRNAseq_template                | AssayScRNAseqTemplate              | nan                    | file       | nan             |\n",
      "| left_only  | assay_whole_genome_sequencing_template | AssayWholeGenomeSequencingTemplate | nan                    | file       | nan             |\n",
      "| left_only  | biospecimen_human_template             | BiospecimenHumanTemplate           | nan                    | record     | nan             |\n",
      "| left_only  | biospecimen_non_human_template         | BiospecimenNonHumanTemplate        | nan                    | record     | nan             |\n",
      "| right_only | bsSeq                                  | nan                                | BsSeq                  | nan        | file            |\n",
      "| left_only  | file_annotation_template               | FileAnnotationTemplate             | nan                    | file       | nan             |\n",
      "| left_only  | genotyping_human_template              | GenotypingHumanTemplate            | nan                    | file       | nan             |\n",
      "| left_only  | individual_human_template              | IndividualHumanTemplate            | nan                    | record     | nan             |\n",
      "| left_only  | individual_non_human_template          | IndividualNonHumanTemplate         | nan                    | record     | nan             |\n",
      "| right_only | scRNAseq                               | nan                                | ScRNAseq               | nan        | file            |\n",
      "| left_only  | study_folder_template                  | StudyFolderTemplate                | nan                    | file       | nan             |\n"
     ]
    }
   ],
   "source": [
    "with open(\"../../dca-template-config.json\", \"r\", encoding=\"UTF-8\") as f:\n",
    "    template_config = json.load(f)\n",
    "\n",
    "# difference\n",
    "m = manifest_schemas_df.merge(\n",
    "    pd.DataFrame(template_config[\"manifest_schemas\"]),\n",
    "    on=\"display_name\",\n",
    "    how=\"outer\",\n",
    "    suffixes=[\"_new\", \"_previous\"],\n",
    "    indicator=True,\n",
    ")\n",
    "\n",
    "\n",
    "print(m[sorted(list(m.columns))].to_markdown(index = False))\n",
    "\n",
    "template_config[\"manifest_schemas\"] = manifest_schemas  # update manifests\n",
    "template_config[\"schema_version\"] = \"v\" + datetime.now().strftime(\n",
    "    \"%Y.%m.%d\"\n",
    ")  # increment schema version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "New Schema: \n",
      "\n",
      "{\n",
      "    \"manifest_schemas\": [\n",
      "        {\n",
      "            \"display_name\": \"biospecimen_human_template\",\n",
      "            \"schema_name\": \"BiospecimenHumanTemplate\",\n",
      "            \"type\": \"record\"\n",
      "        },\n",
      "        {\n",
      "            \"display_name\": \"biospecimen_non_human_template\",\n",
      "            \"schema_name\": \"BiospecimenNonHumanTemplate\",\n",
      "            \"type\": \"record\"\n",
      "        },\n",
      "        {\n",
      "            \"display_name\": \"assay_bsSeq_template\",\n",
      "            \"schema_name\": \"AssayBsSeqTemplate\",\n",
      "            \"type\": \"file\"\n",
      "        },\n",
      "        {\n",
      "            \"display_name\": \"genotyping_human_template\",\n",
      "            \"schema_name\": \"GenotypingHumanTemplate\",\n",
      "            \"type\": \"file\"\n",
      "        },\n",
      "        {\n",
      "            \"display_name\": \"individual_human_template\",\n",
      "            \"schema_name\": \"IndividualHumanTemplate\",\n",
      "            \"type\": \"record\"\n",
      "        },\n",
      "        {\n",
      "            \"display_name\": \"individual_non_human_template\",\n",
      "            \"schema_name\": \"IndividualNonHumanTemplate\",\n",
      "            \"type\": \"record\"\n",
      "        },\n",
      "        {\n",
      "            \"display_name\": \"assay_metabolomics_template\",\n",
      "            \"schema_name\": \"AssayMetabolomicsTemplate\",\n",
      "            \"type\": \"file\"\n",
      "        },\n",
      "        {\n",
      "            \"display_name\": \"assay_metagenomics_template\",\n",
      "            \"schema_name\": \"AssayMetagenomicsTemplate\",\n",
      "            \"type\": \"file\"\n",
      "        },\n",
      "        {\n",
      "            \"display_name\": \"assay_microbiome_template\",\n",
      "            \"schema_name\": \"AssayMicrobiomeTemplate\",\n",
      "            \"type\": \"file\"\n",
      "        },\n",
      "        {\n",
      "            \"display_name\": \"assay_phenotype_human_template\",\n",
      "            \"schema_name\": \"AssayPhenotypeHumanTemplate\",\n",
      "            \"type\": \"file\"\n",
      "        },\n",
      "        {\n",
      "            \"display_name\": \"assay_proteomics_template\",\n",
      "            \"schema_name\": \"AssayProteomicsTemplate\",\n",
      "            \"type\": \"file\"\n",
      "        },\n",
      "        {\n",
      "            \"display_name\": \"assay_RNAseq_template\",\n",
      "            \"schema_name\": \"AssayRNAseqTemplate\",\n",
      "            \"type\": \"file\"\n",
      "        },\n",
      "        {\n",
      "            \"display_name\": \"assay_whole_genome_sequencing_template\",\n",
      "            \"schema_name\": \"AssayWholeGenomeSequencingTemplate\",\n",
      "            \"type\": \"file\"\n",
      "        },\n",
      "        {\n",
      "            \"display_name\": \"assay_scRNAseq_template\",\n",
      "            \"schema_name\": \"AssayScRNAseqTemplate\",\n",
      "            \"type\": \"file\"\n",
      "        },\n",
      "        {\n",
      "            \"display_name\": \"study_folder_template\",\n",
      "            \"schema_name\": \"StudyFolderTemplate\",\n",
      "            \"type\": \"file\"\n",
      "        },\n",
      "        {\n",
      "            \"display_name\": \"file_annotation_template\",\n",
      "            \"schema_name\": \"FileAnnotationTemplate\",\n",
      "            \"type\": \"file\"\n",
      "        }\n",
      "    ],\n",
      "    \"service_version\": \"v23.1.1\",\n",
      "    \"schema_version\": \"v2024.05.30\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print('New Schema: \\n')\n",
    "print(json.dumps(template_config, indent = 4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../../dca-template-config.json\", \"w\", encoding=\"UTF-8\") as f:\n",
    "    json.dump(template_config, f, ensure_ascii=False, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
